{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Предварительная обработка текстовых данных перед их передачей в модели Machine Learning."
      ],
      "metadata": {
        "id": "bmYshdoM-aIL"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Первый корпус (*BIOcorpus*) базируется на материале 251 научной статьи разного типа (research article, review article), которые были опубликованы в период с 2010 по 2019 гг., из междисциплинарных и специализированных периодических журналов по 7 дисциплинарным направлениям (биохимия, биофизика, биотехнология/биоинженерия, клеточная биология, генетика, гидробиология, молекулярная биология).\n",
        "\n",
        "* Предобработка медико-биологических текстов требует особого подхода из-за превалирования в них специальной терминологии, часто представленной цепочками число-буквенных последовательностей и/или аббревиатур."
      ],
      "metadata": {
        "id": "jARSjywc_XGy"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Для предобработки текстов BIOcorpus была использована модель scispaCy, созданная специально для обработки медико-биологических текстов (малая модель для 100тыс. словарных ед). Также был использован EntityLinker - компонент SpaCy, который выполняет связывание с базой знаний (в нашем случае с  the Unified Medical Language System, содержащей ~3M концептов. Linker  выполняет поиск  именованных сущностей (от char до 3grams), сравнивая их с концепциями в базе знаний с помощью алгоритма поиска ближайших соседей.\n",
        "\n",
        "https://github.com/allenai/scispacy\n",
        "\n",
        "Neumann, M., King, D., Beltagy, I., Ammar, W. (2019). ScispaCy: Fast and Robust Models for\n",
        "Biomedical Natural Language Processing, Proceedings of the 18th BioNLP Workshop and Shared Task (pp. 319–327), Florence: Association for Computational Linguistics."
      ],
      "metadata": {
        "id": "M-U35ZBgJVcG"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3uTQr27CrjS9",
        "outputId": "aa8ca046-004a-49e7-8d9f-c31306792c8b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.4/en_core_sci_sm-0.5.4.tar.gz\n",
            "  Downloading https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.4/en_core_sci_sm-0.5.4.tar.gz (14.8 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.8/14.8 MB\u001b[0m \u001b[31m32.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: spacy<3.8.0,>=3.7.4 in /usr/local/lib/python3.10/dist-packages (from en-core-sci-sm==0.5.4) (3.7.5)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.11 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.0.12)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.0.5)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.0.10)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.0.8)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.0.9)\n",
            "Requirement already satisfied: thinc<8.3.0,>=8.2.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (8.2.4)\n",
            "Requirement already satisfied: wasabi<1.2.0,>=0.9.1 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.1.3)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.3 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.4.8)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.0.10)\n",
            "Requirement already satisfied: weasel<0.5.0,>=0.1.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.4.1)\n",
            "Requirement already satisfied: typer<1.0.0,>=0.3.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.12.3)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (4.66.4)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.31.0)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.7.3)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.1.4)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (67.7.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (24.1)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.4.0)\n",
            "Requirement already satisfied: numpy>=1.19.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.25.2)\n",
            "Requirement already satisfied: language-data>=1.2 in /usr/local/lib/python3.10/dist-packages (from langcodes<4.0.0,>=3.2.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.2.0)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.18.4 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.18.4)\n",
            "Requirement already satisfied: typing-extensions>=4.6.1 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (4.12.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.7)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2024.6.2)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.7.8 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.2.2->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.7.11)\n",
            "Requirement already satisfied: confection<1.0.0,>=0.0.1 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.2.2->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.1.5)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (8.1.7)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (13.7.1)\n",
            "Requirement already satisfied: cloudpathlib<1.0.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.18.1)\n",
            "Requirement already satisfied: smart-open<8.0.0,>=5.2.1 in /usr/local/lib/python3.10/dist-packages (from weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (7.0.4)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.1.5)\n",
            "Requirement already satisfied: marisa-trie>=0.7.7 in /usr/local/lib/python3.10/dist-packages (from language-data>=1.2->langcodes<4.0.0,>=3.2.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.2.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (2.16.1)\n",
            "Requirement already satisfied: wrapt in /usr/local/lib/python3.10/dist-packages (from smart-open<8.0.0,>=5.2.1->weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (1.14.1)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.4->en-core-sci-sm==0.5.4) (0.1.2)\n",
            "Building wheels for collected packages: en-core-sci-sm\n",
            "  Building wheel for en-core-sci-sm (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for en-core-sci-sm: filename=en_core_sci_sm-0.5.4-py3-none-any.whl size=14778490 sha256=5bfbb75bc5d4f219409ae926b8e2bada3227beac940f8e478ce88dfe02a05877\n",
            "  Stored in directory: /root/.cache/pip/wheels/0e/d3/6e/e03165bcf8c0fe90c7a41e8a44dd268e4c7779582c5e022707\n",
            "Successfully built en-core-sci-sm\n",
            "Installing collected packages: en-core-sci-sm\n",
            "Successfully installed en-core-sci-sm-0.5.4\n"
          ]
        }
      ],
      "source": [
        "! pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.4/en_core_sci_sm-0.5.4.tar.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "YkY6RnAarn2n",
        "outputId": "1f174736-8feb-4247-af84-c6cecdbd12d1"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting scispacy\n",
            "  Downloading scispacy-0.5.4-py3-none-any.whl (45 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.4/45.4 kB\u001b[0m \u001b[31m865.1 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: spacy<3.8.0,>=3.7.0 in /usr/local/lib/python3.10/dist-packages (from scispacy) (3.7.5)\n",
            "Collecting scipy<1.11 (from scispacy)\n",
            "  Downloading scipy-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.4 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.4/34.4 MB\u001b[0m \u001b[31m17.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: requests<3.0.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scispacy) (2.31.0)\n",
            "Collecting conllu (from scispacy)\n",
            "  Downloading conllu-4.5.3-py2.py3-none-any.whl (16 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from scispacy) (1.25.2)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.10/dist-packages (from scispacy) (1.4.2)\n",
            "Collecting nmslib>=1.7.3.6 (from scispacy)\n",
            "  Downloading nmslib-2.1.1.tar.gz (188 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m188.7/188.7 kB\u001b[0m \u001b[31m22.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: scikit-learn>=0.20.3 in /usr/local/lib/python3.10/dist-packages (from scispacy) (1.2.2)\n",
            "Collecting pysbd (from scispacy)\n",
            "  Downloading pysbd-0.3.4-py3-none-any.whl (71 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m71.1/71.1 kB\u001b[0m \u001b[31m10.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting pybind11<2.6.2 (from nmslib>=1.7.3.6->scispacy)\n",
            "  Using cached pybind11-2.6.1-py2.py3-none-any.whl (188 kB)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.10/dist-packages (from nmslib>=1.7.3.6->scispacy) (5.9.5)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (3.7)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2024.6.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.20.3->scispacy) (3.5.0)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.11 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (3.0.12)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (1.0.5)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (1.0.10)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (2.0.8)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (3.0.9)\n",
            "Requirement already satisfied: thinc<8.3.0,>=8.2.2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (8.2.4)\n",
            "Requirement already satisfied: wasabi<1.2.0,>=0.9.1 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (1.1.3)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.3 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (2.4.8)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (2.0.10)\n",
            "Requirement already satisfied: weasel<0.5.0,>=0.1.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (0.4.1)\n",
            "Requirement already satisfied: typer<1.0.0,>=0.3.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (0.12.3)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (4.66.4)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (2.7.3)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (3.1.4)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (67.7.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (24.1)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.10/dist-packages (from spacy<3.8.0,>=3.7.0->scispacy) (3.4.0)\n",
            "Requirement already satisfied: language-data>=1.2 in /usr/local/lib/python3.10/dist-packages (from langcodes<4.0.0,>=3.2.0->spacy<3.8.0,>=3.7.0->scispacy) (1.2.0)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.0->scispacy) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.18.4 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.0->scispacy) (2.18.4)\n",
            "Requirement already satisfied: typing-extensions>=4.6.1 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy<3.8.0,>=3.7.0->scispacy) (4.12.2)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.7.8 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.2.2->spacy<3.8.0,>=3.7.0->scispacy) (0.7.11)\n",
            "Requirement already satisfied: confection<1.0.0,>=0.0.1 in /usr/local/lib/python3.10/dist-packages (from thinc<8.3.0,>=8.2.2->spacy<3.8.0,>=3.7.0->scispacy) (0.1.5)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (8.1.7)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.10/dist-packages (from typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (13.7.1)\n",
            "Requirement already satisfied: cloudpathlib<1.0.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.0->scispacy) (0.18.1)\n",
            "Requirement already satisfied: smart-open<8.0.0,>=5.2.1 in /usr/local/lib/python3.10/dist-packages (from weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.0->scispacy) (7.0.4)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->spacy<3.8.0,>=3.7.0->scispacy) (2.1.5)\n",
            "Requirement already satisfied: marisa-trie>=0.7.7 in /usr/local/lib/python3.10/dist-packages (from language-data>=1.2->langcodes<4.0.0,>=3.2.0->spacy<3.8.0,>=3.7.0->scispacy) (1.2.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (2.16.1)\n",
            "Requirement already satisfied: wrapt in /usr/local/lib/python3.10/dist-packages (from smart-open<8.0.0,>=5.2.1->weasel<0.5.0,>=0.1.0->spacy<3.8.0,>=3.7.0->scispacy) (1.14.1)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<1.0.0,>=0.3.0->spacy<3.8.0,>=3.7.0->scispacy) (0.1.2)\n",
            "Building wheels for collected packages: nmslib\n",
            "  Building wheel for nmslib (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for nmslib: filename=nmslib-2.1.1-cp310-cp310-linux_x86_64.whl size=13578731 sha256=729eeb52dd9364cc7c2c495ce9c0ed0b8900739de1c82a769f57dbdfc23bc055\n",
            "  Stored in directory: /root/.cache/pip/wheels/21/1a/5d/4cc754a5b1a88405cad184b76f823897a63a8d19afcd4b9314\n",
            "Successfully built nmslib\n",
            "Installing collected packages: scipy, pysbd, pybind11, conllu, nmslib, scispacy\n",
            "  Attempting uninstall: scipy\n",
            "    Found existing installation: scipy 1.11.4\n",
            "    Uninstalling scipy-1.11.4:\n",
            "      Successfully uninstalled scipy-1.11.4\n",
            "Successfully installed conllu-4.5.3 nmslib-2.1.1 pybind11-2.6.1 pysbd-0.3.4 scipy-1.10.1 scispacy-0.5.4\n"
          ]
        }
      ],
      "source": [
        "! pip install scispacy"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_6F7spKNTIk5"
      },
      "outputs": [],
      "source": [
        "# Функция для подсчета словоупотреблений в корпусе\n",
        "def count_words(df):\n",
        "    w = 0\n",
        "    for idx, row in df.iterrows():\n",
        "        w += len(row['text'].split())\n",
        "    return w"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tw3JcP3gGrbY",
        "outputId": "d53bcdb0-3a5c-44fe-a78a-92059b9c9e86"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package stopwords to /root/nltk_data...\n",
            "[nltk_data]   Unzipping corpora/stopwords.zip.\n",
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ],
      "source": [
        "import scispacy.linking\n",
        "import scispacy\n",
        "import spacy\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "import re\n",
        "import string\n",
        "import nltk\n",
        "from nltk.corpus import stopwords\n",
        "\n",
        "nltk.download(\"stopwords\")\n",
        "stops = stopwords.words(\"english\")\n",
        "\n",
        "nltk.download('punkt')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rp4SAVN4siXa",
        "outputId": "4bcc75da-d960-4ef4-c748-5f69afef3146"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/spacy/language.py:2195: FutureWarning: Possible set union at position 6328\n",
            "  deserializers[\"tokenizer\"] = lambda p: self.tokenizer.from_disk(  # type: ignore[union-attr]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/linkers/2023-04-23/umls/tfidf_vectors_sparse.npz not found in cache, downloading to /tmp/tmpe4mbq0ca\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 492M/492M [00:12<00:00, 39.9MiB/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmpe4mbq0ca to cache at /root/.scispacy/datasets/2b79923846fb52e62d686f2db846392575c8eb5b732d9d26cd3ca9378c622d40.87bd52d0f0ee055c1e455ef54ba45149d188552f07991b765da256a1b512ca0b.tfidf_vectors_sparse.npz\n",
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/linkers/2023-04-23/umls/nmslib_index.bin not found in cache, downloading to /tmp/tmpw_vzn1yn\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 724M/724M [00:33<00:00, 22.7MiB/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmpw_vzn1yn to cache at /root/.scispacy/datasets/7e8e091ec80370b87b1652f461eae9d926e543a403a69c1f0968f71157322c25.6d801a1e14867953e36258b0e19a23723ae84b0abd2a723bdd3574c3e0c873b4.nmslib_index.bin\n",
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/linkers/2023-04-23/umls/tfidf_vectorizer.joblib not found in cache, downloading to /tmp/tmp1t8y9nay\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1.32M/1.32M [00:00<00:00, 3.31MiB/s]\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator TfidfTransformer from version 1.1.2 when using version 1.2.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
            "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmp1t8y9nay to cache at /root/.scispacy/datasets/37bc06bb7ce30de7251db5f5cbac788998e33b3984410caed2d0083187e01d38.f0994c1b61cc70d0eb96dea4947dddcb37460fb5ae60975013711228c8fe3fba.tfidf_vectorizer.joblib\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/sklearn/base.py:318: UserWarning: Trying to unpickle estimator TfidfVectorizer from version 1.1.2 when using version 1.2.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
            "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/linkers/2023-04-23/umls/concept_aliases.json not found in cache, downloading to /tmp/tmpqciibawn\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 264M/264M [00:07<00:00, 36.8MiB/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmpqciibawn to cache at /root/.scispacy/datasets/6238f505f56aca33290aab44097f67dd1b88880e3be6d6dcce65e56e9255b7d4.d7f77b1629001b40f1b1bc951f3a890ff2d516fb8fbae3111b236b31b33d6dcf.concept_aliases.json\n",
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/kbs/2023-04-23/umls_2022_ab_cat0129.jsonl not found in cache, downloading to /tmp/tmp_42sg0fq\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 628M/628M [00:55<00:00, 11.9MiB/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmp_42sg0fq to cache at /root/.scispacy/datasets/d5e593bc2d8adeee7754be423cd64f5d331ebf26272074a2575616be55697632.0660f30a60ad00fffd8bbf084a18eb3f462fd192ac5563bf50940fc32a850a3c.umls_2022_ab_cat0129.jsonl\n",
            "https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/data/umls_semantic_type_tree.tsv not found in cache, downloading to /tmp/tmpp045ylzg\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 4.26k/4.26k [00:00<00:00, 8.76MiB/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Finished download, copying /tmp/tmpp045ylzg to cache at /root/.scispacy/datasets/21a1012c532c3a431d60895c509f5b4d45b0f8966c4178b892190a302b21836f.330707f4efe774134872b9f77f0e3208c1d30f50800b3b39a6b8ec21d9adf1b7.umls_semantic_type_tree.tsv\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<scispacy.linking.EntityLinker at 0x7a86a2846e90>"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "nlp = spacy.load(\"en_core_sci_sm\")\n",
        "nlp.add_pipe(\"scispacy_linker\", config={\"resolve_abbreviations\": True, \"linker_name\": \"umls\"})"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## При обработке текста с помощью sciscpaCy стало очевидно, насколько значительную часть академического текста составляют термины: текст сокращался приблизительно в 4 раза, оставляя практически голый скелет из служебных слов и редких ЛЕ литературного языка. Это связано с тем, что многие так называемые общенаучные / академические ЛЕ в научных текстах приобретают особую семантику, а также входят в состав терминологических единиц, являясь \"субстратом\" для создания новых терминов (В.М.Лейчик). Таким образом, было решено удалять термины, оставляя при этом академические слова, даже входящие в терминологические единицы. В качестве списка академических слов был взят AVL (D.Gardner, M.Davies, 2013), состоящий из 1992 академических ЛЕММ, к которому  при помощи Wordnet были добавлены словообразовательные гнезда, что позволило увеличить список \"оставляемых\" слов до 4634 ЛЕ."
      ],
      "metadata": {
        "id": "9bPT5CKLAwh-"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "R55Qc4BxsvkP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1b8b5608-9a82-42d2-c176-64af832911be"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "1992"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "# Достаём академические слова из файла:\n",
        "words = ''\n",
        "with open(\"AVL.txt\", 'r') as f:\n",
        "    for line in f:\n",
        "        words += line\n",
        "words = re.sub(r'\\s', ' ', words)\n",
        "words = set(words.split(' '))\n",
        "len(words)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from nltk.corpus import wordnet as wn\n",
        "\n",
        "# Добавляем к каждому слову списка \"словарные гнезда\", т. е. совокупности слов, обладающих общностью корня(например, indicate - indicating, indication и т.д.)\n",
        "def add_words_derivations(words):\n",
        "    forms = set()\n",
        "    for word in words:\n",
        "        # We'll store the derivational forms in a set to eliminate duplicates\n",
        "        for happy_lemma in wn.lemmas(word): #for each \"happy\" lemma in WordNet\n",
        "            forms.add(happy_lemma.name()) #add the lemma itself\n",
        "            for related_lemma in happy_lemma.derivationally_related_forms(): #for each related lemma\n",
        "                forms.add(related_lemma.name()) #add the related lemma\n",
        "    return forms\n",
        "\n",
        "nltk.download('wordnet')\n",
        "words = add_words_derivations(words)\n",
        "print(len(words))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wbfHpPQKE6jV",
        "outputId": "c09b664b-2755-49d0-c5c3-72a3f399f397"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package wordnet to /root/nltk_data...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "4634\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "D_uYNaaHr4Pv"
      },
      "outputs": [],
      "source": [
        "def remove_biology_terms(text):\n",
        "    text = re.sub(r'\\s', ' ', text)\n",
        "\n",
        "    doc = nlp(text)\n",
        "    doc_no_ents = []\n",
        "    ents = list(set([e.text for e in doc.ents]) - words)\n",
        "\n",
        "    ents_one_word = [ent for ent in ents if len(ent.split(' ')) == 1]\n",
        "    ents_two_words = [ent for ent in ents if len(ent.split(' ')) == 2]\n",
        "    ents_three_words = [ent for ent in ents if len(ent.split(' ')) == 3]\n",
        "    ents_four_words = [ent for ent in ents if len(ent.split(' ')) == 4]\n",
        "\n",
        "    i = 0\n",
        "    while i < len(doc):\n",
        "        #удаление медицинских терминов, состоящих из 1, 2, 3 и 4 слов соответственно:\n",
        "        if  doc[i].text in ents_one_word:\n",
        "            i += 1\n",
        "            continue\n",
        "        if (i + 1 < len(doc)) and (doc[i].text + ' ' + doc[i + 1].text in ents_two_words):\n",
        "            i += 2\n",
        "            continue\n",
        "        if (i + 2 < len(doc)) and (doc[i].text + ' ' + doc[i + 1].text + ' ' + doc[i + 2].text in ents_three_words):\n",
        "            i += 3\n",
        "            continue\n",
        "        if (i + 3 < len(doc)) and (doc[i].text + ' ' + doc[i + 1].text + ' ' + doc[i + 2].text + ' ' + doc[i + 3].text in ents_four_words):\n",
        "            i += 4\n",
        "            continue\n",
        "\n",
        "        #удаление стоп-слов (зависит от задачи):\n",
        "        #if doc[i].text.lower() in stops:\n",
        "           # i += 1\n",
        "           # continue\n",
        "\n",
        "        doc_no_ents.append(doc[i].text)\n",
        "        i += 1\n",
        "    return \" \".join(doc_no_ents)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ddNzBOnfbArz"
      },
      "outputs": [],
      "source": [
        "# Разбиваем тексты на предложения\n",
        "def sent_split(df, col):\n",
        "    sents = []\n",
        "    for _, row in df.iterrows():\n",
        "        text = row[col]\n",
        "        for sent in nltk.sent_tokenize(text):\n",
        "            sents.append(sent)\n",
        "    sents = pd.DataFrame(sents).rename({0: 'text'}, axis=1)\n",
        "    return sents"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "g8534NafcN0A"
      },
      "outputs": [],
      "source": [
        "# NB! Стоит обратить внимание на последовательность действий:\n",
        "def cleantext(df):\n",
        "    # удаляем имена из одной буквы и точки\n",
        "    df['cleaned_text'] = df['text'].replace(r'[A-Z]\\.', \"\", regex=True)\n",
        "\n",
        "    # приводим к нижнему регистру\n",
        "    df['cleaned_text'] = df['cleaned_text'].str.lower()\n",
        "\n",
        "    # Внутритекстовые ссылки в квадратных скобках\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\[[0-9 ,– -]+\\]', \"\", regex=True) # от ссылок в квадратных скобках внутри текста\n",
        "\n",
        "    # Удаление дефиса\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r\"-\\n\", \"\", regex = True)\n",
        "\n",
        "    # удаляем всё, кроме заглавных и строчных букв английского алфавита\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'[^a-zA-Z0-9]', \" \", regex=True)\n",
        "\n",
        "    # Удаление doi | DOI\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'(/doi/10[.][0-9]{4,}.*)', '', regex=True)\n",
        "\n",
        "    # удаляем возможные ссылки URL(x):\n",
        "    df['cleaned_text']  = df['cleaned_text'].replace(r'https.*', \"\", regex = True)\n",
        "\n",
        "    # Удаление цифр\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\d+', '', regex=True)\n",
        "\n",
        "    # Удаление двух и более подряд идущих пробелов\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\s{2,}', ' ', regex=True)\n",
        "    return df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vfKbBKPMMDiH"
      },
      "outputs": [],
      "source": [
        "# Для работы с моделями Bert, которые принимают на вход предложения одной длины (512 токенов по умолчанию)\n",
        "def clean_long_sent(df):\n",
        "    df[\"len\"] = df[\"cleaned_text\"].apply(lambda x: len(x))\n",
        "    return df[df.len <= 512]"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Биологический текст"
      ],
      "metadata": {
        "id": "upXT6IXr3__d"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "bio = pd.read_csv(\"/content/BIOcorpus.csv\")\n",
        "bio"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "WuCNTv0D3_L8",
        "outputId": "e7e91830-b88a-4209-ddca-7ef6aa279a43"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Unnamed: 0                                               text  \\\n",
              "0             0  bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...   \n",
              "1             1  AIMS Bioengineering, 5(1): 1–38. \\n\\nDOI: 10.3...   \n",
              "2             2  Leuba et al. BMC Biophysics 2014, 7:6\\nhttp://...   \n",
              "3             3  Hydrobiologia (2019) 836:35–47\\nhttps://doi.or...   \n",
              "4             4  Woods BMC Biophysics 2014, 7:8\\nhttp://www.bio...   \n",
              "..          ...                                                ...   \n",
              "246         246  J. Biochem. 2019;166(1):97–106 doi:10.1093/jb/...   \n",
              "247         247  Smith et al. BMC Molecular Biol  (2016) 17:22 ...   \n",
              "248         248  Protein Engineering, Design & Selection vol. 2...   \n",
              "249         249  Protein Engineering, Design & Selection vol. 2...   \n",
              "250         250  Volume 2, Issue 3, 249-263. \\nDOI: 10.3934/bio...   \n",
              "\n",
              "                              branch  \n",
              "0                    CELL BIOLOGY 18  \n",
              "1    BIOTECHNOLOGY:BIOENGINEERING 18  \n",
              "2                      BIOPHYSICS 28  \n",
              "3                    HYDROBIOLOGY 11  \n",
              "4                      BIOPHYSICS 29  \n",
              "..                               ...  \n",
              "246                  BIOCHEMISTRY 36  \n",
              "247              MOLECULAR BIOLOGY 2  \n",
              "248  BIOTECHNOLOGY:BIOENGINEERING 23  \n",
              "249  BIOTECHNOLOGY:BIOENGINEERING 35  \n",
              "250   BIOTECHNOLOGY:BIOENGINEERING 9  \n",
              "\n",
              "[251 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7849d93f-ff2b-48dc-8dc9-b72371dcb614\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>text</th>\n",
              "      <th>branch</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...</td>\n",
              "      <td>CELL BIOLOGY 18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>AIMS Bioengineering, 5(1): 1–38. \\n\\nDOI: 10.3...</td>\n",
              "      <td>BIOTECHNOLOGY:BIOENGINEERING 18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Leuba et al. BMC Biophysics 2014, 7:6\\nhttp://...</td>\n",
              "      <td>BIOPHYSICS 28</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Hydrobiologia (2019) 836:35–47\\nhttps://doi.or...</td>\n",
              "      <td>HYDROBIOLOGY 11</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Woods BMC Biophysics 2014, 7:8\\nhttp://www.bio...</td>\n",
              "      <td>BIOPHYSICS 29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>246</th>\n",
              "      <td>246</td>\n",
              "      <td>J. Biochem. 2019;166(1):97–106 doi:10.1093/jb/...</td>\n",
              "      <td>BIOCHEMISTRY 36</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>247</th>\n",
              "      <td>247</td>\n",
              "      <td>Smith et al. BMC Molecular Biol  (2016) 17:22 ...</td>\n",
              "      <td>MOLECULAR BIOLOGY 2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>248</th>\n",
              "      <td>248</td>\n",
              "      <td>Protein Engineering, Design &amp; Selection vol. 2...</td>\n",
              "      <td>BIOTECHNOLOGY:BIOENGINEERING 23</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>249</th>\n",
              "      <td>249</td>\n",
              "      <td>Protein Engineering, Design &amp; Selection vol. 2...</td>\n",
              "      <td>BIOTECHNOLOGY:BIOENGINEERING 35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>250</th>\n",
              "      <td>250</td>\n",
              "      <td>Volume 2, Issue 3, 249-263. \\nDOI: 10.3934/bio...</td>\n",
              "      <td>BIOTECHNOLOGY:BIOENGINEERING 9</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>251 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7849d93f-ff2b-48dc-8dc9-b72371dcb614')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-7849d93f-ff2b-48dc-8dc9-b72371dcb614 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-7849d93f-ff2b-48dc-8dc9-b72371dcb614');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-b2b53408-f140-4811-bd46-424386edf8e3\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b2b53408-f140-4811-bd46-424386edf8e3')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-b2b53408-f140-4811-bd46-424386edf8e3 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bio",
              "summary": "{\n  \"name\": \"bio\",\n  \"rows\": 251,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 72,\n        \"min\": 0,\n        \"max\": 250,\n        \"num_unique_values\": 251,\n        \"samples\": [\n          155,\n          6,\n          164\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 248,\n        \"samples\": [\n          \"Protein Engineering, Design & Selection vol. 25 no. 4 pp. 171\\u2013 178, 2012\\nPublished online February 14, 2012 doi:10.1093/protein/gzs004\\n\\nComparison of random mutagenesis and semi-rational\\ndesigned libraries for improved cytochrome P450\\nBM3-catalyzed hydroxylation of small alkanes\\n\\nMike M.Y.Chen1, Christopher D.Snow2, Christina\\nL.Vizcarra1,3, Stephen L.Mayo1,4 and Frances H.Arnold1,5\\n\\n1Division of Chemistry and Chemical Engineering, California Institute of\\nTechnology, Pasadena, CA 91125, USA, 2Department of Chemical and\\nBiological Engineering, Colorado State University, 1370 Campus Delivery,\\nFort Collins, CO 80523, USA, 3Department of Chemistry and Biochemistry,\\nUniversity of California Los Angeles, Los Angeles, CA 90095, USA and\\n4Division of Biology, California Institute of Technology, Pasadena, CA\\n91125, USA\\n\\n5To whom correspondence should be addressed.\\nE-mail: frances@cheme.caltech.edu\\n\\nReceived January 9, 2012; revised January 9, 2012;\\naccepted January 10, 2012\\n\\nEdited by Alan Berry\\n\\ncombinatorial\\n\\nsemi-rational approaches,\\n\\nThree\\nsite-\\nsaturation mutagenesis (CSSM) using a reduced amino\\nacid set and two libraries based on Corbit and CRAM\\ncomputational design algorithms targeting up to 10\\nactive site residues, were used to engineer cytochrome\\nP450 BM3 to demethylate dimethyl ether and hydroxyl-\\nate propane and ethane. These small\\nlibraries (343 \\u2013\\n1028 variants) were all enriched with respect to the\\nfraction functional and maximal activities compared\\nwith a random mutagenesis library and individual site-\\nsaturation libraries targeting the same residues. Despite\\nhigh average amino acid substitution levels of 2.6, 5 and\\n7.5, the CSSM, Corbit and CRAM libraries had at least\\n75% of library members properly folded. Propane- and\\nethane-hydroxylating P450 BM3 variants were identi\\ufb01ed\\nusing all three mutagenesis approaches, with as few as\\ntwo amino acid substitutions. The library designed using\\nthe CRAM algorithm, which sought to reduce the size\\nof the binding pocket, produced both a higher number\\nof active variants and variants supporting the greatest\\nnumber of catalytic turnovers. The most active variant\\nE32 supports 16 800 propane turnovers at 36% coup-\\nling, which rivals the activity of variants obtained after\\n10 \\u2013 12 rounds of directed evolution using random and\\nsite-saturation mutagenesis. None of the variants in this\\nstudy\\nre-specialization for\\ncomplete\\npropane hydroxylation (including 93% coupling) previ-\\nously obtained via multiple rounds of mutagenesis and\\nscreening. However,\\nsemi-rational approaches\\nallowed for large jumps in sequence space to variants\\nwith the desired functions.\\nKeywords: alkane hydroxylation/C \\u2013 H activation/directed\\nevolution/P450 enzymes/semi-rational library design\\n\\nachieved the\\n\\nthese\\n\\nIntroduction\\n\\na\\n\\nare\\n\\nP450s\\n\\n(P450s)\\n\\nsuperfamily\\n\\nof\\nCytochrome\\nheme-thiolate monooxygenases involved in the biosynthesis\\nof secondary metabolites, degradation of carbon sources and\\ndetoxi\\ufb01cation of xenobiotics (de Montellano, 1986). Using\\nmolecular oxygen and electrons from a nicotinamide adenine\\ndinucleotide (NAD(P)H) cofactor, these enzymes catalyze a\\nvariety of oxidative transformations on a wide array of sub-\\nstrates that range from small hydrocarbons to complex ter-\\npenes (Morant et al., 2003; Isin and Guengerich, 2007; de\\nMontellano, 2010). The breadth of chemical reactivity and\\nsubstrate range of P450s has generated interest in their use as\\nbiocatalysts for industrial synthesis, generation of metabolites\\nfor toxicology studies and bioremediation. CYP102A1 (P450\\nBM3) isolated from Bacillus megaterium is one of the most\\npromising P450s for utilization in biotechnology applications\\nas it is soluble, readily over-expressed in industrial hosts and\\ncatalytically self-suf\\ufb01cient with its hydroxylase and di\\ufb02avin\\nreductase domains fused in a single polypeptide chain\\n(Chefson and Auclair, 2006; Urlacher and Eiben, 2006). Over\\nthe last decade, protein engineering efforts aimed at expand-\\ning P450 BM3\\u2019s substrate range beyond its preferred sub-\\nstrates of C12 \\u2013 C18 fatty acids have produced variants with\\nincreased activity for hydroxylation of small alkanes, regio-\\nselective hydroxylation of indoles, arenes and polycyclic aro-\\nmatic\\nof\\npermethylated monosaccharides, among other activities (Ost\\net al., 2000; Li et al., 2001; Meinhold et al., 2006; Fasan\\net al., 2007; Lewis et al., 2009).\\n\\ndemethylation\\n\\nhydrocarbons,\\n\\nselective\\n\\nand\\n\\nof\\n\\nIn a few instances, engineered variants have been used as\\npractical catalysts for preparative scale synthesis (Lewis\\net al., 2010; Zhang et al., 2011). However, most evolved var-\\niants exhibit low total turnover numbers (TTN) and poor\\ncoupling between cofactor consumption and product forma-\\ntion, typically 5 \\u2013 30%, which hampers their utility as bioca-\\ntalysts. Low coupling ef\\ufb01ciency re\\ufb02ects a poor \\ufb01t between\\nsubstrate and the enzyme active site, which disrupts the\\nusually tightly regulated electron transfer and protonation\\nreactions\\ncycle. Complete\\nre-specialization of P450 BM3 to yield a variant with\\nwild-type-like af\\ufb01nity and coupling for a novel substrate has\\nonly been demonstrated for propane (Fasan et al., 2008). The\\nresulting variant, P450PMO, acquired 22 mutations from 15\\nrounds of directed evolution. The differences in molecular\\nsize and chemical functionality between propane and the\\nP450 BM3-preferred fatty acid substrates certainly contribu-\\nted to the extensive mutagenesis and screening effort\\nrequired to produce P450PMO. However, the lack of compar-\\nable results for other non-natural substrates suggests that al-\\nthough identi\\ufb01cation of P450 BM3 variants with modi\\ufb01ed\\n\\ncatalytic\\n\\nP450\\n\\nthe\\n\\n# The Author 2012. Published by Oxford University Press. All rights reserved.\\nFor Permissions, please e-mail: journals.permissions@oup.com\\n\\n171\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fM.M.Y.Chen et al.\\n\\nsubstrate speci\\ufb01city at modest activity levels can be achieved\\nreadily with a small number of mutations, producing variants\\nthat can be used as practical catalysts may require many\\nrounds of labor-intensive mutagenesis and screening.\\n\\nTo reduce screening effort during directed evolution, a\\nnumber of semi-rational \\u2018library design\\u2019 approaches such as\\nSCOPE (O\\u2019Maille et al., 2002), combinatorial site-saturation\\nmutagenesis (CSSM) (Widersten et al., 2000; Geddie and\\nMatsumura, 2004) and others (Damborsky and Brezovsky,\\n2009; Lassila, 2010) have been described with the aim of gen-\\nerating smaller and/or functionally enriched libraries. These\\napproaches rely on structural and sequence information to\\ndesign libraries with mutations targeted to speci\\ufb01c regions\\nsuch as the enzyme active site. Such methods may be particu-\\nlarly useful for the engineering of P450s, as many of the resi-\\ndues involved in substrate recognition do not participate\\ndirectly in catalysis (Pylypenko and Schlichting, 2004). In this\\nstudy, we sought to determine the extent of re-specialization of\\nfunction that could be obtained through semi-rational muta-\\ngenesis and compare the functional richness of libraries gener-\\nated by these approaches to error-prone polymerase chain\\nreaction (EP-PCR) libraries and site-saturation mutagenesis\\n(SSM) libraries. We applied three semi-rational mutagenesis\\napproaches, CSSM and two structure-based computational\\nlibrary design algorithms, Corbit (Treynor et al., 2007) and\\nCRAM, to engineer P450 BM3 for propane hydroxylation, an\\nactivity not exhibited by the wild-type enzyme. Ef\\ufb01cient\\nP450-catalyzed small alkane hydroxylation could enable\\ngas-to-liquid conversion under mild conditions. Our previous\\neffort to evolve P450 BM3 to accept propane as its preferred\\nsubstrate relied solely on random mutagenesis with EP-PCR\\nand SSM,\\nin conjunction with high-throughput screening\\n(Fasan et al., 2008). That earlier study showed that a step-wise\\npath of accumulating mutations over multiple rounds of evolu-\\ntion existed for re-specialization of P450 BM3 for activity on\\ndimethyl ether (DME), propane and ethane. Here, we deter-\\nmined that a considerable degree of specialization can be\\nobtained through semi-rational library design, circumventing\\nseveral generations of step-wise evolution.\\n\\nMaterials and methods\\nMaterials\\nAll chemicals were purchased from Sigma-Aldrich (St Louis,\\nUSA). Enzymes for DNA manipulation were obtained from\\nNew England Biolabs (Ipswich, USA), Roche (Indianapolis,\\nUSA) and Finnzymes (Woburn, USA). NADP\\u00fe and NADPH\\nwere purchased from Codexis, Inc. (Redwood City, USA)\\nand recombinant\\nisocitrate dehydrogenase was purchased\\nfrom OYC, Inc. (Azusawa Japan). Alcohol oxidase and\\nhorseradish peroxidase were purchased from Sigma-Aldrich.\\nFor cloning and expression of P450 BM3 variants, the bac-\\nterial strain Escherichia coli DH5a was used along with the\\nplasmid pCWori as reported previously (Peters et al., 2003).\\nOligonucleotides were purchased from Invitrogen (Carlsbad,\\nUSA) and Integrated DNA Technology (Coralville, USA).\\nSequencing reactions were carried out by Laragen (Culver\\nCity, USA).\\n\\nExpression and puri\\ufb01cation of P450 BM3 variants\\nEscherichia coli DH5a cells transformed with plasmids con-\\ntaining P450 BM3 variants were cultivated in TB medium\\n\\n172\\n\\nsupplemented with 100 mg/ml of ampicillin. The cultures\\nwere grown at 378C to an optical density of 1 \\u2013 1.5 at 600 nm\\nand cooled to 258C before protein expression was induced by\\nthe addition of IPTG (1 mM) and d-aminolevulinic acid\\nhydrochloride (0.5 mM). Following 16\\u2013 20 h of expression,\\nthe cells were harvested by centrifugation and stored at\\n2208C. An established single-step puri\\ufb01cation was used for\\nall variants (Schwaneberg et al., 1999). Cell pellets were\\nresuspended in 25 mM Tris \\u2013 HCl ( pH 8.0) and lysed by son-\\nication. Cell debris was removed by centrifugation for\\n30 min at 20 000 g, and the crude cell extraction was loaded\\non a pre-equilibrated Toyopearlw Super Q-650 M column for\\nion exchange chromatography. After washing with \\ufb01ve\\ncolumn volumes of both 25 mM Tris \\u2013 HCl ( pH 8.0) and\\n25 mM Tris \\u2013HCl with 0.15 M NaCl, the protein was eluted\\nwith 25 mM Tris \\u2013 HCl ( pH 8.0) with 0.34 M NaCl. The\\npuri\\ufb01ed protein solution was buffer exchanged with potas-\\nsium phosphate buffer (100 mM, pH 8.0) and stored at\\n2808C. The P450 enzyme concentrations were quanti\\ufb01ed by\\nCO-binding difference spectra as described in Omura and\\nSato (1964), using 91/mM/cm as the extinction coef\\ufb01cient.\\n\\nlisted\\n\\nusing\\n\\nprimers\\n\\n(aa1-433)\\n\\nP450 BM3\\n\\nCloning and mutagenesis library construction\\nMutagenesis libraries were constructed of the heme domain\\nin\\nof\\nSupplementary Table SI. All library constructs were rein-\\nserted into pCWori\\nthrough the \\ufb02anking restriction sites\\nBamHI and EcoRI. A random mutagenesis library of the\\nP450 BM3 heme domain was created by EP-PCR (Cadwell\\nand Joyce, 1994) using 150 mM Mn2\\u00fe to induce mutations\\nwith Taq DNA polymerase and primers BamHI-fwd and\\nSacI-rev. An average error rate of 2.1 amino acids/protein\\nwas determined by sequencing 10 randomly selected clones.\\nSSM libraries were constructed at selected residues using\\noverlap extension PCR (Kunkel et al., 1987) with comple-\\nmentary sense and antisense primers bearing the NNK\\ntriplet. This method was similarly used to construct the tar-\\ngeted mutagenesis libraries designed by the CSSM, Corbit\\nand CRAM algorithms with mutagenic primers containing\\ndegenerate codons and mixtures of\\nsuch primers. For\\nexample, a mixture of three primers containing codons: GYN\\n(encodes A and V), WTS (encodes L, I, M and F) and TGG\\n(encodes W) at a ratio of 2 : 4 : 1 was used to achieve the\\ndesired set of allowed amino acid substitutions (L, I, F, V, A,\\nM and W) of the CSSM libraries. Likewise, for the construc-\\ntion of CRAM and Corbit\\nlibraries, primers with codon\\ndegeneracies listed in Supplementary Table SIII were used to\\ngenerate the desired amino acid substitutions.\\n\\nLibrary designs\\nSee supporting information for library design details.\\n\\nHigh-throughput screening\\nSingle colonies grown on LB-agar plates supplemented with\\n100 mg/ml of ampicillin were picked to inoculate 300 ml of\\nLB pre-culture grown in 1-ml 96-well plates. After 16 \\u2013 20 h\\nof incubation at 378C, 250 rpm and 80% relative humidity,\\n50 ml of the pre-culture was used to inoculate expression cul-\\ntures of 650 ml TB grown in 2-ml 96-well plates. After 5 h\\ngrowth at 378C, protein expression was induced with the add-\\nition of IPTG and d-aminolevulinic acid hydrochloride at\\n\\ufb01nal concentrations of 0.5 and 0.25 mM, respectively. The\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fcultures were then cooled to 258C for the duration of protein\\nexpression, 20\\u2013 24 h. The cells were pelleted and stored at\\n2208C until lysis, which was completed by resuspending the\\npellets in 350 ml of 0.1 M phosphate buffer ( pH 8.0) with\\n0.5 mg/ml lysozyme, 2 unit/ml DNaseI and 10 mM MgCl2\\nand incubated at 378C for 60 min. Crude cell extracts\\nobtained after centrifugation were transferred to microtiter\\nplates to be used for all screening assays.\\n\\nand\\n\\n(Otey\\n\\nplates\\n\\nJoern,\\n\\nP450 folding and concentration were determined using a\\nCO-binding difference spectroscopy assay developed for\\n2003).\\n96-well microtiter\\nColorimetric quanti\\ufb01cation of the P450 demethylation of\\nDME was conducted with Purpaldw to detect the resulting\\naldehydes (Peters et al., 2003). In a 96-well microtiter plate,\\n120 ml of 0.1 M, pH 8.0 potassium phosphate buffer pre-\\nsaturated with DME was added to 30 ml of cell lysate, and\\nthe reaction was initiated with the addition of 50 ml of 4 mM\\nNADPH. After 15\\u2013 30 min, 50 ml of 168 mM Purpaldw in\\n2 M NaOH was added to quench the reaction and produce\\nthe purple adduct. The reaction yield was quanti\\ufb01ed by ab-\\nsorbance at 550 nm after an additional 15 min.\\n\\nComparison of random mutagenesis and semi-rational-designed libraries\\n\\nTable I. Targeted residues and allowed amino acids for the semi-rational li-\\nbraries of P450 BM3\\n\\nResidue\\n\\nCSSM\\n\\nCRAM\\n\\nCorbit\\n\\nA74\\nL75\\nV78\\nA82\\nF87\\nL181\\nA184\\nL188\\nA328\\nA330\\n\\n\\u2013\\n\\u2013\\nLIVMFAW\\nLIVMFAW\\n\\u2013\\n\\u2013\\n\\u2013\\n\\u2013\\nLIVMFAW\\n\\u2013\\n\\nLW\\nLF\\nIF\\nVL\\nFA\\nLW\\nAV\\nLW\\nVF\\nLW\\n\\nAV\\nLF\\nVL\\nAS\\nFA\\nLF\\nAT\\nLW\\nAF\\nAV\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\nAlkane hydroxylation reactions\\nP450 ethane and propane hydroxylation reactions were per-\\nformed in a pressurizable 96-well reactor (Accelrys, USA).\\nReactions with 0.5 ml total volume typically contained 20\\u2013\\n40 ml of \\ufb01ltered (2 mm) crude cell extract or puri\\ufb01ed protein\\nat concentrations of 50\\u2013 200 nM, along with an NADPH re-\\ngeneration system consisting of 20 mM DL-isocitric acid tri-\\nsodium salt, 400 mM NADP\\u00fe and 0.5 U/ml\\nisocitrate\\ndehydrogenase in alkane saturated 0.1 M potassium phos-\\nphate buffer. The reactor headspace was pressurized to 30 psi\\nwith alkane and the reactions were stirred at least 20 h at\\n48C. The reactions were quenched with the addition of 15 ml\\nof 5 M HCl and neutralized with 75 ml of 1 M potassium\\nphosphate buffer, pH 8.0. After the removal of precipitated\\ncell debris and denatured protein by centrifugation, 1-penta-\\nnol was added to the reaction mixture as an internal standard\\nand alcohol yield was determined by gas chromatography\\nwith a \\ufb02ame ionization detector.\\n\\nGas chromatography analysis\\nAnalysis for hydroxylation products of P450 reactions with\\nsmall alkanes was performed on a Hewlett Packard 5890\\nSeries II Plus gas chromatograph with a \\ufb02ame ionization de-\\ntector and \\ufb01tted with an HP-7673 autosampler system.\\nAnalysis of ethanol and propanol products was performed on\\na Supelco SPB-1 column (60 m length, 0.32 mm ID,\\n0.25 mm \\ufb01lm thickness). A typical temperature program used\\nfor separating the alcohol products was 2508C injector,\\n3008C detector, 508C oven for 3 min, then 108C/min gradient\\nto 2008C, 258C/min gradient to 2508C and then 2508C for\\n3 min.\\n\\nResults\\n\\nWe constructed mutant\\nlibraries of wild-type P450 BM3\\nheme domain using EP-PCR, SSM, CSSM and two structure-\\nbased computational design algorithms, Corbit and CRAM,\\nwith library compositions listed in Table I. The residues tar-\\ngeted for mutagenesis, A74, L75, V78, A82, F87, L181,\\nA184, L188, A328 and A330, were selected using the crystal\\n\\nFig. 1. The 10 residues targeted for SSM and structure-based computational\\nlibrary design were chosen based on their proximity to the bound\\nN-palmitoyl glycine substrate in the 1JPZ structure of P450 BM3 (Haines\\net al., 2001). Heme is shown in dark gray, N-palmitoyl glycine is shown in\\nlight gray.\\n\\nstructure of\\nthe P450 BM3 heme domain bound with\\nN-palmitoyl glycine, PDB:1JPZ (Haines et al., 2001) (see\\nFig. 1). These 10 residues were included based on their close\\nproximity (,4.5 A\\u02da ) to the bound substrate and were chosen\\nover adjacent candidates due to favorable side chain orienta-\\ntion that point directly into the enzyme active site. Six of\\nthese 10 residues (A74, V78, A82, A184, L188 and A328)\\nwere also mutated in P450PMO (Fasan et al., 2008) and thus\\nwere known sites of bene\\ufb01cial mutations for small alkane hy-\\ndroxylation. Site-saturation libraries were constructed at each\\nof these 10 positions using primers with NNK codon degen-\\neracy to determine the range of activities that could be\\nobtained by single amino acid substitutions.\\n\\nThree different approaches were used to generate mutant\\nlibraries that explore combinations of mutations at these 10\\npositions. The CSSM library focused mutations to three resi-\\ndues, V78, A82 and A328, previously found to be particular-\\nly important for shifting P450 BM3\\u2019s substrate speci\\ufb01city\\ntoward smaller substrates (Peters et al., 2003). Mutating all\\nthree positions with NNK codon degeneracy would yield a\\nlibrary of 323 \\u00bc 32 768 members, which is too large to\\nscreen. To reduce the library to a more manageable size\\n\\n173\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fM.M.Y.Chen et al.\\n\\nTable II. Protein folding and DME demethylation activity of P450 BM3 mutant libraries\\n\\nLibrary properties\\n\\nEP-PCR\\n\\nSite saturation libraries\\n\\nCSSM\\n\\nCRAM\\n\\nCorbit\\n\\nAverage\\n\\nV78\\n\\nA82\\n\\n,MAA.\\nLibrary size (number of unique sequences)\\nFraction of folded variantsb\\nFraction of active variantsc\\nAverage DME activity\\nMaximum DME activity\\n\\n2.1\\n\\n1166a\\n\\n0.52\\n0.36\\n0.06\\n0.52\\n\\n0.9\\n20\\n0.89\\n0.10\\n0.08\\n0.34\\n\\n0.9\\n20\\n0.90\\n0.26\\n0.05\\n0.17\\n\\n0.9\\n20\\n0.96\\n0.21\\n0.09\\n0.21\\n\\nA328\\n\\n0.9\\n20\\n0.62\\n0.36\\n0.11\\n0.34\\n\\n2.6\\n\\n343\\n\\n0.94\\n0.54\\n0.22\\n0.81\\n\\n7.5\\n\\n1024\\n\\n5\\n1024\\n\\n0.75\\n0.34\\n0.33\\n2.6\\n\\n0.84\\n0.31\\n0.19\\n1.9\\n\\n,MAA., average mutation level.\\naEstimate of the number of unique sequences sampled from 1408 clones of the EP-PCR library made using PEDEL (Patrick et al., 2003).\\nbFraction of folded variants determined by CO-binding spectroscopy, corrected for stop codon presence.\\ncFraction of variants active on DME was determined in cell-free extract, corrected for background Purpaldw oxidation.\\n\\n(73 \\u00bc 343 members), the allowed amino acids were restricted\\nto L, I, M, V, F, A and W, following the intuition that intro-\\nducing amino acids with large hydrophobic side chains\\nshould reduce active site volume and promote binding of\\nsmall hydrophobic substrates. To generate libraries with even\\nhigher levels of mutation, two structure-based computational\\ndesign algorithms, Corbit (Hayes et al., 2002) and CRAM,\\nwere used to introduce mutations at all 10 targeted residues\\nallowing for two possible amino acids at each position (see\\nTable I). The Corbit algorithm, which was used previously to\\nalter the green \\ufb02uorescent protein \\ufb02uorescence spectrum\\n(Treynor et al., 2007), models protein stability as a surrogate\\nfor function. Mutations were selected based on the frequency\\nof their appearance in sequences predicted to maintain the\\nP450 BM3 substrate-bound conformation in the absence of\\nsubstrate. The second approach, which we called CRAM,\\nwas developed with the principal objective of reducing the\\nP450 BM3 active site volume by incorporating the largest\\ntolerated amino acid. The amino acid tolerance at each target\\nposition was evaluated as a single amino acid substitution\\nbased on the P450 BM3 substrate-bound conformation\\n(1JPZ) with the substrate removed. The potential for van der\\nWaal clashes introduced by each amino acid substitution was\\nscored using the repulsive van der Waal energy determined\\nby the Rosetta design software after side-chain optimization\\n(Rohl et al., 2004). The mutations were selected to maximize\\nthe side chain size of the amino acid substitution while min-\\nimizing the potential for introducing steric clashes with the\\nprotein backbone and the side chains adjacent\\nto the 10\\ndesigned residues. As the CRAM algorithm evaluates muta-\\ntions based on single amino acid substitutions, it does not\\ntake into account potential van der Waals clashes between\\ndesigned mutations. However, we hypothesized that relax-\\nation of the protein backbone would relieve the energetic\\nstrain of side chain/side chain clashes in the substrate-\\nbinding pocket.\\n\\nLibrary characterization for DME demethylation\\nand protein folding\\nEach mutant library was characterized for DME demethyla-\\ntion and protein folding using high-throughput assays; the\\nresults are summarized in Table II. DME demethylation ac-\\ntivity was quanti\\ufb01ed colorimetrically using Purpaldw dye,\\nwhich reacts with the formaldehyde product of the P450 re-\\naction to form a purple adduct with a UV/Vis peak\\nmaximum at 550 nm (Peters et al., 2003). The folded state of\\n\\n174\\n\\nFig. 2. Pro\\ufb01le of DME demethylation activity for the various libraries.\\nActivities of individual variants are plotted in descending order. All variants\\nfrom the 10 SSM libraries are shown, as are the 1000 most active variants of\\nthe EP-PCR, CSSM, Corbit and CRAM libraries.\\n\\n1166\\n\\nsampling\\n\\ncorresponds\\n\\neach variant was determined from the CO-binding difference\\nspectra of the cell-free extract. All libraries were screened\\nwith at least 90% library coverage, with the exception of the\\nEP-PCR library. The screening of 1408 variants of the\\nEP-PCR library with an error rate of 2.1 amino acid substitu-\\ntions/protein,\\nunique\\nto\\nsequences (Programme for estimating diversity in error-prone\\nPCR libraries (PEDEL); Patrick et al., 2003). A surprisingly\\nlarge fraction of the EP-PCR library variants, 36%, was\\nfound to be active on DME, and 52% were folded. In com-\\nparison, only 8% of variants generated by saturation muta-\\ngenesis at\\nthe 10 individual active site residues were\\nobserved to demethylate DME, even though 89% of these\\nvariants were folded. Variants active on DME were found\\nonly in SSM libraries with mutations at V78, A82, A328 and\\nA330. The high mutational\\ntolerance of these active site\\npositions is also re\\ufb02ected in the high fraction of folded var-\\niants of the CSSM, Corbit and CRAM libraries, which despite\\nhaving average amino acid substitution rates of 2.6, 5 and\\n7.5, contained 91, 84 and 75% folded variants, respectively.\\nThe CSSM library had the largest fraction of variants active\\nfor DME demethylation, 54%, whereas both computationally\\ndesigned libraries had 31\\u2013 34% active variants.\\n\\nAlthough all the mutagenesis strategies generated variants\\nwith DME demethylation activity, the activity levels varied\\n(see Fig. 2). The functional richness of the EP-PCR library\\nand SSM libraries are similar, with average DME activities\\nof 0.06 and 0.08 (A550nm) and maximum DME activities of\\n0.54 and 0.34, respectively. The averages re\\ufb02ect the overall\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fComparison of random mutagenesis and semi-rational-designed libraries\\n\\nlow activity of variants in these libraries. In contrast, the var-\\niants of the CSSM libraries exhibit DME activities up to\\n0.81, with an average activity of 0.22, more than twice those\\nof the EP-PCR and single-site saturation libraries. The Corbit\\nand CRAM libraries contain variants with even higher DME\\nactivity, with maximum activities of 1.9 and 2.6 A550nm.\\nHowever, as the majority of variants in these two libraries\\nhas only mediocre activity, the library averages of 0.19 and\\n0.33 are similar to the CSSM library average.\\n\\nPropane and ethane hydroxylation\\nTop-performing variants were puri\\ufb01ed and characterized for\\npropane and ethane hydroxylation activity (see Table III for\\na representative sample and supporting information for com-\\nplete sequence and activity information). From the 10 SSM\\nlibraries, 12 variants were identi\\ufb01ed that supported propane\\nTTN from 120 to 2200. Mutations at V78 (T, C, S) and A82\\n(E, Q) located in the B\\u2019 helix yielded variants with moderate\\npropane activity, 120\\u2013 370 TTN. More active variants, with\\n.1000 propane TTN, were obtained with mutations at resi-\\ndues A328 (I, P, L, V) and A330 (L, P, V), which are\\nlocated in the loop between the J and K helices. The best\\nsingle active-site variant, A328V, supports 2200 propane\\nTTN with a product formation rate of 7.1/ min and 8.1%\\nformation.\\ncoupling of cofactor consumption to product\\nVariants from the EP-PCR library supported 130\\u2013 3300\\npropane TTN. The best one, WT-F162L, supports 3300\\npropane TTN with a product formation rate of 19/min and\\n15% coupling. The F162L mutation is in the linker between\\nthe E and F helices, outside the active site. Although this\\nresidue was not mutated in P450PMO, several positions in the\\nadjacent F-helix were mutated, which indicates that\\nthis\\nregion is important for substrate speci\\ufb01city.\\n\\nNine variants supporting 380\\u2013 4200 propane TTN were\\nidenti\\ufb01ed from the CSSM library. Four of the nine contained\\nmutations at all three targeted residues, the remaining var-\\niants had mutations at two residues. V78L was the only mu-\\ntation found at position 78, whereas more substitutions were\\nbene\\ufb01cial at A82 (L, W, M and V) and A328 (L, V and F).\\nThe best variant, WT-A82L-A328V, supports 4200 propane\\nTTN with a product formation rate of 40/min and 44% coup-\\nling. In addition, two of the nine propane-active variants,\\n\\nTable III. In vitro small alkane hydroxylation activity of representative\\nP450 BM3 variants\\n\\nVariants Library Mutations\\n\\nPropane\\nTTNa\\n\\nCouplingb\\n(%)\\n\\nEthane\\nTTNa\\n\\nA328V\\n4F9\\n4E10\\nOD2\\n\\nSSM\\nA328V\\nEP-PCR F162L\\nCSSM\\nCorbit\\n\\nA82L, A328V\\nA74V, L181F,\\nA328F\\n\\n2300\\n3300\\n4200\\n11 600\\n\\nE32\\n\\nCRAM A74W, V78I,\\n\\n16 800\\n\\n8.0\\n\\n18\\n44\\n42\\n\\n36\\n\\n0\\n0\\n200\\n660\\n\\n1200\\n\\nA82L, A184V,\\nL188W, A328F,\\nA330W\\n\\naTTN determined as nanomole product/nanomole enzyme. Alkane reactions\\ncontained 25\\u2013250 nM protein, 100 mM potassium phosphate buffer, pH 8.0,\\nsaturated with alkane and an NADPH regeneration system containing\\n100 mM NADP\\u00fe, 2 U/ml isocitrate dehydrogenase and 10 mM isocitrate.\\nStandard errors are within 15% of average.\\nbCoupling is the ratio of product formation rate to NADPH consumption rate\\n(%).\\n\\nWT-A82L-A328L and WT-A82L-A328V, also hydroxylated\\nethane, with 140 and 200 TTN, respectively. The CRAM\\nand Corbit\\nlibraries, which contained variants with much\\nhigher DME demethylation activity in the initial screen, pro-\\nduced 37 variants supporting at least 3500 propane TTN; 16\\nof these hydroxylated ethane with TTN of at least 300. A\\nsigni\\ufb01cantly higher number of active variants for propane\\nand ethane hydroxylation was found in the CRAM library,\\n25 and 13, respectively, compared with the Corbit library,\\nwhich produced only 12 propane-hydroxylating variants and\\nthree ethane-hydroxylating variants. The most active CRAM\\nvariant\\nWT-A74W-V78I-A82L-A184V-L188W-A328F-\\nA330W (E32) supported 16 800 propane TTN and 1200\\nethane TTN. The most active Corbit variant WT-A74V-\\nL181F-A328F (OD2) supported 11 600 propane TTN and 660\\nethane TTN. The coupling of co-factor consumption to pro-\\npanol formation for most of these variants ranged from 36 to\\n52%, with the best variant, E31, exhibiting 68% coupling.\\n\\nThe 25 propane-hydroxylating CRAM library variants\\nform a concise data set for sequence analysis. The distribu-\\ntion of amino acids for all 25 variants, shown in Fig. 3a, dis-\\nplays strong biases at 7 of\\nthe 10 targeted positions.\\nTryptophan appears at residue 74 and 188 in .72% of the\\nsequences. Likewise, strong preferences exist for L at posi-\\ntions 75 (84%), 82 (76%) and 181 (88%), I at 78 (76%) and\\nF at 87 (96%). Of the remaining residues, a weaker prefer-\\nence exists for W at 330 (68%) and V at 184 (60%), and\\nnearly equal representations of F and V were observed at\\nposition 328. In most instances, the representation of the less\\npreferred amino acid decreases with increasing propane hy-\\ndroxylation activity (see Fig. 3c and e). For variants support-\\ning ,7500 propane TTN,\\nthe\\nless-preferred amino acids is found at positions 74, 75, 78,\\nsupporting higher\\n82 and 188. Proceeding to variants\\npropane TTN, the less-preferred amino acids decrease, cul-\\nminating in nearly absolute preference for W at position 74,\\nL at positions 75, 82 and 181, I at position 78 and F at pos-\\nition 87 in variants supporting .10 000 propane TTNs.\\nThese results indicate that the screening process was able to\\n\\ufb01nd a narrow section of the total allowed sequence space\\ncontaining the best solutions for propane hydroxylation.\\n\\nfraction of\\n\\na higher\\n\\nThe amino acid consensus for ethane-hydroxylating var-\\niants of the CRAM library shares many of the same prefer-\\nences (see Fig. 3b). The biggest difference occurs at residue\\n188, where the ethane-hydroxylating variants strongly prefer\\ntryptophan (92%) and the propane-hydroxylating variants\\nhave a weaker preference for tryptophan (72%). Interestingly,\\nthe sequence consensus of the ethane-hydroxylating variants\\nmost strongly resembles that of propane-hydroxylating var-\\niants with intermediate activity,\\nthose supporting 7500 \\u2013\\n10 000 propane TTN, rather than the most active propane-\\nhydroxylating variants. This implies that even for similar\\nsubstrates such as propane and ethane, the optimal active site\\ncon\\ufb01gurations occupy distinct and non-overlapping regions\\nof sequence space. Therefore, regions of sequence space with\\nthe best correlation of activities are occupied by generalist\\nenzymes with intermediate activity for both substrates.\\n\\nDiscussion\\n\\nVariants with DME demethylation and propane hydroxyl-\\nthe mutagenesis\\nation activity were identi\\ufb01ed through all\\n\\n175\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fM.M.Y.Chen et al.\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\nFig. 3. Amino acid distributions at the 10 targeted positions of CRAM library variants that support (a) propane hydroxylation, (b) ethane hydroxylation, (c)\\n,7500 propane TTN, (d) between 7500 and 10 000 propane TTN and (e) .10 000 propane TTN.\\n\\napproaches investigated, with as few as one mutation. The\\nsurprising ease with which these two functions are acquired\\ncontrasts with the results of our early directed evolution\\nstudy (Glieder et al., 2002). One major difference between\\nthis study and the previous effort is this study\\u2019s use of DME\\ndemethylation as a surrogate for activity on propane directly\\non libraries generated from wild-type P450 BM3. In the pre-\\nvious study, P450 BM3 mutant\\nlibraries were initially\\nscreened for activity on para-nitrophenoxy octyl ether, an\\noctane surrogate, in a circuitous route to obtaining propane-\\nhydroxylating variants. The large differences in effort and\\nnumber of accumulated mutations between these two studies\\nis another example for the importance of the screen in direc-\\nted evolution, reaf\\ufb01rming the \\ufb01rst rule of directed evolution:\\n\\u2018You get what you screen for\\u2019 (You and Arnold, 1996).\\n\\nAs a substrate, propane shares many similarities with\\nhexane, the smallest alkane known to be hydroxylated by\\nwild-type P450 BM3. They are both hydrophobic with poor\\nwater solubility and possess sub-terminal alkane C\\u2013 H bonds\\nof bond strength comparable with P450 BM3\\u2019s preferred\\nfatty acid substrates (99 \\u2013 100 kcal/mol). The only difference\\nbetween propane and known P450 BM3 substrates is its\\nsmaller molecular size, which would result\\nin a lower\\nbinding af\\ufb01nity. Poorly bound substrates cannot displace the\\ndistal water ligand to initiate catalysis (Denisov et al., 2005),\\nwhich results in weak activation of the P450 catalytic cycle.\\nFor wild-type P450 BM3, propane binding would be the sole\\ntrigger for catalysis, as the enzyme exhibits very low (2/min)\\nresting state oxidase activity (Kadkhodayan et al., 1995), in-\\ndicating that substrate-independent activation of the catalytic\\n\\n176\\n\\ncycle occurs rarely. However, P450 BM3 variants can exhibit\\nmuch higher resting state oxidase activity, thereby reducing\\nthe requirement of propane binding to induce catalysis. In\\nfact, variants generated in the P450PMO lineage and many of\\nthe variants found in this study have substrate-free cofactor\\nconsumption rates that are orders of magnitude higher than\\nthe wild-type enzyme. The existence of\\nthis alternative\\npathway for propane hydroxylation, which can be achieved\\nwithout appreciable propane-induced activation of the cata-\\nlytic cycle (Fasan et al., 2008), could account for the large\\nthat were\\nnumber\\nidenti\\ufb01ed.\\n\\npropane-hydroxylating\\n\\nvariants\\n\\nof\\n\\nThese targeted mutagenesis libraries also revealed the high\\nmutational tolerance of the P450 BM3 active site, which\\ngoes against the general observation that mutations in the\\ncore of a protein are on average more destabilizing than\\nmutations of solvent-exposed residues (Reidhaarolson and\\nSauer, 1988). Residues in the P450 BM3 active site may\\nhave an atypical degree of \\ufb02exibility compared with typical\\ncore residues, as the active site of P450 BM3 undergoes sig-\\nni\\ufb01cant motion between the \\u2018closed\\u2019 substrate-bound state\\nand its \\u2018open\\u2019 resting state (Arnold and Ornstein, 1997).\\nLocal backbone \\ufb02exibility could allow mutations with\\ngreater probability than highly constrained position buried\\nwithin a rigid protein. The active site environment between\\nthese two states also differs signi\\ufb01cantly in terms of solvent\\naccessibility (Schlichting et al., 2000), which may allow\\nthese residues to tolerate polar or even charged amino acid\\nsubstitutions. Finally, many of the P450 BM3 active site resi-\\ndues that participate in substrate recognition are not directly\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fComparison of random mutagenesis and semi-rational-designed libraries\\n\\ninvolved in enzyme catalysis, which also contributes to the\\ntolerance of these residues for substitution. The malleable\\nnature of the active site in combination with the capacity to\\ndecouple substrate binding from catalysis could account for\\nthe ease with which P450 BM3 variants are able to acquire\\nactivity for non-native substrates.\\n\\nIn comparing the functional richness between the libraries\\ngenerated by the various mutagenesis methods investigated,\\nit was surprising to \\ufb01nd that the EP-PCR library appears to\\ngenerate more variants that catalyze DME demethylation\\nthan the combined efforts of the SSM libraries (see Fig. 2).\\nThe functional richness of a random mutagenesis library has\\nbeen generally characterized as poor, as the majority of var-\\niants with point mutations are expected to exhibit a parent-\\nlike phenotype. Although the number of clones sampled\\nfrom these libraries is similar (1408 for EP-PCR vs. 910 for\\nSSM), the actual genotypic diversity is quite different: 200\\nunique variants for the SSM libraries vs. 1166 expected\\nunique variants for the EP-PCR library. This six-fold differ-\\nence in the number of sequences sampled could account for\\nthe higher number of DME-demethylating variants identi\\ufb01ed\\nin the EP-PCR library. Of the 10 positions that were sub-\\njected to SSM, only four of the positions, V78, A82, A328\\nand A330, yielded variants that were active on DME. A dif-\\nferent selection of SSM targets could yield more active var-\\niants, but the opposite result is also possible. The ease with\\nwhich EP-PCR can generate sequence diversity differentiates\\nthese two approaches. For activities that can be altered by\\nmutations dispersed across the protein sequence, EP-PCR\\nwill be advantageous compared with SSM.\\n\\nprotein\\n\\nThe CSSM, Corbit and CRAM libraries with higher\\naverage mutation levels generated variants with much higher\\nlevels of DME demethylation activity compared with var-\\niants from the SSM and EP-PCR libraries. This is unsurpris-\\ning in that a variant with more bene\\ufb01cial mutations is likely\\nto have greater activity. However, because most mutations\\nare neutral or deleterious,\\nthe fraction of total sequence\\nspace with functional sequences decreases precipitously as a\\nfunction of mutational distance. As a consequence, random\\nmutagenesis at a high mutation rate generally results in li-\\nbraries with lower functional richness due to the exponential\\ndecrease\\nfolding/function with mutation\\n(Drummond et al., 2005). Designed libraries avoid this\\nproblem by targeting mutations to residues that are highly\\ntolerant to mutation and by favoring amino acid substitutions\\nthat are compatible with the target substrates. Restricting\\namino acid substitutions to hydrophobic residues with larger\\nside chains enriched these libraries for functional variants\\ncompared with the SSM libraries that replaced the targeted\\nresidues with all 19 amino acids. Although the CSSM library\\ngenerated the highest fraction of variants active on DME, the\\nmaximum activities lagged behind CRAM and Corbit librar-\\nies. This is likely the result of the limited volume reduction\\nthat was accessible to the CSSM library, which only allowed\\nmutations at three residues compared with the 10 that could\\nlibraries. The Corbit\\nbe mutated in the CRAM and Corbit\\ndesign was able to generate a library with a higher fraction\\nof folded variants (84%) compared with the CRAM library\\n(75%). However, it is dif\\ufb01cult to assess the relative success\\nof the Corbit design at preserving the folded structure as the\\nmutation level of the Corbit library, 5 mutations/protein, is\\n2.5 mutations/protein lower than that of the CRAM library.\\n\\nof\\n\\nAlthough the Corbit library contained more folded variants\\nthan the CRAM library, this advantage was not re\\ufb02ected in\\nthe fraction functional for DME demethylation, which was\\nsimilar (31 vs. 34%). The high neutrality of the targeted resi-\\ndues devalues the need to preserve folding, which was the\\ncore criterion of the Corbit design.\\n\\nThe single mutations at V78, A82, A328 and A330 that\\nresulted in variants with DME demethylation activity gener-\\nally introduced amino acids with bulkier side chains into the\\nactive site, which follows the intuition that reducing active\\nsite volume would promote activity on a smaller substrate.\\nBecause none of the amino acids found in P450PMO (F78,\\nG82 and F328), were found at these positions, one can con-\\nclude that they are not bene\\ufb01cial individually. The presence\\nof proline mutations at A328 and A330 suggests that altering\\nthe orientation of the loop containing these residues can gen-\\nerate DME activity. Of the seven mutations found using\\nrandom mutagenesis, only I260V, located on the I-helix, is\\nnear the active site. Mutations F162L and I153V found in\\nthe more active variants are in the region between the E and\\nF helices. The remaining mutations, E4D, T235M, D232V\\nand Q359R, are at surface-exposed residues, which is typical\\nof mutations found using random mutagenesis (Arnold,\\n1998), whose effects are dif\\ufb01cult to rationalize. Of the muta-\\ntions identi\\ufb01ed from the CSSM variants, V78L, A82 (L, W,\\nM and V) and A328 (L, V and F), only A328F was previous-\\nly found with P450PMO. As this mutation was not found to\\nbe bene\\ufb01cial individually, its presence in P450PMO and these\\nreduced CSSM variants suggests there are synergistic effects\\nbetween this residue and neighboring amino acids. The\\neffect of the V78L mutation on propane TTN also showed a\\ndependence on the surrounding mutations: when V78L was\\nintroduced to WT-A328L or WT-A82L-A328L, the resulting\\nthe parental propane activity (see\\nvariants lost 48% of\\nSupplementary Table SII). However, when V78L was\\nmutated in WT-A82W-A328F,\\nthe resulting variant had\\n6-fold improved propane activity. This illustrates the rugged-\\nness of active site \\ufb01tness landscape, in which the effects of\\nmutations are highly dependent on the identity of neighbor-\\ning amino acids.\\n\\nfrom the Corbit\\n\\nThe most effective libraries for DME demethylation and\\npropane hydroxylation were the CRAM and Corbit structure-\\nbased computationally designed libraries. These two libraries\\nglobally redesigned the P450 BM3 active site by mutating 10\\nresidues to two possible amino acids. From these libraries,\\nwe obtained variants with propane and ethane activity rival-\\ning those achieved by variants of the P450PMO lineage. The\\nbest variant from the CRAM library (E32) supported 16 800\\npropane TTN and 1200 ethane TTN, which represents nearly\\n50% of P450PMO\\u2019s activity on these substrates. The best\\nlibrary (OD2) supported 11 600\\nvariant\\npropane TTN and 660 ethane TTN, or 34 and 27% of\\nP450PMO\\u2019s activity on these substrates. As a comparison, the\\nP450PMO activity levels were obtained after 10\\u2013 12 rounds of\\nmutagenesis and screening. Of the 37 variants isolated from\\nthese two designed libraries, all allowed amino acids were\\nfound at least once. The consensus sequence of propane-\\nhydroxylating CRAM variants, WT-L74W-V78I-A82L-\\nA184V-L188W-A328F-A330W, is actually the sequence of\\nthe most active variant E32, which suggests that the screen-\\ning process was able to \\ufb01nd an optimal solution within the\\nassigning\\nallowed\\n\\n(210). However,\\n\\nsequence\\n\\nspace\\n\\n177\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\fChefson,A. and Auclair,K. (2006) Mol. Biosyst., 2, 462\\u2013469.\\nDamborsky,J. and Brezovsky,J. (2009) Curr. Opin. Chem. Biol., 13, 26\\u2013 34.\\nde Montellano,P.R.O. (1986) Cytochrome P450, 1st edn, Plenum Publishing\\n\\nCorp., New York.\\n\\nde Montellano,P.R.O. (2010) Chem. Rev., 110, 932\\u2013 948.\\nDenisov,I.G., Makris,T.M., Sligar,S.G. and Schlichting,I. (2005) Chem. Rev.,\\n\\n105, 2253\\u20132277.\\n\\nDrummond,D.A., Iverson,B.L., Georgiou,G. and Arnold,F.H. (2005) J. Mol.\\n\\nBiol., 350, 806\\u2013 816.\\n\\nFasan,R., Chen,M.M., Crook,N.C. and Arnold,F.H. (2007) Angew. Chem.\\n\\nInt. Ed., 46, 8414\\u2013 8418.\\n\\nFasan,R., Meharenna,Y.T., Snow,C.D., Poulos,T.L. and Arnold,F.H. (2008)\\n\\nJ. Mol. Biol., 383, 1069\\u20131080.\\n\\nGeddie,M. and Matsumura,I. (2004) J. Mol. Biol., 279, 26462\\u2013 26468.\\nGlieder,A., Farinas,E.T. and Arnold,F.H.\\n\\n(2002) Nat. Biotechnol., 20,\\n\\n1135\\u2013 1139.\\n\\nHaines,D.C., Tomchick,D.R., Machius,M.\\n\\nand\\n\\nPeterson,J.A.\\n\\n(2001)\\n\\nBiochemistry, 40, 13456\\u201313465.\\n\\nHayes,R.J., Bentzien,J., Ary,M.L., Hwang,M.Y., Jacinto,J.M., Vielmetter,J.,\\n(2002) Proc. Natl Acad. Sci. USA, 99,\\n\\nKundu,A. and Dahiyat,B.I.\\n15926\\u201315931.\\n\\nIsin,E.M. and Guengerich,F.P. (2007) Biochim. Biophys. Acta-Gen Subj.,\\n\\n1770, 314\\u2013329.\\n\\nKadkhodayan,S.,\\n\\nCoulter,E.D., Maryniak,D.M.,\\n\\nBryson,T.A.\\n\\nand\\n\\nDawson,J.H. (1995) J. Biol. Chem., 270, 28042\\u201328048.\\n\\nKunkel,T.A., Roberts,J.D. and Zakour,R.A. (1987) Methods Enzymol., 154,\\n\\n367\\u2013382.\\n\\nLassila,J.K. (2010) Curr. Opin. Chem. Biol., 14, 676\\u2013682.\\nLewis,J.C., Bastian,S., Bennett,C.S., Fu,Y., Mitsuda,Y., Chen,M.M.,\\nGreenberg,W.A., Wong,C.H. and Arnold,F.H. (2009) Proc. Natl Acad.\\nSci. USA, 106, 16550\\u201316555.\\n\\nLewis,J.C., Mantovani,S.M., Fu,Y., Snow,C.D., Komor,R.S., Wong,C.H. and\\n\\nArnold,F.H. (2010) Chembiochem, 11, 2502\\u2013 2505.\\n\\nLi,Q.S., Ogawa,J., Schmid,R.D. and Shimizu,S.\\n\\n(2001) Appl. Environ.\\n\\nMicrobiol., 67, 5735\\u2013 5739.\\n\\nMeinhold,P., Peters,M.W., Hartwick,A., Hernandez,A.R. and Arnold,F.H.\\n\\n(2006) Adv. Synth. Catal., 348, 763\\u2013772.\\n\\nMorant,M., Bak,S., Moller,B.L. and Werck-Reichhart,D. (2003) Curr. Opin.\\n\\nBiotechnol., 14, 151\\u2013 162.\\n\\nO\\u2019Maille,P.E., Bakhtina,M. and Tsai,M.D.\\n\\n(2002) J. Mol. Biol., 321,\\n\\n677\\u2013691.\\n\\nOmura,T. and Sato,R. (1964) J. Biol. Chem., 239, 2370\\u2013 2374.\\nOst,T.W.B., Miles,C.S., Murdoch,J., Cheung,Y.F., Reid,G.A., Chapman,S.K.\\n\\nand Munro,A.W. (2000) FEBS Lett., 486, 173\\u2013177.\\n\\nOtey,C. and Joern,J.M.\\n\\n(2003)\\n\\nIn Arnold,F.H. and Georgiou,G.\\n\\n(eds),\\n\\nMethods in Molecular Biology. Humana Press, Totowa, pp. 141\\u2013148.\\n\\nPatrick,W.M., Firth,A.E. and Blackburn,J.M.\\n\\n(2003) Protein Eng., 16,\\n\\n451\\u2013457.\\n\\nPeters,M.W., Meinhold,P., Glieder,A. and Arnold,F.H. (2003) J. Am. Chem.\\n\\nSoc., 125, 13442\\u201313450.\\n\\nPylypenko,O. and Schlichting,I. (2004) Annu. Rev. Biochem., 73, 991\\u20131018.\\nReidhaarolson,J.F. and Sauer,R.T. (1988) Science, 241, 53\\u2013 57.\\nRohl,C.A., Strauss,C.E.M., Misura,K.M.S. and Baker,D. (2004) Numerical\\n\\nComputer Methods, Partt D, Academic Press Inc, San Diego, p. 66.\\n\\nSchlichting,I., Berendzen,J., Chu,K., et al. (2000) Science, 287, 1615\\u2013 1622.\\nSchwaneberg,U., Sprauer,A., Schmidt-Dannert,C. and Schmid,R.D. (1999)\\n\\nJ. Chromatogr. A, 848, 149\\u2013 159.\\n\\nTreynor,T.P., Vizcarra,C.L., Nedelcu,D. and Mayo,S.L. (2007) Proc. Natl\\n\\nAcad. Sci. USA, 104, 48\\u201353.\\n\\nUrlacher,V.B. and Eiben,S. (2006) Trends Biotechnol., 24, 324\\u2013 330.\\nWidersten,M., Hansson,L.O., Tronstad,L. and Mannervik,B. (2000) Methods\\n\\nEnzymol., 328, 389\\u2013 404.\\n\\nYou,L. and Arnold,F.H. (1996) Protein Eng., 9, 77\\u2013 83.\\nZhang,K.D., El Damaty,S. and Fasan,R. (2011) J. Am. Chem. Soc., 133,\\n\\n3242\\u2013 3245.\\n\\nM.M.Y.Chen et al.\\n\\nsigni\\ufb01cance to these amino acid preferences in the context of\\nthe total possible sequence space of these 10 residues (1020)\\nis problematic, as each mutation was compared against only\\none other amino acid within a limited set of surrounding\\nmutations. Of the six positions where the P450PMO residue\\nwas an allowed choice by the library design (75, 78, 87, 181,\\n184 and 328), \\ufb01ve positions converged on that amino acid.\\nThis convergence on P450PMO mutations is not surprising as\\nthe P450PMO active site is a sequence solution for the\\nscreened activities, and it may suggest\\nthe P450PMO\\nactive site sequence is a very accessible solution from wild-\\ntype P450 BM3.\\n\\nthat\\n\\nit\\n\\nThese libraries demonstrate that\\n\\nis possible to jump\\nfrom P450 BM3 to variants with moderate propane hydroxyl-\\nation activity ((cid:2)10 000 TTN). None of\\nthe variants,\\nhowever, reached the level of specialization that was previ-\\nously obtained with P450PMO,\\nin either propane TTN or\\ncoupling of cofactor consumption. In the evolution of P450\\nBM3 to P450PMO, the specialization for propane hydroxyl-\\nation did not occur evenly through the 16 rounds of mutagen-\\nesis. In fact, the variants of the lineage can be categorized\\ninto three groups by their substrate speci\\ufb01city for linear\\nalkanes as (i) preferring longer chain alkanes, (ii) having\\nequal preference for chain lengths C3 \\u2013 C10 and (iii) prefer-\\nring shorter chains the length of propane (Fasan et al., 2008).\\nThese three groups of variants represent a transition from a\\nspecialized fatty acid hydroxylase to generalist P450s with\\nbroad alkane substrate acceptance followed by a second tran-\\nsition to a specialized propane monooxygenase. This last\\ntransition occurs in the \\ufb01nal four rounds of mutagenesis\\nwhere the largest improvements in cofactor coupling (44 \\u2013\\n93%) and propane TTN (10 550\\u2013 33 400) occurred. In add-\\nition, mutations acquired in these \\ufb01nal rounds of mutagenesis\\nappeared in the reductase domain as well as in the heme\\ndomain. This suggests that mutations outside the active site\\nor even the heme domain may be necessary for functional\\noptimization. The ranges of propane TTN (3500\\u2013 16 800)\\nand coupling of cofactor consumption (36 \\u2013 68%) for variants\\nin the designed libraries are those of the generalist intermedi-\\nates found preceding the propane specialization phase of the\\nP450PMO evolution. This suggests that semi-rational library\\ndesign can be an effective strategy to move toward generalist\\nenzymes, but that functional specialization still requires opti-\\nmization through several rounds of random mutagenesis and\\nscreening.\\n\\nSupplementary data\\n\\nSupplementary data are available at PEDS online.\\n\\nAcknowledgements\\n\\nThis work was supported by the Chemical Sciences, Geosciences and\\nBiosciences Division, Of\\ufb01ce of Basic Science, US Department of Energy\\nGrant DE-FG02-06ER15762 and the DARPA Protein Design Processes\\nGrant. C.D.S. was supported by a Jane Cof\\ufb01n Childs postdoctoral fellowship.\\nassistance with gas\\nThe\\nchromatography.\\n\\nthank Dr Nathan Dalleska\\n\\nauthors\\n\\nfor\\n\\nReferences\\nArnold,F.H. (1998) Acc. Chem. Res., 31, 125\\u2013 131.\\nArnold,G.E. and Ornstein,R.L. (1997) Biophys. J., 73, 1147\\u2013 1159.\\nCadwell,R.C. and Joyce,G.F. (1994) PCR-Methods Appl., 3, S136\\u2013S140.\\n\\n178\\n\\nl\\n\\nD\\no\\nw\\nn\\no\\na\\nd\\ne\\nd\\n\\nf\\nr\\no\\nm\\nh\\n\\nt\\nt\\n\\np\\ns\\n:\\n/\\n/\\n\\ni\\n\\n/\\n\\na\\nc\\na\\nd\\ne\\nm\\nc\\n.\\no\\nu\\np\\n.\\nc\\no\\nm\\np\\ne\\nd\\ns\\n/\\na\\nr\\nt\\ni\\nc\\ne\\n-\\na\\nb\\ns\\nt\\nr\\na\\nc\\nt\\n/\\n\\nl\\n\\n/\\n\\n/\\n\\n/\\n\\n2\\n5\\n4\\n1\\n7\\n1\\n1\\n5\\n1\\n4\\n5\\n6\\n4\\nb\\ny\\ng\\nu\\ne\\ns\\nt\\n\\no\\nn\\n2\\n4\\nA\\nu\\ng\\nu\\ns\\nt\\n\\n2\\n0\\n1\\n9\\n\\n \\n \\n \\n \\n \\n \\n \\n \\n\\f\",\n          \"Hydrobiologia (2012) 683:123\\u2013133\\nDOI 10.1007/s10750-011-0946-3\\n\\nP R I M A R Y R E S E A R C H P A P E R\\n\\nCan elevated CO2 modify regeneration from seed banks\\nof \\ufb02oating freshwater marshes subjected to rising sea-level?\\n\\nBeth A. Middleton \\u2022 Karen L. McKee\\n\\nReceived: 20 June 2011 / Revised: 26 October 2011 / Accepted: 31 October 2011 / Published online: 22 November 2011\\n(cid:2) The Author(s) 2012. This article is published with open access at Springerlink.com\\n\\nAbstract Higher atmospheric concentrations of CO2\\ncan offset the negative effects of \\ufb02ooding or salinity on\\nplant species, but previous studies have focused on\\nmature, rather than regenerating vegetation. This study\\nexamined how interacting environments of CO2, water\\nregime, and salinity affect seed germination and\\nseedling biomass of \\ufb02oating freshwater marshes in\\nthe Mississippi River Delta, which are dominated by\\nC3 grasses, sedges, and forbs. Germination density and\\nseedling growth of the dominant species depended on\\nmultifactor interactions of CO2 (385 and 720 ll l-1)\\nwith \\ufb02ooding (drained, ?8-cm depth, ?8-cm depth-\\ngradual) and salinity (0, 6% seawater) levels. Of the\\nthree factors tested, salinity was the most important\\ndeterminant of seedling response patterns. Species\\nrichness (total = 19) was insensitive to CO2. Our\\n\\ufb01ndings suggest that for freshwater marsh communi-\\nties, seedling response to CO2 is species-speci\\ufb01c and\\nsecondary to salinity and \\ufb02ooding effects. Elevated\\nCO2 did not ameliorate \\ufb02ooding or salinity stress.\\nConsequently, climate-related changes in sea level or\\nhuman-caused alterations in hydrology may override\\n\\nHandling editor: Sidinei Magela Thomaz\\n\\nB. A. Middleton (&) (cid:2) K. L. McKee\\nU.S. Geological Survey, National Wetlands Research\\nCenter, 700 Cajundome Boulevard, Lafayette, LA 70506,\\nUSA\\ne-mail: middletonb@usgs.gov\\n\\natmospheric CO2 concentrations in driving shifts in\\nthis plant community. The results of this study suggest\\ncaution in making extrapolations from species-speci\\ufb01c\\nresponses to community-level predictions without\\ndetailed attention to the nuances of multifactor\\nresponses.\\n\\nKeywords Climate change (cid:2) Coastal land loss (cid:2) CO2\\nenhancement (cid:2) Floating marsh (cid:2) Gulf of Mexico (cid:2)\\nSalinity intrusion (cid:2) Sea level rise (cid:2) Seed bank\\n\\nIntroduction\\n\\nThe regeneration of plant species during climate\\nchange may hold the key to the future species patterns\\nin ecosystems (Woodward et al., 1991). In coastal\\necosystems, regeneration may be dictated by distur-\\nbances related to hurricanes, water \\ufb02uctuation, and/or\\nsaltwater intrusion (Platt & Connell, 2003; Middleton,\\n2009a, b, c, d). These disturbances may increase in the\\nfuture with climate change, particularly because of\\nincreased hurricane activity, sea-level rise (Michener\\net al., 1997), or declining freshwater input because of\\ninland freshwater diversion or abstraction along rivers\\n(Middleton, 1999). Experimental studies of species in\\nfreshwater wetlands indicate that water depth and\\nsalinity changes can greatly alter seed germination\\nand seedling growth of species (Middleton, 1999,\\n2009a, d; Howard & Mendelssohn, 2000; Brock et al.,\\n2005).\\n\\n123\\n\\n\\f124\\n\\nHydrobiologia (2012) 683:123\\u2013133\\n\\nEspecially for freshwater marsh species, the out-\\ncomes of post-disturbance regeneration may depend\\non species-speci\\ufb01c responses of seeds and seedlings to\\nfuture conditions of rainfall, sea level, and associated\\nfactors. While \\u2018\\u2018regeneration\\u2019\\u2019 refers to the reestab-\\nlishment of vegetation after disturbance either by seed\\nor vegetative means (Middleton, 1999), in this study,\\nonly materials originating from seed banks were\\nconsidered. Knowledge of the potential role of seed\\nbanks in supplying regenerative material to wetlands\\nafter disturbance is important; after large-scale distur-\\nbance, seed banks can be an important source of\\nregenerative material (Leck, 1989; Middleton et al.,\\n1991), supplying as much as 80% (Jarvis & Moore,\\n2010). Nevertheless, detailed information regarding\\nspecies responses to important abiotic and biotic factors\\nis mostly lacking for wetland species (Middleton, 1999).\\nRegeneration studies undertaken in controlled environ-\\nments could improve our understanding of the potential\\nmaintenance of species in future climate settings, and\\ninform modeling efforts.\\n\\nFuture environments for regeneration in freshwater\\nmay involve novel combinations of increased atmo-\\nspheric CO2 (IPCC, 2007), temperature, sea level,\\nsalinity (Michener et al., 1997), precipitation (Easterling\\net al., 2000), and hurricane activity (Webster et al.,\\n2005). Careful assessment of any interactive effects of\\nthese environmental factors is important in projecting\\nthe future responses of freshwater wetland species to\\nclimate change. Of these factors, CO2 has received\\nmuch emphasis in ecosystem modeling (Mayle et al.,\\n2004; Hickler et al., 2009) and empirical \\ufb01eld studies in\\nwetlands (Rasse et al., 2005), but only rarely studied in\\ntandem with multifactor interactions (McKee & Rooth,\\n2008; Cherry et al., 2009; Langley et al., 2009).\\n\\nWhile single-factor studies of CO2 effects may miss\\ncritical details of\\nthe potential multiple-factor\\nresponses of freshwater wetland species to climate\\nchange, considerations of the \\ufb01ndings of these studies\\nare important. Single-factor studies of CO2 often report\\nthat the carbon assimilation strategies of plant species\\ndominate outcomes, with C3 species more responsive\\nthan C4 species to elevated atmospheric CO2 (Osmond\\net al., 1980; Bazzaz, 1990). For example, C3 brackish\\nmarsh species such as Schoenoplectus americanus\\nincrease in biomass when subjected to elevated CO2\\nlevels (Curtis et al., 1989, 1990; Ziska et al., 1990;\\nRasse et al., 2003). While most studies \\ufb01nd that\\nterrestrial C3 species grow faster under elevated CO2,\\n\\n123\\n\\nthe growth response of C3 seedlings to elevated CO2\\ncan be negative (Garbutt et al., 1990), or positive for C4\\nspecies (Wand et al., 1999).\\n\\nThere is a logical expectation that elevated CO2\\ncould increase the growth of C3 species under stressful\\nconditions, including anaerobic or saline environments\\n(i.e., an interactive effect). For example, elevated CO2\\noffsets the effects of salinity on the biomass of\\nS. americanus in brackish marshes (Cherry et al.,\\n2009). C3 species of salt marshes generally had higher\\nbiomass in elevated CO2, but the response varied for\\nspecies depending on water regime and/or salinity\\nlevels (Cherry et al., 2009). In other studies, C4 species\\ndid not increase growth in enhanced CO2 in any water\\nregime or salinity level (Rozema et al., 1991; Lenssen\\net al., 1993). In elevated CO2,\\nthe seedlings of\\nTaxodium distichum (C3), a freshwater swamp species,\\ndid not have a higher biomass in \\ufb02ooded conditions,\\nbut seedlings of Orontium aquaticum (C3) had higher\\nbiomass when exposed to elevated CO2 in both \\ufb02ooded\\nand un\\ufb02ooded conditions (Megonigal et al., 2005).\\n\\nAs a further complication in understanding the\\neffects of CO2 on the regeneration of wetland species,\\nsingle-species experiments could miss\\nimportant\\nbiotic interactions. In one experiment, growth of the\\nmangrove Avicennia germinans (C3) was higher in\\nelevated CO2 when grown in monoculture, but when\\ngrown in mixed cultures with Spartina alterni\\ufb02ora\\n(C4), the growth was suppressed (McKee & Rooth,\\n2008). While both C3 and C4 species may have\\nenhanced growth in elevated CO2,\\nthe speci\\ufb01c\\nresponse of species may depend on a large array of\\nbiotic and abiotic factors. Therefore, freshwater wet-\\nland species subjected to multiple interacting factors\\nmay or may not bene\\ufb01t\\nfrom CO2 enrichment.\\n\\u2018\\u2018Bene\\ufb01t\\u2019\\u2019 is de\\ufb01ned for the purposes of this study as\\nan increase in seed germination density or seedling\\nbiomass. Many factors other than CO2 may affect the\\noutcome of\\nregeneration in freshwater wetlands\\nundergoing climate change. Nevertheless, basic infor-\\nmation for species related to either single-factor or\\ninteracting multifactor responses to CO2, water regime\\nand salinity is limited (Middleton, 1999).\\n\\nFew multifactorial studies have been conducted to\\nexplore the responses of wetland species to interac-\\ntions of CO2, water regime and salinity (e.g., Cherry\\net al., 2009; Langley & Megonigal, 2010), even\\nthough enhanced CO2 might have the potential to\\noffset the negative impact of \\ufb02ooding and salinity on\\n\\n\\fHydrobiologia (2012) 683:123\\u2013133\\n\\n125\\n\\nseed germination and seedling growth in wetland\\nspecies. Controlled greenhouse, laboratory, and \\ufb01eld\\nstudies could help us better understand species\\u2019\\nresponses to environmental changes related to climate\\nchange factors, to aid modeling, and/or assist efforts to\\ncategorize functional types. The species of the \\ufb02oating\\nmarshes of Lake Salvador in coastal Louisiana are\\nexcellent candidates for studies of potential multifac-\\ntorial interactions in wetland environments. These\\nwetlands constitute some of the southernmost fresh-\\nwater wetlands in central North America (Swarzenski\\net al., 1991), and are thus likely to be directly impacted\\nin the future by shifts in levels of CO2, water level, and\\nsalinity along the Gulf Coast. These wetlands lie\\nwithin the Barataria Basin of the Mississippi River\\nDeltaic Plain in the most recently abandoned channel\\nof the Mississippi River (Swarzenski et al., 1991).\\n\\nThe overall goal of this study was to examine\\nregeneration from the seed bank of these freshwater\\n\\ufb02oating marshes in response to various combinations\\nof CO2, water regime, and salinity. Our speci\\ufb01c\\nobjectives were to determine if (1) elevated CO2 might\\nameliorate stress effects of sea-level rise (increased\\n\\ufb02ooding and salinity) on species richness, density, or\\nbiomass of seedlings emerging from the seed bank;\\nand (2) species responses to CO2 and other factors\\nwere generally similar or individualistic.\\n\\nMaterials and methods\\n\\nStudy area\\n\\nFloating freshwater marshes are common in Louisiana\\n(Russell, 1942), India (Middleton et al., 1991; Mid-\\ndleton, 1999), Africa (Migahid, 1947; Gaudet, 1977;\\nThompson, 1985), and South America (Junk, 1970). In\\nLouisiana, \\ufb02oating marshes often lie on the coastal\\nside of T. distichum swamps (Middleton, personal\\nobservation). In the Lake Salvador marshes, the water\\nlevels are the highest in the winter, and in\\ufb02uenced only\\nminimally by tides (Sasser et al., 1996). The vegeta-\\ntion \\ufb02oats on a buoyant mat of peat\\nis not\\nanchored. The mat may settle on the underlying\\nsubstrate during low water events. During \\ufb02ooding,\\nwater may \\ufb02ow above or below the mat (Swarzenski\\net al., 1991). Saline water enters Lake Salvador\\nfrom the Gulf of Mexico\\nthrough tidal\\n\\ninlets\\n\\nthat\\n\\n(maximum salinity = 4; Swarzenski et al., 1991) so\\nthat \\ufb02oating marshes near Lake Salvador have a range\\nof salinity from freshwater to brackish (Swarzenski &\\nSwenson, 1994). We studied freshwater marshes in the\\nLake Salvador Management Area, which had lower\\ntype\\nlevels of\\n(Swarzenski et al., 1991). The dominant species in the\\nLake Salvador Management Area are of the C3 assim-\\nilation type including Cyperus haspan, Hydrocotyle\\nverticillata, Panicum hemitomum, Sagittaria lancifolia,\\nand S. americanus (Swarzenski et al., 1991).\\n\\nsalinity than the\\n\\nintermediate\\n\\nExperimental approach\\n\\nSeed bank material was collected in the Lake Salvador\\nManagement Area (29(cid:3)48.513N; 90(cid:3)18.677W) on\\nApril 17, 2003. The peat for the seed bank study was\\nlifted from 10 randomly selected points in the marsh.\\nWith shears,\\nthe shoots were separated from the\\nrhizomes and roots, just below the marsh surface.\\nThe entire top 10 cm of peat was removed from each of\\nthe sample quadrats in a 1-m2 area, and placed into a\\nbag. The collected peat was held for 1.5 months in a\\ncold room (at ca. 5(cid:3)C). On June 7, 2003, rhizomes were\\nremoved from the peat, and the peat samples from all\\ncollection sites were thoroughly homogenized together\\nby hand. The homogenized peat (1 cm) was placed\\nover a layer of vermiculite and sand in 10 9 15 cm\\npots, and placed into the Wetland Elevated CO2\\nExperimental Facility (WECEF), USGS National\\nWetlands Research Center, Lafayette, LA (30(cid:3)22.3N;\\n92(cid:3)1.9W).\\n\\nWECEF consisted of four greenhouses and a control\\nroom that housed CO2 and environmental monitoring\\nequipment. Two greenhouses were randomly assigned\\nto elevated CO2 (720 ll l-1 = elevated) and the other\\ntwo to ambient CO2 (385 ll l-1) treatments. CO2\\nconcentrations were continuously monitored and kept\\nwithin \\u00b15% of the set-point through automated feed-\\nback regulation of \\ufb02ow meters supplying CO2 gas to the\\ngreenhouses (see McKee & Rooth, 2008). One model of\\nthe IPCC (2007) projects a doubling of CO2 by 2050,\\nwhich was the basis for selecting the elevated level of\\nCO2 in the experiment. Photosynthetically active radi-\\nation, humidity, and temperature were also monitored\\n(data not shown). Greenhouse temperatures were\\nallowed to \\ufb02uctuate naturally, although evaporative\\ncooling systems prevented higher-than-ambient tem-\\nperatures during the warmer months. Temperature\\n\\n123\\n\\n\\f126\\n\\nHydrobiologia (2012) 683:123\\u2013133\\n\\ncontrol is impossible without shading during hottest\\nmonths at this latitude in south Louisiana. Shade cloth\\nwas present on the roof (not sides) of the greenhouse and\\nonly during the hottest months (June\\u2013September); light\\nwas mainly attenuated from 10 am to 2 pm while the\\nsun was highest (June\\u2013September 600 lmol m-2 s-1 at\\nmidday).\\n\\nSix pots of peat were placed into each of 24 plastic\\nbins, which were assigned to three water and two\\nsalinity regimes. Salinity levels were set at 0 and 2\\n(low and high salinity, respectively), which simulated\\nfreshwater and 6% sea-strength conditions using\\nInstant Ocean(cid:4). The \\ufb02ooding treatments from low to\\nhigh included freely drained, gradually increasing\\nwater levels to 8 cm above soil surface, and initially\\n\\ufb02ooded to 8 cm (low, gradual, and high, respectively).\\nTo prevent any \\ufb02oating of materials, all of the seed\\nbank pots in all the treatments were watered to the top\\nof the pot to allow the peat to settle, and then, the water\\ntreatments were applied. For the gradually increasing\\nwater regime, the water level was increased by 1 cm\\nper week for 8 weeks, starting on June 20, 2003 after\\nseeds had germinated;\\nthe \\ufb02ooded treatment was\\nraised to 8 cm slowly on the same day to avoid\\ndislodging peat from the pot. For our purposes, the\\ngradually increasing water regime was an intermediate\\nlevel treatment between low and high water levels.\\nEach greenhouse (2 ambient, 2 elevated CO2) con-\\ntained all combinations of three water regimes and two\\nsalinity levels (6 bins).\\n\\nEach bin was watered daily with distilled water to\\nmaintain the water depths and salinity levels in the\\nbins. A pot of vermiculite was placed into each bin to\\ntest for seed contamination. To fertilize the plants, a\\nmodi\\ufb01ed Hoagland solution was added to the water\\nweekly (1 mol m-3 MgSO4, 2 mol m-3 KH2PO4,\\n2.5 mol m-3 CaCl2, 2.0 mol m-3 NH4Cl; trace ele-\\nments as mg dm3: Fe (2.5), Mn (0.25), Cu (0.01), Zn\\n(0.025), Mo (0.025), B (0.25).\\n\\nThe seedlings were harvested starting on November\\n18, 2003. Each seedling was separated from the peat,\\nplaced into a marked bag, counted, and dried in an\\noven at 70(cid:3)C. Roots and shoots were weighed\\nseparately. While seed contamination within the study\\nwas found to be minimal based on seedling emergence\\nfrom vermiculate samples, Utricularia grew between\\nsubmerged pots, so that we were unable to determine\\nits origin. Nomenclature follows the U.S. Department\\nof Agriculture (2011a, b).\\n\\n123\\n\\nStatistical analyses\\n\\nSeed densities were calculated per m2 based on the pot\\nsize (m2 conversion = 62). This common procedure is\\nused to standardize reported seed germination values,\\nand is based on the idea that the density of seeds is\\nrelated to the surface volume in the upper 1 cm2 of the\\nsoil sample (Middleton et al., 1991). In this way,\\nthe technique of using a thin layer of peat in the\\ngreenhouse pots best characterizes the seeds in the\\nupper 10 cm of the peat collected from the \\ufb01eld site.\\nThe experimental design was a split-plot with CO2\\nas the main plot, and salinity and water\\nlevel\\ntreatments as the subplots. Thus, each pair of\\nelevated-ambient greenhouses constituted a complete\\nblock (total of 2 blocks). Three-way ANOVAs were\\nperformed on mean seed density, seedling mass, and\\nspecies richness with main effects of CO2 level,\\nsalinity and water regime, and their interactions\\nexamined using a mixed model. CO2, salinity, and\\nwater level were \\ufb01xed effects in the model; green-\\nhouse (nested within CO2) was a random effect in the\\nmodel. The CO2 effect (main plot) was tested against\\ngreenhouse (nested within CO2), and the subplot\\neffects and interactions were tested with subplot error\\n(Scheiner & Gurevich, 1993). Also, for each dominant\\nspecies,\\nthree way ANOVAs were performed on\\ndensity, mean total biomass, and the mean biomass per\\nindividual. All other species were grouped into a\\nminor category. Using SAS JMP (2007), ANOVA of\\nmain effects and their interactions were tested. To\\nexamine differences among multiple treatment levels,\\nsingle degree-freedom contrasts of\\ninterest were\\nconducted when there was a signi\\ufb01cant main effect\\nor interaction. Data were log transformed to meet the\\nnormality and homogeneity assumptions of ANOVA.\\n\\nResults\\n\\nSeed bank responses\\n\\nA total of 19 species germinated from the seed banks.\\nSpecies richness was not affected by the main effects\\nof CO2, water\\nregime, salinity, or any of their\\ninteractions (Table 1A). In contrast, the density of\\ngerminating seeds was in\\ufb02uenced by the interaction of\\nCO2 with water\\nIn\\nelevated (high) CO2\\nthe gradual water\\n\\nregime (Table 1B; Fig. 1a).\\n\\nlevels,\\n\\n\\fHydrobiologia (2012) 683:123\\u2013133\\n\\n127\\n\\nTable 1 ANOVA analysis of\\ntotal species responses in\\nfreshwater \\ufb02oating marshes in coastal Louisiana for the main\\neffects of CO2 level (ambient vs. elevated), 385 and 720 ll l-1,\\nrespectively, salinity level (salinity = 0 vs. 2, respectively),\\nand water\\nregime (freely drained, 8 cm, and gradually\\nincreased by 1 cm a week to 8 cm) and their interactions, for\\n(A) log species richness, (B) log seedling density, and (C) log\\nmean total of individual seedling biomass\\n\\nSource\\n\\n(A) Species richness\\n\\nCO2\\nSalinity\\n\\nWater regime\\nCO2 9 salinity\\nSalinity 9 water regime\\nCO2 9 water regime\\nCO2 9 salinity 9 water regime\\n\\n(B) Seedling density\\n\\nCO2\\nSalinity\\n\\nWater regime\\nCO2 9 salinity\\nSalinity 9 water regime\\nCO2 9 water regime\\nCO2 9 salinity 9 water regime\\n(C) Individual seedling biomass\\n\\nCO2\\nSalinity\\n\\nWater regime\\nCO2 9 salinity\\nCO2 9 water regime\\nSalinity 9 water regime\\nCO2 9 salinity 9 water regime\\n\\nF\\n\\n0.3\\n\\n0.4\\n\\n0.6\\n\\n0.6\\n\\\\0.1\\n\\n2.8\\n\\n0.5\\n\\n\\\\0.1\\n\\n0.3\\n\\n0.5\\n\\n0.6\\n\\n3.8\\n\\n12.5\\n\\n0.4\\n\\n\\\\0.1\\n\\n28.6\\n\\n0.8\\n\\n1.2\\n0.3\\n\\n0.9\\n\\n3.6\\n\\nP\\n\\nNS\\n\\nNS\\n\\nNS\\n\\nNS\\nNS\\n\\nNS\\n\\nNS\\n\\nNS\\n\\nNS\\n\\nNS\\n\\nNS\\n\\n*\\n\\n***\\n\\nNS\\n\\nNS\\n\\n***\\n\\nNS\\n\\nNS\\nNS\\n\\nNS\\n\\n*\\n\\nSigni\\ufb01cant\\n** P \\\\ 0.01, and *** P \\\\ 0.001\\n\\nrelationships\\n\\nare designated as: * P \\\\ 0.05,\\n\\nhigh water\\n\\n(881.0 \\u00b1 48.0\\n\\ntreatment had higher seedling densities than the low\\nand\\nvs.\\ntreatments\\n701.1 \\u00b1 40.3 seedlings m-2, respectively; t = 9.5,\\nP = 0.0024), while in ambient (low) CO2 levels,\\nseedling densities were higher in high water than in\\nlow and gradual waters (975.5 \\u00b1 79.8 vs. 729.5 \\u00b1\\n34.5 seedlings m-2,\\nt = 13.5, P =\\n0.0003). Water regime also interacted with salinity\\nlevel, so that in higher salinity levels (salinity = 2),\\nseedling density was higher in high than in low and\\ngradual water regimes (917.1 \\u00b1 78.8 vs. 744.8 \\u00b1\\n58.3, respectively; t = -2.12, P = 0.0357; Table 1B;\\n\\nrespectively;\\n\\nFig. 1 Bar graphs of seven dominant species and \\u2018\\u2018other\\u2019\\u2019 minor\\nspecies for all\\ntreatment combinations showing responses\\nincluding a total seedling number m-2 (density: based on pot\\nsize), b seedling total mass per pot (g), and c individual seedling\\nmass (g). Standard errors are omitted for clarity. Treatments\\nincluded CO2 levels of ambient and elevated (385 and 720 ll l-1,\\nrespectively; A and E, respectively), water regimes of low, gradual\\nand high (freely drained, gradually increased by 1 cm a week to\\n8 cm, and 8 cm; L, G, H, respectively), and salinities of 0 and 2\\n\\nFig. 1a). C. haspan had higher total mass in elevated\\nCO2 in low \\ufb02ooding and low salinity conditions, and\\nthe response of this species dominated the overall\\n\\n123\\n\\n\\f128\\n\\nHydrobiologia (2012) 683:123\\u2013133\\n\\nin\\n\\nand\\n\\nhigher\\n\\nresponse in the experiment (Fig. 1b). CO2, water, and\\nto the effects on\\nsalinity interacted with respect\\nindividual seedling biomass (Table 1C). The nature\\nof this interactive response of individual seedling\\nbiomass was that in elevated CO2, seedling biomass\\nlow salinity\\nlow water\\nwas\\n(0.105 \\u00b1 0.01 per seedling; Table 1C; Fig. 1c), but\\nin gradual water and ambient CO2 at either low or high\\nsalinity, the individual seedling weight did not differ\\n(0.06 \\u00b1 0.01 g per seedling; P [ 0.05). Overall,\\nseedling responses to CO2 with respect to species\\nrichness, density and biomass were not necessarily\\npositive, and depended on interactions with other\\ntreatment factors including salinity and water regime\\n(Fig. 1a\\u2013c).\\n\\nSpecies-speci\\ufb01c responses\\n\\nSeedling density\\n\\n(not\\n\\nsalinity)\\n\\n(CO2 9 water\\n\\nFor certain dominant species, the seedling density\\nresponse to CO2 interacted with water regime and\\nsalinity level (Table 2A). For Eleocharis palustris, the\\nseedling density varied across water regime and CO2\\nlevels\\ninteraction;\\nTable 2A; Fig. 1a). For C. haspan, seedling densities\\nwere higher in low CO2 in low salinity and gradual or\\nlow water, and in high CO2 in high or low salinity in\\nlow or gradual water, than in other combinations\\ntreatments (t = 16.9, P \\\\ 0.00001; Table 2A;\\nof\\nFig. 1a). For Polygonum punctatum, the pattern of\\nresponse was very similar to C. haspan in high CO2 in\\nall water regimes; however, in low CO2, the density\\nwas lower in gradually increasing water depths in high\\ntreatments (t = 2.95, P =\\nsalinity than in other\\n0.0038; Table 2A; Fig. 1a). For S. lancifolia, seedling\\ndensities were similar at all water levels, salinities, and\\nCO2 levels, except that, in low CO2 in high water and\\nhigh salinity, the seedling densities were higher than in\\nother treatments (t = 3.2, P = 0.0036; Table 2A;\\nFig. 1a). The seedling densities of E. cellulosa,\\nH. verticillata, and Leersia oryzoides did not respond\\nto main effects or combinations of CO2, salinity, or\\nwater regime (Table 2A).\\n\\nBiomass\\n\\nThe individual seedling biomass of several of the\\ndominant\\nthese\\n\\nresponded to CO2, but\\n\\nspecies\\n\\n123\\n\\nresponses often interacted with water regime and\\nsalinity level (Table 2B). For E. cellulosa, higher\\nseedling biomasses were found at\\nlow salinities,\\ngradually increasing water regime and low CO2 levels\\nthan in other treatment combinations (t = -1.98,\\nP = 0.05; Table 2B; Fig. 1c). For P. punctatum,\\nbiomass increased in the treatments of gradually\\nincreasing water regime, low CO2, and high salinity\\nlevels, but not low water in low CO2 and high salinity\\n(t = -5.58, P \\\\ 0.0001; Table 2B; Fig. 1c). For\\nC. haspan, the biomass response generally increased\\nin treatments with low rather than high salinity\\nregardless of CO2 level or water regime (t = 7.85,\\nP \\\\ 0.0001; Table 2B; Fig. 1c); however, the indi-\\nvidual mean weight of seedlings for C. haspan was\\nhighest in a treatment combination of high CO2, low\\nsalinity and gradual water regime (0.7 g individual-1;\\nt = 6.46, P = 0.0034; Fig. 1c). For E. palustris, mean\\nseedling weight was higher in high CO2 in low and\\ngradual water, rather than in high water (t = 2.36,\\nP = 0.0199; Table 2B; Fig. 1c). In high salinity,\\nE. palustris had higher mean biomass in low water\\nthan in high or gradual water regimes (t = -2.24,\\nP = 0.0269; Table 2B; Fig. 1c). S.\\nlancifolia and\\nL. oryzoides did not respond to CO2 level; instead the\\nweights of seedlings decreased with increasing salinity\\n(Table 2B). For S. lancifolia, mean seedling weight\\nwas higher in low water, than in gradual or high water\\nregimes (Table 2B). H. verticillata did not respond\\ndifferently to any treatment or treatment combination\\nof salinity, CO2 or water depth (Table 2B).\\n\\nDiscussion\\n\\nClimate change may present species with interacting\\nenvironments, so that the overall effect may be more\\nsevere than anticipated based on the analysis of single-\\nenvironmental factors (Ghosn et al., 2010). This study\\nempirically tested seed germination and seedling\\nresponses to interacting environments, which may be\\nassociated with changes in sea level and CO2 in coastal\\nfreshwater marshes. We asked the question in the\\nintroduction: can the species richness, density, and\\nbiomass of freshwater \\ufb02oating marshes be affected by\\nchanges in CO2 concentration and its interactions with\\nwater regime and salinity?\\n\\nSalinity, which reduced total seedling mass by 68%\\nand individual seedling mass by 59%, was the main\\n\\n\\fHydrobiologia (2012) 683:123\\u2013133\\n\\n129\\n\\nTable 2 ANOVA with F ratios of log responses of dominant\\nseed bank species of freshwater \\ufb02oating marshes in coastal\\nLouisiana to CO2 level (ambient vs. elevated), salinity level (0\\nvs. 2) and water regime (freely drained, 8 cm, and gradually\\n\\nincreased by 1 cm a week to 8 cm) for (A) log seedling density\\nm-2 and (B) log individual seedling weight (g); Dominant\\nspecies exceed a total mean density of 0.1 individuals per pot,\\nor, total mean seedling weight of 0.1 g per pot, respectivelyA\\n\\nSource of variation\\n\\nSpecies\\n\\nC. haspan\\n\\nE.\\ncellulosa\\n\\nE. palustris\\n\\nH.\\nverticillata\\n\\nL.\\noryzoidesa\\n\\nP.\\npunctatum\\n\\nS. lancifolia\\n\\n(A) Seedling density\\n\\nCO2\\nSalinity\\nCO2 9 salinity\\nH2O\\nCO2 9 water regime\\nSalinity 9 water regime\\nCO2 9 salinity 9 water\\nregime\\nMean \\u00b1 SE density pot-1\\n(B) Individual seedling weight\\n\\nCO2\\nSalinity\\nCO2 9 salinity\\nH2O\\nCO2 9 water regime\\nSalinity 9 water regime\\nCO2 9 salinity 9 water\\nregime\\n\\nMean \\u00b1 SE individual\\nweight (mg)\\n\\n1.1\\n\\n9.6**\\n\\n0.6\\n\\n5.8**\\n\\n0.2\\n\\n0.1\\n\\n7.8***\\n\\n0.6\\n\\n0.8\\n\\n0.6\\n\\n0.8\\n\\n0.4\\n\\n0.3\\n\\n0.6\\n\\n0.6\\n\\\\0.1\\n\\n0.1\\n\\n0.1\\n\\n5.9**\\n\\n0.7\\n\\n0.2\\n\\n1.6\\n\\n0.5\\n\\n1.2\\n\\n0.9\\n\\n0.6\\n\\n2.2\\n\\n1.3\\n\\n0.4\\n\\\\0.1\\n\\n4.0\\n\\n1.2\\n\\n0.5\\n\\n1.1\\n\\n2.5\\n\\n0.4\\n\\\\0.1\\n\\n3.1\\n\\n1.9\\n\\n5.4**\\n\\n2.5\\n\\n3.3*\\n\\n2.3\\n\\n8.1**\\n\\n\\\\0.1\\n\\\\0.1\\n\\\\0.1\\n\\n3.3*\\n\\n3.1**\\n\\n178.4 \\u00b1 13.7\\n\\n39.8 \\u00b1 5.1\\n\\n138.0 \\u00b1 11.0\\n\\n12.5 \\u00b1 2.4\\n\\n18.1 \\u00b1 3.0\\n\\n101.9 \\u00b1 8.2\\n\\n224.2 \\u00b1 12.1\\n\\n\\\\0.1\\n\\n25.3***\\n\\n5.8*\\n\\n3.4*\\n\\n1.1\\n\\n1.2\\n\\n2.4\\n\\n3.0\\n\\n3.2\\n\\\\0.1\\n\\n0.4\\n\\n2.9\\n\\n1.5\\n\\n3.1*\\n\\n1.3\\n\\n1.6\\n\\n0.6\\n\\n0.8\\n\\n3.8*\\n\\n3.2*\\n\\n1.6\\n\\n0.1\\n\\n0.6\\n\\n0.5\\n\\n2.6\\n\\n0.7\\n\\n1.4\\n\\n0.1\\n\\n0.7\\n\\n5.3*\\n\\n0.8\\n\\\\0.1\\n\\n0.6\\n\\n1.0\\n\\n0.7\\n\\n7.9\\n\\\\0.1\\n\\n0.2\\n\\n0.4\\n\\n0.5\\n\\n1.6\\n\\n3.2*\\n\\n0.2\\n\\n12.1***\\n\\n1.5\\n\\n4.5**\\n\\n\\\\0.1\\n\\n0.7\\n\\n0.7\\n\\n73.8 \\u00b1 6.6\\n\\n16.2 \\u00b1 3.3\\n\\n2.2 \\u00b1 0.4\\n\\n5.6 \\u00b1 2.9\\n\\n35.5 \\u00b1 8.2\\n\\n12.0 \\u00b1 1.9\\n\\n3.4 \\u00b1 0.4\\n\\nValues per m2 were based on the pot dimensions. Signi\\ufb01cant relationships are designated as: * P \\\\ 0.05, ** P \\\\ 0.01 and\\n*** P \\\\ 0.001. The carbon assimilation strategy of all species was C3, with the possible exception of the unknown dicot and\\nmonocot. Species that were not dominant included Alternanthera philoxeroides, Ambrosia artemisifoliaa, Carex sp., Juncus sp.,\\nP. hemitomon, Pluchea odorataa, Ptilimnium capillaceuma and Vigna luteola (mean density \\u00b1 SE = 0.5 \\u00b1 0.5, 0.9 \\u00b1 0.7,\\n1.4 \\u00b1 1.4, 0.9 \\u00b1 0.7, 1.9 \\u00b1 1.1, 0.5 \\u00b1 0.5, 0.5 \\u00b1 0.5, 0.5 \\u00b1 0.5 seedlings m-2, respectively). In addition, there were \\ufb01ve unknowns\\nthat ranged in density from\\\\0.1 to 51.4 seedlings m-2. Species designated with superscript \\u2018\\u2018a\\u2019\\u2019, are annuals or short-lived perennials\\n(USDA, 2011a, b). The mean biomass per individual for all of the minor species was \\\\0.01 g per pot. The total species richness was\\n19, with a total mean density of 786.0 \\u00b1 26.3 seedlings m-2 and a mean total biomass of 159.8 \\u00b1 12.5 mg pot-1. Except for P.\\nhemitomon and the unknowns, these species are likely to use a C3 carbon assimilation strategy\\n\\ntreatment driving seedling response patterns in this\\nstudy, not CO2 level (minimum and maximum value:\\ntotal\\nseedling weight per pot = 0.3\\u2013733.3 mg).\\nFlooding also had an effect on seedling mass, but less\\nthan salinity. Previous studies indicate that some\\nfreshwater species are inhibited by \\ufb02ooding and\\nsalinity levels above 4 ppt (Baldwin et al., 1996).\\nExperimental studies of species from coastal fresh-\\nwater wetlands indicate that the seed germination and/\\nor seedling growth of most species decrease with\\n\\nincreased salinity (Howard & Mendelssohn, 2000;\\nBrock et al., 2005; Middleton, 2009a). As a rule,\\nsalinities above 1,000 mg l-1 decrease the species\\nrichness of seeds germinating from the seed bank\\n(Nielsen et al., 2003; Brock et al., 2005), and this is\\ntrue of a wide range of fresh to salt marsh species\\n(Middleton, 2009a). The negative effect of salinity on\\nseed germination for most species may be caused by:\\n(1) reduced seed viability, (2) blocked cues for\\nemergence, or (3) seedling mortality immediately\\n\\n123\\n\\n\\f130\\n\\nHydrobiologia (2012) 683:123\\u2013133\\n\\nafter emergence (i.e., germination not directly\\naffected) (Nielsen et al., 2003). After storms, coastal\\nfreshwater wetlands may increase in salinity and\\nbecome dominated by different species, even if the\\nspecies involved do not change (Michener et al., 1997;\\nMiddleton, 2009a).\\n\\nThe general characteristics of the seed bank in this\\n\\ufb02oating marsh are roughly similar to those of an\\noligohaline marsh in Louisiana (Baldwin et al., 1996).\\nWhile these studies are really not comparable in terms\\nof the study design and treatments, the general results\\nindicate similar seed bank characteristics. The seed\\nbanks for the oligohaline marsh had 3\\u201310 species, with\\na total of 606\\u20132,960 seeds m-2. In comparison, our\\nstudy had 20 species with a total of 753\\u2013828 seedlings\\nm-2, depending on the water regime and salinity\\ntreatment. For S.\\nthe seed banks held\\n202\\u2013331 seedlings m-2 in the Baldwin et al. study\\n(1996), while our study had 224 seedlings m-2. We\\nknow of one other seed bank study of freshwater and\\nfreshwater \\ufb02oating marshes in Louisiana, but these\\nmarshes had much lower total mean seedling m-2\\nnumber (106\\u2013207 seedlings m-2) in post-hurricane\\nconditions with excessive \\ufb02ooding and/or hurricane\\nsedimentation (Middleton, 2009a).\\n\\nlatifolia,\\n\\nOur study indicates that CO2, water regime, and\\nsalinity in combination produce seed bank responses\\nnot readily predictable from single-factor studies or\\nfunctional type category species (e.g., graminoid vs.\\nherb). Nevertheless, many climate models use rela-\\ntively simple assumptions of functional responses,\\noften based on the carbon assimilation strategy of\\nspecies (Cramer et al., 2001; Krinner et al., 2005), e.g.,\\nCO2 enrichment bene\\ufb01ts the growth of C3 over C4\\nspecies (Bazzaz, 1990; Sage & Kubien, 2003). In\\ncontrast, extensive study shows variable responses of\\nspecies to CO2 treatment depending on other environ-\\nmental constraints (Ko\\u00a8rner, 2006). Similarly in this\\nstudy, the CO2 responses of C3 species in this study\\nfollowed individualistic patterns.\\n\\nTo further explore potential effects of CO2 on\\nthese species, we examined how CO2 enrichment\\nmight offset the effects of stressful environments.\\nSpecies richness was unaffected by CO2, but density\\nand biomass were affected by CO2, depending on\\n\\ufb02ooding or salinity conditions. The implication is that\\nunder a future elevated CO2 scenario, species dom-\\ninance may change even if the total number of species\\ndoes not.\\n\\n123\\n\\nOur \\ufb01ndings differ somewhat from germination\\nstudies of non-wetland species, which generally\\nconclude that germination is little affected by elevated\\nCO2 (Garbutt et al., 1990; Huxman et al., 1998;\\nSteinger et al., 2000; Poorter & Navas, 2003). In these\\nterrestrial studies, a lack of germination density\\nresponse to elevated CO2 has been attributed partly\\nto the fact that the CO2 levels of terrestrial soils are\\nhigher than in the atmosphere (Bazzaz, 1979). Nev-\\nertheless, CO2 levels higher than those of the soil\\ncould promote germination of some species (Baskin &\\nBaskin, 1998). For example,\\nthe germination of\\nPlantago lanceolata is increased by high CO2 levels\\n(Wulff & Alexander, 1985). Ethylene synthesis is\\nincreased in the seeds of Helianthus annuus in higher\\nCO2 levels (Corbineau et al., 1990). Our \\ufb01ndings\\nshowed some effect on seed germination, but gener-\\nally support the broader pattern that CO2 primarily\\nin\\ufb02uences seedling growth, rather than germination,\\nsimilar to \\ufb01ndings of other freshwater wetlands studies\\n(e.g., Megonigal et al., 2005).\\n\\nIndividual species responses to CO2 were inconsis-\\ntent, with some species showing an increase in density or\\nbiomass under elevated CO2 (depending on \\ufb02ooding or\\nsalinity) and others showing no response or a decrease\\n(Fig. 1; Table 2). CO2 enrichment might cause seed-\\nlings to grow better under stressful edaphic conditions,\\nand any CO2-related increase in survival could be\\nimportant to species exposed to rising sea levels. In our\\nstudy, seedlings did not bene\\ufb01t from higher CO2 in\\nstressed conditions, although elevated CO2 promoted\\nhigher total biomass for one species under low stress,\\ne.g., C. haspan under low \\ufb02ooding and low salinity.\\nThus, our results do not support the prediction that\\nhigher CO2 concentrations will ameliorate stressful\\nenvironments related to \\ufb02ooding or salinity levels\\nduring early growth of this seed bank community.\\n\\nClearly,\\n\\nthe responses of these C3 freshwater\\nwetland species in this study were individualistic and\\nnot readily generalized by functional type (graminoid\\nvs. herb) to the entire community. For the purposes of\\nmodeling, the regeneration responses of each species\\nshould be evaluated separately as a set of potential\\noutcomes related to multifactor interactions. Environ-\\nmental change in atmospheric CO2 and sea-level could\\nplay a major role in shaping the vegetation of coastal\\nwetlands. Responses of wetland plant species to future\\nenvironments may be highly individualistic, however,\\ncomplicating speci\\ufb01c predictions of wetland response.\\n\\n\\fHydrobiologia (2012) 683:123\\u2013133\\n\\n131\\n\\nOur \\ufb01ndings also showed that early regeneration\\npatterns re\\ufb02ected the interactions of CO2 with other\\nfactors associated with climate change such as sea-level\\nrise; however, not all species in this freshwater\\ncommunity were able to bene\\ufb01t from higher CO2 levels\\nbecause of overriding effects of salinity and/or \\ufb02ooding.\\nOf the three environmental factors tested, salinity\\nemerged as the most important determinant of seedling\\nresponse pattern. This \\ufb01nding suggests that changes in\\nsea-level, leading to salinity increases, may be more\\nimportant to seedling recruitment in this freshwater\\ncommunity. Flooding can also combine with salinity to\\nfurther impact regeneration patterns, but this effect may\\nbe a minor issue for marshes that \\ufb02oat compared with\\nattached freshwater marshes that are more vulnerable to\\ninundation. CO2 may play a greater role during later\\ngrowth stages and in relation to biotic interactions of\\ncompetition, herbivory, or disease. The entire subject is\\nripe for\\nfuture study because the individualistic\\nresponses of species to abiotic (and biotic) factors can\\nprovide the critical\\nframework to inform models\\npredicting impact of climate change on ecosystems.\\n\\nAcknowledgments This study was part of the USGS program\\non Global Climate Change. The authors wish to express their\\nthanks to all those who helped us in the \\ufb01eld and the greenhouse,\\nincluding Evelyn Anemaet, Andrea Anteau, Kari Foster, Ben\\nHandley, Jill Rooth, Justin Stelly, and Inyoung Yang. Rassa\\nDale and Darren Johnson provided statistical advice. Rebecca\\nHoward, Joy Marburger, and Beth Vairin gave advice to\\nimprove upon the earlier versions of the manuscript. Any use of\\ntrade, product, or \\ufb01rm names is for descriptive purposes only\\nand does not imply endorsement by the U.S. Government.\\n\\nOpen Access This article is distributed under the terms of the\\nCreative Commons Attribution Noncommercial License which\\npermits any noncommercial use, distribution, and reproduction in\\nany medium, provided the original author(s) and source are credited.\\n\\nReferences\\n\\nBaldwin, A. H., K. L. McKee & I. A. Mendelssohn, 1996. The\\nin\\ufb02uence of vegetation, salinity, and inundation on seed\\nbanks of oligohaline coastal marshes. American Journal of\\nBotany 8: 470\\u2013479.\\n\\nBaskin, C. C. & J. M. Baskin, 1998. Seeds. Ecology, Biogeog-\\nraphy, and Evolution of Dormancy and Germination.\\nAcademic Press, San Diego, CA.\\n\\nBazzaz, F. A., 1979. The physiological ecology of plant suc-\\ncession. Annual Review of Ecology and Systematics 10:\\n351\\u2013371.\\n\\nBazzaz, F. A., 1990. The response of natural ecosystems to the\\n\\nrising global CO2 levels. Annual Reviews 21: 167\\u2013196.\\n\\nBrock, M. A., D. L. Nielsen & K. Crossle\\u00b4, 2005. Changes in\\nbiotic communities developing from freshwater wetland\\nsediments under experimental salinity and water regimes.\\nFreshwater Biology 50: 1376\\u20131390.\\n\\nCherry, J. A., K. L. McKee & J. B. Grace, 2009. Elevated CO2\\nenhances biological contributions to elevation change in\\ncoastal wetland by offsetting stressors associated with sea-\\nlevel rise. Journal of Ecology 97: 67\\u201377.\\n\\nCorbineau, F., S. Bagniol & D. Come, 1990. Sun\\ufb02ower\\n(Helianthus annuus L.) seed dormancy and its regulation\\nby ethylene. Israel Journal of Botany 39: 313\\u2013325.\\nCramer, W., A. Bondeau, F. I. Woodward, I. C. Prentice, R.\\nA. Betts, V. Brovkin, P. M. Cox, V. Fisher, J. A. Foley, A.\\nD. Friend, C. Kucharik, M. R. Lomas, N. Ramankutty, S.\\nSitch, B. Smith, A. White & C. Young-Molling, 2001.\\nGlobal response of terrestrial ecosystem structure and\\nfunction to CO2 and climate change: results from six\\ndynamic global vegetation models. Global Change Biology\\n7: 357\\u2013373.\\n\\nCurtis, P. S., B. G. Drake, P. W. Leadley, W. J. Arp & D.\\nF. Whigham, 1989. Growth and senescence in plant com-\\nmunities exposed to elevated CO2 concentrations on an\\nestuarine marsh. Oecologia 78: 20\\u201326.\\n\\nCurtis, P. S., L. M. Balduman, B. G. Drake & D. F. Whigham,\\n1990. Elevated atmospheric CO2 effects on below-ground\\nprocesses in C3 and C4 estuarine marsh communities.\\nEcology 71: 2001\\u20132006.\\n\\nEasterling, D. R., G. A. Meehl, C. Parmesan, S. A. Chagnon, R.\\nR. Karl & L. O. Mearns, 2000. Climate extremes: obser-\\nvation, modeling and impacts. Science 289: 2068\\u20132074.\\n\\nGarbutt, K., W. E. Williams & F. A. Bazzaz, 1990. Analysis of\\nthe differential response of \\ufb01ve annuals to elevated CO2\\nduring growth. Ecology 71: 1185\\u20131194.\\n\\nGaudet, J. J., 1977. Uptake, accumulation and loss of nutrients\\nby papyrus in tropical swamps. Ecology 58: 415\\u2013422.\\nGhosn, D., I. N. Vogiatzakis, G. Kazakis, E. Dimitriou, E.\\nMoussoulis, V. Maliaka & I. Zacharias, 2010. Ecological\\nchanges in the highest temporary pond of western Crete\\n(Greece): past, present and future. Hydrobiologia 648:\\n3\\u201318.\\n\\nHickler, T., S. Fronzek, M. B. Araujo, O. Schweiger, W. Thuiller\\n& M. T. Sykes, 2009. An ecosystem model-based estimate\\nof changes in water availability differs from water proxies\\nthat are commonly used in species distribution models.\\nGlobal Ecology and Biogeography 18: 304\\u2013313.\\n\\nHoward, R. J. & I. A. Mendelssohn, 2000. Structure and com-\\nposition of oligohaline marsh plant communities exposed\\nto salinity pulses. Aquatic Botany 68: 142\\u2013164.\\n\\nHuxman, T. E., E. P. Hamerlynck, D. N. Jordan, K. J. Salsman &\\nS. D. Smith, 1998. The effects of parental CO2 environ-\\nment on seed quality and subsequent seedling performance\\nin Bromus rubens. Oecologia 114: 202\\u2013208.\\n\\nIPCC, Intergovernmental Panel on Climate Change, 2007. Cli-\\nmate Change 2007: The Synthesis Report. Cambridge\\nUniversity Press, Cambridge, UK.\\n\\nJarvis, J. C. & K. A. Moore, 2010. The role of seedlings and seed\\nbank viability in the recovery of Chesapeake Bay, USA,\\nZostera marina populations\\nfollowing a large-scale\\ndecline. Hydrobiologia 649: 55\\u201368.\\n\\nJunk, W. J., 1970. Investigations on the ecology and production\\nbiology of the \\u2018\\ufb02oating meadows\\u2019 (Paspalo-Echinochloetum)\\n\\n123\\n\\n\\f132\\n\\nHydrobiologia (2012) 683:123\\u2013133\\n\\non the Middle Amazon. I. The \\ufb02oating vegetation and its\\necology. Amazonia 2: 449\\u2013495.\\n\\nKo\\u00a8rner, C., 2006. Plant CO2 responses: an issue of de\\ufb01nition,\\ntime and resource supply. New Phytologist 172: 393\\u2013411.\\nKrinner, G., N. Viovy, N. de Noblet-Ducoudre\\u00b4, J. Oge\\u00b4e, J.\\nPolcher, P. Friedlingstein, P. Philippe Ciais, S. Sitch & I.\\nC. Colin Prentice, 2005. A dynamic global vegetation\\nmodel for studies of the coupled atmosphere\\u2013biosphere\\nsystem. Global Biogeochemical Cycles 19: GB1015.\\nLangley, J. A. & J. P. Megonigal, 2010. Ecosystem response to\\nelevated CO2 levels limited by nitrogen-induced plant\\nspecies shift. Nature 466: 96\\u201399.\\n\\nLangley, J. A., K. L. McKee, D. R. Cahoon, J. A. Cherry & J.\\nP. Megonigal, 2009. Elevated CO2 stimulates marsh ele-\\nvation gain, counterbalancing sea-level rise. PNAS 106:\\n6182\\u20136186.\\n\\nLeck, M. A., 1989. Wetland seed banks. In Leck, M. A., V.\\nT. Parker & R. Simpson (eds), Ecology of Soil Seed Banks.\\nAcademic Press, New York: 283\\u2013305.\\n\\nLenssen, G. M., J. Lamers, M. Stroetenga & J. Rozema, 1993.\\nInteractive effects of atmospheric CO2 enrichment, salinity\\nand \\ufb02ooding on growth of C3 (Elymus athericus) and C4\\n(Spartina anglica) salt marsh species. Vegetatio 104(105):\\n379\\u2013388.\\n\\nof Amazonian\\n\\nMayle, F. E., D. J. Beerling, W. D. Gosling & M. B. Bush, 2004.\\nResponses\\nclimatic\\nand atmospheric carbon dioxide changes since the last\\nglacial maximum. Philosophical Transactions of the Royal\\nSociety of London Series B-Biological Sciences 359:\\n499\\u2013514.\\n\\necosystems\\n\\nto\\n\\nMcKee, K. L. & J. E. Rooth, 2008. Where temperate meets\\ntropical: multifactorial effects of elevated CO2, nitrogen\\nenrichment, and competition on a mangrove\\u2013salt marsh\\ncommunity. Global Change Biology 14: 971\\u2013984.\\n\\nMegonigal, J. P., C. D. Vann & A. A. Wolf, 2005. Flooding\\nconstraints on tree (Taxodium distichum) and herb growth\\nresponses to elevated CO2. Wetlands 25: 430\\u2013438.\\nMichener, W. K., E. R. Blood, K. L. Bildstein, M. M. Brinson &\\nL. R. Gardner, 1997. Climate change, hurricanes and\\ntropical storms, and rising sea level in coastal wetlands.\\nEcological Applications 7: 770\\u2013801.\\n\\nMiddleton, B. A., 1999. Succession and herbivory in monsoonal\\nwetlands. Wetland Ecology and Management 6: 189\\u2013202.\\nMiddleton, B. A., 2009a. Regeneration of coastal marsh vege-\\ntation impacted by Hurricanes Katrina and Rita. Wetlands\\n29: 54\\u201365.\\n\\nMiddleton, B. A., 2009b. Effects of Hurricane Katrina on the\\nforest structure of baldcypress swamps of the Gulf Coast.\\nWetlands 29: 80\\u201387.\\n\\nMiddleton, B. A., 2009c. Effects of Hurricane Katrina tree\\nregeneration in baldcypress swamps of the Gulf Coast.\\nWetlands 29: 135\\u2013141.\\n\\nMiddleton, B. A., 2009d. Regeneration potential of baldcypress\\n(Taxodium distichum) swamps and climate change. Plant\\nEcology 202: 257\\u2013274.\\n\\nMiddleton, B. A., A. G. van der Valk, R. L. Williams, D.\\nJ. Mason & C. B. Davis, 1991. Vegetation dynamics and\\nseed banks of a monsoonal wetland overgrown with\\nPaspalum distichum in northern India. Aquatic Botany 40:\\n239\\u2013259.\\n\\n123\\n\\nMigahid, A. M., 1947. An ecological survey of the \\u2018Sudd\\u2019\\nswamp of the Upper Nile. Proceedings of the Egyptian\\nAcademy of Sciences 7: 57\\u201386.\\n\\nNielsen, D. L., M. A. Brock, K. Crossle\\u00b4, K. Harris, M. Healey &\\nI. Jarosinski, 2003. The effects of salinity of aquatic plant\\ngermination and zooplankton hatching from two wetland\\nsediments. Freshwater Biology 48: 2214\\u20132223.\\n\\nOsmond, C. B., O. Bjo\\u00a8rkman & D. J. Anderson, 1980. Physio-\\nlogical Processes in Plant Ecology. Springer-Verlag,\\nBerlin.\\n\\nPlatt, W. J. & J. J. Connell, 2003. Natural disturbances and\\ndirectional replacement of species. Ecological Mono-\\ngraphs 73: 507\\u2013522.\\n\\nPoorter, H. & M.-L. Navas, 2003. Plant growth and competition\\nat elevated CO2: on winners, losers and functional groups.\\nNew Phytologist 157: 175\\u2013198.\\n\\nRasse, D. P., J.-H. Li & B. G. Drake, 2003. Carbon dioxide\\nassimilation by a wetland sedge canopy exposed to ambient\\nand elevated CO2: measurements and model analysis.\\nFunctional Ecology 17: 222\\u2013230.\\n\\nRasse, D. P., G. Peresta & B. G. Drake, 2005. Seventeen years of\\nelevated CO2 exposure in a Chesapeake Bay Wetland:\\nsustained but contrasting responses of plant growth and\\nCO2 uptake. Global Change Biology 11: 369\\u2013377.\\nRozema, J., F. Dorel, R. Janissen, G. Lenssen, R. Broekman, W.\\nArp & B. G. Drake, 1991. Effect of elevated atmospheric\\nCO2 on growth, photosynthesis and water relations of salt\\nmarsh grass species. Aquatic Botany 39: 45\\u201355.\\n\\nRussell, R. J., 1942. Flotant. Geographical Review 32: 74\\u201398.\\nSage, R. F. & D. S. Kubien, 2003. Quo vadis C4? An ecophys-\\niological perspective on global change and the future of C4\\nplants. Photosynthesis Research 77: 209\\u2013225.\\n\\nSAS JMP, 2007. JMP SAS, Version 7.0.1. Statistical Analysis\\n\\nSystem, Cary, NC.\\n\\nSasser, C. E., J. G. Gosselink, E. M. Swenson, C. M. Swarzenski\\n& N. C. Leibowitz, 1996. Vegetation, substrate and\\nhydrology in \\ufb02oating marshes in the Mississippi River\\ndelta plain wetlands, USA. Vegetatio 122: 129\\u2013142.\\nScheiner, S. & J. Gurevich (eds), 1993. Design and Analysis of\\nEcological Experiments. CRC Press, Boca Raton, FL.\\nSteinger, T., R. Gall & B. Schmid, 2000. Maternal and direct\\neffects of elevated CO2 on seed provisioning, germination\\nand seedling growth in Bromus erectus. Oecologia 123:\\n475\\u2013480.\\n\\nSwarzenski, C. M. & E. M. Swenson, 1994. Effect of mat\\nmovement on soil salinity \\ufb02uctuations in some coastal\\nLouisiana marshes. Estuarine, Coastal and Shelf Science\\n39: 143\\u2013156.\\n\\nSwarzenski, C. M., E. M. Swenson, C. E. Sasser & J. G.\\nGosselink, 1991. Marsh mat \\ufb02otation in the Louisiana\\nDelta Plain. Journal of Ecology 79: 999\\u20131011.\\n\\nThompson, K., 1985. Emergent plants of permanent and sea-\\nsonally \\ufb02ooded wetlands. In Denny, P. (ed.), The Ecology\\nand Management of African Wetlands. Dr. W. Junk, The\\nHague, Netherlands: 43\\u2013107.\\n\\nU.S. Department of Agriculture, 2011a. Integrated Taxonomic\\nInformation System. USDA, Washington, DC. www.itis.\\nusda.gov/index.html.\\n\\nU.S. Department of Agriculture, 2011b. Plants Database.\\nUSDA, Washington, DC. http://plants.usda.gov/java/.\\n\\n\\fHydrobiologia (2012) 683:123\\u2013133\\n\\n133\\n\\nWand, S. J. E., G. F. Midgley, M. H. Jones & P. S. Curtis, 1999.\\nResponses of wild C4 and C3 grass (Poaceae) species to\\nelevated atmospheric CO2 concentration: a meta-analytic\\ntest of current theories and perceptions. Global Change\\nBiology 5: 723\\u2013741.\\n\\nWebster, P. J., G. J. Holland, J. A. Curry & H.-R. Chang, 2005.\\nChanges in tropical cyclone number, duration, and inten-\\nsity in a warming environment. Science 309: 1844\\u20131846.\\nWoodward, F. I., G. B. Thompson & I. F. McKee, 1991. The\\neffects of elevated concentration of carbon dioxide on\\n\\nindividual plants, communities and ecosystems. Annals of\\nBotany 67: 23\\u201338.\\n\\nWulff, R. D. & H. M. Alexander, 1985. Intraspeci\\ufb01c variation in\\nthe response to CO2 enrichment in seeds and seedlings of\\nPlantago lanceolata L. Oecologia 66: 458\\u2013460.\\n\\nZiska, L. H., B. G. Drake & S. Chamberlain, 1990. Long-term\\nphotosynthetic response in single leaves of a C3 and C4 salt\\nmarsh species grown at elevated atmospheric CO2 in situ.\\nOecologia 83: 469\\u2013473.\\n\\n123\\n\\n\\f\",\n          \"Manuscript submitted to:   \\n\\nAIMS Biophysics  \\n\\n          Volume 1, Issue 1, 16-30.   \\n\\n        DOI: 10.3934/biophy.2014.1.16 \\n\\nReceived date 4 November 2014, Accepted date 1 December 2014, Published date 4 December 2014 \\n\\nResearch article \\n\\nMechanoelectric feedback does not contribute to the Frank-Starling \\n\\nrelation in the rat and guinea pig heart \\n\\nD Kelly, L Mackenzie and David A. Saint * \\n\\nDiscipline of Physiology, School of Medical Sciences, University of Adelaide, Adelaide, SA 5005 \\nAustralia \\n\\n*  Correspondence: Email:\\u00a0david.saint@adelaide.edu.au; Tel: +6-188-313-3931.   \\n\\nAbstract:  Mechanoelectric  feedback  (MEF)  is  the  process  by  which  mechanical  forces  on  the \\nmyocardium can alter its electrical properties. The effect can be large enough to induce ectopic beats \\nor fibrillation. However, the role of MEF at physiological levels of mechanical stress is not clear. We \\nhave investigated alteration in action potential morphology in rat and guinea pig ventricle and in rat \\natrial tissue at levels of stretch near the plateau of the Frank-Starling curve. Stretch of >100 mm.Hg \\nEnd Diastolic Left Ventricular Pressure (EDLVP) or rapidly applied stretch (EDLVP increased by 25 \\nmm.Hg within 100 ms) often triggered ectopic beats in isolated rat and guinea-pig hearts. However, \\nventricular epicardial monophasic action potentials (MAPs) recorded during stretch to EDLVP up to \\n30 mm. Hg showed no consistent changes in action potential duration (at APD20, APD50 or APD80) \\nin either species. MAP recording detected APD prolongation with very small concentrations of 4-AP \\n(10  \\u00b5M),  confirming  the  discrimination  of  the  recording  technique.  In  isolated  rat  atrial  strips,  no \\nchanges  in  intracellular  action  potential  morphology  or  membrane  potential  were  seen  when \\nstretched  to  levels  producing  an  optimum  increase  in  contractility.  We  conclude  that  alteration  in \\naction  potential  morphology  with  stretch  does  not  contribute  to  the  Frank-Starling  relation  in \\nventricle of rat or guinea-pig isolated heart, or in rat atrial tissue. \\n\\nKeywords: Mechano-electric feedback; Frank-Starling relation; Cardiac electrophysiology; Stretch   \\n\\n1.    Introduction   \\n\\nNearly  a  century  after  Frank,  Starling,  and  colleagues  demonstrated  that  an  increase  in \\nventricular  filling  (sarcomere  length)  promotes  an  increase  in  developed  pressure  (tension),  the \\ncellular  and  molecular  mechanisms  underlying  this  effect  remain  elusive.  A  sudden  increase  in \\nmyocardial stretch produces an immediate (within a beat or two) increase in force that is not due to \\n\\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n\\f17\\u00a0\\n\\nan increase in the calcium transient [1], followed by a slower increase in tension accompanied by a \\nrise in the calcium transient [39]. The rapid change in force is larger than can be accounted for by the \\nchange in overlap of the contractile filaments, and this extra increase in force appears to be due, in \\nlarge  part,  to  an  increase  in  sensitivity  of  the  contractile  proteins.  The  mechanisms  of  this  are  still \\nbeing investigated, but it has been suggested that a closer myofilament lattice spacing is one of the \\nmajor underlying factors [13].   \\n\\nHowever,  as  well  as  these  intracellular  cellular  responses  to  stretch,  stretch  produces \\nelectrophysiological effects on myocardial cells through modulation of stretch-sensitive ion channels \\nin the sarcolemmal membrane [21]. For example, as early as 1915, it was reported that heart rate is \\nincreased  when  the  right  atrium  is  distended  [3]  and  later  studies  showed  this  to  be  a  direct  (ie \\nnon-reflex mediated) effect [4,24]. Mechanical stretch also produces electrical responses in ventricles \\n[14,33]. The responses are complex, and depend upon the timing of the stretch relative to the cardiac \\ncycle;  stretch  produces  a  rapid  hyperpolarisation  if  applied  during  systole  but  a  depolarisation  if \\napplied during diastole [43]. Again, these responses are not due to reflex loops or other extracardiac \\neffects,  since  they  can  be  demonstrated  in  isolated  tissues  [14],  and  indeed  at  the  level  of  single \\nmyocytes  [41].  This  electrophysiological  response  to  stretch  is  called  mechano-electric  feedback \\n(MEF). \\n\\nThe physiological role of MEF in the heart is not clear. It has been proposed that MEF may aid \\nin  the  beat  to  beat  regulation  of  contraction  in  the  face  of  haemodynamic  perturbations  [22]. \\nAlternatively, it has been suggested that it acts to reduce the dispersion of repolarisation that would \\notherwise occur because of regional inhomogeneities in action potential propagation velocity arising \\nas a consequence of, for example, ischaemia [22]. Fundamental to these ideas is the idea that stretch \\ncan  alter  action  potential  morphology  essentially  instantaneously,  and  that  this  in  turn  alters \\ncontractility. Note that stretch has been reported to either prolong, shorten, or produce a \\u201ccross over\\u201d \\nof the AP [21], presumably depending on which ion channels are activated and their current density.   \\nHence, the effect of MEF on the Frank-Starling relation could potentially be additive (a prolongation \\nof the action potential acting to add further tension increases) or negative (shortening of the action \\npotential with stretch attenuating the increase in force produced by other mechanisms). An important \\ncondition for either effect of MEF on the Frank-Starling relation is that MEF, in the form of action \\npotential alterations, should be demonstrable at levels of stretch which are normally encountered by \\nthe heart, i.e. levels of stretch that produce a maximal increase in contractility (ie, near the plateau of \\nthe Frank-Starling relation). Whether changes in action potential occur over this range of stretch is \\nstill  equivocal.  Therefore,  in  the  present  study  we  have  examined  whether  action  potential \\nmorphology is altered in rat and guinea-pig cardiac muscle at levels of stretch within the range of the \\nFrank-Starling relation. \\n\\n2.    Methods \\n\\n2.1.   Isolated hearts \\n\\nAnimals used throughout this study were cared for according to the Australian National Health \\nand  Medical  Council  Guidelines  for  the  Care  and  Use  of  Animals.  The  University  of  Adelaide \\nAnimal Ethics Committee approved all experimental procedures.     \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n\\f18\\u00a0\\n\\nMale  Sprague-Dawley  rats  weighing  350  to  450  g  were  anaesthetized  by  I.P  injection  of  30 \\nmg/kg sodium pentobarbitone with 2000 units of heparin. Guinea pigs weighing 850 to 950 g were \\nanaesthetized  by  I.P  injection  of  60  mg/kg  Ketamine  and  10  mg/kg  Xylazine  with  2000  units  of \\nheparin. Under full anesthesia animals were killed by exsanguination and hearts from either species \\nwere  rapidly  excised  and  placed  in  ice  cold  (5  \\u00b0C)  HEPES  buffer  consisting  of  (in  mmol/L)  NaCl \\n133.5,  NaH2PO4  1.2,  KCl  4,  HEPES  10.0,  MgCl2  1.2,  and  D-glucose  11  (pH  7.4).  The  aorta  was \\ncannulated  with  an  18  gauge  needle  and  retrogradely  perfused  with  bicarbonate  buffered  solution \\n(average  time  to  perfusion  <3  min).  Perfusate  was  equilibrated  with  95%  O2  and  5%  CO2  and \\nconsisted of (in mmol/L) CaCl2 1.5, NaCl 111.0, KCl 4.0, MgCl2 0.6, NaHCO3 23.9, NaH2PO4 1.2, \\nD-glucose  12.0  at  37oC.  Through  an  incision  in  the  left  atrium  a  deflated  thin  latex  balloon  was \\nanchored within the left ventricle. The balloon was then inflated with water until the desired diastolic \\npressure (typically 5 to 30 mmHg) was attained. The balloon was designed to have a considerably \\nlarger  volume  than  the  ventricle,  so  that  its  compliance  did  not  contribute  to  the  diastolic  pressure \\nrecorded at any volume used in the experiments (figure 1). The \\u201czero pressure\\u201d volume of the left \\nventricle was about 200 \\uf06dl and the compliance was considerably greater than that reported by some \\nother  groups  e.g.  [31],  indicating  that  abnormal  stiffness  of  the  ventricular  wall  was  not  a \\nconfounding  factor  in  our  experiments.  Intraventricular  pressure  was  measured  with  a  force \\ntransducer (ADInstruments, Australia) connected to the balloon via a 38 mm long 18 gauge needle.   \\nHearts  were  perfused  at  a  constant  coronary  perfusion  pressure  in  the  range  of  50\\u201380  mmHg  and \\npaced at 4 Hz via the left atrium at twice threshold with 10 ms bipolar pulses. \\n\\n.\\n\\n)\\ng\\nH\\nm\\nm\\n\\n(\\ne\\nr\\nu\\ns\\ns\\ne\\nr\\nP\\n\\n120\\n\\n100\\n\\n80\\n\\n60\\n\\n40\\n\\n20\\n\\n0\\n\\nBalloon + Heart\\n\\nBalloon alone\\n\\n0.0\\n\\n0.5\\n\\n1.0\\n\\n1.5\\n\\n2.0\\n\\n2.5\\n\\nVolume (ml)\\n\\nFigure 1. Compliance of left ventricle of isolated rat hearts and the balloon alone. \\nFor a typical sized rat heart, pressure was measured in the intraventricular balloon \\nfor  small  increments  in  volume,  either  with  (circles)  or  without  (triangles)  the \\nballoon  being  placed  in  the  left  ventricle.  The  balloon  is  partially  collapsed  until \\nslightly over 1 ml is injected, and so generates no pressure at volumes less than this. \\n\\n2.2.   Ventricular monophasic action potential recordings: \\n\\nMonophasic  action  potential  (MAP)  recordings  were  made  using  a  modified  floating  suction \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n\\f19\\u00a0\\n\\nMAP catheter consisting of a small AgCl ball (~1 mm in diameter) in an outer suction cup held on \\nthe epicardial surface with ~50 mm.Hg suction. The MAP electrograms were amplified, digitised at a \\nrate of 2 kHz and analysed using Chart V and peak parameter software (ADInstruments, Australia).   \\nRecordings  were  considered  reproducible  and  acceptable  for  analysis  only  if  they  had  stable  base \\nlines and amplitudes with a variation of less than 10%, as well as a stable action potential duration \\nduring the control period.   \\n\\n2.3.   Rat atrial and ventricular strips \\n\\nWhole rat left atria and epicardial rat left ventricular strips cut from the apex to the base of the   \\n\\nleft  ventricle  were  paced  at  0.5  Hz  at  twice  threshold  with  1  ms  bipolar  pulses.  Tissue  was  then \\nstretched  between  a  fixed  point  and  a  micro-tension  transducer  attached  to  a  micro-manipulator. \\nAction potentials were recorded during intracellular impalements performed on areas under tension \\nusing a modified floating glass microelectrode (1\\u20133 M) filled with 3 M KCl.   \\n\\n3.    Results \\n\\n3.1.   Ventricle\\u00a0\\n\\nIn  isolated  rat  hearts,  MEF,  in  the  form  of  ectopic  beats,  could  be  readily  demonstrated  by \\nincreasing end-diastolic left ventricular pressure (EDLVP). An increase in the volume of the balloon \\nin the left ventricle sufficient to produce an increase in EDLVP from about 10 mmHg to about 100 \\nmmHg  (an  injected  volume  of  approx  750  \\uf06dl)  within  2  to  3  seconds  typically  induced  a  series  of \\nectopic  beats  lasting  up  to  10  seconds,  after  which  the  rhythm  returned  to  normal  (figure  2A). \\nAlthough this pressure is obviously beyond the physiological range, it did not cause damage to the \\nheart since the contractility at normal LVEDP appeared unaffected even after repeated applications of \\nthis  pressure.  Lower  pressures  could  evoke  arrhythmias  if  applied  more  rapidly:  an  increase  in \\ndiastolic pressure from near zero to approx 20 mm.Hg (an injection volume of approx 300 \\uf06dl) within \\nabout  250  ms  produced  a  single  ectopic  beat  (figure  2B).  Intriguingly,  a  sudden  release  of  stretch \\nfrom approx 20 mm.Hg to near zero also produced an ectopic beat (figure 2C). Smaller changes in \\npressure, even if very rapid, did not induce ectopic beats. When EDLVP was increased stepwise in \\nshort  steps  of  a  few  seconds  each  from  0  mm.Hg  to  between  25  and  30  mm.Hg  and  then  reduced \\nonce  more,  a  substantial  increase  in  contractility  of  the  ventricle  was  seen,  as  evidenced  by  an \\nincrease in pulse pressure and dP/dt (figure 3A). An EDLVP of 25 to 30 mm.Hg was well into the \\nplateau phase of the Frank-Starling relation (figure 3B). Ectopic beats were very rarely observed at \\nthese  levels  of  ventricular  stretch  with  this  protocol  and  within  this  pressure  range  there  were  no \\nobvious  differences  in  MAPs  recorded  at  <5  mm.Hg  and  >20  mm.Hg  duration  (figure  3C). \\nMeasurement of APD at 20%, 50% and 80% repolarisation and AP amplitude at number of pressures \\nover  this  range  of  LVEDP  showed  no  change  in  either  rat  or  guinea-pig  isolated  hearts  (figure  4). \\nSuperimposition of individual action potentials recorded at near zero and at 25 to 30 mm.Hg in these \\nhearts showed that they were almost identical for both species (figure 4B and 4D). Stable amplitude \\nMAP recordings obtained from animals were also observed over time by looking at control,   \\n20\\u201325 mm.Hg EDLVP and re-control conditions following the completion of a Frank-Starling curve. \\nThe data is summarised for 4 rat and 4 guinea pig hearts at these EDLVPs in figure 5. There was no \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n\\fsignificant  difference  in  MAP  duration  at  any  level  of  repolarisation  during  stretch,  compared  to \\neither control or recontrol, in either species. \\n\\n20\\u00a0\\n\\n250\\n250\\n250\\n250\\n\\n200\\n200\\n200\\n200\\n\\n150\\n150\\n150\\n150\\n\\n100\\n100\\n100\\n100\\n\\n50\\n50\\n50\\n50\\n\\n0\\n0\\n0\\n0\\n\\n10\\n10\\n10\\n10\\n\\n5\\n5\\n5\\n5\\n\\n0\\n0\\n0\\n0\\n\\n-5\\n-5\\n-5\\n-5\\n\\n)\\n)\\n)\\ng\\ng\\ng\\nH\\nH\\nH\\nm\\nm\\nm\\nm\\nm\\nm\\n\\n(\\n(\\n(\\n\\nP\\nP\\nP\\nV\\nV\\nV\\nL\\nL\\nL\\n\\n)\\n)\\n)\\n\\nV\\nV\\nV\\nm\\nm\\nm\\n\\n(\\n(\\n(\\nP\\nP\\nP\\nA\\nA\\nA\\nM\\nM\\nM\\n\\n1 s\\n1 s\\n\\n)\\n)\\n)\\ng\\ng\\ng\\nH\\nH\\nH\\nm\\nm\\nm\\nm\\nm\\nm\\n\\n(\\n(\\n(\\n\\nP\\nP\\nP\\nV\\nV\\nV\\nL\\nL\\nL\\n\\n)\\n)\\n)\\n\\nV\\nV\\nV\\nm\\nm\\nm\\n\\n(\\n(\\n(\\nP\\nP\\nP\\nA\\nA\\nA\\nM\\nM\\nM\\n\\n120\\n120\\n120\\n100\\n100\\n100\\n80\\n80\\n80\\n60\\n60\\n60\\n40\\n40\\n40\\n20\\n20\\n20\\n0\\n0\\n0\\n\\n4\\n4\\n4\\n\\n2\\n2\\n2\\n\\n0\\n0\\n0\\n\\n-2\\n-2\\n-2\\n\\nA \\n\\n)\\n)\\ng\\ng\\nH\\nH\\nm\\nm\\nm\\nm\\n\\n(\\n(\\n\\nP\\nP\\nV\\nV\\nL\\nL\\n\\n)\\n)\\n\\nV\\nV\\nm\\nm\\n\\n(\\n(\\nP\\nP\\nA\\nA\\nM\\nM\\n\\n100\\n100\\n\\n50\\n50\\n\\n0\\n0\\n\\n4\\n4\\n\\n2\\n2\\n\\n0\\n0\\n\\n-2\\n-2\\n\\n0.5 s0.5 s\\n0.5 s\\n\\n0.5 s\\n0.5 s\\n\\n                B                                                                  C \\n\\nFigure 2. Ectopic beats evoked in isolated rat heart by stretch. Panel A: Top trace; \\nintraventricular pressure, lower trace; MAP recorded from the epicardial  surface. \\nHeart  paced  at  4  Hz.  Increase  in  EDLVP  from  near  0  to \\uf0bb100  mm.Hg  over  2  to  3 \\nseconds  produced  a  series  of  ectopic  beats.  Panel  B:  effect  of  a  sudden  increase  in \\nEDLVP  pressure  from  near  zero  to  20  mm.Hg.  Upper  trace;  ventricular  pressure, \\nlower trace is ECG recording from surface electrodes. Pressure was increased at the \\narrow. Panel C: Effect of sudden release of stretch (at arrow). Traces as in Panel B. \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n\\fControl FS3 starting commment 4\\nControl FS3 starting commment 4\\n\\n21\\u00a0\\n\\na\\na\\n\\nb\\nb\\n\\nc\\nc\\n\\n20\\n20\\n\\n4:25\\n4:25\\n\\n4:30\\n4:30\\n\\n4:35\\n4:35\\n\\n4:40\\n4:40\\n\\n4:45\\n4:45\\n\\n4:50\\n4:50\\n\\n4:55\\n4:55\\n\\n5:00\\n5:00\\n\\n5:05\\n5:05\\n\\n5:10\\n5:10\\n\\n5:15\\n5:15\\n\\ne\\ne\\nr\\nr\\nu\\nu\\ns\\ns\\ns\\ns\\ne\\ne\\nr\\nr\\nP\\nP\\n\\nw\\nw\\no\\no\\nl\\nl\\nF\\nF\\n\\nP\\nP\\nV\\nV\\nL\\nL\\n\\nt\\nt\\nd\\nd\\n/\\n/\\nP\\nP\\nV\\nV\\nL\\nL\\nd\\nd\\n\\n.\\n.\\n.\\n.\\n.\\n.\\n\\n(\\n(\\n\\n) 80\\n) 80\\n80\\n80\\nm\\nm\\ng\\ng\\n60\\n60\\nH\\nH\\nm\\nm\\nm\\nm\\n\\ns\\ns\\ne\\ne\\nr\\nr\\nP\\nP\\nn\\nn\\no\\no\\ni\\ni\\ns\\ns\\nu\\nu\\nf\\nf\\nr\\nr\\ne\\ne\\n(\\n(\\nP\\nP\\n\\n40\\n40\\n\\n20\\n20\\n0\\n0\\n0\\n0\\n20\\n20\\n20\\n20\\n15\\n15\\n\\n10\\n10\\n\\n5\\n5\\n0\\n0\\n0\\n0\\n\\n)\\n)\\nn\\nn\\ni\\ni\\nm\\nm\\nL/\\nL/\\nn\\nn\\nm\\nm\\ni\\ni\\n(\\n(\\nm\\nm\\n\\ne\\ne\\nt\\nt\\na\\na\\n/\\n/\\nR\\nR\\nl\\nl\\nw\\nw\\no\\no\\nl\\nl\\nF\\nF\\n\\nm\\nm\\n\\n(\\n(\\n\\n)\\n)\\ng\\ng\\ng 100\\ng 100\\nH\\nH\\n100\\n100\\nm\\nm\\nm\\nm\\nH\\nH\\nm\\nm\\nm\\nm\\n\\ns\\ns\\ne\\ne\\nr\\nr\\nP\\nP\\nV \\nV \\n\\n50\\n50\\n\\nt\\nt\\nf\\nf\\ne\\ne\\nL\\nL\\n\\n0\\n0\\n0\\n0\\n\\n(\\n(\\n\\ns2000\\ns2000\\n2000\\n2000\\n)\\n)\\ns\\ns\\n/\\n/\\ng\\ng\\nH\\nH\\n/\\n/\\n1000\\n1000\\nm\\nm\\ng\\ng\\n1000\\n1000\\nm\\nm\\nH\\nH\\n0\\n0\\n0\\n0\\nm\\nm\\nm\\nm\\n-1000\\n-1000\\n-1000\\n-1000\\n\\nP\\nP\\nV\\nV\\nL\\nL\\nd\\nd\\n\\nt\\nt\\nd\\nd\\n/\\n/\\n\\nP\\nP\\nA\\nA\\nM\\nM\\n\\n)\\n)\\nV)\\nV)\\nV\\nV\\nm\\nm\\n(\\n(\\nm\\nm\\nP\\nP\\nA\\nA\\nM\\nM\\n(\\n(\\n\\n15\\n15\\n15\\n15\\n10\\n10\\n5\\n5\\n5\\n5\\n0\\n0\\n-5\\n-5\\n-5\\n-5\\n\\n)\\ng\\nH\\nm\\nm\\n\\n.\\n\\n(\\n\\ne\\nr\\nu\\ns\\ns\\ne\\nr\\np\\ne\\ns\\nl\\nu\\nP\\n\\nb\\n\\n90\\n\\n80\\n\\n70\\n\\n60\\n\\na, c\\n\\n0\\n\\n5\\n\\n10\\n\\n15\\n\\n20\\n\\n25\\n\\nDiastolic pressure (mm.Hg)\\n\\nA \\n\\na\\na\\n\\n15\\n15\\n15\\n\\n10\\n10\\n10\\n\\n5\\n5\\n5\\n\\n0\\n0\\n0\\n\\nV)\\nV)\\nV)\\nm\\nm\\nm\\nMAP (\\nMAP (\\nMAP (\\n\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\n\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\n\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\nControl FS3 starting commment 4\\n\\nV\\nV\\nm\\nm\\n5\\n5\\n\\n50 ms\\n50 ms\\n\\nV)\\nV)\\nV)\\nm\\nm\\nm\\nMAP (\\nMAP (\\nMAP (\\n\\nb\\nb\\n\\n15\\n15\\n15\\n\\n10\\n10\\n10\\n\\n5\\n5\\n5\\n\\n0\\n0\\n0\\n\\nc\\nc\\n\\n15\\n15\\n15\\n\\n10\\n10\\n10\\n\\n5\\n5\\n5\\n\\n0\\n0\\n0\\n\\nV)\\nV)\\nV)\\nm\\nm\\nm\\nMAP (\\nMAP (\\nMAP (\\n\\nB                                                                C \\n\\nFigure  3.  Effect  of  increasing  Left  Ventricular  End  Diastolic  Pressure  on \\ncontractility and Monophasic action potential in rat heart. Panel A: from top down: \\nCoronary  perfusion  pressure  (  \\uf0bb62  mm.Hg),  coronary  flow  (maintained  at  15.5 \\nml.min-1  by  feedback  control),  Left  ventricular  pressure,  left  ventricular  dP/dt, \\nMAP recording. The horizontal scale bar indicates 5 seconds. Panel B: plot of pulse \\npressure  against  LVEDP.  5  pressure  pulses  were  measured  at  each  LVEDP  and \\nplotted as mean \\uf0b1 SEM. Panel C: MAPs recorded at the times indicated by a,b,c in \\nthe upper section (at LVEDP of <5, >20 and <5 mm.Hg respectively).     \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n\\f22\\u00a0\\n\\nAPD20\\nAPD20\\nAPD20\\n\\nAPD50\\nAPD50\\nAPD50\\n\\nAPD80\\nAPD80\\nAPD80\\n\\n5 mV\\n5 mV\\n5 mV\\n\\n50 ms\\n50 ms\\n50 ms\\n50 ms\\n\\n0\\n0\\n0\\n\\n5\\n5\\n5\\n\\n10\\n10\\n10\\n\\n15\\n15\\n15\\n\\n20\\n20\\n20\\n\\n25\\n25\\n25\\n\\nDiastolic Pressure (mmHg)\\nDiastolic Pressure (mmHg)\\nDiastolic Pressure (mmHg)\\n\\n)\\n)\\n)\\ns\\ns\\ns\\n\\nm\\nm\\nm\\n\\n(\\n(\\n(\\nn\\nn\\nn\\no\\no\\no\\ni\\ni\\ni\\nt\\nt\\nt\\na\\na\\na\\nr\\nr\\nr\\nu\\nu\\nu\\nD\\nD\\nD\\nP\\nP\\nP\\nA\\nA\\nA\\nM\\nM\\nM\\n\\n50\\n50\\n50\\n45\\n45\\n45\\n40\\n40\\n40\\n35\\n35\\n35\\n30\\n30\\n30\\n25\\n25\\n25\\n20\\n20\\n20\\n15\\n15\\n15\\n10\\n10\\n10\\n5\\n5\\n5\\n0\\n0\\n0\\n\\n                                                       A   \\n\\n                                                          B \\n\\n)\\n)\\ns\\ns\\n\\nm\\nm\\n\\n(\\n(\\n\\nn\\nn\\no\\no\\ni\\ni\\nt\\nt\\na\\na\\nr\\nr\\nu\\nu\\nD\\nD\\n\\nP\\nP\\nA\\nA\\nM\\nM\\n\\n140\\n140\\n\\n120\\n120\\n\\n100\\n100\\n\\n80\\n80\\n\\n60\\n60\\n\\n40\\n40\\n\\n20\\n20\\n\\n0\\n0\\n\\n0\\n0\\n\\n5\\n5\\n\\nAPD20\\nAPD20\\nAPD50\\nAPD50\\nAPD80\\nAPD80\\n\\n25\\n25\\n\\n30\\n30\\n\\n5mV\\n5mV\\n5mV\\n\\n50 ms\\n50 ms\\n50 ms\\n\\n10\\n10\\nDiastolic Pressure (mmHg)\\nDiastolic Pressure (mmHg)\\n\\n20\\n20\\n\\n15\\n15\\n\\n                                                   C       \\n\\n                                                          D \\n\\nFigure  4.  Effect  of  increasing  Left  Ventricular  End  Diastolic  Pressure  on \\nMonophasic action potentials in rat and guinea pig heart. Panel A: MAP durations \\nin  a  rat  heart  measured  at  20%,  50%  and  80%  repolarisation  plotted  against  the \\nEDLVP at the time of recording. Dotted lines show bet fit linear regression lines for \\neach data set: none of these lines has a slope significantly different from zero. Panel \\nB: MAPs recorded at EDLVP of 0, 25 and 0 mm.Hg shown superimposed. Panel C \\nand D: as in Panel A and Panel B, for Guinea pig heart (again, slopes of regression \\nlines not significantly different from zero).   \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n\\f23\\u00a0\\n\\nA \\n\\nB \\n\\nFigure  5.  Effect  of  increasing  Left  Ventricular  End  Diastolic  Pressure  on \\nMonophasic action potential durations at different levels of repolarisation in rat and \\nguinea pig hearts. MAP durations at 20%, 50% and 80% repolarisation plotted for \\ncontrol  (close  to  zero  EDLVP),  EDLVP  >20  mm.Hg  and  re-control  (EDLVP \\nreturned  to  close  to  zero).  Error  bars  indicate  \\uf0b1  SEM,  n  =  4  hearts  (each \\nmeasurement in each heart is an average of at least 15 action potentials at each level \\nof stretch). There are no significant differences in MAP durations between control,   \\n>20 mm.Hg and re-control. Panel A: data for rat and Panel B: data for Guinea pig. \\n\\nSimilar  experiments  in  which  intracellular  action  potentials  were  recorded  in  rat  ventricular \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n\\f24\\u00a0\\n\\nstrips  also  revealed  no  consistent  differences  in  action  potential  morphology  when  the  tissue  was \\nstretched (table 1). This result is consistent with that previously reported by Kiseleva et al [20], who \\nalso reported afterdepolarisations (presumably the correlate of our ectopic beats) at higher levels of \\nstretch. \\n\\nTable  1.  Changes  in  intracellular  action  potential  parameters  in  rat  atrial  and \\nventricular strips. \\n\\nAtrial\\n\\nAPD20 APD50 APD80 RMP Amplitude Ventricular APD20 APD50 APD80 RMP Amplitude\\n\\nNo Change 20 (11)\\nIncrease\\nDecrease\\n\\n4 (3)\\n\\n-\\n\\n14 (9)\\n\\n8 (7)\\n\\n1 (1)\\n\\n9 (7)\\n\\n-\\n\\n14 (8)\\n\\n9 (8)\\n\\n8 (7)\\n\\n7 (5)\\n\\n6 (6)\\n\\n6 (5)\\n\\n12 (8)\\n\\nNo Change\\n\\n2 (2)\\n\\n2 (2)\\n\\nIncrease\\n\\nDecrease\\n\\n-\\n\\n-\\n\\n1 (1)\\n\\n1 (1)\\n\\n1 (1)\\n\\n1 (1)\\n\\n-\\n\\n-\\n\\n2 (2)\\n\\n-\\n\\n1 (1)\\n\\n-\\n\\n2 (2)\\n\\nTable  1.  Results  from  \\u201cx\\u201d  experiments  and  (y)  rats.  Data  show  variation \\nobserved  in  Action  Potential  Duration  (APD)  measurements  following  a  minimum \\n0.3 g step in resting tension (from 0\\u20130.5 g to 0.8 g). Decreases in Resting Membrane \\nPotential  (RMP)  denote  hyperpolarization  while  increases  in  RMP  represent \\ndepolarisation.  P<0.05  (un-paired  t-tests  for  separate  impalements,  paired  for \\ncontinuous  impalements).  Multiple  steps  in  resting  tension  were  made  in  each \\npreparation.  Where  differences  were  observed  in  response,  the  number  of \\npreparations and animals that exhibited the response is also shown. \\n\\n3.2.   Atria \\n\\nIn  intact  rat  atria,  previous  studies  have  shown  that  showed  that  modest  stretch  alters  action \\npotential  morphology  [17,38].  We  attempted  to  reproduce  this  observation  in  rat  atrial  strips. \\nAlthough increasing the tension over the range 0.5 to 3 g produced an obvious increase in dT/dt, no \\nclear  change  in  action  potential  morphology  could  be  seen  (figure  6A  and  6B).  In  a  single \\nimpalement retained for over 2 minutes, no obvious change in action potential morphology could be \\nseen at stretches up to 4 g (which produced an increase in developed tension of about 50%), although \\na  small  depolarisation  of  resting  membrane  potential  could  be  discerned  (figure  6C).  The  passive \\ntensions  that  we  used  in  these  experiments,  although  apparently  within  the  range  of  the \\nFrank-Starling relation, are much higher than those used by Kamkin et al (1.7 mN) [17]. It may be \\nthat this is because of the different orientation of the tissue relative to the stretch axis (ie stretch is \\nnot  being  applied  axially  to  the  muscle  fibres).  Nevertheless,  for  levels  of  stretch  that  produced \\nobvious changes in contractility, we recorded no change in action potential morphology. The results \\nof  our  recordings  of  intracellular  action  potentials  in  isolated  atrial  tissue  are  also  summarised  in \\ntable  1.  Since  the  action  potential  morphology  remains  essentially  constant  throughout  the \\nexperiment,  and  contractility  is  similarly  maintained,  we  conclude  that  tissue  ischaemia  is  not \\naffecting  the  data  (rat  atrial  wall  is  about  200  \\u00b5m  thick  so  diffusion  of  oxygen  should  not  be  a \\nlimiting factor.     \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n\\fResting tension (g)\\nResting tension (g)\\n\\nMinimum dT /dt\\nMinimum dT /dt\\nMaximum dT /dt\\nMaximum dT /dt\\n\\n0.5\\n0.5\\n\\n1\\n1\\n\\n1.5\\n1.5\\n\\n2\\n2\\n\\n2.5\\n2.5\\n\\n3\\n3\\n\\n3.5\\n3.5\\n\\n)\\n)\\ns\\ns\\n\\nm\\nm\\n\\n(\\n(\\n\\nn\\nn\\no\\no\\ni\\ni\\nt\\nt\\na\\na\\nr\\nr\\nu\\nu\\nD\\nD\\n\\nl\\nl\\na\\na\\ni\\ni\\nt\\nt\\nn\\nn\\ne\\ne\\nt\\nt\\no\\no\\nP\\nP\\nn\\nn\\no\\no\\ni\\ni\\nt\\nt\\nc\\nc\\nA\\nA\\n\\n35\\n35\\n35\\n\\n30\\n30\\n30\\n\\n25\\n25\\n25\\n\\n20\\n20\\n20\\n\\n15\\n15\\n15\\n\\n10\\n10\\n10\\n\\n5\\n5\\n5\\n\\n0\\n0\\n0\\n0\\n0\\n0\\n\\n)\\n)\\ns\\ns\\n/\\n/\\ng\\ng\\n(\\n(\\n\\nn\\nn\\no\\no\\ni\\ni\\ns\\ns\\nn\\nn\\ne\\ne\\nt\\nt\\n\\nn\\nn\\n\\ni\\ni\\n\\ne\\ne\\ng\\ng\\nn\\nn\\na\\na\\nh\\nh\\nc\\nc\\n\\nf\\nf\\no\\no\\n\\ne\\ne\\nt\\nt\\na\\na\\nR\\nR\\n\\n4\\n4\\n\\n3\\n3\\n\\n2\\n2\\n\\n1\\n1\\n\\n0\\n0\\n\\n-1\\n-1\\n\\n-2\\n-2\\n\\n-3\\n-3\\n\\n-4\\n-4\\n\\n25\\u00a0\\n\\nAPD20\\nAPD20\\nAPD20\\n\\nAPD50\\nAPD50\\nAPD50\\nAPD80\\nAPD80\\nAPD80\\n\\n0.5\\n0.5\\n0.5\\n\\n1\\n1\\n1\\n\\n1.5\\n1.5\\n1.5\\n\\n2\\n2\\n2\\n\\n2.5\\n2.5\\n2.5\\n\\n3\\n3\\n3\\n\\n3.5\\n3.5\\n3.5\\n\\nResting Tension (g)\\nResting Tension (g)\\nResting Tension (g)\\n\\n                                          A                                                                            B \\n\\n)\\n)\\n\\nV\\nV\\nm\\nm\\n\\n(\\n(\\n\\n(\\n(\\nl\\nl\\na\\na\\ni\\ni\\nt\\nt\\nn\\nn\\nV)\\nV)\\ne\\ne\\nm\\nm\\nt\\nt\\no\\no\\nr\\nr\\na\\na\\nP\\nP\\nu\\nu\\nll\\nll\\ne\\ne\\nc\\nc\\ne\\ne\\na\\na\\nr\\nr\\nt\\nt\\nn\\nn\\nn\\nn\\nI\\nI\\na\\na\\nr\\nr\\nb\\nb\\nm\\nm\\ne\\ne\\nM\\nM\\n\\n)\\n)\\ng\\ng\\n(\\n(\\nn\\nn\\no\\no\\ni\\ni\\ns\\ns\\nn\\nn\\n)\\n)\\ng\\ng\\ne\\ne\\n(\\n(\\nT\\nT\\nn\\nn\\nio\\nio\\ns\\ns\\nn\\nn\\ng\\ng\\ne\\ne\\nT\\nT\\nn\\nn\\ni\\ni\\nt\\nt\\ns\\ns\\ne\\ne\\nR\\nR\\n\\n20\\n20\\n\\n20\\n20\\n\\n0\\n0\\n\\n0\\n0\\n\\n-20\\n-20\\n\\n-20\\n-20\\n\\n-40\\n-40\\n\\n-40\\n-40\\n\\n-60\\n-60\\n\\n-60\\n-60\\n\\n5\\n5\\n\\n5\\n5\\n\\n4\\n4\\n\\n4\\n4\\n\\n3\\n3\\n\\n3\\n3\\n\\n2\\n2\\n\\n2\\n2\\n\\n1\\n1\\n\\n1\\n1\\n\\n0\\n0\\n\\nControl FS Part 4\\nControl FS Part 4\\n\\nc\\nc\\n\\nb\\nb\\n\\na\\na\\n\\nd\\nd\\n\\n33:50\\n33:50\\n\\n1:34:00\\n1:34:00\\n\\n1:34:10\\n1:34:10\\n\\n1:34:30\\n1:34:30\\n\\n1:34:50\\n1:34:50\\n\\n1:35:00\\n1:35:00\\n\\n1:35:10\\n1:35:10\\n\\n1:35:20\\n1:35:20\\n\\n1:35:30\\n1:35:30\\n\\n1:35:40\\n1:35:40\\n\\n1:35:50\\n1:35:50\\n\\n1:36:00\\n1:36:00\\n\\n1:36:10\\n1:36:10\\n\\n1:36:20\\n1:36:20\\n\\n1:36:40\\n1:36:40\\n\\n10 s\\n10 s\\n\\na\\u00a0\\n\\nb\\u00a0\\n\\nC \\n\\nc\\n\\nIntr\\nIntr\\n\\nIntr\\nIntr\\n\\nIntr\\nIntr\\n\\nIntr\\nIntr\\n\\nd\\n\\n80\\u00a0mV\\u00a0\\n\\n25\\u00a0ms\\u00a0\\n\\nD \\n\\nFigure  6.  Effect  of  stretch  on  contractility  and  intracellular  action  potential \\nmorphology  in  rat  atrial  strips.  Panel  A:  dT/dt  (both  positive  and  negative) \\nrecorded  in  a  rat  atrial  strip  plotted  against  the  level  of  resting  tension.  Panel  B: \\nAction  potential  duration  for  intracellular  action  potentials  in  the  same  tissue  as \\nPanel A at 20%, 50% and 80% repolarisation at different levels of resting tension. \\nEach point represents an action potential from a continuous impalement. Panel C: \\nexample of a record in which a single impalement was maintained in an atrial strip. \\nTop panel shows the intracellular membrane potential, below it is the tension record. \\nAction potentials recorded at points a, b, c and d are shown in Panel D. \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n\\f26\\u00a0\\n\\n3.3.   Discrimination of the recording technique \\n\\nWe were concerned that the failure to see changes in action potential morphology at low levels \\nof  stretch  may  have  been  due  to  poor  discrimination  of  the  recording  technique.  We  tested  this  by \\nrecording  the  action  potential  changes  produced  by  10  \\uf06dM  4-AP,  a  concentration  which  would  be \\nexpected  to  produce  barely  perceptible  changes  in  action  potential  morphology  (EC50  for  4-AP  to \\nblock Ito in rat heart is 1 mM [40]). We find that even with this very low concentration of 4-AP we \\ncan record significant differences in APD (figure 7). This confirms that the failure to see changes in \\nAPD with stretch is not due to poor discrimination of the recording technique. \\n\\n*\\n\\n5 mV\\n\\nControl\\n10\\u791b  4-AP\\n\\n20 ms\\n\\nn\\no\\n\\ni\\nt\\na\\nr\\nu\\nD\\n\\nl\\na\\ni\\nt\\nn\\ne\\nt\\no\\nP\\nn\\no\\n\\ni\\nt\\nc\\nA\\nc\\ni\\ns\\na\\nh\\np\\no\\nn\\no\\nM\\n\\n)\\ns\\nm\\n\\n(\\n\\n50\\n\\n45\\n\\n40\\n\\n35\\n\\n30\\n\\n25\\n\\n20\\n\\n15\\n\\n10\\n\\n5\\n\\n0\\n\\nAPD20\\n\\nAPD50\\n\\nAPD80\\n\\nMonophasic Action Potential Measure\\n\\nFigure 7. Sensitivity of MAP recording to detect change in action potential duration \\nwith  low  dose  4-AP.  Main  Panel:  Action  potential  duration  measured  by \\nmonophasic action potential recording in an isolated rat heart paced at 5 Hz at 20%, \\n50%  and  80%  repolarisation  in  control  (open  bars)  and  in  the  presence  of  10  \\u00b5M \\n4-AP  (shaded  bars)  (n  =  4).  *  indicates  significant  difference  (P  <  0.05).  Inset- \\nsuperimposed  typical  traces  of  MAP  recording  in  control  and  in  the  presence  of   \\n10 \\u00b5M 4-AP. \\n\\n4.    Discussion \\n\\nMany experimental and theoretical studies have sought to demonstrate a role for MEF in cardiac \\nphysiology  [21].  One  of  the  more  consistent  observations  is  that  stretch  produces  slight \\ndepolarisation [20] and a rise in intracellular calcium [42], presumably by activation of non-selective \\ncation  channels,  and  this  appears  to  be  involved  in  the  slow  inotropic  response  and  possibly  the \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n \\n \\n \\n \\n \\n\\f27\\u00a0\\n\\nhypertrophic  response  to  stretch  [6\\u20138].  However,  these  responses  do  not  involve  changes  in  action \\npotential morphology, and do not appear to contribute to the immediate increase in contractility with \\nstretch. We find that, in isolated heart preparations and isolated tissues, it is difficult to demonstrate \\nMEF at levels of tissue stretch comparable to those normally encountered by the myocardium. This \\nresult reflects the inconsistencies in the literature; some studies have demonstrated shortening of the \\naction potential with stretch [10,12,29], an effect also seen in human heart [35], while others report \\nno effect [15,17] or a \\u201ccrossover\\u201d effect [17,26]: see reviews [28] and [11]. Similar inconsistencies \\ncan  be  seen  in  experiments  examining  MEF  at  the  cellular  level  [30].  This  lack  of  clear  MEF  at \\n\\u201cphysiological\\u201d levels of stretch is a perplexing result, since there is no doubt that stretch activated \\nchannels are present in the membrane of both ventricular and atrial cardiac myocytes [16,23,32,36].   \\nThis failure to demonstrate MEF in our experiments may be related to the observation that it appears \\nto  be  necessary  to  rapidly  stretch  the  myocardium  in  order  to  elicit  obvious  electrophysiological \\nchanges (figure 2B). This suggests that the rate of stretch is more important than the degree of stretch \\nper se, as suggested by Calkins, Levine and Kass [9], perhaps due to the visco-elastic properties of \\nthe  myocardium,  or  perhaps  due  to  MEF  being  elicited  by  sheer  stress  between  planes  of  muscle \\nfibres [25]. More intriguingly, ectopic beats can be elicited by rapid release of the myocardium from \\nstretch  (figure  2C),  and  observation  made  some  time  ago  [18]  but  still  not  fully  understood.  The \\nmechanism may be related to the dissociation of calcium from the unloaded myofilaments, due to the \\ntension-dependence  of  the  affinity  of  troponin  for  calcium  [2],  the  consequent  rise  in  intracellular \\ncalcium  accelerating  sodium-calcium  exchange  or  other  electrogenic  calcium  dependent  processes \\nwhich in turn result in afterdepolarisations [22,23]. If such changes in calcium handling are occurring \\nat physiological levels of stretch, they are not prominent enough to alter action potential morphology, \\nas might be predicted by effects on calcium dependent channels or the sodium-calcium exchanger. \\n\\n5.  Conclusion \\n\\nThere are other possible reasons why we may fail to see MEF in these experiments. MEF may \\nbe a labile process, subject to continuous modulation by hormonal or biochemical influences [5,37] \\nand,  in  the  process  of  preparing  the  isolated  heart  or  tissues,  this  influence  is  lost.    Perhaps  more \\nimportantly, MEF is not uniform throughout the myocardium. It has been shown that the distribution \\nof stretch activated channels is heterogeneous in the myocardium [34,36] and that the distribution of \\nstrain  is  also  inhomogeneous  [27].  Consistent  with  this,  we  have  previously  shown  in  rat  ventricle \\nthat  sub-epicardial  action  potentials  (recorded  with  a  different  electrode  technique)  are  much  less \\naffected by stretch than are sub-endocardial [19]. One might also expect similar regional disparities \\nbetween apex and base [10]. Hence, it may be that our recordings do not reveal AP changes because \\nwe are recording from the epicardial surface near the equator, whereas more obvious changes may be \\noccurring  elsewhere  in  the  myocardium.  These  considerations  also  apply  to  guinea-pig  heart, \\nalthough the subtleties of MEF in this species is less well understood. \\n\\nAcknowledgments \\n\\nSupported by grants from the National Health and Medical Research Council of Australia.   \\n\\nReferences \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n\\f28\\u00a0\\n\\n1.  Allen DG, Kentish JC (1985) The cellular basis of the length-tension relation in cardiac muscle. J \\n\\nMol Cell Cardiol 17: 821\\u2013840. \\n\\n2.  Babu A, Sonnenblick E, Gulati J (1988) Molecular basis for the influence of muscle length on \\n\\nmyocardial performance. Science 240: 74\\u201376. \\n\\n3.  Bainbridge  FA  (1915)  The  influence  of  venous  filling  upon  the  rate  of  the  heart.  J  Physiol  50: \\n\\n65\\u201384. \\n\\n4.  Blinks  JR  (1956)  Positive  chronotropic  effect  of  increasing  right  atrial  pressure  in  the  isolated \\n\\nmammalian heart. Am J Physiol 186: 299\\u2013303. \\n\\n5.  Bockenhauer  D,  Zilberberg  N,  Goldstein  SA  (2001)  KCNK2:  reversible  conversion  of  a \\n\\nhippocampal potassium leak into a voltage-dependent channel. Nat Neurosci 4: 486\\u2013491. \\n\\n6.  Bustamante JO, Ruknudin A, Sachs F (1991) Stretch-activated channels in heart cells: relevance \\n\\nto cardiac hypertrophy. J Cardiovasc Pharmacol 17 Suppl 2: S110\\u2013113. \\n\\n7.  Calaghan  SC,  Belus  A,  White  E  (2003)  Do  stretch-induced  changes  in  intracellular  calcium \\n\\nmodify the electrical activity of cardiac muscle? Prog Biophys Mol Biol 82: 81\\u201395. \\n\\n8.  Calaghan  SC,  Le  Guennec  JY,  White  E  (2004)  Cytoskeletal  modulation  of  electrical  and \\n\\nmechanical activity in cardiac myocytes. Prog Biophys Mol Biol 84: 29\\u201359. \\n\\n9.  Calkins H, Levine JH, Kass DA  (1991) Electrophysiological effect of varied rate and  extent of \\n\\nacute in vivo left ventricular load increase. Cardiovasc Res 25: 637\\u2013644. \\n\\n10. Dean JW, Lab MJ (1990) Regional changes in ventricular excitability during load manipulation \\n\\nof the in situ pig heart. J Physiol 429: 387\\u2013400. \\n\\n11. Franz  MR  (1996)  Mechano-electrical  feedback  in  ventricular  myocardium.  Cardiovasc  Res  32: \\n\\n15\\u20134. \\n\\n12. Franz MR, Cima R, Wang D, et al. (1992) Electrophysiological effects of myocardial stretch and \\n\\nmechanical determinants of stretch-activated arrhythmias. Circulation 86: 968\\u2013978. \\n\\n13. Fuchs  F,  Martyn  DA  (2005)  Length-dependent  Ca(2+)  activation  in  cardiac  muscle:  some \\n\\nremaining questions. J Muscle Res Cell Motil 26: 199\\u2013212. \\n\\n14. Hansen  DE,  Craig  CS,  Hondeghem  LM  (1990)  Stretch-induced  arrhythmias  in  the  isolated \\ncanine  ventricle.  Evidence  for  the  importance  of  mechanoelectrical  feedback.  Circulation  81: \\n1094\\u20131105. \\n\\n15. Hennekes R, Kaufmann R, Lab M (1981) The dependence of cardiac membrane excitation and \\ncontractile ability on active muscle shortening (cat papillary muscle). Pflugers Arch 392: 22\\u201328. \\n16. Hu  H,  Sachs  F  (1997)  Stretch-activated  ion  channels  in  the  heart.  J  Mol  Cell  Cardiol  29: \\n\\n1511\\u20131523. \\n\\n17. Kamkin A, Kiseleva I, Wagner KD, et al. (2000) Mechano-electric feedback in right atrium after \\n\\nleft ventricular infarction in rats. J Mol Cell Cardiol 32: 465\\u2013477. \\n\\n18. Kaufmann  RL,  Lab  MJ,  Hennekes  R,  et  al.  (1971)  Feedback  interaction  of  mechanical  and \\nelectrical  events  in  the  isolated  mammalian  ventricular  myocardium  (cat  papillary  muscle). \\nPflugers Arch 324: 100\\u2013123. \\n\\n19. Kelly D, Mackenzie L, Hunter P, et al. (2006) Gene expression of stretch-activated channels and \\n\\nmechanoelectric feedback in the heart. Clin Exp Pharmacol Physiol 33: 642\\u2013648. \\n\\n20. Kiseleva I, Kamkin A, Wagner KD, et al. (2000) Mechanoelectric feedback after left ventricular \\n\\ninfarction in rats. Cardiovasc Res 45: 370\\u2013378. \\n\\n21. Kohl P, Bollensdorff C, Garny A (2006) Effects of mechanosensitive ion channels on ventricular \\n\\nelectrophysiology: experimental and theoretical models. Exp Physiol 91: 307\\u2013321. \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n\\f29\\u00a0\\n\\n22. Lab MJ (1999) Mechanosensitivity as an integrative system in heart: an audit. Prog Biophys Mol \\n\\nBiol 71: 7\\u201327. \\n\\n23. Lab  MJ,  Allen  DG,  Orchard  CH  (1984)  The  effects  of  shortening  on  myoplasmic  calcium \\nconcentration  and  on  the  action  potential  in  mammalian  ventricular  muscle.  Circ  Res  55: \\n825\\u2013829. \\n\\n24. Lange G, Lu HH, Chang A, et al. (1966) Effect of stretch on the isolated cat sinoatrial node. Am J \\n\\nPhysiol 211: 1192\\u20131196. \\n\\n25. Legrice IJ, Hunter PJ, Smaill BH (1997) Laminar structure of the heart: a mathematical model. \\n\\nAm J Physiol 272: H2466\\u20132476. \\n\\n26.Nazir SA, Lab MJ (1996) Mechanoelectric feedback in the atrium of the isolated guinea-pig heart. \\n\\nCardiovasc Res 32: 112\\u2013119. \\n\\n27. Nickerson  D,  Niederer  S,  Stevens  C,  et  al.  (2006)  A  computational  model  of  cardiac \\n\\nelectromechanics. Conf Proc IEEE Eng Med Biol Soc 1: 5311\\u20135314. \\n\\n28. Ravens  U  (2003)  Mechano-electric  feedback  and  arrhythmias.  Prog  Biophys  Mol  Biol  82: \\n\\n255\\u2013266. \\n\\n29. Reiter  MJ,  Synhorst  DP,  Mann  DE  (1988)  Electrophysiological  effects  of  acute  ventricular \\n\\ndilatation in the isolated rabbit heart. Circ Res 62: 554\\u2013562. \\n\\n30. Riemer  TL,  Tung  L  (2003)  Stretch-induced  excitation  and  action  potential  changes  of  single \\n\\ncardiac cells. Prog Biophys Mol Biol 82: 97\\u2013110. \\n\\n31. Rozenberg  S,  Tavernier  B,  Riou  B,  et  al.  (2006)  Severe  impairment  of  ventricular  compliance \\naccounts  for  advanced  age-associated  hemodynamic  dysfunction  in  rats.  Exp  Gerontol  41: \\n289\\u2013295. \\n\\n32. Sackin H (1995) Mechanosensitive channels. Annu Rev Physiol 57: 333\\u2013353. \\n33. Stacy  GP,  Jr,  Jobe  RL,  Taylor  LK,  et  al.  (1992)  Stretch-induced  depolarizations  as  a  trigger  of \\n\\narrhythmias in isolated canine left ventricles. Am J Physiol 263: H613\\u2013621. \\n\\n34. Stones  R,  Calaghan  SC,  Billeter  R,  et  al.  (2007)  Transmural  variations  in  gene  expression  of \\n\\nstretch-modulated proteins in the rat left ventricle. Pflugers Arch 454: 545\\u2013549. \\n\\n35. Taggart P (1996) Mechano-electric feedback in the human heart. Cardiovasc Res 32: 38\\u201343. \\n36. Tan  JH,  Liu  W,  Saint  DA  (2004)  Differential  expression  of  the  mechanosensitive  potassium \\nchannel  TREK-1  in  epicardial  and  endocardial  myocytes  in  rat  ventricle.  Exp  Physiol  89: \\n237\\u2013242. \\n\\n37. Tan JH, Liu W, Saint DA (2002) Trek-like potassium channels in rat cardiac ventricular myocytes \\n\\nare activated by intracellular ATP. J Membr Biol 185: 201\\u2013207. \\n\\n38. Tavi  P,  Han  C,  Weckstrom  M  (1998)  Mechanisms  of  stretch-induced  changes  in  [Ca2+]i  in  rat \\natrial myocytes: role of increased troponin C affinity and stretch-activated ion channels. Circ Res \\n83: 1165\\u20131177. \\n\\n39. Todaka  K,  Ogino  K,  Gu  A,  et  al.  (1998)  Effect  of  ventricular  stretch  on  contractile  strength, \\n\\ncalcium transient, and cAMP in intact canine hearts. Am J Physiol 274: H990\\u20131000. \\n\\n40. Volk T, Nguyen TH, Schultz JH, et al. (1999) Relationship between transient outward K+ current \\nand  Ca2+  influx  in  rat  cardiac  myocytes  of  endo-  and  epicardial  origin.  J  Physiol  519  Pt  3: \\n841\\u2013850. \\n\\n41. White  E,  Boyett  MR,  Orchard  CH  (1995)  The  effects  of  mechanical  loading  and  changes  of \\n\\nlength on single guinea-pig ventricular myocytes. J Physiol 482 ( Pt 1): 93\\u2013107. \\n\\n42. Wilhelm J, Kondratev D, Christ A, et al. (2006) Stretch induced accumulation of total Ca and Na \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n\\fin cytosol and nucleus: a comparison between cardiac trabeculae and isolated myocytes. Can J \\nPhysiol Pharmacol 84: 487\\u2013498. \\n\\n43. Zabel  M,  Koller  BS,  Sachs  F,  et  al.  (1996)  Stretch-induced  voltage  changes  in  the  isolated \\nbeating  heart:  importance  of  the  timing  of  stretch  and  implications  for  stretch-activated  ion \\nchannels. Cardiovasc Res 32: 120\\u2013130. \\n\\n30\\u00a0\\n\\n\\u00a9  2014,  David  A.  Saint,  et  al.  licensee  AIMS  Press.  This  is  an  open  access  article  distributed \\nunder \\nLicense \\nCreative \\nterms \\n(http://creativecommons.org/licenses/by/4.0) \\n\\nAttribution \\n\\nCommons \\n\\nthe \\n\\nthe \\n\\nof \\n\\nAIMS Biophysics \\n\\nVolume 1, Issue 1, 16-30. \\n\\n \\n\\f\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"branch\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 251,\n        \"samples\": [\n          \"BIOPHYSICS 20\",\n          \"HYDROBIOLOGY 30\",\n          \"GENETICS 28\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Немного разбираемся с регулярками"
      ],
      "metadata": {
        "id": "h40iqyipwiFY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "re.findall(r'\\[[0-9 ,– -]+\\]', bio[\"text\"].iloc[1])\n",
        "# [0-9 ,– -] -- набор символов, который мы ищем\n",
        "# [0-9 ,– -]+ -- любой из символов встречается хотя бы 1 раз (\"abcd6 9-,1\" -> \"6 9-,1\")\n",
        "# \\[[0-9 ,– -]+\\] -- ищем строчки из п.2 в квадратных скобках, \\ -- символ экранирования"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "9mKIZ7QNlPOD",
        "outputId": "281bf5f8-6276-419e-9e5c-4143504bff0a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['[1,2]',\n",
              " '[3]',\n",
              " '[4]',\n",
              " '[5]',\n",
              " '[6]',\n",
              " '[6]',\n",
              " '[7]',\n",
              " '[1]',\n",
              " '[8]',\n",
              " '[1]',\n",
              " '[6]',\n",
              " '[1]',\n",
              " '[6]',\n",
              " '[9]',\n",
              " '[10]',\n",
              " '[2]',\n",
              " '[11]',\n",
              " '[2]',\n",
              " '[10]',\n",
              " '[4]',\n",
              " '[4]',\n",
              " '[4]',\n",
              " '[12]',\n",
              " '[13]',\n",
              " '[14]',\n",
              " '[3]',\n",
              " '[15]',\n",
              " '[3]',\n",
              " '[16]',\n",
              " '[17]',\n",
              " '[18]',\n",
              " '[19]',\n",
              " '[20]',\n",
              " '[4]',\n",
              " '[21,22]',\n",
              " '[21]',\n",
              " '[23]',\n",
              " '[24]',\n",
              " '[24]',\n",
              " '[25]',\n",
              " '[5]',\n",
              " '[26]',\n",
              " '[27]',\n",
              " '[28]',\n",
              " '[29]',\n",
              " '[30]',\n",
              " '[31]',\n",
              " '[32]',\n",
              " '[5]',\n",
              " '[33]',\n",
              " '[34]',\n",
              " '[34]',\n",
              " '[5,35]',\n",
              " '[36,37]',\n",
              " '[38,39]',\n",
              " '[40,41]',\n",
              " '[30]',\n",
              " '[42]',\n",
              " '[43]',\n",
              " '[44]',\n",
              " '[45]',\n",
              " '[44,46]',\n",
              " '[47]',\n",
              " '[48]',\n",
              " '[49]',\n",
              " '[50]',\n",
              " '[51]',\n",
              " '[47,52]',\n",
              " '[47]',\n",
              " '[53,54]',\n",
              " '[47,55–57]',\n",
              " '[58,59]',\n",
              " '[60]',\n",
              " '[61,62]',\n",
              " '[61,62]',\n",
              " '[63,64]',\n",
              " '[57]',\n",
              " '[65]',\n",
              " '[66–68]',\n",
              " '[69,70]',\n",
              " '[71]',\n",
              " '[72]',\n",
              " '[73]',\n",
              " '[72]',\n",
              " '[74]',\n",
              " '[75,76]',\n",
              " '[63,64]',\n",
              " '[77]',\n",
              " '[78]',\n",
              " '[79]',\n",
              " '[80]',\n",
              " '[81]',\n",
              " '[63]',\n",
              " '[82]',\n",
              " '[82,83]',\n",
              " '[83–85]',\n",
              " '[85]',\n",
              " '[86,87]',\n",
              " '[88]',\n",
              " '[89]',\n",
              " '[88]',\n",
              " '[90]',\n",
              " '[90]',\n",
              " '[79]',\n",
              " '[91]',\n",
              " '[90]',\n",
              " '[90]',\n",
              " '[92]',\n",
              " '[92]',\n",
              " '[91]',\n",
              " '[93]',\n",
              " '[87]',\n",
              " '[93]',\n",
              " '[91]',\n",
              " '[91]',\n",
              " '[35,94]',\n",
              " '[35]',\n",
              " '[35,83]',\n",
              " '[27]',\n",
              " '[95]',\n",
              " '[83]',\n",
              " '[96]',\n",
              " '[97]',\n",
              " '[83]',\n",
              " '[35]',\n",
              " '[96]',\n",
              " '[79,80,83,85,92,93,96,98,99]',\n",
              " '[98]',\n",
              " '[85]',\n",
              " '[92]',\n",
              " '[79]',\n",
              " '[96]',\n",
              " '[96]',\n",
              " '[35,94]',\n",
              " '[94,100]',\n",
              " '[94]',\n",
              " '[100]',\n",
              " '[27,28,101]',\n",
              " '[100]',\n",
              " '[102]',\n",
              " '[103,104]',\n",
              " '[105]',\n",
              " '[106]',\n",
              " '[107]',\n",
              " '[108]',\n",
              " '[109,110]',\n",
              " '[93]',\n",
              " '[96]',\n",
              " '[96]',\n",
              " '[96]',\n",
              " '[96]',\n",
              " '[96]',\n",
              " '[27]',\n",
              " '[84,105]',\n",
              " '[111]',\n",
              " '[112]',\n",
              " '[113–115]',\n",
              " '[116]',\n",
              " '[48,117,118]',\n",
              " '[113]',\n",
              " '[91]',\n",
              " '[116]',\n",
              " '[50]',\n",
              " '[119]',\n",
              " '[120]',\n",
              " '[61,62]',\n",
              " '[121]',\n",
              " '[90]',\n",
              " '[122]',\n",
              " '[90]',\n",
              " '[85]',\n",
              " '[122]',\n",
              " '[90]',\n",
              " '[90]',\n",
              " '[35]',\n",
              " '[92]',\n",
              " '[80]',\n",
              " '[123]',\n",
              " '[124]',\n",
              " '[32]',\n",
              " '[125]',\n",
              " '[123]',\n",
              " '[35]',\n",
              " '[126]',\n",
              " '[126]',\n",
              " '[90]',\n",
              " '[127]',\n",
              " '[127]',\n",
              " '[98]',\n",
              " '[98]',\n",
              " '[99]',\n",
              " '[85]',\n",
              " '[85]',\n",
              " '[83]',\n",
              " '[83]',\n",
              " '[35]',\n",
              " '[35]',\n",
              " '[35]',\n",
              " '[92]',\n",
              " '[121]',\n",
              " '[93]',\n",
              " '[93]',\n",
              " '[93]',\n",
              " '[79]',\n",
              " '[79]',\n",
              " '[91]',\n",
              " '[91]',\n",
              " '[80]',\n",
              " '[96]']"
            ]
          },
          "metadata": {},
          "execution_count": 73
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for i in range(len(bio)):\n",
        "    if len(re.findall(r'https.*', bio['text'].iloc[i])) > 0:\n",
        "        print(re.findall(r'https.*', bio['text'].iloc[i]))\n",
        "\n",
        "# https. -- ищем подстроку \"https{любой символ кроме начала строки}\"\n",
        "# https.* -- ищем подстроку \"https{все символы кроме начала строки}\"\n",
        "# https.*$ -- ищем подстроку \"https{все символы кроме начала строки}\",  которая заканчивается символом конца строки\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "OWz8wUBXnj2h",
        "outputId": "1cf3b131-7655-4a76-9782-ddad19cd887b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['https://doi.org/10.1007/s10750-019-3939-2 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-019-3939-2) con-']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201707168', 'https ://doi .org /10 .1093 /molehr /gap092', 'https ://doi .org /10 .1093 /hmg /ddp483', 'https ://doi .org ', 'https ://doi .org /10 .1042 /BST20110609', 'https ://doi .org /10 .1016 /j .exger .2012 .09 .009', 'https ://doi ', 'https ://doi .org /10 .1126 /science .1065768', 'https ://doi .org /10 .1016 /', 'https ://doi .org /10 .1111 /j .1474 -9726 .2007 .00348 .x', 'https ://doi .org /10 .1016 /j .cell .2006 .01 .039', 'https ://doi ', 'https ://doi .org /10 .1126 /science .1078223', 'https ://doi .org /10 ', 'https ://doi .org /10 .18632 /', 'https ://doi ', 'https ://doi .org ', 'https ', 'https ://doi .org /10 .1016 /S0092 -8674(00)80595 ', 'https ://doi .org /10 .1126 /science .289 .5487 .2122', 'https ://doi .org /10 .1126 /science .277 .5322 .99', 'https ://doi .org /10 ', 'https ://doi .org /10 .1038 /35030105', 'https ://doi .org /10 ', 'https ://doi .org /10 .1046 /j .1474 -9728 .2003 .00049 .x', 'https ://doi .org /10 .1126 /science ', 'https ://doi ', 'https ://doi .org /10 .1074 /jbc .M114 ', 'https ://doi ', 'https ://doi .org /10 .1126 /science .1057991', 'https ://doi .org /10 .1017 ', 'https ://doi .org /10 ', 'https ://doi .org ', 'https ://doi .org ', 'https ://doi .org /10 .1016 /S1047 ', 'https ://doi .org /10 .1002 /mrd .22806', 'https ://doi ', 'https ://doi .org /10 .1038 /', 'https ://doi .org /10 .1371 /journal .pgen .1004225', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1073 /pnas .0709128105', 'https ://doi .org /10 ', 'https ://', 'https ://doi .org /10 .1016 ', 'https ://doi .org /10 .2174 /157339913804143225', 'https ://doi .org /10 .1242 /bio ', 'https ://doi .org /10 .1016 /j .fertnstert .2015 .03 .030', 'https ://doi .org /10 ', 'https ://doi .org /10 .1126 /', 'https ://doi .org /10 ', 'https ://doi .org /10 .1126 /science .1098219', 'https ://doi .org /10 .1101 /gad .13 .11 .1422', 'https ://doi .org /10 .1038 /nature08619', 'https ://doi .org /10 .1016 /j .cub .2007 .08 .047', 'https ://doi .org /10 .1074 /jbc .M705325200', 'https ://doi .org /10 .1371 /', 'https ', 'https ://doi .org /10 .1093 /hmg /ddh241', 'https ://doi .org /10 .1111 /j ', 'https ://doi .org /10 .1038 /nrm3311', 'https ://doi .org /10 .1016 /j .tcb .2015 .10 .013', 'https ://doi ', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1038 ', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1126 /science .1083701', 'https ://doi .org /10 ', 'https ://doi .org /10 .1371 /journal .pgen .0030025', 'https ://doi .org /10 .1038 /nature02549', 'https ://doi .org /10 .1073 /pnas .0710787105', 'https ://doi .org /10 .1038 /ncb839', 'https ://doi .org /10 ', 'https ://doi .org /10 .1016 /j .annepidem .2014 .12 .002', 'https ://doi .org /10 ', 'https ://doi .org ', 'https ', 'https ://doi .org /10 .1016 /j .cub ', 'https ://doi .org /10 .1371 ', 'https ', 'https ', 'https ://doi .org /10 .1038 /ncb2152', 'https ://doi .org /10 .1016 /0047 -6374(77)90043 -4', 'https ://doi .org /10 .1016 /j .exger .2004 .05 ', 'https ://doi .org /10 .1242 /dev .074047', 'https ://doi .org /10 .1126 /science .1111410', 'https ://', 'https ://doi .org /10 .1073 /pnas .95 .22 .13091', 'https ://doi .org /10 ', 'https ://doi .org /10 .1016 /j .cell .2012 .03 .017', 'https ://doi .org /10 ', 'https ://doi .org /10 .1093 /hmg /ddp459', 'https ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1038 ', 'https ://', 'https ://doi .org /10 .1016 /j ', 'https ://doi .org ', 'https ://doi .org /10 .1111 /j .1558 -5646 .1984 .tb00369 .x', 'https ://doi .org /10 .1186 /1471 -2350 -8 ', 'https ://doi .org /10 .1371 /journal .pgen .1000789', 'https ://doi .org /10 ', 'https ', 'https ://doi .org ', 'https ://doi .org /10 .1016 /j .cub .2008 .06 .059', 'https ://doi .org /10 .1016 /S1097 -2765(02)00568 -3', 'https ://doi .org /10 .1038 /nature11432', 'https ://doi .org /10 .1095 /', 'https ://', 'https ://doi .org ', 'https ://doi .org /10 .1111 /acel .12610', 'https ://doi .org /10 .1242 /dev .042523', 'https ://doi .org /10 .1016 /j .cub .2012 .06 .036', 'https ://doi .org /10 .1038 /', 'https ://doi .org /10 .1038 /38148', 'https ://doi .org /10 .1101 /gad .867301', 'https ://doi .org /10 .1038 ', 'https ://doi .org /10 .1016 /j ', 'https ://doi .org /10 .1038 /ncomms8107', 'https ://doi .org /10 .1038 /nature01789', 'https ://doi .org /10 .1089 /rej .2005 .8 .86', 'https ://doi .org /10 ', 'https ://doi .org /10 .1126 /science .1152257', 'https ://doi .org /10 .1111 /j .1474 -9726 ', 'https ://doi .org /10 .1242 /dev ', 'https ://doi .org /10 .1093 /hmg /ddp217', 'https ://doi .org /10 ', 'https ://doi .org /10 .1038 /nature07536', 'https ://doi .org /10 .1016 /j .mce .2014 .09 .027', 'https ', 'https ://doi ', 'https ://doi .org /10 .1038 /40194', 'https ://doi .org /10 .1126 /', 'https ://doi .org /10 .1016 /j .cmet .2012 .02 .001', 'https ://doi .org /10 .1111 ', 'https ://doi .org /10 .1016 /j .cell .2005 .01 ', 'https ://doi .org /10 .1210 /en .2009 -0096', 'https ://doi .org /10 .1016 /j .cmet .2012 .04 .007', 'https ://doi .org /10 .1038 /287141a0', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1016 /j .arr .2011 .12 .005', 'https ://', 'https ://doi .org /10 .1016 /j .cell .2017 .02 .004', 'https ://doi .org /10 .1111 /j .1474 -9726 .2008 .00409 .x', 'https ://doi .org /10 .1111 /j .1474 -9726 .2009 .00493 .x', 'https ://doi .org /10 ', 'https ://doi .org /10 .1126 /science .1177221', 'https ://doi .org /10 .1016 /j .cub .2007 .08 .058', 'https ://doi .org /10 .1111 /j .1474 -9726 .2011 .00707 .x', 'https ://doi .org /10 .1016 /j .cell .2015 .06 .023', 'https ://doi .org /10 .1016 /j .physbeh .2014 .05 ', 'https ://doi .org /10 .1016 /j .cmet .2014 .02 .009', 'https ://doi .org /10 .1073 /pnas .1422041112', 'https ://doi .org /10 .1016 /j .cmet .2012 .12 ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1073 /pnas .0705467105', 'https ://doi .org /10 .1895 /wormbook .1 .80 .2', 'https ://doi .org /10 .1371 /journal .pone .0022640', 'https ://doi .org /10 .1126 /science .1142179', 'https ://doi ', 'https ://doi .org /10 .1038 /nrm1837', 'https ://doi .org ', 'https ://doi .org /10 .1016 /j .celrep .2017 .06 .048', 'https ://doi .org /10 .1016 /j .cell .2013 .07 .006', 'https ://doi .org /10 .1093 /', 'https ://doi .org /10 .1210 /en .2005 -0301', 'https ://doi .org /10 .1016 /j .cell .2008 .01 .030', 'https ://doi .org /10 .1016 /j .celrep .2014 .08 .006', 'https ://doi .org /10 .1038 /ncb1547', 'https ', 'https ://doi .org /10 .1038 /426620a', 'https ://doi .org ', 'https ://doi .org ', 'https ://doi .org /10 .1016 /j .rbmo ', 'https ://doi .org /10 .1016 /j .fertnstert .2006 .03 .054', 'https ://doi .org /10 .1111 /acel .12479', 'https ://doi .org /10 .1080 ', 'https ://doi .org /10 .1073 /pnas .0801030105', 'https ://doi .org /10 .1016 /j .cub .2014 .09 ', 'https ://', 'https ://doi .org /10 ']\n",
            "['https://doi.org/10.1186/s12863-019-0718-z', 'https://grenada.-', 'https://grenada.lumc.nl/LSDB_list/', 'https://www.gwascentral.org) is also available,', 'https://databases.lovd.nl/shared/genes/NR3C1), Lei-', 'https://www.ncbi.nlm.nih.gov/books/NBK269031/. Accessed 3 Mar 2017.', 'https://www.ncbi.nlm.nih.gov/books/NBK3848/.']\n",
            "['https:// creativecommons .org/ licenses/ by/ 4 .0/ ).', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https://doi.org/10.1083/jcb.201808014', 'https:// doi .org/ 10 .1016/ S0092 -8674(01)00410 -X', 'https:// doi .org/ 10 .1038/ nature19083', 'https:// doi .org/ 10 .1083/ ', 'https:// doi .org/ 10 .1083/ jcb .201008108', 'https:// doi .org/ 10 .1128/ MCB .06525 -11', 'https:// doi .org/ 10 .1038/ nprot .2009 .21', 'https:// doi .org/ 10 .1073/ pnas .0914191107', 'https:// doi .org/ 10 .1091/ mbc .e09 -07 -0643', 'https:// doi .org/ 10 .1016/ j .cell .2006 ', 'https:// doi .org/ 10 .1021/ jacs .7b09368', 'https:// doi .org/ 10 .1038/ 10049', 'https:// doi .org/ 10 .1038/ ncb2855', 'https:// doi .org/ 10 .1021/ ', 'https:// ', 'https:// doi .org/ 10 .1083/ jcb .201606033', 'https:// doi .org/ 10 .1101/ gad .267603', 'https:// doi .org/ 10 .1083/ jcb .201205119', 'https:// doi .org/ 10 .1016/ j .cell .2006 ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1371/ ', 'https:// doi .org/ 10 .1073/ pnas .91 ', 'https:// doi .org/ 10 .1371/ journal .pone .0018793', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1083/ jcb .201406109', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1038/ nrm3494', 'https:// doi .org/ 10 .1038/ ', 'https:// doi .org/ 10 .1083/ jcb .201301006', 'https:// doi .org/ 10 .1126/ science .1197048', 'https:// ', 'https:// doi .org/ 10 .1073/ pnas ', 'https:// doi .org/ 10 .1038/ ncb3634', 'https://doi.org/10.1083/jcb.201808014', 'https:// doi .org/ 10 .1126/ science .1163300', 'https:// doi .org/ 10 .1126/ science ', 'https:// doi .org/ 10 .1016/ j .cub .2009 .02 ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1083/ jcb .201003038', 'https:// doi .org/ 10 .1126/ science .aaa4029', 'https:// doi .org/ 10 .7554/ eLife .22513', 'https:// doi ', 'https:// doi .org/ 10 .1038/ ncb3065', 'https:// doi .org/ 10 .1038/ ncb0707 -724', 'https:// doi .org/ 10 .1038/ 344503a0', 'https:// doi .org/ 10 .7554/ eLife .01030', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1016/ j .cub .2014 .01 .033', 'https:// doi .org/ 10 .1016/ j .celrep .2014 .02 .034', 'https:// doi .org/ ', 'https:// doi .org/ 10 .1016/ j .cub .2014 .02 .020', 'https:// doi .org/ 10 .1242/ jcs .122481', 'https:// doi .org/ 10 .1073/ ', 'https:// doi .org/ 10 .1038/ emboj ', 'https:// doi .org/ 10 .1016/ j .tcb .2010 .10 .007', 'https:// doi .org/ 10 .1146/ annurev -biochem -061611 ', 'https:// doi .org/ 10 .1091/ ', 'https:// doi .org/ 10 .1016/ j .cub .2012 .03 .052', 'https:// doi .org/ 10 .1016/ j .ccr .2010 .10 .028', 'https:// doi .org/ 10 .1126/ science .1195689', 'https:// doi .org/ 10 .1016/ j .cub .2011 .12 .048', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1038/ nmeth .2019', 'https:// doi .org/ 10 .1016/ ', 'https:// doi .org/ 10 .1038/ nrm3934', 'https:// doi .org/ 10 .1016/ j .celrep .2016 .01 .045', 'https:// doi .org/ 10 .1093/ emboj/ 21 .7 ', 'https:// doi .org/ 10 .1007/ s00412 -004 -0288 -2', 'https:// doi ', 'https:// doi .org/ 10 .1083/ jcb .200712028', 'https:// doi .org/ 10 .1038/ nature09390', 'https:// doi .org/ 10 .1016/ j .cub .2010 ', 'https:// doi .org/ 10 .1042/ ', 'https:// doi .org/ 10 .1038/ 35048096', 'https://doi.org/10.1083/jcb.201808014', 'https:// doi .org/ 10 .1016/ j .cub .2014 .01 ', 'https:// doi ', 'https:// doi .org/ 10 .1128/ MCB .00753 -10', 'https:// doi .org/ 10 .1016/ j .febslet .2015 .02 ', 'https:// doi .org/ 10 .4161/ cc .9 .17 .12832', 'https:// doi .org/ 10 .1016/ j .molcel .2010 .02 .034', 'https:// doi .org/ 10 .7554/ eLife .01695', 'https:// doi .org/ 10 .1038/ ncb2515', 'https:// doi .org/ 10 .1016/ j .molcel .2016 .07 .003', 'https:// doi .org/ 10 .1016/ j .molcel .2006 .02 .022', 'https:// doi .org/ ', 'https://doi.org/10.1083/jcb.201808014']\n",
            "['https://github.com/joey711/phyloseq).']\n",
            "['https://doi.org/10.1083/jcb.201705214', 'https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https ://doi .org /10 .1016 /j .semcdb .2009 .03 .007', 'https ://doi .org /10 .1038 /nrm2999', 'https ://doi .org /10 .1016 /j .cell .2013 .01 .030', 'https ://doi .org /10 ', 'https ://doi .org /10 ', 'https ://doi .org /10 ', 'https ://doi .org /10 .4161 ', 'https ://doi .org /10 .1111 /tra .12095', 'https ://doi .org /10 .1038 /', 'https ://doi .org /10 ', 'https ://doi .org /10 .1091 /mbc .E14 -04 -0917', 'https ://doi .org /10 .1111 /j .1600 -0854 .2011 .01181 .x', 'https ://doi .org /10 .1534 /genetics ', 'https ://doi .org /10 .1016 /j .devcel .2017 .03 .015', 'https ://doi .org /10 .1016 ', 'https ://doi .org /10 .1016 /j .molcel .2008 .07 .020', 'https ://doi .org /10 .1093 /', 'https ://doi .org /10 .1074 /jbc .M112 .411454', 'https ', 'https ://doi .org /10 .3389 /fcell ', 'https ://doi .org /10 .1083 /jcb .200901145', 'https ://doi .org /10 ', 'https ://doi .org /10 .1091 ', 'https ://doi .org /10 .4161 /21592780 .2014 .954870', 'https ://doi .org /10 ', 'https ://doi .org /10 .1091 /mbc .E13 -05 -0269', 'https ://doi .org /10 .1534 /genetics .116 .194910', 'https ://doi .org /10 ', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1016 /j .devcel .2014 .07 .016', 'https ://doi .org /10 .1074 /jbc ', 'https ://doi .org /10 .1016 ', 'https ://doi .org /10 .1111 /j ', 'https ://doi .org /10 .1038 /ncb1489', 'https ://doi .org /10 .1042 /', 'https ://doi .org /10 .1083 /jcb .115 .1 .31', 'https ://doi .org /10 .1016 /j .devcel .2012 .02 .006', 'https ://doi .org /10 .1073 /pnas ', 'https ://doi .org /10 .1093 /', 'https ://doi .org /10 .1016 /', 'https ', 'https ://doi .org /10 .1016 /j .cell .2008 .06 .025', 'https ://doi .org /10 .1111 /j .1600 -0854 .2010 .01031 .x', 'https ://doi .org /10 .1091 /mbc .E10 -11 -0873', 'https ://doi .org /10 .1242 /jcs .131318', 'https ://doi .org /10 .1016 /j .febslet .2013 .05 .015', 'https ://doi .org /10 .1038 /nrm2728', 'https ://doi .org /10 .1073 /pnas .1308627110', 'https ://doi .org /10 .1083 /jcb .200111081', 'https ://doi .org /10 .1073 /pnas .1316356110', 'https ://doi .org /10 .1083 /jcb .201608123', 'https ://doi .org /10 .1083 /jcb .119 .4 .749', 'https ://doi .org /10 .1016 ', 'https ://doi .org ', 'https ://doi .org /10 ', 'https ://doi .org /10 ', 'https ://doi .org ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1083 /jcb ', 'https ://doi .org /10 .1083 /jcb .200603044', 'https ://doi .org /10 .1038 /srep43207', 'https ://doi .org /10 .1111 ', 'https ://doi .org /10 ', 'https ://doi .org /10 ']\n",
            "['https://labs.genetics.ucla.edu/horvath/dnamage/).']\n",
            "['https://doi.org/10.1155/2017/7315351', 'https://www.hindawi.com']\n",
            "['https://zenodo.org/record/45167,', 'https://zenodo.org/record/45168, https://zenodo.org/record/', 'https://zenodo.org/record/45196, https://zenodo.']\n",
            "['https://doi.org/10.1007/s10750-019-3918-7 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-019-3918-7) con-', 'https://toxics.usgs.gov/hypoxia/', 'https://data.gulfresearchinitiative.orgdata/R4.', 'https://doi.org/10.3133/', 'https://doi.org/10.1890/100008.', 'https://', 'https://doi.', 'https://doi.org/10.1007/s12237-016-', 'https://doi.org/10.3354/ame01607.']\n",
            "['https://doi.org/10.1016/j.molcel.2017.12.007', 'https://doi.org/10.1016/j.molcel.2017.12.007', 'https://doi.org/', 'https://support.illumina.com/content/dam/illumina-support/', 'https://bitbucket.org/csmaxwell/crispr-txtl-pam-counting-script']\n",
            "['https://doi.org/10.1007/s10750-019-3927-6 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1139/cjfas-2018-']\n",
            "['https://doi.org/10.1016/j.molcel.2018.07.017']\n",
            "['https://doi.org/10.1007/s10750-018-3559-2']\n",
            "['https://doi.org/10.1016/j.molcel.2019.05.023', 'https://doi.org/10.1016/j.molcel.2019.05.023', 'https://doi.org/10.1016/j.', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://usegalaxy.org', 'https://github.com/naomipetela/', 'https://github.com/naomipetela/', 'https://github.com/naomipetela/nasmythlab-ngs.']\n",
            "['https://doi.org/10.1016/j.molcel.2018.08.002']\n",
            "['https://doi.org/10.1007/s10750-018-3796-4 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-018-3796-4) con-', 'https://github.com/', 'https://doi.org/10.5281/zenodo.1463690).', 'https://CRAN.']\n",
            "['https://', 'https://www.aacc.org/publications/cln/articles/2014/april/cervical-cancer.', 'https://doi.org/10.6084/m9.ﬁgshare.3971883 (2016).']\n",
            "['https://doi.org/10.1155/2018/9046891', 'https://imagej.nih.gov/ij/). The']\n",
            "['https://doi.org/10.1007/s00249-018-1340-0']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201612068']\n",
            "['https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1136/bmj.329.7456.15', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1001/jamainternmed.2017.3601', 'https://doi.org/10.1002/hep.28886', 'https://doi.org/10.7326/', 'https://doi.org/10.1016/j.bbamcr.2015.10.014', 'https://doi.org/10.1136/bmj.', 'https://doi.org/10.3791/50837', 'https://doi.org/10.1002/ar.', 'https://doi.org/10.1038/s41598-017-06433-3', 'https://doi.org/10.1002/sctm.16-0029', 'https://doi.org/10.1016/j.cell.2013.07.004', 'https://doi.org/10.1016/j.cell.2014.11.050', 'https://doi.org/10.1053/j.gastro.2011.09.001', 'https://doi.org/10.2131/jts.43.241', 'https://doi.org/10.1289/ehp.10727', 'https://doi.org/10.1089/adt.2015.659', 'https://doi.org/10.1038/4641116a', 'https://doi.org/10.1007/s40139-016-0113-7', 'https://doi.org/', 'https://doi.org/10.1172/JCI43122', 'https://doi.org/10.1007/', 'https://doi.org/10.1111/gtc.12101', 'https://doi.org/', 'https://doi.org/10.1016/j.cell.2008.07.041', 'https://doi.org/10.1089/scd.2013.0157', 'https://doi.org/10.1002/hep.', 'https://doi.org/', 'https://doi.org/10.1093/toxsci/kft176', 'https://doi.org/10.1038/srep35434', 'https://doi.org/10.1002/hep.26160', 'https://doi.org/10.', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1016/j.cell.2018.11.013', 'https://doi.org/', 'https://doi.org/10.1159/000335753', 'https://doi.org/10.1016/j.stem.2012.10.010', 'https://doi.org/10.1161/CIRCULATIONAHA.', 'https://doi.org/10.1038/nature09747', 'https://doi.org/10.1038/nature09855', 'https://doi.org/10.1161/01.CIR.102.6.706', 'https://doi.org/10.1038/bjp.2008.191', 'https://doi.org/10.1111/j.1476-5381.2009.00191.x', 'https://doi.org/10.1093/cvr/28.6.923', 'https://doi.org/10.', 'https://doi.org/10.1016/j.taap.2012.09.', 'https://doi.org/10.1111/jcmm.13702', 'https://doi.org/10.2174/138161211796904812', 'https://doi.org/10.1038/nm.4087', 'https://doi.org/10.1016/j.taap.2013.', 'https://doi.org/10.1016/j.tox.2014.', 'https://doi.org/10.1007/s12265-012-9396-1', 'https://doi.org/10.1093/toxsci/', 'https://doi.org/10.1007/s12015-010-9113-x', 'https://doi.org/10.1111/j.1476-5381.2011.01558.x', 'https://doi.org/10.1038/nature06196', 'https://doi.org/10.1038/ncomms1260', 'https://doi.org/10.', 'https://doi.org/10.1042/BCJ20170780', 'https://doi.org/10.1016/j.stemcr.2017.07.024', 'https://doi.org/10.3791/52483', 'https://doi.org/10.1016/j.ctrv.2004.05.002', 'https://doi.org/10.1007/s00216-008-2291-6', 'https://doi.org/10.1124/dmd.117.075945', 'https://doi.org/10.1177/1535370214529398', 'https://doi.org/10.1126/scitranslmed.aad8278', 'https://doi.org/10.1016/j.stem.2013.11.002', 'https://doi.org/10.1016/S0140-6736(13)60844-2', 'https://doi.org/10.1038/nrgastro.2015.111', 'https://doi.org/10.1128/mBio.00159-12', 'https://doi.org/10.1016/j.cell.2015.03.053', 'https://doi.org/10.1186/s13059-014-0428-9', 'https://doi.org/10.7554/eLife.18489', 'https://doi.org/10.1002/jps.', 'https://doi.org/10.1101/cshperspect.a014027', 'https://doi.org/10.1093/cvr/cvs206', 'https://doi.org/10.1093/eurheartj/eht067', 'https://doi.org/10.1093/eurheartj/ehr073', 'https://doi.org/10.1242/dmm.008409', 'https://doi.org/10.1093/eurheartj/ehs226', 'https://doi.org/10.1161/CIRCGENETICS.113.000188']\n",
            "['https://doi.org/10.1155/2017/7156489', 'https://www.hindawi.com']\n",
            "['https:// creativecommons .org/ licenses/ by -nc -sa/ 4 .0/ ).', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https:// doi .org/ 10 .1083/ jcb .201301090', 'https:// doi ', 'https:// doi .org/ 10 .1016/ j .cub .2004 .09 .022', 'https:// doi .org/ 10 .1038/ 35099020', 'https:// doi .org/ 10 .1083/ jcb .201008108', 'https:// doi .org/ 10 .1016/ ', 'https:// doi .org/ ', 'https:// doi ', 'https:// doi .org/ 10 .1128/ MCB ', 'https:// doi .org/ 10 .1126/ science .1176495', 'https:// doi .org/ 10 .1016/ ', 'https:// doi .org/ 10 .1016/ S0378 ', 'https:// doi .org/ 10 .1099/ 00221287 ', 'https:// doi .org/ 10 .1091/ mbc .e11 -05 -0434', 'https:// doi .org/ 10 .1016/ S0092 -8674(01)00596 ', 'https:// doi .org/ 10 .1083/ jcb ', 'https:// doi .org/ 10 .1126/ science ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1242/ jcs .109157', 'https:// doi .org/ 10 .1101/ gad .6 .1 .93', 'https:// doi .org/ 10 .1111/ j .1742 -4658 ', 'https:// doi .org/ 10 .1091/ mbc .e06 -09 -0822', 'https:// doi ', 'https:// doi .org/ 10 .1101/ cshperspect .a022343', 'https:// doi ', 'https:// doi .org/ 10 .1016/ j .semcdb .2015 .12 .012', 'https:// doi .org/ 10 .1146/ annurev -cellbio ', 'https:// doi .org/ 10 .1016/ j .cell .2005 .07 ', 'https:// doi ', 'https:// doi .org/ 10 .1016/ 0014 -4827(71)90223 -0', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https:// doi .org/ 10 .1126/ ', 'https:// doi .org/ 10 .1091/ mbc .e04 -12 -1114', 'https:// doi .org/ 10 .1038/ nature02026', 'https:// doi .org/ 10 .1038/ ', 'https:// doi .org/ 10 .1016/ j .gene ', 'https:// doi .org/ 10 .1083/ jcb .200604107', 'https:// doi .org/ 10 .1016/ j .cub .2009 .10 ', 'https:// ', 'https:// doi .org/ 10 .1186/ jbiol195', 'https:// doi .org/ 10 .1038/ nsmb .2706', 'https:// doi .org/ 10 .1016/ j .cell .2012 .05 ', 'https:// doi .org/ 10 .15252/ embj .201488958', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1091/ mbc ', 'https:// doi .org/ 10 .1242/ jcs .206334', 'https:// doi .org/ 10 .1128/ MMBR ', 'https:// doi .org/ 10 .1128/ MMBR .69 ', 'https:// doi .org/ 10 .1016/ 0092 -8674(81)90015 -5', 'https:// doi .org/ 10 .1002/ (SICI)1097 ', 'https:// doi .org/ 10 .1016/ j .cub .2013 .07 .032', 'https:// doi ', 'https:// doi .org/ 10 .1091/ mbc .e17 -06 -0404', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1016/ S0962 -8924(00)01901 -2', 'https:// doi .org/ 10 .1242/ jcs .063891', 'https:// doi .org/ 10 .4161/ cc .11 .2 .18857', 'https:// ', 'https:// doi .org/ 10 .1038/ ncb1855', 'https:// doi .org/ 10 ', 'https:// doi ', 'https:// doi .org/ 10 ', 'https:// ', 'https:// doi ', 'https:// doi .org/ 10 .1242/ jcs .083949', 'https:// doi ', 'https:// doi .org/ 10 .1083/ jcb .200903125', 'https:// doi .org/ 10 .1038/ ', 'https:// doi .org/ 10 .1016/ j .cell .2006 .01 .045', 'https:// ', 'https:// doi .org/ 10 .1016/ j .cub .2017 .08 .032', 'https:// doi .org/ 10 .1083/ jcb .200603003', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100', 'https:// doi .org/ 10 .1083/ jcb .201302001', 'https:// doi .org/ 10 .1371/ journal .pone .0045895', 'https:// doi .org/ 10 .1242/ ', 'https:// doi .org/ 10 ', 'https:// doi .org/ ', 'https:// doi .org/ 10 .1016/ 0076 ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1093/ emboj/ 19 .17 .4524', 'https:// doi .org/ 10 .1016/ j .cell .2008 .09 .039', 'https:// doi .org/ 10 .1016/ S0003 -9861(02)00614 -8', 'https:// doi .org/ 10 .1083/ jcb .201410076', 'https:// doi .org/ 10 .1038/ ncb1701', 'https:// doi .org/ 10 .1002/ yea ', 'https:// doi .org/ ', 'https:// doi .org/ ', 'https:// doi .org/ 10 ', 'https:// ', 'https:// doi .org/ 10 .1091/ mbc .e02 -09 -0558', 'https:// doi .org/ 10 .1091/ mbc .e04 -12 -1090', 'https:// doi .org/ 10 .1371/ journal .pone .0085212', 'https:// doi .org/ 10 .1126/ science .3291120', 'https:// doi .org/ 10 .1534/ genetics .112 .145516', 'https:// doi .org/ 10 .1083/ jcb .200203094', 'https:// doi .org/ 10 .1016/ j .cell .2017 .05 .028', 'https:// doi ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1083/ jcb .200604094', 'https:// doi .org/ 10 ', 'https:// doi .org/ 10 .1091/ mbc .e06 -01 -0015', 'https://doi.org/10.1083/jcb.201805100', 'https://doi.org/10.1083/jcb.201805100']\n",
            "['https://doi.org/10.1016/j.molcel.2018.03.016', 'https://doi.org/10.1016/j.molcel.2018.03.016', 'https://doi.org/10.1016/j.molcel.', 'https://data.mendeley.com/datasets/', 'https://imagej.nih.gov/ij', 'https://pymol.org/2/', 'https://www.microsoft.com/store/b/home', 'https://systatsoftware.com/products/']\n",
            "['https://doi.org/10.1007/s10750-017-3423-9', 'https://doi.org/10.1007/s10750-017-3423-9) con-', 'https://doi.org/', 'https://www-01.ibm.com/common/ssi/']\n",
            "['https://www.kyazma.nl/index.']\n",
            "['https://doi.org/10.1155/2018/8510620', 'https://www.brown.edu/research/labs/wessel/wally-']\n",
            "['https://cran.r-project.']\n",
            "['https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:35:42, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231200002X']\n",
            "['https://doi.org/10.1155/2018/7382351', 'https://sciex.com/software-downloads-x2110). The']\n",
            "['https://doi.org/10.1016/j.clinbiochem.2019.07.007', 'https://www.who.int/biologicals/BS.2018.2339_Free_PSA.pdf.', 'https://www.who.int/biologicals/BS.2018.2340_']\n",
            "['https://www.nbstrn.org/supported-', 'https://www.team-share.net/NDSI-IEM-Initiative/Overview.aspx.']\n",
            "['https://zhanglab.ccmb.med.umich.edu/GPCR-EXP/. ']\n",
            "['https://doi.org/10.1016/j.clinbiochem.2018.10.011']\n",
            "['https://doi.org/10.1007/s10750-017-3364-3', 'https://CRAN.R-project.org/package=maptools.', 'https://', 'https://www.R-project.org/.']\n",
            "['https://dx.doi.org/10.6084/m9.']\n",
            "['https://doi.org/10.1007/s10750-018-3572-5', 'https://doi.org/10.1007/s10750-018-3572-5) con-', 'https://cran.r-project.']\n",
            "['https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:37:04, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672311000097']\n",
            "['https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1016/j.tibs.2008.06.002', 'https://doi.org/10.4161/cl.21912', 'https://doi.org/10.4161/cl.21882', 'https://doi.org/10.4161/sgtp.', 'https://doi.org/10.1007/s12640-016-9613-9', 'https://doi.org/10.1073/pnas.0705003104', 'https://doi.org/10.1021/acs.jmedchem.', 'https://doi.org/10.1016/j.cellsig.2008.11.001', 'https://doi.org/10.4161/cl.21602', 'https://doi.org/10.', 'https://doi.org/10.1186/s13041-016-0230-6', 'https://doi.org/10.1038/1675', 'https://doi.org/10.1016/j.ejmg.2017.01.004', 'https://doi.org/10.1161/CIRCULATIONAHA.111.048785', 'https://doi.org/10.1111/bph.13872', 'https://doi.org/10.1016/j.yjmcc.2011.09.016', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1016/j.yjmcc.2009.', 'https://doi.org/10.1152/ajpheart.00612.2011', 'https://doi.org/10.1016/j.cellsig.2014.04.020', 'https://doi.org/10.1128/MCB.21.15.5179-5189.2001', 'https://doi.org/10.1016/S0092-8674(00)00043-X', 'https://doi.org/10.1038/embor.2012.75', 'https://doi.org/10.1016/j.jmb.2006.06.017', 'https://doi.org/10.1074/', 'https://doi.org/10.1074/jbc.M110.156505', 'https://doi.org/10.1073/pnas.1214447109', 'https://doi.org/10.1073/pnas.1717437115', 'https://doi.org/10.1016/j.str.2007.01.001', 'https://doi.org/10.1371/journal.pone.0077818', 'https://doi.org/10.1016/j.bmcl.2016.04.037', 'https://doi.org/10.1016/j.', 'https://doi.org/10.1016/j.bbapap.2017.10.004', 'https://doi.org/10.1038/ncomms9681', 'https://doi.org/10.1016/j.str.2011.10.013', 'https://doi.org/10.1002/anie.201108865', 'https://doi.org/10.1021/acsmedchemlett.5b00102', 'https://doi.org/10.1021/acsmedchemlett.6b00322', 'https://doi.org/10.1021/acsmedchemlett.', 'https://doi.org/10.1021/ja805555a', 'https://doi.org/10.1016/j.bcp.2010.03.012', 'https://doi.org/10.', 'https://doi.org/10.1016/j.jmb.2006.02.027', 'https://doi.org/10.1038/nature09593', 'https://doi.org/10.1016/j.jsb.2010.06.008', 'https://doi.org/10.1007/s12575-009-9008-x', 'https://doi.org/10.1016/S0006-3495(00)76713-0', 'https://doi.org/10.1042/BCJ20180867', 'https://doi.org/10.1016/S0003-2697(03)00289-6', 'https://doi.org/10.1107/S0021889810008289', 'https://doi.', 'https://doi.org/10.1007/', 'https://doi.org/10.1002/prot.22588', 'https://doi.org/10.1093/nar/gkw389', 'https://doi.org/10.1107/S1399004715002576', 'https://doi.org/10.1016/S0006-3495(01)76260-1', 'https://doi.org/', 'https://doi.org/10.1107/', 'https://doi.org/10.1107/S0021889807021206', 'https://doi.org/10.1107/S0907444909052925', 'https://doi.org/10.1107/S0907444909042073', 'https://doi.org/10.1016/S1097-2765(01)00428-2', 'https://doi.org/10.1002/bms.1200220902', 'https://doi.org/10.1002/oms.1210270709', 'https://doi.org/10.1002/ajmg.a.20131', 'https://doi.org/10.1093/hmg/', 'https://doi.org/10.1002/ajmg.a.31956', 'https://doi.org/10.1038/mp.2014.189', 'https://doi.org/10.1080/10826076.2012.743724', 'https://doi.org/10.1016/j.ymeth.2008.10.020', 'https://doi.org/10.1016/S0165-022X(03)00053-8', 'https://doi.org/10.1007/978-981-10-6038-0_3']\n",
            "['https://github.com/cgevans/', 'https://books.google.com/books?id=sk3OBQAAQBAJ&pgis=1', 'https://books.google.com/books?hl=en&lr=&id=n7lLTzD6PHYC&pgis=1', 'https://www.researchgate.net/profile/John_Conery/', 'https://doi.', 'https://doi.org/10.1021/es049822c']\n",
            "['https://', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:31:34, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000016', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:31:34, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000016', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:31:34, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000016', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:31:34, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000016', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:31:34, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000016']\n",
            "['https://doi.org/10.1155/2018/3868070']\n",
            "['https://doi.org/10.1017/', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://pymol.org/2/; Schrödinger, LLC).', 'https://github.', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://orcid.org/0000-0002-6803-4456', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:21:52, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672319000065']\n",
            "['https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/', 'https://github.com/', 'https://', 'https://doi.', 'https://doi.org/', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio', 'https://doi.org/10.1038/s42003-019-0555-7', 'https://doi.org/', 'https://doi.org/10.5281/', 'https://doi.org/10.1038/s42003-', 'https://doi.org/10.1038/s42003-019-0555-7 | www.nature.com/commsbio']\n",
            "['https://doi.org/10.1016/j.molcel.2017.10.030']\n",
            "['https://doi.org/10.1016/j.molcel.2017.10.022', 'https://doi.org/10.1016/j.molcel.2017.10.022', 'https://doi.org/10.1016/j.molcel.2017.10.022.', 'https://www.zeiss.com/', 'https://imagej.nih.gov/ij/']\n",
            "['https://doi.org/10.1007/s10750-017-3441-7', 'https://', 'https://']\n",
            "['https://cran.r-pro-', 'https://cran.r-project.org/web/packages/mcr/index.html2014.']\n",
            "['https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:34:21, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672314000226']\n",
            "['https://doi.org/10.1007/s10750-018-3840-4 (0123456789().,-volV)(0123456789().,-volV)', 'https://waterdata.usgs.gov/nwis/']\n",
            "['https://doi.org/10.1016/j.clinbiochem.2017.10.013', 'https://www.cadth.ca/protein-testing-patients-multiple-myeloma-review-clinical-']\n",
            "['https://doi.org/10.1186/s12863-018-0709-5', 'https://github.com/cbkmephisto/SNPipeline) and FIm-', 'https://github.com/', 'https://ucanr.edu/sites/UCCE_LR/files/152458.pdf.', 'https://www.aphis.usda.gov/']\n",
            "['https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6', 'https://doi.org/10.1038/s41598-019-48335-6']\n",
            "['https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://www.who.int/malaria/publications/', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications', 'https://doi.org/10.1038/s41467-019-11790-w', 'https://doi.org/10.1101/251082 (2018).', 'https://doi.org/10.1111/', 'https://doi.org/10.1038/s41467-', 'https://doi.org/10.1038/s41467-019-11790-w | www.nature.com/naturecommunications']\n",
            "['https://doi.org/10.1007/s10750-018-3750-5 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-018-3750-5) con-', 'https://doi.org/10.1111/', 'https://doi.org/10.1101/278309.', 'https://www.woking.gov.uk/environment/greeninf/']\n",
            "['https://doi.org/10.1007/s10750-018-3644-6', 'https://doi.org/']\n",
            "['https://doi.org/10.1186/s12863-019-0750-z', 'https://scholar.google.com).']\n",
            "['https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6', 'https://doi.org/10.1038/s41598-019-48181-6.', 'https://doi.org/10.1038/s41598-019-48181-6']\n",
            "['https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://ecvam-dbalm.', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1080/02652030701744538 (2008).', 'https://doi.org/10.3109/17435390.2', 'https://doi.org/10.1289/ehp.7339 (2005).', 'https://doi.org/10.1080/02652030802007553 (2008).', 'https://doi.org/10.1002/wnan.153 (2011).', 'https://doi.org/10.1111/joim.12109 (2013).', 'https://ssl.biomax.de/nanomile/cgi/login_bioxm_portal.cgi (2018).', 'https://cordis.europa.eu/project/rcn/106943/reporting/en (2017).', 'https://doi.org/10.1124/dmd.107.020016 (2008).', 'https://doi.org/10.3390/microarrays5020008 (2016).', 'https://doi.org/10.1016/j.nano.2013.04.010 ', 'https://doi.org/10.1371/journal.', 'https://doi.org/10.1038/s41597-019-0053-2', 'https://doi.org/10.1021/acsami.5b11216 (2016).', 'https://ec.europa.eu/jrc/en/scientific-tool/jrc-', 'https://doi.org/10.1007/s00204-017-2106-7 (2017).', 'https://doi.org/10.2905/', 'https://doi.org/10.1073/pnas.0604348103 (2006).', 'https://doi.org/10.1088/1749-4699/6/1/014006 (2013).', 'https://doi.org/10.1177/108705719900400206 (1999).', 'https://doi.org/10.6084/m9.figshare.c.4434041 (2019).', 'https://doi.org/10.1038/s41597-019-0053-2']\n",
            "['https://creativecommons.org/licenses/by/4.0), which permits unrestricted use,']\n",
            "['https://doi.org/10.1186/s12863-018-0685-9', 'https://doi.org/', 'https://serval.unil.ch/notice/serval:BIB_DB589E420FA0 (2018).']\n",
            "['https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:26:36, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672313000220']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201612106']\n",
            "['https://doi.org/10.1007/s10750-018-3863-x (0123456789().,-volV)(0123456789().,-volV)']\n",
            "['https://doi.org/10.1007/s10750-018-3816-4 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-018-3816-4) con-']\n",
            "['https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1073/pnas.92.6.1941', 'https://doi.org/10.1074/jbc.271.28.16662', 'https://doi.org/10.1111/j.1432-1033.1996.00438.x', 'https://doi.org/10.1002/1873-3468.12214', 'https://doi.org/10.1074/jbc.M400958200', 'https://doi.org/10.1038/', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/10.1021/bi034082q', 'https://doi.org/10.1042/bj3390127', 'https://doi.org/10.1006/jmbi.2001.4834', 'https://doi.org/10.1021/bi300149z', 'https://doi.org/10.1021/jacs.6b02827', 'https://doi.org/10.1074/jbc.M113.477943', 'https://doi.org/10.1021/acs.biochem.5b01080', 'https://doi.org/10.1105/tpc.106.042374', 'https://doi.org/10.1016/j.bbrc.2018.02.146', 'https://doi.org/10.1021/bi010562a', 'https://doi.org/10.1074/jbc.M709172200', 'https://doi.org/10.1038/nplants.2015.125', 'https://doi.org/10.1016/S0092-8674(02)00971-6', 'https://doi.org/10.1111/j.1432-1033.2004.04143.x', 'https://doi.org/10.1021/bi050240x', 'https://doi.org/10.1042/bj2990277', 'https://doi.org/10.1104/pp.95.4.1189', 'https://doi.org/10.1042/BJ20130834', 'https://doi.org/10.1042/bj3340335', 'https://doi.org/10.1021/bi701515y', 'https://doi.org/10.1099/00221287-111-1-1', 'https://doi.org/10.1074/jbc.M112.352526', 'https://doi.org/10.1371/journal.pone.0091484', 'https://doi.org/10.1074/jbc.M411884200', 'https://doi.org/10.1007/s11120-006-9065-9', 'https://doi.org/10.1126/', 'https://doi.org/10.3389/fpls.2016.00292', 'https://doi.org/10.1093/bib/3.3.265', 'https://doi.org/10.1016/j.jmb.2017.12.007', 'https://doi.org/', 'https://doi.org/10.1042/BCJ20190095', 'https://doi.org/', 'https://doi.org/10.1371/journal.pone.0011203', 'https://doi.org/10.', 'https://doi.org/10.1046/j.1365-313X.1997.12050981.x', 'https://doi.org/10.1073/pnas.96.5.1941']\n",
            "['https://doi.org/10.1007/s10750-018-3776-8 (0123456789().,-volV)(0123456789().,-volV)', 'https://generalhydroponics.com/', 'https://www.ebi.ac.uk/Tools/msa/clustalo/).', 'https://', 'https://doi.org/10.']\n",
            "['https://dx.doi.org/10.6084/m9.']\n",
            "['https://www.fda.gov/newsevents/newsroom/pressannouncements/ucm587547.htm. ', 'https://www.accessdata.fda.gov/scripts/cdrh/cfdocs/cfpmn/pmn.cfm?ID=K123486. ', 'https://www.spritam.com/#/hcp/zipdose-technology/what-is-zipdose-technology. ', 'https://futurism.com/3d-printed-tritanium-spinal-implant-simulates-actual-bone. ', 'https://www.stratasys.com/materials/search. ', 'https://www.ethicon.com/emea/products/wound-closure/absorbable-sutures/monocryl-', 'https://www.crbard.com/Davol/en-', 'https://www.spineology.com/united-', 'https://www.accessdata.fda.gov/cdrh_docs/pdf16/K163673.pdf. ', 'https://www.carbon3d.com/industry/dental-', 'https://formlabs.com/materials/dental/. ', 'https://envisiontec.com/wp-', 'https://formlabs.com/industries/dental/ https://formlabs.com/materials/dental/. ', 'https://3dprint.nih.gov/collections/prosthetics. ', 'https://3dprintingindustry.com/news/materialise-partners-with-university-of-', 'https://ois.net/jjs-plans-', 'https://www.accessdata.fda.gov/drugsatfda_docs/nda/2015/207958Orig1s000TOC.cfm. ', 'https://www.spritam.com/#/patient/about-']\n",
            "['https:// doi .org/ 10 .1083/ jcb .201805003) explore the three-dimensional architecture of vertebrate neocentromeres, leading to a model ', 'https:// creativecommons .org/ licenses/ by -nc -sa/ 4 .0/ ).', 'https://doi.org/10.1083/jcb.201811172', 'https://doi.org/10.1083/jcb.201811172', 'https:// doi .org/ 10 ', 'https://doi.org/10.1083/jcb.201811172', 'https://doi.org/10.1023/jcb.201811172J. Cell Biol. 2019 Vol. 218 No. 2\\u2003722Rockefeller University Press\\x0c']\n",
            "['https://doi.org/10.1083/jcb.201705070', 'https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).']\n",
            "['https://doi.org/10.1155/2017/4701481', 'https://www.hindawi.com']\n",
            "['https://doi.org/10.1155/2018/4963942', 'https://doi.org/10.1016/j.atherosclerosissup']\n",
            "['https://dx.doi.org/10.6084/m9.']\n",
            "['https://cran.r-project.org/package=']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201702177']\n",
            "['https://familyhistory.hhs.gov']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201612068']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201702177']\n",
            "['https://doi.org/10.1016/j.molcel.2018.04.022', 'https://doi.org/10.1016/j.molcel.2018.04.022', 'https://doi.org/10.1016/j.molcel.2018.04.022.', 'https://imagej.nih.gov/ij/']\n",
            "['https://public-registry.jbei.org/ ', 'https://www.denovodna.com/software/ ', 'https://j5.jbei.org ', 'https://chopchop.rc.fas.harvard.edu/ ', 'https://synbiotools.bbn.com/ ']\n",
            "['https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099', 'https://www.google.com/url?', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:36:16, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S0016672315000099']\n",
            "['https://doi.org/10.1007/s10750-018-3647-3']\n",
            "['https://doi.org/10.1083/jcb.201704048', 'https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https ://doi .org /10 .1074 /jbc ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1016 /j .ejcb ', 'https ://doi .org /10 .1186 /bcr1329', 'https ://doi .org /10 .1016 /j .ejcb .2014 .07 .003', 'https ://doi .org /10 .1016 /S0076 -6879(02)45028 -8', 'https ://doi .org /10 .4161 /cam .28349', 'https ://doi .org /10 .1016 /j .ejcb .2005 .09 .018', 'https ://doi ', 'https ://doi .org /10 .1038 /sj .emboj ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1242 /jcs .123547', 'https ://doi .org /10 .1007 /s10555 -008 ', 'https ://doi .org /10 .1146 /annurev -cellbio -111315 -125227', 'https ://doi .org /10 .1038 /', 'https ://doi .org /10 .1002 /jez .1402510206', 'https ://doi .org /10 .1111 /j .1749 -6632 .1999 .tb07695 .x', 'https ://doi .org /10 .1158 /0008 -5472 .CAN -06 -3928', 'https ://doi .org /10 .1146 /annurev .cellbio .21 .122303 ', 'https ://doi .org /10 .1096 ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1074 /jbc ', 'https ://doi .org /10 .1016 /j .ceb .2011 .04 .002', 'https ://doi .org /10 .1016 /j .tcb .2017 .03 .003', 'https ://doi .org /10 .1007 /978 -1 -4899 -4541 -9', 'https ://doi .org ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1021 /bi400678y', 'https ://doi .org /10 .1083 /jcb .201403127', 'https ://doi .org /10 .1186 /1471 -2407 -13 -63', 'https ://doi .org /10 .1016 /j .ceb .2008 .01 .005', 'https ://doi .org /10 .1242 /jcs .092726', 'https ', 'https ://doi .org /10 .1074 /jbc .M105363200', 'https ://doi .org /10 .1083 ', 'https ://doi .org /10 .1242 /jcs .03364', 'https ', 'https ://doi .org /10 .1042 /BST0330891', 'https ://', 'https ://doi .org /10 ', 'https ://doi .org /10 .1016 /S0091 -679X(08)85004 -2', 'https ://doi .org ', 'https ://doi .org /10 .1083 /', 'https ://doi .org /10 .1091 /mbc ', 'https ', 'https ://doi .org /10 .4161 /cam .28406', 'https ://doi ', 'https ://doi .org /10 .1158 /0008 ', 'https ://doi .org /10 .1083 /jcb .201103045', 'https ://doi .org ', 'https ://doi .org /10 .1021 /pr060140t', 'https ://doi .org /10 .4049 /jimmunol .1501655', 'https ://doi .org ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1038 /ncb2972', 'https ://doi .org /10 .1002 /ijc .2910280612', 'https ', 'https ://doi .org /10 .1016 /j .bpj .2010 .12 .3733', 'https ://doi ', 'https ://doi .org /10 .1038 /nature04665', 'https ://doi .org /10 .1038 /srep25502', 'https ://doi .org /10 .1038 /onc .2013 .377', 'https ://doi .org /10 .1091 /mbc .E16 -12 -0885', 'https ://doi .org ', 'https ://doi .org /10 .1083 ', 'https ://', 'https ://doi ', 'https ://doi .org /10 .1126 /science .1247005', 'https ://doi .org /10 .1038 /ncomms12405', 'https ://doi ', 'https ://doi .org ', 'https ://doi .org /10 .1016 /B978 -0 ', 'https ', 'https ://doi .org /10 .1093 /cercor /bhs402', 'https ://doi .org /10 .1074 /jbc ', 'https ://doi .org /10 .1074 /jbc .275 .18 .13901', 'https ://doi .org /10 .1186 /1472 -6750 -3 -13', 'https ://doi .org /10 .1016 /0092 ', 'https ://doi .org /10 ', 'https ://doi .org /10 ', 'https ://doi .org /10 .1016 /j .ceb .2005 .08 ', 'https ://doi .org /10 .1083 ', 'https ://doi .org /10 .1074 /jbc .M807845200', 'https ://doi .org /10 .1038 /onc .2013 .363', 'https ://doi .org /10 .1074 /jbc ']\n",
            "['https://doi.org/10.1017/S001667231900003X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:25:05, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231900003X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:25:05, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231900003X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:25:05, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231900003X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:25:05, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231900003X', 'https://www.cambridge.org/core. IP address: 213.79.89.166, on 16 Aug 2019 at 09:25:05, subject to the Cambridge Core terms of use, available at', 'https://www.cambridge.org/core/terms. https://doi.org/10.1017/S001667231900003X']\n",
            "['https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8', 'https://doi.org/10.1038/s41598-019-47337-8']\n",
            "['https://doi.org/10.1007/s10750-019-3956-1 (0123456789().,-volV)(0123456789().,-volV)', 'https://doi.org/10.1007/s10750-019-3956-1) con-', 'https://cran.r-project.org/web/packages/vegan/index.html.', 'https://doi.org/10.1017/s0022149x18000482.']\n",
            "['https://github.com/harmslab/wheeler-harms-S100A5-Ca-Cu-', 'https://academic.oup.com/chemse/article/39/5/', 'https://github.com/harmslab/pytc.', 'https://', 'https://link.springer.com/article/10.']\n",
            "['https://www.cabi.org/', 'https://doi.org/10.', 'https://doi.org/10.1186/s13007-018-0323-6.', 'https://doi.org/10.', 'https://www.engineeringtoolbox.com/maxi']\n",
            "['https://doi.org/10.1155/2019/4625279']\n",
            "['https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1016/j.plantsci.2012.04.013', 'https://doi.org/10.1074/jbc.R700029200', 'https://doi.org/10.1002/bip.21256', 'https://doi.org/10.12688/f1000research.8133.1', 'https://doi.org/10.1023/A:1012440926982', 'https://doi.org/10.1104/pp.115.3.1259', 'https://doi.org/10.1074/jbc.272.27.16746', 'https://doi.org/10.1073/pnas.94.22.12230', 'https://doi.org/10.1016/S1360-1385', 'https://doi.org/10.1105/tpc.104.025379', 'https://doi.org/10.1021/bi300570v', 'https://doi.org/10.1021/bi2004312', 'https://doi.org/10.1021/bi201425f', 'https://doi.org/10.1371/journal.pone.0002039', 'https://doi.org/10.1371/', 'https://doi.org/10.1016/j.plantsci.2016.03.013', 'https://doi.org/10.1007/s00425-003-1192-3', 'https://doi.org/10.1093/pcp/pcu027', 'https://doi.org/10.1007/s12355-010-0041-5', 'https://doi.org/10.4172/2155-9600.1000497', 'https://doi.org/10.1093/aob/', 'https://doi.org/10.1093/', 'https://doi.org/10.', 'https://doi.org/10.1042/BCJ20190154', 'https://doi.org/10.1002/(SICI)1097-0231(19990315)13:5<315::AID-RCM483>3.0.CO;2-C', 'https://doi.org/10.1172/JCI24650', 'https://doi.org/10.1016/S0969-2126(00)00194-5', 'https://doi.org/10.1016/j.bbapap.2013.02.024', 'https://doi.org/10.1021/bi051902l', 'https://doi.org/10.', 'https://doi.org/10.1021/bi0100790', 'https://doi.org/10.3390/ijms17050640', 'https://doi.org/10.1074/jbc.M110.159541', 'https://doi.org/10.1007/BF00197106', 'https://academicpublishingplatforms.com/journal.php?journal=PARB', 'https://doi.org/10.1073/pnas.0800585105', 'https://doi.org/10.1016/j.niox.2014.04.003', 'https://doi.org/10.1104/pp.106.086918', 'https://doi.org/10.1007/PL00008128', 'https://doi.org/10.1086/336870', 'https://www.seedtest.org/en/content—1–1084.html', 'https://doi.org/10.1093/jxb/', 'https://doi.org/10.1007/s11738-017-2569-4', 'https://doi.org/10.1111/j.1469-8137.2007.02226.x', 'https://doi.org/10.1529/biophysj.105.060582', 'https://doi.org/10.1111/tpj.12340', 'https://doi.org/10.3389/fpls.', 'https://doi.org/10.1016/j.jmb.2007.05.029', 'https://doi.org/10.1021/bi060716s', 'https://doi.org/10.1021/ja066638d']\n",
            "['https://doi.org/10.1016/j.molcel.2017.10.033', 'https://doi.org/10.1016/j.molcel.2017.10.033', 'https://doi.org/10.1016/j.molcel.2017.10.033.', 'https://doi.org/10.17632/xgbjnmxnbr.1', 'https://doi.org/10.17632/']\n",
            "['https://doi.org/10.1016/j.molcel.2017.12.019', 'https://doi.org/10.1016/j.molcel.2017.12.019', 'https://doi.org/10.17632/k2t8nx8kh5.1', 'https://doi.org/10.17632/k2t8nx8kh5.1', 'https://portal.thermo-brims.com/', 'https://www.rstudio.com/', 'https://www.graphpad.com', 'https://doi.org/10.17632/k2t8nx8kh5.1).']\n",
            "['https://doi.org/10.1038/s41398-019-0539-x', 'https://breenms.shinyapps.io/DEXPTSD/.', 'https://doi.org/']\n",
            "['https ://creativecommons .org /licenses /by -nc -sa /4 .0 /).', 'https://doi.org/10.1083/jcb.201612169']\n",
            "['https://www.intechopen.com/books/next-generation-sequencing-advances-applications-and-challeng']\n",
            "['https://doi.org/10.1186/s12863-019-0722-3', 'https://doi.', 'https://doi.org/10.1186/1297-9686-46-34.']\n",
            "['https://doi.org/10.1016/j.clinbiochem.2019.06.001', 'https://www.']\n",
            "['https://www.avacta.com).']\n",
            "['https://doi.org/10.1016/j.molcel.2017.11.028']\n",
            "['https://doi.org/10.1007/s10750-018-3864-9 (0123456789().,-volV)(0123456789().,-volV)', 'https://nrfa.ceh.ac.uk/', 'https://vemco.com/wp-']\n",
            "['https://www.asminternational.org/documents/10192/1849770/06974G_Chapter_1.pdf ', 'https://grabcad.com/library/femur-bone-2. ', 'https://forum.solidworks.com/thread/189281. ', 'https://www.healthpages.org/anatomy-', 'https://healthfully.com/do-ligaments-do-skeletal-system-']\n",
            "['https://imagej.nih.gov/ij/', 'https://www.ﬂowjo.com/', 'https://github.com/BradnerLab/pipeline/', 'https://github.com/BradnerLab/pipeline/', 'https://pypi.python.org/pypi/coltron', 'https://github.com/BradnerLab/netseq', 'https://github.com/BradnerLab/pipeline/wiki/bamliquidator). Brieﬂy, ChIP-Seq reads aligning to the region were', 'https://github.com/BradnerLab/pipeline/ (ROSE2_main.py). Default parameters, including exclusion of', 'https://github.com/BradnerLab/netseq. Brieﬂy, the six 50 end nucleotides of the', 'https://', 'https://github.com/BradnerLab/', 'https://pypi.python.org/pypi/coltron.']\n",
            "['https://doi.org/10.1016/j.clinbiochem.2018.01.020', 'https://']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# re.findall(r'.[A-Z]+[a-z]+', bio[\"text\"].iloc[0])\n",
        "# # [A-Z]+[a-z]+ -- ищем подстроку вида \"{какое-то кол-во заглавных букв}{какое-то кол-во строчных букв}\"\n",
        "# # .[A-Z]+[a-z]+ -- ищем подстроку вида \"{любой символ кроме начала строки}{какое-то кол-во заглавных букв}{какое-то кол-во строчных букв}\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "iNwQilf_rZiU",
        "outputId": "c13ca630-7b44-41fd-a1d6-441cf31cd9b2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[' Do',\n",
              " ' Health',\n",
              " ' Research',\n",
              " ' Centre',\n",
              " ' Department',\n",
              " ' Pharmacology',\n",
              " ' Lung',\n",
              " ' Health',\n",
              " ' Research',\n",
              " ' Centre',\n",
              " ' University',\n",
              " ' Melbourne',\n",
              " ' Victoria',\n",
              " ' Australia',\n",
              " ' In',\n",
              " ' We',\n",
              " ' The',\n",
              " '‘Omnis',\n",
              " ' Virchow',\n",
              " ' Bertoncello',\n",
              " ' Lung',\n",
              " ' Health',\n",
              " ' Research',\n",
              " ' Department',\n",
              " ' Pharmacology',\n",
              " ' Lung',\n",
              " ' Health',\n",
              " ' Research',\n",
              " ' Level',\n",
              " ' Medical',\n",
              " ' Building',\n",
              " ' The',\n",
              " ' University',\n",
              " ' Melbourne',\n",
              " ' Australia',\n",
              " ' Email',\n",
              " ' Authors',\n",
              " ' Associate',\n",
              " ' Professor',\n",
              " ' Ivan',\n",
              " ' Bertoncello',\n",
              " ' Lung',\n",
              " ' Regeneration',\n",
              " ' Laboratory',\n",
              " ' Dr',\n",
              " ' Mc',\n",
              " ' Lung',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Microenvi',\n",
              " ' Group',\n",
              " ' January',\n",
              " ' January',\n",
              " ' February',\n",
              " ' February',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " ' Rare',\n",
              " ' Relatively',\n",
              " ' This',\n",
              " ' In',\n",
              " ' The',\n",
              " ' Multipotent',\n",
              " ' Bertoncello',\n",
              " ' Mc',\n",
              " '(Omnis',\n",
              " ' The',\n",
              " ' The',\n",
              " ' Airway',\n",
              " ' However',\n",
              " ' Accordingly',\n",
              " '(Fig',\n",
              " ' As',\n",
              " ' Day',\n",
              " ' In',\n",
              " ' Day',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " '\\x0cEpithelial',\n",
              " ' Analysis',\n",
              " ' Regional',\n",
              " ' There',\n",
              " ' Subpopulations',\n",
              " ' Till',\n",
              " ' Like',\n",
              " ' Lajtha',\n",
              " ' He',\n",
              " ' In',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " ' Scho',\n",
              " ' He',\n",
              " ' The',\n",
              " ' In',\n",
              " ' For',\n",
              " ' Speci',\n",
              " ' Robust',\n",
              " ' Transplant',\n",
              " ' Lineage',\n",
              " ' Genetically',\n",
              " '(Fig',\n",
              " ' The',\n",
              " ' However',\n",
              " ' Bertoncello',\n",
              " ' Mc',\n",
              " ' Stem',\n",
              " ' The',\n",
              " ' Discord',\n",
              " ' Transplant',\n",
              " ' Evidence',\n",
              " ' The',\n",
              " ' While',\n",
              " ' The',\n",
              " ' Distal',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " '\\x0cEpithelial',\n",
              " ' Regenerative',\n",
              " '(Clara',\n",
              " ' Brie',\n",
              " '(Fig',\n",
              " ' These',\n",
              " '(Krt',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " '(CCSPpos',\n",
              " '(Scgb',\n",
              " '(Clara',\n",
              " ' Krt',\n",
              " ' Cy',\n",
              " ' Scgb',\n",
              " ' Scgb',\n",
              " '(Clara',\n",
              " '(Clara',\n",
              " '-Cpos',\n",
              " ' However',\n",
              " ' Bertoncello',\n",
              " ' Mc',\n",
              " ' On',\n",
              " ' In',\n",
              " ' Krt',\n",
              " ' Likewise',\n",
              " '(NGFRposa',\n",
              " ' In',\n",
              " '(Clara',\n",
              " '(Clara',\n",
              " ' Other',\n",
              " ' Scgb',\n",
              " 'sKrt',\n",
              " 'sKrt',\n",
              " ' In',\n",
              " '-Cneg',\n",
              " '(CGRPpos',\n",
              " '(Fig',\n",
              " ' However',\n",
              " ' In',\n",
              " ' This',\n",
              " ' These',\n",
              " ' Bertoncello',\n",
              " 'cQualter',\n",
              " ' More',\n",
              " ' Multiparameter',\n",
              " ' NGFRposa',\n",
              " ' Krt',\n",
              " ' Krt',\n",
              " ' Krt',\n",
              " ' Oeztuerk',\n",
              " '-Winder',\n",
              " '-Cadpos',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " '\\x0cEpithelial',\n",
              " '-Cadpos',\n",
              " ' While',\n",
              " ' Murine',\n",
              " ' Conse',\n",
              " ' The',\n",
              " ' Regular',\n",
              " ' These',\n",
              " ' Australian',\n",
              " ' Health',\n",
              " ' Medical',\n",
              " ' Research',\n",
              " ' Council',\n",
              " '(Grant',\n",
              " ' Victorian',\n",
              " ' Cancer',\n",
              " ' Agency',\n",
              " ' Ramalho',\n",
              " '-Santos',\n",
              " ' Willenbring',\n",
              " ' On',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Potten',\n",
              " ' Cell',\n",
              " ' In',\n",
              " ' Mc',\n",
              " ' Isaacson',\n",
              " ' Wright',\n",
              " ' Oxford',\n",
              " ' Textbook',\n",
              " ' Pathology',\n",
              " ' Oxford',\n",
              " ' University',\n",
              " ' Press',\n",
              " ' Rando',\n",
              " ' Stem',\n",
              " ' Moore',\n",
              " ' Overview',\n",
              " ' In',\n",
              " ' Morstyn',\n",
              " ' Sheridan',\n",
              " ' Cell',\n",
              " ' Therapy',\n",
              " ' Stem',\n",
              " ' Transplantation',\n",
              " ' Gene',\n",
              " ' Therapy',\n",
              " ' Cellular',\n",
              " ' Immuno',\n",
              " ' Cambridge',\n",
              " ' University',\n",
              " ' Press',\n",
              " ' New',\n",
              " ' York',\n",
              " ' Marshman',\n",
              " ' Booth',\n",
              " ' Potten',\n",
              " ' The',\n",
              " ' Bioessays',\n",
              " ' Blanpain',\n",
              " ' Horsley',\n",
              " ' Fuchs',\n",
              " ' Epithelial',\n",
              " ' Cell',\n",
              " ' Rawlins',\n",
              " ' Hogan',\n",
              " ' Epithelial',\n",
              " ' Development',\n",
              " ' Brown',\n",
              " ' Rannels',\n",
              " ' Rannels',\n",
              " ' Implications',\n",
              " ' Respir',\n",
              " ' Res',\n",
              " ' Potten',\n",
              " ' Loef',\n",
              " ' Stem',\n",
              " ' Lessons',\n",
              " ' Devel',\n",
              " ' Lajtha',\n",
              " ' Stem',\n",
              " ' Differentiation',\n",
              " ' Jordan',\n",
              " ' Lemischka',\n",
              " ' Clonal',\n",
              " ' Genes',\n",
              " ' Dev',\n",
              " ' Rando',\n",
              " ' The',\n",
              " ' Cell',\n",
              " ' Ho',\n",
              " ' Wagner',\n",
              " ' The',\n",
              " ' Curr',\n",
              " ' Hematol',\n",
              " ' Morrison',\n",
              " ' Kimble',\n",
              " ' Asymmetric',\n",
              " ' Nature',\n",
              " ' Haan',\n",
              " ' Nijhof',\n",
              " ' Van',\n",
              " ' Zant',\n",
              " ' Mouse',\n",
              " ' Blood',\n",
              " ' Avagyan',\n",
              " ' Glouchkova',\n",
              " ' Choi',\n",
              " ' Immunol',\n",
              " ' Henckaerts',\n",
              " ' Langer',\n",
              " ' Snoeck',\n",
              " ' Quantitative',\n",
              " ' Blood',\n",
              " ' Ousset',\n",
              " ' Van',\n",
              " ' Keymeulen',\n",
              " ' Bouvencourt',\n",
              " ' Multipotent',\n",
              " ' Nat',\n",
              " ' Cell',\n",
              " ' Biol',\n",
              " ' Barker',\n",
              " ' Bartfeld',\n",
              " ' Clevers',\n",
              " ' Tissue',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Barker',\n",
              " ' Epithelial',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Mc',\n",
              " ' Bertoncello',\n",
              " ' Concise',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Doupe',\n",
              " ' Alcolea',\n",
              " ' Roshan',\n",
              " ' Science',\n",
              " ' Becher',\n",
              " ' Holland',\n",
              " ' Evidence',\n",
              " ' Genes',\n",
              " ' Stripp',\n",
              " ' Hierarchical',\n",
              " ' Proc',\n",
              " ' Am',\n",
              " ' Thorac',\n",
              " ' Soc',\n",
              " ' Yanger',\n",
              " ' Stanger',\n",
              " ' Facultative',\n",
              " ' Dev',\n",
              " ' Dyn',\n",
              " ' Zhang',\n",
              " ' Li',\n",
              " ' Lee',\n",
              " ' Dedifferentiation',\n",
              " ' PLo',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " ' Bertoncello',\n",
              " ' Mc',\n",
              " ' Kubin',\n",
              " ' Poling',\n",
              " ' Kostin',\n",
              " ' Oncostatin',\n",
              " ' Cell',\n",
              " ' Cell',\n",
              " ' Till',\n",
              " ' Regulation',\n",
              " ' In',\n",
              " ' Carnie',\n",
              " ' Lala',\n",
              " ' Osmond',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Renewing',\n",
              " ' Cell',\n",
              " ' Populations',\n",
              " ' Press',\n",
              " ' Inc',\n",
              " ' New',\n",
              " ' York',\n",
              " ' Lensch',\n",
              " ' An',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Rev',\n",
              " ' Quesenberry',\n",
              " ' Dooner',\n",
              " ' Goldberg',\n",
              " ' Trans',\n",
              " ' Am',\n",
              " ' Clin',\n",
              " ' Cli',\n",
              " ' Assoc',\n",
              " ' Schier',\n",
              " ' Nodal',\n",
              " ' Ann',\n",
              " ' Rev',\n",
              " ' Biol',\n",
              " ' Wolf',\n",
              " ' Trentin',\n",
              " ' Hemopoietic',\n",
              " ' Effect',\n",
              " ' Exp',\n",
              " ' Med',\n",
              " ' Scho',\n",
              " ' The',\n",
              " ' Blood',\n",
              " ' Cells',\n",
              " ' Nelson',\n",
              " ' Bissell',\n",
              " ' Of',\n",
              " ' Ann',\n",
              " ' Rev',\n",
              " ' Dev',\n",
              " ' Biol',\n",
              " ' Morrison',\n",
              " ' Spradling',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Scadden',\n",
              " ' The',\n",
              " ' Nature',\n",
              " ' Chen',\n",
              " ' Lewallen',\n",
              " ' Xie',\n",
              " ' Adhesion',\n",
              " ' Development',\n",
              " ' Brody',\n",
              " ' Atkinson',\n",
              " ' Epithelial',\n",
              " ' In',\n",
              " ' Molecular',\n",
              " ' Pathology',\n",
              " ' Lung',\n",
              " ' Diseases',\n",
              " ' Springer',\n",
              " ' York',\n",
              " ' Holgate',\n",
              " ' Davies',\n",
              " ' Lackie',\n",
              " ' Epithelial',\n",
              " ' Allergy',\n",
              " ' Clin',\n",
              " ' Immun',\n",
              " ' Holgate',\n",
              " ' Davies',\n",
              " ' Rethinking',\n",
              " ' Shi',\n",
              " ' Xu',\n",
              " ' Warburton',\n",
              " ' Development',\n",
              " ' Respirology',\n",
              " ' Buckley',\n",
              " ' Shi',\n",
              " ' Carraro',\n",
              " ' The',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Bryder',\n",
              " ' Rossi',\n",
              " ' Weissman',\n",
              " ' Hematopoietic',\n",
              " ' Am',\n",
              " ' Pathol',\n",
              " ' Shackleton',\n",
              " ' Vaillant',\n",
              " ' Simpson',\n",
              " ' Generation',\n",
              " ' Nature',\n",
              " ' Stingl',\n",
              " ' Eirew',\n",
              " ' Ricketson',\n",
              " ' Puri',\n",
              " ' Nature',\n",
              " ' Raiser',\n",
              " ' Kim',\n",
              " ' Commentary',\n",
              " ' Sca',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Alexander',\n",
              " ' Puchalski',\n",
              " ' Klos',\n",
              " ' Separating',\n",
              " ' Cell',\n",
              " ' Cell',\n",
              " ' Holmes',\n",
              " ' Stanford',\n",
              " ' Concise',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Grassinger',\n",
              " ' Haylock',\n",
              " ' Williams',\n",
              " ' Phenotypically',\n",
              " ' Blood',\n",
              " ' Roobrouck',\n",
              " ' Vanuytsel',\n",
              " ' Verfaillie',\n",
              " ' Concise',\n",
              " ' Cells',\n",
              " ' Wion',\n",
              " ' Christen',\n",
              " ' Barbier',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Csaszar',\n",
              " ' Kirouac',\n",
              " ' Yu',\n",
              " ' Rapid',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Unwin',\n",
              " ' Smith',\n",
              " ' Blinco',\n",
              " ' Quantitative',\n",
              " ' Blood',\n",
              " ' Unwin',\n",
              " ' Whetton',\n",
              " ' Systematic',\n",
              " ' Cell',\n",
              " ' Cycle',\n",
              " ' Habibian',\n",
              " ' Peters',\n",
              " ' Hsieh',\n",
              " ' The',\n",
              " ' Exp',\n",
              " ' Med',\n",
              " ' Matute',\n",
              " '-Bello',\n",
              " ' Downey',\n",
              " ' Moore',\n",
              " ' An',\n",
              " ' Ameri',\n",
              " ' Thoracic',\n",
              " ' Society',\n",
              " ' Am',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Rawlins',\n",
              " ' Perl',\n",
              " ' The',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Baron',\n",
              " ' Choi',\n",
              " ' Owen',\n",
              " ' Genetically',\n",
              " ' Am',\n",
              " ' Lung',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Morrisey',\n",
              " ' Hogan',\n",
              " ' Preparing',\n",
              " ' Dev',\n",
              " ' Cell',\n",
              " ' Weiss',\n",
              " ' Bertoncello',\n",
              " ' Borok',\n",
              " ' Stem',\n",
              " ' Proc',\n",
              " ' Am',\n",
              " ' Thorac',\n",
              " ' Soc',\n",
              " ' Weibel',\n",
              " ' What',\n",
              " ' Swiss',\n",
              " ' Med',\n",
              " '.Wkly',\n",
              " ' Rackley',\n",
              " ' Stripp',\n",
              " ' Building',\n",
              " ' Clin',\n",
              " ' Invest',\n",
              " ' Kauffman',\n",
              " ' Cell',\n",
              " ' Int',\n",
              " ' Rev',\n",
              " ' Pathol',\n",
              " ' Bowden',\n",
              " ' Cell',\n",
              " ' Am',\n",
              " ' Rev',\n",
              " ' Respir',\n",
              " ' Dis',\n",
              " ' Beers',\n",
              " ' Morrisey',\n",
              " ' The',\n",
              " ' Clin',\n",
              " ' Invest',\n",
              " ' Bertoncello',\n",
              " ' Mc',\n",
              " ' Endogenous',\n",
              " ' Expert',\n",
              " ' Rev',\n",
              " ' Med',\n",
              " ' Rock',\n",
              " ' Hogan',\n",
              " ' Epithelial',\n",
              " ' Ann',\n",
              " ' Rev',\n",
              " ' Dev',\n",
              " ' Biol',\n",
              " ' Borthwick',\n",
              " ' Shahbazian',\n",
              " ' Krantz',\n",
              " ' Evidence',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Biol',\n",
              " ' Hong',\n",
              " ' Reynolds',\n",
              " ' Watkins',\n",
              " ' Basal',\n",
              " ' Am',\n",
              " ' Pathol',\n",
              " ' Rock',\n",
              " ' Onaitis',\n",
              " ' Rawlins',\n",
              " ' Basal',\n",
              " ' Proc',\n",
              " ' Natl',\n",
              " ' Sci',\n",
              " ' Cole',\n",
              " ' Smith',\n",
              " ' Jenkins',\n",
              " ' Tracheal',\n",
              " ' Am',\n",
              " ' Pathol',\n",
              " ' Hegab',\n",
              " ' Ha',\n",
              " ' Gilbert',\n",
              " ' Novel',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Reynolds',\n",
              " ' Giangreco',\n",
              " ' Power',\n",
              " ' Neuroepithelial',\n",
              " ' Am',\n",
              " ' Pathol',\n",
              " ' Reynolds',\n",
              " ' Hong',\n",
              " ' Giangreco',\n",
              " ' Conditional',\n",
              " ' Clara',\n",
              " ' Am',\n",
              " ' Physiol',\n",
              " ' Lung',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Giangreco',\n",
              " ' Reynolds',\n",
              " ' Stripp',\n",
              " ' Terminal',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology',\n",
              " '\\x0cEpithelial',\n",
              " ' Am',\n",
              " ' Pathol',\n",
              " ' Am',\n",
              " ' Physiol',\n",
              " ' Lung',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Kim',\n",
              " ' Jackson',\n",
              " ' Woolfenden',\n",
              " ' Identi',\n",
              " ' Nolen',\n",
              " '-Walston',\n",
              " ' Kim',\n",
              " ' Mazan',\n",
              " ' Cellular',\n",
              " ' Am',\n",
              " ' Physiol',\n",
              " ' Lung',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Zacharek',\n",
              " ' Fillmore',\n",
              " ' Lau',\n",
              " ' Lung',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Adamson',\n",
              " ' Bowden',\n",
              " ' Derivation',\n",
              " ' Lab',\n",
              " ' Invest',\n",
              " ' Evans',\n",
              " ' Cabral',\n",
              " ' Stephens',\n",
              " ' Transformation',\n",
              " ' Exp',\n",
              " ' Pathol',\n",
              " ' Giangreco',\n",
              " ' Arwert',\n",
              " ' Rosewell',\n",
              " ' Stem',\n",
              " ' Natl',\n",
              " ' Acad',\n",
              " ' Sci',\n",
              " ' Teisanu',\n",
              " ' Chen',\n",
              " ' Matsumoto',\n",
              " ' Functional',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Mc',\n",
              " ' Yuen',\n",
              " ' Williams',\n",
              " ' Evidence',\n",
              " ' Proc',\n",
              " ' Acad',\n",
              " ' Sci',\n",
              " ' Chen',\n",
              " ' Matsumoto',\n",
              " ' Brockway',\n",
              " ' Airway',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Lee',\n",
              " ' Kim',\n",
              " ' Gludish',\n",
              " ' Am',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " '[EPub',\n",
              " ' Rawlins',\n",
              " ' Okubo',\n",
              " ' Xue',\n",
              " ' The',\n",
              " ' Scgb',\n",
              " ' Clara',\n",
              " ' Cell',\n",
              " ' Stem',\n",
              " ' Cell',\n",
              " ' Ghosh',\n",
              " ' Brechbuhl',\n",
              " ' Smith',\n",
              " ' Context',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Song',\n",
              " ' Yao',\n",
              " ' Lin',\n",
              " ' Functional',\n",
              " ' Proc',\n",
              " ' Natl',\n",
              " ' Acad',\n",
              " ' Sci',\n",
              " ' Tropea',\n",
              " ' Leder',\n",
              " ' Aslam',\n",
              " ' Bronchioalveolar',\n",
              " ' Kumar',\n",
              " ' Hu',\n",
              " ' Yamamoto',\n",
              " ' Distal',\n",
              " ' Cell',\n",
              " ' Chapman',\n",
              " ' Li',\n",
              " ' Alexander',\n",
              " ' Integrin',\n",
              " ' Clin',\n",
              " ' Invest',\n",
              " ' Hynds',\n",
              " ' Giangreco',\n",
              " ' The',\n",
              " ' Cells',\n",
              " ' Bertoncello',\n",
              " ' Bradford',\n",
              " ' Surrogate',\n",
              " ' In',\n",
              " ' Garland',\n",
              " ' Quesenberry',\n",
              " ' Hilton',\n",
              " ' Stimulating',\n",
              " ' Factor',\n",
              " ' Marcel',\n",
              " ' Dekker',\n",
              " ' Inc',\n",
              " ' New',\n",
              " ' Coulombel',\n",
              " ' Identi',\n",
              " ' Oncogene',\n",
              " ' Franzdottir',\n",
              " ' Axelsson',\n",
              " ' Arason',\n",
              " ' Airway',\n",
              " ' Respir',\n",
              " ' Res',\n",
              " ' Wu',\n",
              " ' Peters',\n",
              " '-Hall',\n",
              " ' Bose',\n",
              " ' Human',\n",
              " ' Am',\n",
              " ' Respir',\n",
              " ' Cell',\n",
              " ' Mol',\n",
              " ' Biol',\n",
              " ' Hegab',\n",
              " ' Ha',\n",
              " ' Darmawan',\n",
              " ' Isolation',\n",
              " ' Stem',\n",
              " ' Cells',\n",
              " ' Med',\n",
              " ' Fessart',\n",
              " ' Begueret',\n",
              " ' Delom',\n",
              " ' Respir',\n",
              " '[Epub',\n",
              " ' Oeztuerk',\n",
              " '-Winder',\n",
              " ' Guinot',\n",
              " ' Ochalek',\n",
              " ' Regulation',\n",
              " ' Kajstura',\n",
              " ' Rota',\n",
              " ' Hall',\n",
              " ' Evidence',\n",
              " ' Engl',\n",
              " ' Med',\n",
              " ' Lung',\n",
              " ' Nat',\n",
              " ' Med',\n",
              " ' Kipling',\n",
              " ' Cooke',\n",
              " ' Hypervariable',\n",
              " ' Nature',\n",
              " ' Jungebluth',\n",
              " ' Alici',\n",
              " ' Baiguera',\n",
              " ' Tracheobronchial',\n",
              " ' Lancet',\n",
              " ' The',\n",
              " ' Authors',\n",
              " ' Asian',\n",
              " ' Paci',\n",
              " ' Society',\n",
              " ' Respirology']"
            ]
          },
          "metadata": {},
          "execution_count": 94
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "re.findall(r'([article]*/doi/10[.][0-9]{4,}[^\\s\"/<>]*/[^\\s\"<>]+)', bio[\"text\"].iloc[10])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Kid7jUtntmLW",
        "outputId": "107187e5-857e-4481-b286-07798452dc32"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[]"
            ]
          },
          "metadata": {},
          "execution_count": 98
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for i in range(len(bio)):\n",
        "    if len(re.findall(r'(/doi/10[.][0-9]{4,}.*)', bio[\"text\"].iloc[i])) > 0:\n",
        "        print(re.findall(r'(/doi/10[.][0-9]{4,}.*)', bio[\"text\"].iloc[i]))\n",
        "\n",
        "# [0-9]{4,6} -- ищем строчку, состояющую из цифр (и только из них), длина которой -- минимум 4 и максимум 6"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "QO_2Decjp-Ww",
        "outputId": "116ded8a-e0e0-4322-e7f1-ab88f373b774"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['/doi/10.1242/dev.02732.']\n",
            "['/doi/10.1242/bio.038232.supplemental']\n",
            "['/doi/10.1046/j.0305-1846.2003.00525.x/abstract.']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Продолжаем работу над текстом"
      ],
      "metadata": {
        "id": "fqRvBCs8wmY7"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Tf5nmQvdkDnU"
      },
      "outputs": [],
      "source": [
        "bio['text'] = bio['text'].astype(str)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "bio_sents = sent_split(bio, \"text\")\n",
        "bio_sents"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "hc_CWaUSAx2J",
        "outputId": "ffd1f473-c39b-4f89-ebda-c32c0f7f4fa3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                     text\n",
              "0       bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...\n",
              "1       IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...\n",
              "2       In this\\nreview, we describe the attributes of...\n",
              "3       We describe the power and\\nlimitations of expe...\n",
              "4       The review summarizes\\nrecent progress and obs...\n",
              "...                                                   ...\n",
              "121159                   Sheth RD, Jin M, Bhut BV, et al.\n",
              "121160  (2014) Affinity precipitation of a monoclonal ...\n",
              "121161               Biotechnol  Bioeng \\n111: 1595–1603.\n",
              "121162   © 2015 Rachel Chen, et al., licensee AIMS Press.\n",
              "121163  This is an open \\naccess  article  distributed...\n",
              "\n",
              "[121164 rows x 1 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-105a21ac-addf-40d5-af23-0719461965cd\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>In this\\nreview, we describe the attributes of...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>We describe the power and\\nlimitations of expe...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>The review summarizes\\nrecent progress and obs...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121159</th>\n",
              "      <td>Sheth RD, Jin M, Bhut BV, et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121160</th>\n",
              "      <td>(2014) Affinity precipitation of a monoclonal ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121161</th>\n",
              "      <td>Biotechnol  Bioeng \\n111: 1595–1603.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121162</th>\n",
              "      <td>© 2015 Rachel Chen, et al., licensee AIMS Press.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121163</th>\n",
              "      <td>This is an open \\naccess  article  distributed...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>121164 rows × 1 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-105a21ac-addf-40d5-af23-0719461965cd')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-105a21ac-addf-40d5-af23-0719461965cd button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-105a21ac-addf-40d5-af23-0719461965cd');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-ab4af3d2-7345-4189-86d3-180076db986d\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ab4af3d2-7345-4189-86d3-180076db986d')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-ab4af3d2-7345-4189-86d3-180076db986d button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bio_sents"
            }
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "clean_bio = cleantext(bio_sents)\n",
        "clean_bio"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "Av6CyfBPBFwL",
        "outputId": "880ed615-b1df-4bdf-e5c7-2b0c987d201e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                     text  \\\n",
              "0       bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...   \n",
              "1       IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...   \n",
              "2       In this\\nreview, we describe the attributes of...   \n",
              "3       We describe the power and\\nlimitations of expe...   \n",
              "4       The review summarizes\\nrecent progress and obs...   \n",
              "...                                                   ...   \n",
              "121159                   Sheth RD, Jin M, Bhut BV, et al.   \n",
              "121160  (2014) Affinity precipitation of a monoclonal ...   \n",
              "121161               Biotechnol  Bioeng \\n111: 1595–1603.   \n",
              "121162   © 2015 Rachel Chen, et al., licensee AIMS Press.   \n",
              "121163  This is an open \\naccess  article  distributed...   \n",
              "\n",
              "                                             cleaned_text  \n",
              "0       bs bs banner invited review series stem cells ...  \n",
              "1       ivan bertoncello and jonathan mcqualter lung h...  \n",
              "2       in this review we describe the attributes of a...  \n",
              "3       we describe the power and limitations of exper...  \n",
              "4       the review summarizes recent progress and obst...  \n",
              "...                                                   ...  \n",
              "121159                      sheth rd jin m bhut bv et al   \n",
              "121160   affinity precipitation of a monoclonal antibo...  \n",
              "121161                                 biotechnol bioeng   \n",
              "121162             rachel chen et al licensee aims press   \n",
              "121163  this is an open access article distributed und...  \n",
              "\n",
              "[121164 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-aa648068-9271-4153-90c7-1ae2d518415a\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...</td>\n",
              "      <td>bs bs banner invited review series stem cells ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...</td>\n",
              "      <td>ivan bertoncello and jonathan mcqualter lung h...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>In this\\nreview, we describe the attributes of...</td>\n",
              "      <td>in this review we describe the attributes of a...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>We describe the power and\\nlimitations of expe...</td>\n",
              "      <td>we describe the power and limitations of exper...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>The review summarizes\\nrecent progress and obs...</td>\n",
              "      <td>the review summarizes recent progress and obst...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121159</th>\n",
              "      <td>Sheth RD, Jin M, Bhut BV, et al.</td>\n",
              "      <td>sheth rd jin m bhut bv et al</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121160</th>\n",
              "      <td>(2014) Affinity precipitation of a monoclonal ...</td>\n",
              "      <td>affinity precipitation of a monoclonal antibo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121161</th>\n",
              "      <td>Biotechnol  Bioeng \\n111: 1595–1603.</td>\n",
              "      <td>biotechnol bioeng</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121162</th>\n",
              "      <td>© 2015 Rachel Chen, et al., licensee AIMS Press.</td>\n",
              "      <td>rachel chen et al licensee aims press</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121163</th>\n",
              "      <td>This is an open \\naccess  article  distributed...</td>\n",
              "      <td>this is an open access article distributed und...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>121164 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-aa648068-9271-4153-90c7-1ae2d518415a')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-aa648068-9271-4153-90c7-1ae2d518415a button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-aa648068-9271-4153-90c7-1ae2d518415a');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-9cfc5792-7491-42f2-88ca-4e35b0bb8242\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-9cfc5792-7491-42f2-88ca-4e35b0bb8242')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-9cfc5792-7491-42f2-88ca-4e35b0bb8242 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bio_sents"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "clean_bio.cleaned_text.iloc[1] #просто привели предложения в нормальный вид и выкинули пунктуацию + \"непонятные\" символы"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 70
        },
        "id": "wDDwVKpvBX22",
        "outputId": "b300d2ca-6972-4183-9bd0-00bb2b6a63a8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'ivan bertoncello and jonathan mcqualter lung health research centre department of pharmacology lung health research centre university of melbourne melbourne victoria australia abstract recognition of the potential of stem cell based therapies for alleviating intractable lung diseases has provided the impetus for research aimed at identifying regenerative cells in the adult lung understanding how they are organized and regulated and how they could be harnessed in lung regenerative medicine '"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "remove_biology_terms(clean_bio.iloc[1].cleaned_text) # удаляем биологические термины,\n",
        "# оставляем при этом академическую лексику - слова литературного языка, которые в научном контексте не только приобретают иную семантику,\n",
        "# но также являются \"субстратом для создания новых терминов\" (Лейчик)\n",
        "# НЕ удаляем / удаляем стоп-слова"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        },
        "id": "7fRXmglnDkw-",
        "outputId": "48e8a6f3-750d-4c40-fc8d-84481a976ded"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'jonathan health research centre department health research centre university victoria australia potential based alleviating intractable diseases provided impetus research aimed identifying adult understanding could regenerative medicine'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "jK9BmaX5uzpg",
        "outputId": "f2fa9c88-f4db-4900-ee0c-5e1ee9ce2e20"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                     text  \\\n",
              "0       bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...   \n",
              "1       IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...   \n",
              "2       In this\\nreview, we describe the attributes of...   \n",
              "3       We describe the power and\\nlimitations of expe...   \n",
              "4       The review summarizes\\nrecent progress and obs...   \n",
              "...                                                   ...   \n",
              "121159                   Sheth RD, Jin M, Bhut BV, et al.   \n",
              "121160  (2014) Affinity precipitation of a monoclonal ...   \n",
              "121161               Biotechnol  Bioeng \\n111: 1595–1603.   \n",
              "121162   © 2015 Rachel Chen, et al., licensee AIMS Press.   \n",
              "121163  This is an open \\naccess  article  distributed...   \n",
              "\n",
              "                                             cleaned_text  \\\n",
              "0       bs bs banner invited review series stem cells ...   \n",
              "1       ivan bertoncello and jonathan mcqualter lung h...   \n",
              "2       in this review we describe the attributes of a...   \n",
              "3       we describe the power and limitations of exper...   \n",
              "4       the review summarizes recent progress and obst...   \n",
              "...                                                   ...   \n",
              "121159                      sheth rd jin m bhut bv et al    \n",
              "121160   affinity precipitation of a monoclonal antibo...   \n",
              "121161                                 biotechnol bioeng    \n",
              "121162             rachel chen et al licensee aims press    \n",
              "121163  this is an open access article distributed und...   \n",
              "\n",
              "                                       removed_terms_text  \n",
              "0       bs bs banner invited review series stem cells ...  \n",
              "1       and jonathan health research centre department...  \n",
              "2       in this review we describe the of adult stem a...  \n",
              "3       we describe the power and of strategies and us...  \n",
              "4       the review summarizes recent progress and in d...  \n",
              "...                                                   ...  \n",
              "121159                                        rd bv et al  \n",
              "121160    affinity precipitation of a from an harvest ...  \n",
              "121161                                                     \n",
              "121162                                    chen et al aims  \n",
              "121163  this is an open access article distributed und...  \n",
              "\n",
              "[121164 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7f82e8c2-781b-40ad-a9ab-837b316ba2ef\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>removed_terms_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...</td>\n",
              "      <td>bs bs banner invited review series stem cells ...</td>\n",
              "      <td>bs bs banner invited review series stem cells ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...</td>\n",
              "      <td>ivan bertoncello and jonathan mcqualter lung h...</td>\n",
              "      <td>and jonathan health research centre department...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>In this\\nreview, we describe the attributes of...</td>\n",
              "      <td>in this review we describe the attributes of a...</td>\n",
              "      <td>in this review we describe the of adult stem a...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>We describe the power and\\nlimitations of expe...</td>\n",
              "      <td>we describe the power and limitations of exper...</td>\n",
              "      <td>we describe the power and of strategies and us...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>The review summarizes\\nrecent progress and obs...</td>\n",
              "      <td>the review summarizes recent progress and obst...</td>\n",
              "      <td>the review summarizes recent progress and in d...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121159</th>\n",
              "      <td>Sheth RD, Jin M, Bhut BV, et al.</td>\n",
              "      <td>sheth rd jin m bhut bv et al</td>\n",
              "      <td>rd bv et al</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121160</th>\n",
              "      <td>(2014) Affinity precipitation of a monoclonal ...</td>\n",
              "      <td>affinity precipitation of a monoclonal antibo...</td>\n",
              "      <td>affinity precipitation of a from an harvest ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121161</th>\n",
              "      <td>Biotechnol  Bioeng \\n111: 1595–1603.</td>\n",
              "      <td>biotechnol bioeng</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121162</th>\n",
              "      <td>© 2015 Rachel Chen, et al., licensee AIMS Press.</td>\n",
              "      <td>rachel chen et al licensee aims press</td>\n",
              "      <td>chen et al aims</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121163</th>\n",
              "      <td>This is an open \\naccess  article  distributed...</td>\n",
              "      <td>this is an open access article distributed und...</td>\n",
              "      <td>this is an open access article distributed und...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>121164 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7f82e8c2-781b-40ad-a9ab-837b316ba2ef')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-7f82e8c2-781b-40ad-a9ab-837b316ba2ef button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-7f82e8c2-781b-40ad-a9ab-837b316ba2ef');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-5778b009-5219-4415-a0a1-f0a7668dff01\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5778b009-5219-4415-a0a1-f0a7668dff01')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-5778b009-5219-4415-a0a1-f0a7668dff01 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bio_sents"
            }
          },
          "metadata": {},
          "execution_count": 18
        }
      ],
      "source": [
        "clean_bio[\"removed_terms_text\"] = clean_bio['cleaned_text'].apply(remove_biology_terms)\n",
        "clean_bio"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "koo0dkGawVrt",
        "outputId": "3a158fad-a5a8-4bf3-9955-0e2a66409c19"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                     text  \\\n",
              "0       bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...   \n",
              "1       IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...   \n",
              "2       In this\\nreview, we describe the attributes of...   \n",
              "3       We describe the power and\\nlimitations of expe...   \n",
              "4       The review summarizes\\nrecent progress and obs...   \n",
              "...                                                   ...   \n",
              "121159                   Sheth RD, Jin M, Bhut BV, et al.   \n",
              "121160  (2014) Affinity precipitation of a monoclonal ...   \n",
              "121161               Biotechnol  Bioeng \\n111: 1595–1603.   \n",
              "121162   © 2015 Rachel Chen, et al., licensee AIMS Press.   \n",
              "121163  This is an open \\naccess  article  distributed...   \n",
              "\n",
              "                                             cleaned_text  \\\n",
              "0       bs bs banner invited review series stem cells ...   \n",
              "1       ivan bertoncello and jonathan mcqualter lung h...   \n",
              "2       in this review we describe the attributes of a...   \n",
              "3       we describe the power and limitations of exper...   \n",
              "4       the review summarizes recent progress and obst...   \n",
              "...                                                   ...   \n",
              "121159                      sheth rd jin m bhut bv et al    \n",
              "121160   affinity precipitation of a monoclonal antibo...   \n",
              "121161                                 biotechnol bioeng    \n",
              "121162             rachel chen et al licensee aims press    \n",
              "121163  this is an open access article distributed und...   \n",
              "\n",
              "                                       removed_terms_text  len  \n",
              "0       bs bs banner invited review series stem cells ...  140  \n",
              "1       and jonathan health research centre department...  494  \n",
              "2       in this review we describe the of adult stem a...  191  \n",
              "3       we describe the power and of strategies and us...  214  \n",
              "4       the review summarizes recent progress and in d...  161  \n",
              "...                                                   ...  ...  \n",
              "121159                                        rd bv et al   29  \n",
              "121160    affinity precipitation of a from an harvest ...  131  \n",
              "121161                                                      18  \n",
              "121162                                    chen et al aims   39  \n",
              "121163  this is an open access article distributed und...  173  \n",
              "\n",
              "[120445 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5f5b20cc-7050-4464-b3ac-3cb0363f68d9\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>removed_terms_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>bs_bs_banner\\n\\nINVITED REVIEW SERIES:\\nSTEM C...</td>\n",
              "      <td>bs bs banner invited review series stem cells ...</td>\n",
              "      <td>bs bs banner invited review series stem cells ...</td>\n",
              "      <td>140</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>IVAN BERTONCELLO AND JONATHAN L. MCQUALTER\\n\\n...</td>\n",
              "      <td>ivan bertoncello and jonathan mcqualter lung h...</td>\n",
              "      <td>and jonathan health research centre department...</td>\n",
              "      <td>494</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>In this\\nreview, we describe the attributes of...</td>\n",
              "      <td>in this review we describe the attributes of a...</td>\n",
              "      <td>in this review we describe the of adult stem a...</td>\n",
              "      <td>191</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>We describe the power and\\nlimitations of expe...</td>\n",
              "      <td>we describe the power and limitations of exper...</td>\n",
              "      <td>we describe the power and of strategies and us...</td>\n",
              "      <td>214</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>The review summarizes\\nrecent progress and obs...</td>\n",
              "      <td>the review summarizes recent progress and obst...</td>\n",
              "      <td>the review summarizes recent progress and in d...</td>\n",
              "      <td>161</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121159</th>\n",
              "      <td>Sheth RD, Jin M, Bhut BV, et al.</td>\n",
              "      <td>sheth rd jin m bhut bv et al</td>\n",
              "      <td>rd bv et al</td>\n",
              "      <td>29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121160</th>\n",
              "      <td>(2014) Affinity precipitation of a monoclonal ...</td>\n",
              "      <td>affinity precipitation of a monoclonal antibo...</td>\n",
              "      <td>affinity precipitation of a from an harvest ...</td>\n",
              "      <td>131</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121161</th>\n",
              "      <td>Biotechnol  Bioeng \\n111: 1595–1603.</td>\n",
              "      <td>biotechnol bioeng</td>\n",
              "      <td></td>\n",
              "      <td>18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121162</th>\n",
              "      <td>© 2015 Rachel Chen, et al., licensee AIMS Press.</td>\n",
              "      <td>rachel chen et al licensee aims press</td>\n",
              "      <td>chen et al aims</td>\n",
              "      <td>39</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121163</th>\n",
              "      <td>This is an open \\naccess  article  distributed...</td>\n",
              "      <td>this is an open access article distributed und...</td>\n",
              "      <td>this is an open access article distributed und...</td>\n",
              "      <td>173</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>120445 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5f5b20cc-7050-4464-b3ac-3cb0363f68d9')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5f5b20cc-7050-4464-b3ac-3cb0363f68d9 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5f5b20cc-7050-4464-b3ac-3cb0363f68d9');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-896fcd07-6a35-4d5b-8512-c8652c821757\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-896fcd07-6a35-4d5b-8512-c8652c821757')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-896fcd07-6a35-4d5b-8512-c8652c821757 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "prep_bio"
            }
          },
          "metadata": {},
          "execution_count": 19
        }
      ],
      "source": [
        "prep_bio = clean_long_sent(clean_bio)\n",
        "prep_bio"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "mkQv6rGo4bqS"
      },
      "outputs": [],
      "source": [
        "prep_bio.to_csv('prep_SW_BIO.csv')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# KJV BIBLE\n",
        "King James Bible, 1611, канонический перевод Библии, обычно называемый Авторизованной версией (the Authorized Version).\n",
        "\n",
        "Отказались от использования этой версии перевода Библии (несмотря на все еще лидирующую популярность King James Version), так как она содержит большое количество архаизмов, встретить которые в современных научно-полпулярных текстах мы не ожидаем."
      ],
      "metadata": {
        "id": "I71RGBalmKdk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "import re\n",
        "import string\n",
        "import nltk\n",
        "from nltk.corpus import stopwords\n",
        "\n",
        "nltk.download(\"stopwords\")\n",
        "stops = stopwords.words(\"english\")\n",
        "\n",
        "nltk.download('punkt')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2U0nJ7FScyif",
        "outputId": "d4b304e7-5a98-4b50-b644-70fe90057028"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package stopwords to /root/nltk_data...\n",
            "[nltk_data]   Unzipping corpora/stopwords.zip.\n",
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "bible = pd.read_csv(\"/content/KJV.csv\")\n",
        "bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 545
        },
        "id": "8n33UBaND2GY",
        "outputId": "90470b4f-a975-4d46-a3a9-a8c1197762ac"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Unnamed: 0          citation        book  chapter  verse  \\\n",
              "0               0       Genesis 1:1     Genesis        1      1   \n",
              "1               1       Genesis 1:2     Genesis        1      2   \n",
              "2               2       Genesis 1:3     Genesis        1      3   \n",
              "3               3       Genesis 1:4     Genesis        1      4   \n",
              "4               4       Genesis 1:5     Genesis        1      5   \n",
              "...           ...               ...         ...      ...    ...   \n",
              "31097       31097  Revelation 22:17  Revelation       22     17   \n",
              "31098       31098  Revelation 22:18  Revelation       22     18   \n",
              "31099       31099  Revelation 22:19  Revelation       22     19   \n",
              "31100       31100  Revelation 22:20  Revelation       22     20   \n",
              "31101       31101  Revelation 22:21  Revelation       22     21   \n",
              "\n",
              "                                                    text  book_code  \\\n",
              "0      In the beginning God created the heaven and th...          1   \n",
              "1      And the earth was without form, and void; and ...          1   \n",
              "2      And God said, Let there be light: and there wa...          1   \n",
              "3      And God saw the light, that it was good: and G...          1   \n",
              "4      And God called the light Day, and the darkness...          1   \n",
              "...                                                  ...        ...   \n",
              "31097  And the Spirit and the bride say, Come. And le...         66   \n",
              "31098  For I testify unto every man that heareth the ...         66   \n",
              "31099  And if any man shall take away from the words ...         66   \n",
              "31100  He which testifieth these things saith, Surely...         66   \n",
              "31101  The grace of our Lord Jesus Christ be with you...         66   \n",
              "\n",
              "           testament  testament_code  \n",
              "0      Old Testament               0  \n",
              "1      Old Testament               0  \n",
              "2      Old Testament               0  \n",
              "3      Old Testament               0  \n",
              "4      Old Testament               0  \n",
              "...              ...             ...  \n",
              "31097  New Testament               1  \n",
              "31098  New Testament               1  \n",
              "31099  New Testament               1  \n",
              "31100  New Testament               1  \n",
              "31101  New Testament               1  \n",
              "\n",
              "[31102 rows x 9 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6a59e35b-2533-4196-808f-540769d97e51\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>citation</th>\n",
              "      <th>book</th>\n",
              "      <th>chapter</th>\n",
              "      <th>verse</th>\n",
              "      <th>text</th>\n",
              "      <th>book_code</th>\n",
              "      <th>testament</th>\n",
              "      <th>testament_code</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Genesis 1:1</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>In the beginning God created the heaven and th...</td>\n",
              "      <td>1</td>\n",
              "      <td>Old Testament</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Genesis 1:2</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>And the earth was without form, and void; and ...</td>\n",
              "      <td>1</td>\n",
              "      <td>Old Testament</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Genesis 1:3</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>And God said, Let there be light: and there wa...</td>\n",
              "      <td>1</td>\n",
              "      <td>Old Testament</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Genesis 1:4</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>And God saw the light, that it was good: and G...</td>\n",
              "      <td>1</td>\n",
              "      <td>Old Testament</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Genesis 1:5</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>5</td>\n",
              "      <td>And God called the light Day, and the darkness...</td>\n",
              "      <td>1</td>\n",
              "      <td>Old Testament</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31097</th>\n",
              "      <td>31097</td>\n",
              "      <td>Revelation 22:17</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>22</td>\n",
              "      <td>17</td>\n",
              "      <td>And the Spirit and the bride say, Come. And le...</td>\n",
              "      <td>66</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31098</th>\n",
              "      <td>31098</td>\n",
              "      <td>Revelation 22:18</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>22</td>\n",
              "      <td>18</td>\n",
              "      <td>For I testify unto every man that heareth the ...</td>\n",
              "      <td>66</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31099</th>\n",
              "      <td>31099</td>\n",
              "      <td>Revelation 22:19</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>22</td>\n",
              "      <td>19</td>\n",
              "      <td>And if any man shall take away from the words ...</td>\n",
              "      <td>66</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31100</th>\n",
              "      <td>31100</td>\n",
              "      <td>Revelation 22:20</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>22</td>\n",
              "      <td>20</td>\n",
              "      <td>He which testifieth these things saith, Surely...</td>\n",
              "      <td>66</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31101</th>\n",
              "      <td>31101</td>\n",
              "      <td>Revelation 22:21</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>22</td>\n",
              "      <td>21</td>\n",
              "      <td>The grace of our Lord Jesus Christ be with you...</td>\n",
              "      <td>66</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>31102 rows × 9 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6a59e35b-2533-4196-808f-540769d97e51')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6a59e35b-2533-4196-808f-540769d97e51 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6a59e35b-2533-4196-808f-540769d97e51');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-f8e129a1-1740-44cf-92bf-5977a08ebb0b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f8e129a1-1740-44cf-92bf-5977a08ebb0b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-f8e129a1-1740-44cf-92bf-5977a08ebb0b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bible",
              "summary": "{\n  \"name\": \"bible\",\n  \"rows\": 31102,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 8978,\n        \"min\": 0,\n        \"max\": 31101,\n        \"num_unique_values\": 31102,\n        \"samples\": [\n          1520,\n          30806,\n          2932\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"citation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 31102,\n        \"samples\": [\n          \"Genesis 50:14\",\n          \"Revelation 6:13\",\n          \"Leviticus 8:15\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"book\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 66,\n        \"samples\": [\n          \"2 Timothy\",\n          \"2 John\",\n          \"Genesis\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chapter\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 23,\n        \"min\": 1,\n        \"max\": 150,\n        \"num_unique_values\": 150,\n        \"samples\": [\n          74,\n          19,\n          119\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"verse\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 14,\n        \"min\": 1,\n        \"max\": 176,\n        \"num_unique_values\": 176,\n        \"samples\": [\n          20,\n          46,\n          140\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 30834,\n        \"samples\": [\n          \"I will arise and go to my father, and will say unto him, Father, I have sinned against heaven, and before thee, \\n\",\n          \"But of these who seemed to be somewhat, (whatsoever they were, it maketh no matter to me: God accepteth no man's person:) for they who seemed to be somewhat in conference added nothing to me: \\n\",\n          \"So Jesus came again into Cana of Galilee, where he made the water wine. And there was a certain nobleman, whose son was sick at Capernaum. \\n\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"book_code\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 16,\n        \"min\": 1,\n        \"max\": 66,\n        \"num_unique_values\": 66,\n        \"samples\": [\n          55,\n          63,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"testament\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"New Testament\",\n          \"Old Testament\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"testament_code\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 0,\n        \"max\": 1,\n        \"num_unique_values\": 2,\n        \"samples\": [\n          1,\n          0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Включаем в \"библейский\" корпус только New Testament. В самом начале исследования была идея использовать только NT, однако в процессе мы от этого отказались, включив в \"библейский\" корпус оба Завета. Кроме того, решили использовать American Standard Version."
      ],
      "metadata": {
        "id": "0eu5HZIaDo-w"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "bible = bible[bible.testament_code == 1]\n",
        "bible.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "10sXZ48yErJY",
        "outputId": "e6663b19-e123-4a96-ff38-ba50bfefe1c6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Unnamed: 0     citation     book  chapter  verse  \\\n",
              "23145       23145  Matthew 1:1  Matthew        1      1   \n",
              "23146       23146  Matthew 1:2  Matthew        1      2   \n",
              "23147       23147  Matthew 1:3  Matthew        1      3   \n",
              "23148       23148  Matthew 1:4  Matthew        1      4   \n",
              "23149       23149  Matthew 1:5  Matthew        1      5   \n",
              "\n",
              "                                                    text  book_code  \\\n",
              "23145  The book of the generation of Jesus Christ, th...         40   \n",
              "23146  Abraham begat Isaac; and Isaac begat Jacob; an...         40   \n",
              "23147  And Judas begat Phares and Zara of Thamar; and...         40   \n",
              "23148  And Aram begat Aminadab; and Aminadab begat Na...         40   \n",
              "23149  And Salmon begat Booz of Rachab; and Booz bega...         40   \n",
              "\n",
              "           testament  testament_code  \n",
              "23145  New Testament               1  \n",
              "23146  New Testament               1  \n",
              "23147  New Testament               1  \n",
              "23148  New Testament               1  \n",
              "23149  New Testament               1  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ee9d28e0-c8aa-4df5-b1bf-307347121329\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>citation</th>\n",
              "      <th>book</th>\n",
              "      <th>chapter</th>\n",
              "      <th>verse</th>\n",
              "      <th>text</th>\n",
              "      <th>book_code</th>\n",
              "      <th>testament</th>\n",
              "      <th>testament_code</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>23145</th>\n",
              "      <td>23145</td>\n",
              "      <td>Matthew 1:1</td>\n",
              "      <td>Matthew</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>The book of the generation of Jesus Christ, th...</td>\n",
              "      <td>40</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23146</th>\n",
              "      <td>23146</td>\n",
              "      <td>Matthew 1:2</td>\n",
              "      <td>Matthew</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>Abraham begat Isaac; and Isaac begat Jacob; an...</td>\n",
              "      <td>40</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23147</th>\n",
              "      <td>23147</td>\n",
              "      <td>Matthew 1:3</td>\n",
              "      <td>Matthew</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>And Judas begat Phares and Zara of Thamar; and...</td>\n",
              "      <td>40</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23148</th>\n",
              "      <td>23148</td>\n",
              "      <td>Matthew 1:4</td>\n",
              "      <td>Matthew</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>And Aram begat Aminadab; and Aminadab begat Na...</td>\n",
              "      <td>40</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23149</th>\n",
              "      <td>23149</td>\n",
              "      <td>Matthew 1:5</td>\n",
              "      <td>Matthew</td>\n",
              "      <td>1</td>\n",
              "      <td>5</td>\n",
              "      <td>And Salmon begat Booz of Rachab; and Booz bega...</td>\n",
              "      <td>40</td>\n",
              "      <td>New Testament</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ee9d28e0-c8aa-4df5-b1bf-307347121329')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ee9d28e0-c8aa-4df5-b1bf-307347121329 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ee9d28e0-c8aa-4df5-b1bf-307347121329');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-4f29d6fb-66b0-4291-9273-ddb6a7d3d351\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-4f29d6fb-66b0-4291-9273-ddb6a7d3d351')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-4f29d6fb-66b0-4291-9273-ddb6a7d3d351 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bible",
              "summary": "{\n  \"name\": \"bible\",\n  \"rows\": 7957,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2297,\n        \"min\": 23145,\n        \"max\": 31101,\n        \"num_unique_values\": 7957,\n        \"samples\": [\n          28729,\n          23882,\n          25462\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"citation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 7957,\n        \"samples\": [\n          \"1 Corinthians 15:11\",\n          \"Matthew 22:10\",\n          \"Luke 12:3\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"book\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 27,\n        \"samples\": [\n          \"Galatians\",\n          \"2 Thessalonians\",\n          \"Ephesians\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chapter\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 7,\n        \"min\": 1,\n        \"max\": 28,\n        \"num_unique_values\": 28,\n        \"samples\": [\n          10,\n          26,\n          9\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"verse\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 13,\n        \"min\": 1,\n        \"max\": 80,\n        \"num_unique_values\": 80,\n        \"samples\": [\n          31,\n          1,\n          23\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 7938,\n        \"samples\": [\n          \"And shall go out to deceive the nations which are in the four quarters of the earth, Gog, and Magog, to gather them together to battle: the number of whom is as the sand of the sea. \\n\",\n          \"Lord, they have killed thy prophets, and digged down thine altars; and I am left alone, and they seek my life. \\n\",\n          \"And I heard a great voice out of the temple saying to the seven angels, Go your ways, and pour out the vials of the wrath of God upon the earth. \\n\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"book_code\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 7,\n        \"min\": 40,\n        \"max\": 66,\n        \"num_unique_values\": 27,\n        \"samples\": [\n          48,\n          53,\n          49\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"testament\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"New Testament\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"testament_code\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "C6jcN2SpESNE"
      },
      "outputs": [],
      "source": [
        "def sent_split(df): # просто разбили на предложения!!!\n",
        "    sents = []\n",
        "    for _, row in df.iterrows():\n",
        "        text = row.text\n",
        "        for sent in nltk.sent_tokenize(text):\n",
        "            sents.append(sent)\n",
        "    sents = pd.DataFrame(sents).rename({0: 'text'}, axis=1)\n",
        "    return sents"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "bible_sents = sent_split(bible)"
      ],
      "metadata": {
        "id": "9GJiK5WrEbfo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QLaeHgY-ESNF"
      },
      "outputs": [],
      "source": [
        "def clean_bible_text(df):\n",
        "    # приводим к нижнему регистру\n",
        "    df['cleaned_text'] = df['text'].str.lower()\n",
        "\n",
        "    # удаляем символы\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'[^a-zA-Z0-9]', \" \", regex=True)\n",
        "\n",
        "    # Удаление цифр\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\d+', '', regex=True)\n",
        "    return df"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible = clean_bible_text(bible_sents)"
      ],
      "metadata": {
        "id": "jf2KF8aUGn9q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "9TQiKs9VHYm2",
        "outputId": "b504d6f4-ebe2-4288-a3d1-d07699f567d8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   text  \\\n",
              "0     The book of the generation of Jesus Christ, th...   \n",
              "1     Abraham begat Isaac; and Isaac begat Jacob; an...   \n",
              "2     And Judas begat Phares and Zara of Thamar; and...   \n",
              "3     And Aram begat Aminadab; and Aminadab begat Na...   \n",
              "4     And Salmon begat Booz of Rachab; and Booz bega...   \n",
              "...                                                 ...   \n",
              "9025  He which testifieth these things saith, Surely...   \n",
              "9026                                              Amen.   \n",
              "9027                         Even so, come, Lord Jesus.   \n",
              "9028  The grace of our Lord Jesus Christ be with you...   \n",
              "9029                                              Amen.   \n",
              "\n",
              "                                           cleaned_text  \n",
              "0     the book of the generation of jesus christ  th...  \n",
              "1     abraham begat isaac  and isaac begat jacob  an...  \n",
              "2     and judas begat phares and zara of thamar  and...  \n",
              "3     and aram begat aminadab  and aminadab begat na...  \n",
              "4     and salmon begat booz of rachab  and booz bega...  \n",
              "...                                                 ...  \n",
              "9025  he which testifieth these things saith  surely...  \n",
              "9026                                              amen   \n",
              "9027                         even so  come  lord jesus   \n",
              "9028  the grace of our lord jesus christ be with you...  \n",
              "9029                                              amen   \n",
              "\n",
              "[9030 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-708b6b8d-ce71-45a3-9998-3fdefcfecb96\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>The book of the generation of Jesus Christ, th...</td>\n",
              "      <td>the book of the generation of jesus christ  th...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Abraham begat Isaac; and Isaac begat Jacob; an...</td>\n",
              "      <td>abraham begat isaac  and isaac begat jacob  an...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And Judas begat Phares and Zara of Thamar; and...</td>\n",
              "      <td>and judas begat phares and zara of thamar  and...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And Aram begat Aminadab; and Aminadab begat Na...</td>\n",
              "      <td>and aram begat aminadab  and aminadab begat na...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And Salmon begat Booz of Rachab; and Booz bega...</td>\n",
              "      <td>and salmon begat booz of rachab  and booz bega...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9025</th>\n",
              "      <td>He which testifieth these things saith, Surely...</td>\n",
              "      <td>he which testifieth these things saith  surely...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9026</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9027</th>\n",
              "      <td>Even so, come, Lord Jesus.</td>\n",
              "      <td>even so  come  lord jesus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9028</th>\n",
              "      <td>The grace of our Lord Jesus Christ be with you...</td>\n",
              "      <td>the grace of our lord jesus christ be with you...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9029</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>9030 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-708b6b8d-ce71-45a3-9998-3fdefcfecb96')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-708b6b8d-ce71-45a3-9998-3fdefcfecb96 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-708b6b8d-ce71-45a3-9998-3fdefcfecb96');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-4b455b8d-1f5f-4a00-ae56-3d45a2af6ee8\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-4b455b8d-1f5f-4a00-ae56-3d45a2af6ee8')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-4b455b8d-1f5f-4a00-ae56-3d45a2af6ee8 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bible_sents",
              "summary": "{\n  \"name\": \"bible_sents\",\n  \"rows\": 9030,\n  \"fields\": [\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8912,\n        \"samples\": [\n          \"And said unto them, Why sleep ye?\",\n          \"And whithersoever he entered, into villages, or cities, or country, they laid the sick in the streets, and besought him that they might touch if it were but the border of his garment: and as many as touched him were made whole.\",\n          \"for ye have received your consolation.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8909,\n        \"samples\": [\n          \"and peter called to mind the word that jesus said unto him  before the cock crow twice  thou shalt deny me thrice \",\n          \"but if ye be led of the spirit  ye are not under the law \",\n          \"i am that bread of life \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import string\n",
        "\n",
        "def stopwords_remover(words):\n",
        "\n",
        "  without_stopwords = \"\"\n",
        "  for word in words.split(\" \"):\n",
        "    if word in stops or word in string.punctuation:\n",
        "      continue\n",
        "    else:\n",
        "      without_stopwords += word + \" \"\n",
        "  return without_stopwords"
      ],
      "metadata": {
        "id": "gB4NEhQVdUPr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible['prep_text'] = prep_bible['cleaned_text'].apply(stopwords_remover)\n"
      ],
      "metadata": {
        "id": "S5PT1s_8Ofjg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "KONeeHZ3fg-o",
        "outputId": "b4d7ac6e-8545-4bdf-f1cb-98bc6c80ced7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   text  \\\n",
              "0     The book of the generation of Jesus Christ, th...   \n",
              "1     Abraham begat Isaac; and Isaac begat Jacob; an...   \n",
              "2     And Judas begat Phares and Zara of Thamar; and...   \n",
              "3     And Aram begat Aminadab; and Aminadab begat Na...   \n",
              "4     And Salmon begat Booz of Rachab; and Booz bega...   \n",
              "...                                                 ...   \n",
              "9025  He which testifieth these things saith, Surely...   \n",
              "9026                                              Amen.   \n",
              "9027                         Even so, come, Lord Jesus.   \n",
              "9028  The grace of our Lord Jesus Christ be with you...   \n",
              "9029                                              Amen.   \n",
              "\n",
              "                                           cleaned_text  \\\n",
              "0     the book of the generation of jesus christ  th...   \n",
              "1     abraham begat isaac  and isaac begat jacob  an...   \n",
              "2     and judas begat phares and zara of thamar  and...   \n",
              "3     and aram begat aminadab  and aminadab begat na...   \n",
              "4     and salmon begat booz of rachab  and booz bega...   \n",
              "...                                                 ...   \n",
              "9025  he which testifieth these things saith  surely...   \n",
              "9026                                              amen    \n",
              "9027                         even so  come  lord jesus    \n",
              "9028  the grace of our lord jesus christ be with you...   \n",
              "9029                                              amen    \n",
              "\n",
              "                                              prep_text  \n",
              "0     book generation jesus christ son david son abr...  \n",
              "1     abraham begat isaac isaac begat jacob jacob be...  \n",
              "2     judas begat phares zara thamar phares begat es...  \n",
              "3     aram begat aminadab aminadab begat naasson naa...  \n",
              "4     salmon begat booz rachab booz begat obed ruth ...  \n",
              "...                                                 ...  \n",
              "9025       testifieth things saith surely come quickly   \n",
              "9026                                              amen   \n",
              "9027                              even come lord jesus   \n",
              "9028                           grace lord jesus christ   \n",
              "9029                                              amen   \n",
              "\n",
              "[9030 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3b447266-8d62-4b93-a1d5-91b0476bc935\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>The book of the generation of Jesus Christ, th...</td>\n",
              "      <td>the book of the generation of jesus christ  th...</td>\n",
              "      <td>book generation jesus christ son david son abr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Abraham begat Isaac; and Isaac begat Jacob; an...</td>\n",
              "      <td>abraham begat isaac  and isaac begat jacob  an...</td>\n",
              "      <td>abraham begat isaac isaac begat jacob jacob be...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And Judas begat Phares and Zara of Thamar; and...</td>\n",
              "      <td>and judas begat phares and zara of thamar  and...</td>\n",
              "      <td>judas begat phares zara thamar phares begat es...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And Aram begat Aminadab; and Aminadab begat Na...</td>\n",
              "      <td>and aram begat aminadab  and aminadab begat na...</td>\n",
              "      <td>aram begat aminadab aminadab begat naasson naa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And Salmon begat Booz of Rachab; and Booz bega...</td>\n",
              "      <td>and salmon begat booz of rachab  and booz bega...</td>\n",
              "      <td>salmon begat booz rachab booz begat obed ruth ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9025</th>\n",
              "      <td>He which testifieth these things saith, Surely...</td>\n",
              "      <td>he which testifieth these things saith  surely...</td>\n",
              "      <td>testifieth things saith surely come quickly</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9026</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9027</th>\n",
              "      <td>Even so, come, Lord Jesus.</td>\n",
              "      <td>even so  come  lord jesus</td>\n",
              "      <td>even come lord jesus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9028</th>\n",
              "      <td>The grace of our Lord Jesus Christ be with you...</td>\n",
              "      <td>the grace of our lord jesus christ be with you...</td>\n",
              "      <td>grace lord jesus christ</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9029</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>9030 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3b447266-8d62-4b93-a1d5-91b0476bc935')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3b447266-8d62-4b93-a1d5-91b0476bc935 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3b447266-8d62-4b93-a1d5-91b0476bc935');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-150f062c-3cff-41f6-8e94-56f2af21c950\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-150f062c-3cff-41f6-8e94-56f2af21c950')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-150f062c-3cff-41f6-8e94-56f2af21c950 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "bible_sents",
              "summary": "{\n  \"name\": \"bible_sents\",\n  \"rows\": 9030,\n  \"fields\": [\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8912,\n        \"samples\": [\n          \"And said unto them, Why sleep ye?\",\n          \"And whithersoever he entered, into villages, or cities, or country, they laid the sick in the streets, and besought him that they might touch if it were but the border of his garment: and as many as touched him were made whole.\",\n          \"for ye have received your consolation.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8909,\n        \"samples\": [\n          \"and peter called to mind the word that jesus said unto him  before the cock crow twice  thou shalt deny me thrice \",\n          \"but if ye be led of the spirit  ye are not under the law \",\n          \"i am that bread of life \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8857,\n        \"samples\": [\n          \"ye sometimes darkness ye light lord walk children light \",\n          \"taketh away first may establish second \",\n          \"carnally minded death spiritually minded life peace \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "R-u67nIxESNF"
      },
      "outputs": [],
      "source": [
        "def clean_long_sent(df):\n",
        "    df[\"len\"] = df[\"prep_text\"].apply(lambda x: len(x))\n",
        "    return df[df.len <= 512]"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible = clean_long_sent(prep_bible)\n",
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "g_Tox82kPK23",
        "outputId": "cec21f0e-02df-451f-816b-09593de34675"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   text  \\\n",
              "0     The book of the generation of Jesus Christ, th...   \n",
              "1     Abraham begat Isaac; and Isaac begat Jacob; an...   \n",
              "2     And Judas begat Phares and Zara of Thamar; and...   \n",
              "3     And Aram begat Aminadab; and Aminadab begat Na...   \n",
              "4     And Salmon begat Booz of Rachab; and Booz bega...   \n",
              "...                                                 ...   \n",
              "9025  He which testifieth these things saith, Surely...   \n",
              "9026                                              Amen.   \n",
              "9027                         Even so, come, Lord Jesus.   \n",
              "9028  The grace of our Lord Jesus Christ be with you...   \n",
              "9029                                              Amen.   \n",
              "\n",
              "                                           cleaned_text  \\\n",
              "0     the book of the generation of jesus christ  th...   \n",
              "1     abraham begat isaac  and isaac begat jacob  an...   \n",
              "2     and judas begat phares and zara of thamar  and...   \n",
              "3     and aram begat aminadab  and aminadab begat na...   \n",
              "4     and salmon begat booz of rachab  and booz bega...   \n",
              "...                                                 ...   \n",
              "9025  he which testifieth these things saith  surely...   \n",
              "9026                                              amen    \n",
              "9027                         even so  come  lord jesus    \n",
              "9028  the grace of our lord jesus christ be with you...   \n",
              "9029                                              amen    \n",
              "\n",
              "                                              prep_text  len  \n",
              "0     book generation jesus christ son david son abr...   51  \n",
              "1     abraham begat isaac isaac begat jacob jacob be...   65  \n",
              "2     judas begat phares zara thamar phares begat es...   67  \n",
              "3     aram begat aminadab aminadab begat naasson naa...   64  \n",
              "4     salmon begat booz rachab booz begat obed ruth ...   63  \n",
              "...                                                 ...  ...  \n",
              "9025       testifieth things saith surely come quickly    44  \n",
              "9026                                              amen     5  \n",
              "9027                              even come lord jesus    21  \n",
              "9028                           grace lord jesus christ    24  \n",
              "9029                                              amen     5  \n",
              "\n",
              "[9030 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b451b355-cd22-419f-b45f-ab52a32d0a93\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>The book of the generation of Jesus Christ, th...</td>\n",
              "      <td>the book of the generation of jesus christ  th...</td>\n",
              "      <td>book generation jesus christ son david son abr...</td>\n",
              "      <td>51</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Abraham begat Isaac; and Isaac begat Jacob; an...</td>\n",
              "      <td>abraham begat isaac  and isaac begat jacob  an...</td>\n",
              "      <td>abraham begat isaac isaac begat jacob jacob be...</td>\n",
              "      <td>65</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And Judas begat Phares and Zara of Thamar; and...</td>\n",
              "      <td>and judas begat phares and zara of thamar  and...</td>\n",
              "      <td>judas begat phares zara thamar phares begat es...</td>\n",
              "      <td>67</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And Aram begat Aminadab; and Aminadab begat Na...</td>\n",
              "      <td>and aram begat aminadab  and aminadab begat na...</td>\n",
              "      <td>aram begat aminadab aminadab begat naasson naa...</td>\n",
              "      <td>64</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And Salmon begat Booz of Rachab; and Booz bega...</td>\n",
              "      <td>and salmon begat booz of rachab  and booz bega...</td>\n",
              "      <td>salmon begat booz rachab booz begat obed ruth ...</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9025</th>\n",
              "      <td>He which testifieth these things saith, Surely...</td>\n",
              "      <td>he which testifieth these things saith  surely...</td>\n",
              "      <td>testifieth things saith surely come quickly</td>\n",
              "      <td>44</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9026</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9027</th>\n",
              "      <td>Even so, come, Lord Jesus.</td>\n",
              "      <td>even so  come  lord jesus</td>\n",
              "      <td>even come lord jesus</td>\n",
              "      <td>21</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9028</th>\n",
              "      <td>The grace of our Lord Jesus Christ be with you...</td>\n",
              "      <td>the grace of our lord jesus christ be with you...</td>\n",
              "      <td>grace lord jesus christ</td>\n",
              "      <td>24</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9029</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>9030 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b451b355-cd22-419f-b45f-ab52a32d0a93')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b451b355-cd22-419f-b45f-ab52a32d0a93 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b451b355-cd22-419f-b45f-ab52a32d0a93');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-1b47a668-c6c3-42d6-9b12-65e3011da4d3\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1b47a668-c6c3-42d6-9b12-65e3011da4d3')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-1b47a668-c6c3-42d6-9b12-65e3011da4d3 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "prep_bible",
              "summary": "{\n  \"name\": \"prep_bible\",\n  \"rows\": 9030,\n  \"fields\": [\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8912,\n        \"samples\": [\n          \"And said unto them, Why sleep ye?\",\n          \"And whithersoever he entered, into villages, or cities, or country, they laid the sick in the streets, and besought him that they might touch if it were but the border of his garment: and as many as touched him were made whole.\",\n          \"for ye have received your consolation.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8909,\n        \"samples\": [\n          \"and peter called to mind the word that jesus said unto him  before the cock crow twice  thou shalt deny me thrice \",\n          \"but if ye be led of the spirit  ye are not under the law \",\n          \"i am that bread of life \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8857,\n        \"samples\": [\n          \"ye sometimes darkness ye light lord walk children light \",\n          \"taketh away first may establish second \",\n          \"carnally minded death spiritually minded life peace \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 28,\n        \"min\": 0,\n        \"max\": 203,\n        \"num_unique_values\": 167,\n        \"samples\": [\n          13,\n          83,\n          77\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible[prep_bible.prep_text != 'amen ']"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "jCKeOIbvWzW3",
        "outputId": "ba27fdd7-b2d7-469c-892e-d9d70c96534c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   text  \\\n",
              "0     The book of the generation of Jesus Christ, th...   \n",
              "1     Abraham begat Isaac; and Isaac begat Jacob; an...   \n",
              "2     And Judas begat Phares and Zara of Thamar; and...   \n",
              "3     And Aram begat Aminadab; and Aminadab begat Na...   \n",
              "4     And Salmon begat Booz of Rachab; and Booz bega...   \n",
              "...                                                 ...   \n",
              "9023  For I testify unto every man that heareth the ...   \n",
              "9024  And if any man shall take away from the words ...   \n",
              "9025  He which testifieth these things saith, Surely...   \n",
              "9027                         Even so, come, Lord Jesus.   \n",
              "9028  The grace of our Lord Jesus Christ be with you...   \n",
              "\n",
              "                                           cleaned_text  \\\n",
              "0     the book of the generation of jesus christ  th...   \n",
              "1     abraham begat isaac  and isaac begat jacob  an...   \n",
              "2     and judas begat phares and zara of thamar  and...   \n",
              "3     and aram begat aminadab  and aminadab begat na...   \n",
              "4     and salmon begat booz of rachab  and booz bega...   \n",
              "...                                                 ...   \n",
              "9023  for i testify unto every man that heareth the ...   \n",
              "9024  and if any man shall take away from the words ...   \n",
              "9025  he which testifieth these things saith  surely...   \n",
              "9027                         even so  come  lord jesus    \n",
              "9028  the grace of our lord jesus christ be with you...   \n",
              "\n",
              "                                              prep_text  len  \n",
              "0     book generation jesus christ son david son abr...   51  \n",
              "1     abraham begat isaac isaac begat jacob jacob be...   65  \n",
              "2     judas begat phares zara thamar phares begat es...   67  \n",
              "3     aram begat aminadab aminadab begat naasson naa...   64  \n",
              "4     salmon begat booz rachab booz begat obed ruth ...   63  \n",
              "...                                                 ...  ...  \n",
              "9023  testify unto every man heareth words prophecy ...  117  \n",
              "9024  man shall take away words book prophecy god sh...  105  \n",
              "9025       testifieth things saith surely come quickly    44  \n",
              "9027                              even come lord jesus    21  \n",
              "9028                           grace lord jesus christ    24  \n",
              "\n",
              "[8987 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-bd635856-99dc-45ee-9992-e69521b598c0\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>The book of the generation of Jesus Christ, th...</td>\n",
              "      <td>the book of the generation of jesus christ  th...</td>\n",
              "      <td>book generation jesus christ son david son abr...</td>\n",
              "      <td>51</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Abraham begat Isaac; and Isaac begat Jacob; an...</td>\n",
              "      <td>abraham begat isaac  and isaac begat jacob  an...</td>\n",
              "      <td>abraham begat isaac isaac begat jacob jacob be...</td>\n",
              "      <td>65</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And Judas begat Phares and Zara of Thamar; and...</td>\n",
              "      <td>and judas begat phares and zara of thamar  and...</td>\n",
              "      <td>judas begat phares zara thamar phares begat es...</td>\n",
              "      <td>67</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And Aram begat Aminadab; and Aminadab begat Na...</td>\n",
              "      <td>and aram begat aminadab  and aminadab begat na...</td>\n",
              "      <td>aram begat aminadab aminadab begat naasson naa...</td>\n",
              "      <td>64</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And Salmon begat Booz of Rachab; and Booz bega...</td>\n",
              "      <td>and salmon begat booz of rachab  and booz bega...</td>\n",
              "      <td>salmon begat booz rachab booz begat obed ruth ...</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9023</th>\n",
              "      <td>For I testify unto every man that heareth the ...</td>\n",
              "      <td>for i testify unto every man that heareth the ...</td>\n",
              "      <td>testify unto every man heareth words prophecy ...</td>\n",
              "      <td>117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9024</th>\n",
              "      <td>And if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>man shall take away words book prophecy god sh...</td>\n",
              "      <td>105</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9025</th>\n",
              "      <td>He which testifieth these things saith, Surely...</td>\n",
              "      <td>he which testifieth these things saith  surely...</td>\n",
              "      <td>testifieth things saith surely come quickly</td>\n",
              "      <td>44</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9027</th>\n",
              "      <td>Even so, come, Lord Jesus.</td>\n",
              "      <td>even so  come  lord jesus</td>\n",
              "      <td>even come lord jesus</td>\n",
              "      <td>21</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9028</th>\n",
              "      <td>The grace of our Lord Jesus Christ be with you...</td>\n",
              "      <td>the grace of our lord jesus christ be with you...</td>\n",
              "      <td>grace lord jesus christ</td>\n",
              "      <td>24</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>8987 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-bd635856-99dc-45ee-9992-e69521b598c0')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-bd635856-99dc-45ee-9992-e69521b598c0 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-bd635856-99dc-45ee-9992-e69521b598c0');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-06e9c8dc-64fc-4545-82b4-54793d1954b0\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-06e9c8dc-64fc-4545-82b4-54793d1954b0')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-06e9c8dc-64fc-4545-82b4-54793d1954b0 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"prep_bible[prep_bible\",\n  \"rows\": 8987,\n  \"fields\": [\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8911,\n        \"samples\": [\n          \"(For as yet he was fallen upon none of them: only they were baptized in the name of the Lord Jesus.)\",\n          \"Then came together unto him the Pharisees, and certain of the scribes, which came from Jerusalem.\",\n          \"Then Paul stood in the midst of Mars' hill, and said, Ye men of Athens, I perceive that in all things ye are too superstitious.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8908,\n        \"samples\": [\n          \"then answered peter  and said unto jesus  lord  it is good for us to be here  if thou wilt  let us make here three tabernacles  one for thee  and one for moses  and one for elias \",\n          \"and the napkin  that was about his head  not lying with the linen clothes  but wrapped together in a place by itself \",\n          \"destroy not him with thy meat  for whom christ died \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 8856,\n        \"samples\": [\n          \"ye bear witness said christ sent \",\n          \"man taken jews killed came army rescued understood roman \",\n          \"even us hath called jews also gentiles \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 28,\n        \"min\": 0,\n        \"max\": 203,\n        \"num_unique_values\": 167,\n        \"samples\": [\n          91,\n          19,\n          77\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible.to_csv('prep_KJV.csv')"
      ],
      "metadata": {
        "id": "1JeooGp_Wy-1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# American Standard Version (asb - в нашем исследовании)\n",
        "* Пересмотр the Authorised Version (King James Bible) произошел благодаря сделанным с  XVII в. многочисленным открытиям  в области древних манускриптов. Основные принципы нового перевода состояли в следующем: как можно меньше изменять текст the Authorised Version, максимально ограничить воздействие таких изменений на язык the Authorised Version и более ранних версий; адаптировать текст к современному (конец ХIX - начало ХХ вв.) состоянию языка. Обновленная редакции вышла в Англии в 1881 г. (Новый Завет) и 1885 г. (Ветхий Завет). В Америке этот же перевод появился в 1901 году под названием Американская стандартная версия (American Standard Version).\n",
        "\n",
        "Форостенко, А В. Топология современных переводов Библии на английский язык. Дисс. ... канд. филол. наук. М., 2003."
      ],
      "metadata": {
        "id": "FqHdhn5mPg4H"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "import re\n",
        "import string\n",
        "import nltk\n",
        "from nltk.corpus import stopwords\n",
        "\n",
        "nltk.download(\"stopwords\")\n",
        "stops = stopwords.words(\"english\")\n",
        "\n",
        "nltk.download('punkt')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "40dc159b-fdff-46a3-a1a2-9cdb92f11f16",
        "id": "hAaVHhReQDcr"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package stopwords to /root/nltk_data...\n",
            "[nltk_data]   Unzipping corpora/stopwords.zip.\n",
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 1
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "asb = pd.read_csv('/content/asb.csv')\n",
        "asb"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "outputId": "5887f0be-79fb-42b1-df15-e3048111d3a7",
        "id": "DxCq3DKsQDct"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       Verse ID   Book Name  Book Number  Chapter  Verse  \\\n",
              "0             1     Genesis            1        1      1   \n",
              "1             2     Genesis            1        1      2   \n",
              "2             3     Genesis            1        1      3   \n",
              "3             4     Genesis            1        1      4   \n",
              "4             5     Genesis            1        1      5   \n",
              "...         ...         ...          ...      ...    ...   \n",
              "31097     31098  Revelation           66       22     17   \n",
              "31098     31099  Revelation           66       22     18   \n",
              "31099     31100  Revelation           66       22     19   \n",
              "31100     31101  Revelation           66       22     20   \n",
              "31101     31102  Revelation           66       22     21   \n",
              "\n",
              "                                                    Text  \n",
              "0      In the beginning God created the heavens and t...  \n",
              "1      And the earth was waste and void; and darkness...  \n",
              "2      And God said, Let there be light: and there wa...  \n",
              "3      And God saw the light, that it was good: and G...  \n",
              "4      And God called the light Day, and the darkness...  \n",
              "...                                                  ...  \n",
              "31097  And the Spirit and the bride say, Come. And he...  \n",
              "31098  I testify unto every man that heareth the word...  \n",
              "31099  and if any man shall take away from the words ...  \n",
              "31100  He who testifieth these things saith, Yea: I c...  \n",
              "31101  The grace of the Lord Jesus be with the saints...  \n",
              "\n",
              "[31102 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d5b7ec0e-8825-4b02-bf30-abe4966c9db9\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Verse ID</th>\n",
              "      <th>Book Name</th>\n",
              "      <th>Book Number</th>\n",
              "      <th>Chapter</th>\n",
              "      <th>Verse</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>In the beginning God created the heavens and t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>2</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>And the earth was waste and void; and darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>3</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>And God said, Let there be light: and there wa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>4</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>4</td>\n",
              "      <td>And God saw the light, that it was good: and G...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>5</td>\n",
              "      <td>Genesis</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>5</td>\n",
              "      <td>And God called the light Day, and the darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31097</th>\n",
              "      <td>31098</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>66</td>\n",
              "      <td>22</td>\n",
              "      <td>17</td>\n",
              "      <td>And the Spirit and the bride say, Come. And he...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31098</th>\n",
              "      <td>31099</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>66</td>\n",
              "      <td>22</td>\n",
              "      <td>18</td>\n",
              "      <td>I testify unto every man that heareth the word...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31099</th>\n",
              "      <td>31100</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>66</td>\n",
              "      <td>22</td>\n",
              "      <td>19</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31100</th>\n",
              "      <td>31101</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>66</td>\n",
              "      <td>22</td>\n",
              "      <td>20</td>\n",
              "      <td>He who testifieth these things saith, Yea: I c...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31101</th>\n",
              "      <td>31102</td>\n",
              "      <td>Revelation</td>\n",
              "      <td>66</td>\n",
              "      <td>22</td>\n",
              "      <td>21</td>\n",
              "      <td>The grace of the Lord Jesus be with the saints...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>31102 rows × 6 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d5b7ec0e-8825-4b02-bf30-abe4966c9db9')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-d5b7ec0e-8825-4b02-bf30-abe4966c9db9 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-d5b7ec0e-8825-4b02-bf30-abe4966c9db9');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-5a15d418-1464-4a5a-a30c-ad2d97388a98\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5a15d418-1464-4a5a-a30c-ad2d97388a98')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-5a15d418-1464-4a5a-a30c-ad2d97388a98 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_df23b656-380d-43c8-a5ba-e9576bb9a646\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('asb')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_df23b656-380d-43c8-a5ba-e9576bb9a646 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('asb');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "asb",
              "summary": "{\n  \"name\": \"asb\",\n  \"rows\": 31102,\n  \"fields\": [\n    {\n      \"column\": \"Verse ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 8978,\n        \"min\": 1,\n        \"max\": 31102,\n        \"num_unique_values\": 31102,\n        \"samples\": [\n          1521,\n          30807,\n          2933\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Book Name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 66,\n        \"samples\": [\n          \"2 Timothy\",\n          \"2 John\",\n          \"Genesis\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Book Number\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 16,\n        \"min\": 1,\n        \"max\": 66,\n        \"num_unique_values\": 66,\n        \"samples\": [\n          55,\n          63,\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Chapter\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 23,\n        \"min\": 1,\n        \"max\": 150,\n        \"num_unique_values\": 150,\n        \"samples\": [\n          74,\n          19,\n          119\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Verse\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 14,\n        \"min\": 1,\n        \"max\": 176,\n        \"num_unique_values\": 176,\n        \"samples\": [\n          20,\n          46,\n          140\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 30700,\n        \"samples\": [\n          \"Hast thou utterly rejected Judah? hath thy soul loathed Zion? why hast thou smitten us, and there is no healing for us? We looked for peace, but no good came; and for a time of healing, and, behold, dismay!\",\n          \"And when the servants of David were come to Abigail to Carmel, they spake unto her, saying, David hath sent us unto thee, to take thee to him to wife.\",\n          \"and sixteen thousand persons),\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "asb.iloc[29]['Text']"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        },
        "id": "ZNXyOO_rgEEH",
        "outputId": "02829041-3f51-483b-e1d2-991738e1dd38"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'and to every beast of the earth, and to every bird of the heavens, and to everything that creepeth upon the earth, wherein there is life, [I have given] every green herb for food: and it was so.'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "XV6rAHDp5LK6"
      },
      "outputs": [],
      "source": [
        "def sent_split(df, col):\n",
        "    sents = []\n",
        "    for _, row in df.iterrows():\n",
        "        text = row[col]\n",
        "        for sent in nltk.sent_tokenize(text):\n",
        "            sents.append(sent)\n",
        "    sents = pd.DataFrame(sents).rename({0: 'Text'}, axis=1)\n",
        "    return sents"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "asb['Text'] = asb['Text'].astype(str)"
      ],
      "metadata": {
        "id": "sIlhr7N4tq4V"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "asb_sents = sent_split(asb, \"Text\")"
      ],
      "metadata": {
        "id": "1aL22oCk5LK7"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "K5ytFHeF5LK7"
      },
      "outputs": [],
      "source": [
        "def clean_bible_text(df):\n",
        "    # приводим к нижнему регистру\n",
        "    df['cleaned_text'] = df['Text'].str.lower()\n",
        "\n",
        "    # удаляем символы\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'[^a-zA-Z0-9]', \" \", regex=True)\n",
        "\n",
        "    # Удаление цифр\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\d+', '', regex=True)\n",
        "    return df"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible = clean_bible_text(asb_sents)"
      ],
      "metadata": {
        "id": "CUnOlkfo5LK7"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "iuNwKKRm5LK7",
        "outputId": "39a3ffbd-c1e6-46fc-c65c-d7b965bf0a00"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      In the beginning God created the heavens and t...   \n",
              "1      And the earth was waste and void; and darkness...   \n",
              "2      And God said, Let there be light: and there wa...   \n",
              "3      And God saw the light, that it was good: and G...   \n",
              "4      And God called the light Day, and the darkness...   \n",
              "...                                                  ...   \n",
              "35345  and if any man shall take away from the words ...   \n",
              "35346  He who testifieth these things saith, Yea: I c...   \n",
              "35347                            Amen: come, Lord Jesus.   \n",
              "35348    The grace of the Lord Jesus be with the saints.   \n",
              "35349                                              Amen.   \n",
              "\n",
              "                                            cleaned_text  \n",
              "0      in the beginning god created the heavens and t...  \n",
              "1      and the earth was waste and void  and darkness...  \n",
              "2      and god said  let there be light  and there wa...  \n",
              "3      and god saw the light  that it was good  and g...  \n",
              "4      and god called the light day  and the darkness...  \n",
              "...                                                  ...  \n",
              "35345  and if any man shall take away from the words ...  \n",
              "35346  he who testifieth these things saith  yea  i c...  \n",
              "35347                            amen  come  lord jesus   \n",
              "35348    the grace of the lord jesus be with the saints   \n",
              "35349                                              amen   \n",
              "\n",
              "[35350 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-14a45e80-5103-4ed8-9847-e9e9a5f23a95\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>In the beginning God created the heavens and t...</td>\n",
              "      <td>in the beginning god created the heavens and t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>And the earth was waste and void; and darkness...</td>\n",
              "      <td>and the earth was waste and void  and darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And God said, Let there be light: and there wa...</td>\n",
              "      <td>and god said  let there be light  and there wa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And God saw the light, that it was good: and G...</td>\n",
              "      <td>and god saw the light  that it was good  and g...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And God called the light Day, and the darkness...</td>\n",
              "      <td>and god called the light day  and the darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35345</th>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35346</th>\n",
              "      <td>He who testifieth these things saith, Yea: I c...</td>\n",
              "      <td>he who testifieth these things saith  yea  i c...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35347</th>\n",
              "      <td>Amen: come, Lord Jesus.</td>\n",
              "      <td>amen  come  lord jesus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35348</th>\n",
              "      <td>The grace of the Lord Jesus be with the saints.</td>\n",
              "      <td>the grace of the lord jesus be with the saints</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35349</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35350 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-14a45e80-5103-4ed8-9847-e9e9a5f23a95')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-14a45e80-5103-4ed8-9847-e9e9a5f23a95 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-14a45e80-5103-4ed8-9847-e9e9a5f23a95');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-b4e79e1b-f513-4e87-90cd-445bd29647a7\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b4e79e1b-f513-4e87-90cd-445bd29647a7')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-b4e79e1b-f513-4e87-90cd-445bd29647a7 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_8b7adc07-cf9c-4726-93ac-4014a2a7c376\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('asb_sents')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_8b7adc07-cf9c-4726-93ac-4014a2a7c376 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('asb_sents');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "asb_sents",
              "summary": "{\n  \"name\": \"asb_sents\",\n  \"rows\": 35350,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34448,\n        \"samples\": [\n          \"And Joseph, who by the apostles was surnamed Barnabas (which is, being interpreted, Son of exhortation), a Levite, a man of Cyprus by race,\",\n          \"he blasphemeth: who can forgive sins but one, [even] God?\",\n          \"and thither ye shall bring your burnt-offerings, and your sacrifices, and your tithes, and the heave-offering of your hand, and your vows, and your freewill-offerings, and the firstlings of your herd and of your flock:\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34394,\n        \"samples\": [\n          \"then he suffereth him \",\n          \"and the sons of sheshan  ahlai \",\n          \"then the tent of meeting shall set forward  with the camp of the levites in the midst of the camps  as they encamp  so shall they set forward  every man in his place  by their standards \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# решили не удалять стоп-слова\n",
        "\"\"\"\n",
        "import string\n",
        "\n",
        "def stopwords_remover(words):\n",
        "\n",
        "  without_stopwords = \"\"\n",
        "  for word in words.split(\" \"):\n",
        "    if word in stops or word in string.punctuation:\n",
        "      continue\n",
        "    else:\n",
        "      without_stopwords += word + \" \"\n",
        "  return without_stopwords\n",
        "  \"\"\""
      ],
      "metadata": {
        "id": "vHT16iA45LK7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# prep_bible['prep_text'] = prep_bible['cleaned_text'].apply(stopwords_remover)\n"
      ],
      "metadata": {
        "id": "m_q9-lHl5LK8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible['prep_text'] = prep_bible['cleaned_text']"
      ],
      "metadata": {
        "id": "4f56lAM1uW_Q"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "IA7-6TP55LK8",
        "outputId": "73934dbe-0a36-4085-9ffa-d37e9cfc6970"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      In the beginning God created the heavens and t...   \n",
              "1      And the earth was waste and void; and darkness...   \n",
              "2      And God said, Let there be light: and there wa...   \n",
              "3      And God saw the light, that it was good: and G...   \n",
              "4      And God called the light Day, and the darkness...   \n",
              "...                                                  ...   \n",
              "35345  and if any man shall take away from the words ...   \n",
              "35346  He who testifieth these things saith, Yea: I c...   \n",
              "35347                            Amen: come, Lord Jesus.   \n",
              "35348    The grace of the Lord Jesus be with the saints.   \n",
              "35349                                              Amen.   \n",
              "\n",
              "                                            cleaned_text  \\\n",
              "0      in the beginning god created the heavens and t...   \n",
              "1      and the earth was waste and void  and darkness...   \n",
              "2      and god said  let there be light  and there wa...   \n",
              "3      and god saw the light  that it was good  and g...   \n",
              "4      and god called the light day  and the darkness...   \n",
              "...                                                  ...   \n",
              "35345  and if any man shall take away from the words ...   \n",
              "35346  he who testifieth these things saith  yea  i c...   \n",
              "35347                            amen  come  lord jesus    \n",
              "35348    the grace of the lord jesus be with the saints    \n",
              "35349                                              amen    \n",
              "\n",
              "                                               prep_text  \n",
              "0      in the beginning god created the heavens and t...  \n",
              "1      and the earth was waste and void  and darkness...  \n",
              "2      and god said  let there be light  and there wa...  \n",
              "3      and god saw the light  that it was good  and g...  \n",
              "4      and god called the light day  and the darkness...  \n",
              "...                                                  ...  \n",
              "35345  and if any man shall take away from the words ...  \n",
              "35346  he who testifieth these things saith  yea  i c...  \n",
              "35347                            amen  come  lord jesus   \n",
              "35348    the grace of the lord jesus be with the saints   \n",
              "35349                                              amen   \n",
              "\n",
              "[35350 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3d8ac070-b193-4722-9dff-c37b63429e01\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>In the beginning God created the heavens and t...</td>\n",
              "      <td>in the beginning god created the heavens and t...</td>\n",
              "      <td>in the beginning god created the heavens and t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>And the earth was waste and void; and darkness...</td>\n",
              "      <td>and the earth was waste and void  and darkness...</td>\n",
              "      <td>and the earth was waste and void  and darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And God said, Let there be light: and there wa...</td>\n",
              "      <td>and god said  let there be light  and there wa...</td>\n",
              "      <td>and god said  let there be light  and there wa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And God saw the light, that it was good: and G...</td>\n",
              "      <td>and god saw the light  that it was good  and g...</td>\n",
              "      <td>and god saw the light  that it was good  and g...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And God called the light Day, and the darkness...</td>\n",
              "      <td>and god called the light day  and the darkness...</td>\n",
              "      <td>and god called the light day  and the darkness...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35345</th>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35346</th>\n",
              "      <td>He who testifieth these things saith, Yea: I c...</td>\n",
              "      <td>he who testifieth these things saith  yea  i c...</td>\n",
              "      <td>he who testifieth these things saith  yea  i c...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35347</th>\n",
              "      <td>Amen: come, Lord Jesus.</td>\n",
              "      <td>amen  come  lord jesus</td>\n",
              "      <td>amen  come  lord jesus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35348</th>\n",
              "      <td>The grace of the Lord Jesus be with the saints.</td>\n",
              "      <td>the grace of the lord jesus be with the saints</td>\n",
              "      <td>the grace of the lord jesus be with the saints</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35349</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35350 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3d8ac070-b193-4722-9dff-c37b63429e01')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3d8ac070-b193-4722-9dff-c37b63429e01 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3d8ac070-b193-4722-9dff-c37b63429e01');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-804442ec-548a-4b43-bab5-90a8f19e8885\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-804442ec-548a-4b43-bab5-90a8f19e8885')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-804442ec-548a-4b43-bab5-90a8f19e8885 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_1e65a945-8f2b-46eb-849a-864285378a9a\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('asb_sents')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_1e65a945-8f2b-46eb-849a-864285378a9a button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('asb_sents');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "asb_sents",
              "summary": "{\n  \"name\": \"asb_sents\",\n  \"rows\": 35350,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34448,\n        \"samples\": [\n          \"And Joseph, who by the apostles was surnamed Barnabas (which is, being interpreted, Son of exhortation), a Levite, a man of Cyprus by race,\",\n          \"he blasphemeth: who can forgive sins but one, [even] God?\",\n          \"and thither ye shall bring your burnt-offerings, and your sacrifices, and your tithes, and the heave-offering of your hand, and your vows, and your freewill-offerings, and the firstlings of your herd and of your flock:\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34394,\n        \"samples\": [\n          \"then he suffereth him \",\n          \"and the sons of sheshan  ahlai \",\n          \"then the tent of meeting shall set forward  with the camp of the levites in the midst of the camps  as they encamp  so shall they set forward  every man in his place  by their standards \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34394,\n        \"samples\": [\n          \"then he suffereth him \",\n          \"and the sons of sheshan  ahlai \",\n          \"then the tent of meeting shall set forward  with the camp of the levites in the midst of the camps  as they encamp  so shall they set forward  every man in his place  by their standards \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "sA6o8gln5LK8"
      },
      "outputs": [],
      "source": [
        "def clean_long_sent(df):\n",
        "    df[\"len\"] = df[\"prep_text\"].apply(lambda x: len(x))\n",
        "    return df[df.len <= 512]"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible = clean_long_sent(prep_bible)\n",
        "prep_bible"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "IM-JxdhS5LK8",
        "outputId": "5c128f4b-7b0c-4736-d76c-e5d19bedafa9"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      In the beginning God created the heavens and t...   \n",
              "1      And the earth was waste and void; and darkness...   \n",
              "2      And God said, Let there be light: and there wa...   \n",
              "3      And God saw the light, that it was good: and G...   \n",
              "4      And God called the light Day, and the darkness...   \n",
              "...                                                  ...   \n",
              "35345  and if any man shall take away from the words ...   \n",
              "35346  He who testifieth these things saith, Yea: I c...   \n",
              "35347                            Amen: come, Lord Jesus.   \n",
              "35348    The grace of the Lord Jesus be with the saints.   \n",
              "35349                                              Amen.   \n",
              "\n",
              "                                            cleaned_text  \\\n",
              "0      in the beginning god created the heavens and t...   \n",
              "1      and the earth was waste and void  and darkness...   \n",
              "2      and god said  let there be light  and there wa...   \n",
              "3      and god saw the light  that it was good  and g...   \n",
              "4      and god called the light day  and the darkness...   \n",
              "...                                                  ...   \n",
              "35345  and if any man shall take away from the words ...   \n",
              "35346  he who testifieth these things saith  yea  i c...   \n",
              "35347                            amen  come  lord jesus    \n",
              "35348    the grace of the lord jesus be with the saints    \n",
              "35349                                              amen    \n",
              "\n",
              "                                               prep_text  len  \n",
              "0      in the beginning god created the heavens and t...   55  \n",
              "1      and the earth was waste and void  and darkness...  134  \n",
              "2      and god said  let there be light  and there wa...   54  \n",
              "3      and god saw the light  that it was good  and g...   85  \n",
              "4      and god called the light day  and the darkness...   63  \n",
              "...                                                  ...  ...  \n",
              "35345  and if any man shall take away from the words ...  185  \n",
              "35346  he who testifieth these things saith  yea  i c...   58  \n",
              "35347                            amen  come  lord jesus    23  \n",
              "35348    the grace of the lord jesus be with the saints    47  \n",
              "35349                                              amen     5  \n",
              "\n",
              "[35350 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-2c8e1ccd-3052-4a5f-abfa-d435ddaf9c98\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>In the beginning God created the heavens and t...</td>\n",
              "      <td>in the beginning god created the heavens and t...</td>\n",
              "      <td>in the beginning god created the heavens and t...</td>\n",
              "      <td>55</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>And the earth was waste and void; and darkness...</td>\n",
              "      <td>and the earth was waste and void  and darkness...</td>\n",
              "      <td>and the earth was waste and void  and darkness...</td>\n",
              "      <td>134</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>And God said, Let there be light: and there wa...</td>\n",
              "      <td>and god said  let there be light  and there wa...</td>\n",
              "      <td>and god said  let there be light  and there wa...</td>\n",
              "      <td>54</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>And God saw the light, that it was good: and G...</td>\n",
              "      <td>and god saw the light  that it was good  and g...</td>\n",
              "      <td>and god saw the light  that it was good  and g...</td>\n",
              "      <td>85</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>And God called the light Day, and the darkness...</td>\n",
              "      <td>and god called the light day  and the darkness...</td>\n",
              "      <td>and god called the light day  and the darkness...</td>\n",
              "      <td>63</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35345</th>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>and if any man shall take away from the words ...</td>\n",
              "      <td>185</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35346</th>\n",
              "      <td>He who testifieth these things saith, Yea: I c...</td>\n",
              "      <td>he who testifieth these things saith  yea  i c...</td>\n",
              "      <td>he who testifieth these things saith  yea  i c...</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35347</th>\n",
              "      <td>Amen: come, Lord Jesus.</td>\n",
              "      <td>amen  come  lord jesus</td>\n",
              "      <td>amen  come  lord jesus</td>\n",
              "      <td>23</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35348</th>\n",
              "      <td>The grace of the Lord Jesus be with the saints.</td>\n",
              "      <td>the grace of the lord jesus be with the saints</td>\n",
              "      <td>the grace of the lord jesus be with the saints</td>\n",
              "      <td>47</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35349</th>\n",
              "      <td>Amen.</td>\n",
              "      <td>amen</td>\n",
              "      <td>amen</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35350 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2c8e1ccd-3052-4a5f-abfa-d435ddaf9c98')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-2c8e1ccd-3052-4a5f-abfa-d435ddaf9c98 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-2c8e1ccd-3052-4a5f-abfa-d435ddaf9c98');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-4bc8ad65-bbd5-4569-99b2-abfa8370242b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-4bc8ad65-bbd5-4569-99b2-abfa8370242b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-4bc8ad65-bbd5-4569-99b2-abfa8370242b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_5bda526c-9513-4610-b413-1cae0c23d3b1\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('prep_bible')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_5bda526c-9513-4610-b413-1cae0c23d3b1 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('prep_bible');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "prep_bible",
              "summary": "{\n  \"name\": \"prep_bible\",\n  \"rows\": 35350,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34448,\n        \"samples\": [\n          \"And Joseph, who by the apostles was surnamed Barnabas (which is, being interpreted, Son of exhortation), a Levite, a man of Cyprus by race,\",\n          \"he blasphemeth: who can forgive sins but one, [even] God?\",\n          \"and thither ye shall bring your burnt-offerings, and your sacrifices, and your tithes, and the heave-offering of your hand, and your vows, and your freewill-offerings, and the firstlings of your herd and of your flock:\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34394,\n        \"samples\": [\n          \"then he suffereth him \",\n          \"and the sons of sheshan  ahlai \",\n          \"then the tent of meeting shall set forward  with the camp of the levites in the midst of the camps  as they encamp  so shall they set forward  every man in his place  by their standards \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 34394,\n        \"samples\": [\n          \"then he suffereth him \",\n          \"and the sons of sheshan  ahlai \",\n          \"then the tent of meeting shall set forward  with the camp of the levites in the midst of the camps  as they encamp  so shall they set forward  every man in his place  by their standards \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 59,\n        \"min\": 1,\n        \"max\": 508,\n        \"num_unique_values\": 373,\n        \"samples\": [\n          313,\n          132,\n          162\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible[prep_bible.prep_text != 'amen ']"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "_2aKS81j5LK8",
        "outputId": "9cb8a05c-2e5d-4d29-8dc7-22c64b85d332"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      The book of the genealogy of Jesus Christ, the...   \n",
              "1                    Abraham became the father of Isaac.   \n",
              "2                      Isaac became the father of Jacob.   \n",
              "3      Jacob became the father of Judah and his broth...   \n",
              "4      Judah became the father of Perez and Zerah by ...   \n",
              "...                                                  ...   \n",
              "10590  I testify to everyone who hears the words of t...   \n",
              "10591  If anyone takes away from the words of the boo...   \n",
              "10592  He who testifies these things says, \"Yes, I co...   \n",
              "10594                             Yes, come, Lord Jesus.   \n",
              "10595  The grace of the Lord Jesus Christ be with all...   \n",
              "\n",
              "                                            cleaned_text  \\\n",
              "0      the book of the genealogy of jesus christ  the...   \n",
              "1                    abraham became the father of isaac    \n",
              "2                      isaac became the father of jacob    \n",
              "3      jacob became the father of judah and his broth...   \n",
              "4      judah became the father of perez and zerah by ...   \n",
              "...                                                  ...   \n",
              "10590  i testify to everyone who hears the words of t...   \n",
              "10591  if anyone takes away from the words of the boo...   \n",
              "10592  he who testifies these things says   yes  i co...   \n",
              "10594                             yes  come  lord jesus    \n",
              "10595  the grace of the lord jesus christ be with all...   \n",
              "\n",
              "                                               prep_text  len  \n",
              "0      book genealogy jesus christ son david son abra...   50  \n",
              "1                           abraham became father isaac    28  \n",
              "2                             isaac became father jacob    26  \n",
              "3                    jacob became father judah brothers    35  \n",
              "4                 judah became father perez zerah tamar    38  \n",
              "...                                                  ...  ...  \n",
              "10590  testify everyone hears words prophecy book any...   88  \n",
              "10591  anyone takes away words book prophecy may god ...   94  \n",
              "10592            testifies things says yes come quickly    39  \n",
              "10594                               yes come lord jesus    20  \n",
              "10595                    grace lord jesus christ saints    31  \n",
              "\n",
              "[10555 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ec65dde9-f075-4c66-9c53-3e96341fce2d\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>prep_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>The book of the genealogy of Jesus Christ, the...</td>\n",
              "      <td>the book of the genealogy of jesus christ  the...</td>\n",
              "      <td>book genealogy jesus christ son david son abra...</td>\n",
              "      <td>50</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Abraham became the father of Isaac.</td>\n",
              "      <td>abraham became the father of isaac</td>\n",
              "      <td>abraham became father isaac</td>\n",
              "      <td>28</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Isaac became the father of Jacob.</td>\n",
              "      <td>isaac became the father of jacob</td>\n",
              "      <td>isaac became father jacob</td>\n",
              "      <td>26</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Jacob became the father of Judah and his broth...</td>\n",
              "      <td>jacob became the father of judah and his broth...</td>\n",
              "      <td>jacob became father judah brothers</td>\n",
              "      <td>35</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Judah became the father of Perez and Zerah by ...</td>\n",
              "      <td>judah became the father of perez and zerah by ...</td>\n",
              "      <td>judah became father perez zerah tamar</td>\n",
              "      <td>38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10590</th>\n",
              "      <td>I testify to everyone who hears the words of t...</td>\n",
              "      <td>i testify to everyone who hears the words of t...</td>\n",
              "      <td>testify everyone hears words prophecy book any...</td>\n",
              "      <td>88</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10591</th>\n",
              "      <td>If anyone takes away from the words of the boo...</td>\n",
              "      <td>if anyone takes away from the words of the boo...</td>\n",
              "      <td>anyone takes away words book prophecy may god ...</td>\n",
              "      <td>94</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10592</th>\n",
              "      <td>He who testifies these things says, \"Yes, I co...</td>\n",
              "      <td>he who testifies these things says   yes  i co...</td>\n",
              "      <td>testifies things says yes come quickly</td>\n",
              "      <td>39</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10594</th>\n",
              "      <td>Yes, come, Lord Jesus.</td>\n",
              "      <td>yes  come  lord jesus</td>\n",
              "      <td>yes come lord jesus</td>\n",
              "      <td>20</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10595</th>\n",
              "      <td>The grace of the Lord Jesus Christ be with all...</td>\n",
              "      <td>the grace of the lord jesus christ be with all...</td>\n",
              "      <td>grace lord jesus christ saints</td>\n",
              "      <td>31</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10555 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ec65dde9-f075-4c66-9c53-3e96341fce2d')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ec65dde9-f075-4c66-9c53-3e96341fce2d button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ec65dde9-f075-4c66-9c53-3e96341fce2d');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-683f402f-dc67-4398-b3e5-61704cfe97f3\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-683f402f-dc67-4398-b3e5-61704cfe97f3')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-683f402f-dc67-4398-b3e5-61704cfe97f3 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"prep_bible[prep_bible\",\n  \"rows\": 10555,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10434,\n        \"samples\": [\n          \"However one of the soldiers pierced his side with a spear, and immediately blood and water came out.\",\n          \"These are those who were not defiled with women, for they are virgins.\",\n          \"So also you, since you are zealous for spiritual gifts, seek that you may abound to the building up of the assembly.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10427,\n        \"samples\": [\n          \"if anyone was not found written in the book of life  he was cast into the lake of fire \",\n          \"what fruit then did you have at that time in the things of which you are now ashamed \",\n          \" behold  i have told you all things beforehand \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"prep_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 10257,\n        \"samples\": [\n          \"people came see happened \",\n          \"lord servant must quarrel gentle towards able teach patient \",\n          \"second said hallelujah \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 25,\n        \"min\": 0,\n        \"max\": 179,\n        \"num_unique_values\": 155,\n        \"samples\": [\n          6,\n          137,\n          66\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 61
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "prep_bible.to_csv('prep_asb_SW.csv') # American Standard Bible with stop-words"
      ],
      "metadata": {
        "id": "ipd74Rrn5LK9"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Корпус статей из журнала \"The Economist\""
      ],
      "metadata": {
        "id": "CR9gEZUQ_rZE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "df = pd.read_csv(\"/content/bio_Economist.csv\")\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "fVezceS379d8",
        "outputId": "7e962f41-912f-49f6-841a-7f16e0e437ba"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Unnamed: 0.1  Unnamed: 0  \\\n",
              "0               0           0   \n",
              "1               1           1   \n",
              "2               2           2   \n",
              "3               3           3   \n",
              "4               4           4   \n",
              "..            ...         ...   \n",
              "957           957         957   \n",
              "958           958         958   \n",
              "959           959         959   \n",
              "960           960         960   \n",
              "961           961         961   \n",
              "\n",
              "                                                  text  \\\n",
              "0    FOR THE past four billion years or so the only...   \n",
              "1    IN A former leatherworks just off Euston Road ...   \n",
              "2    “How many cells are there in a human being?” I...   \n",
              "3    LIVING creatures are jolly useful. Farmers rea...   \n",
              "4    A broken brain, hidden inside a skull, is hard...   \n",
              "..                                                 ...   \n",
              "957  Science can be a little scary. Its potential t...   \n",
              "958  IT MAY challenge the spirit to think of caulif...   \n",
              "959  MAJOR DEPRESSION is a serious illness, but als...   \n",
              "960  Holt StudiosA beautiful crop of plasticAt the ...   \n",
              "961  ONE of the most valuable weapons in the war on...   \n",
              "\n",
              "                                                 title            date  \\\n",
              "0          The promise and perils of synthetic biology    Apr 4th 2019   \n",
              "1    Will artificial intelligence help to crack bio...    Jan 7th 2017   \n",
              "2    The idea of “holobionts” represents a paradigm...   Jun 14th 2023   \n",
              "3          The remarkable promise of cell-free biology    May 4th 2017   \n",
              "4    Better brain biology will deliver better medic...   Sep 21st 2022   \n",
              "..                                                 ...             ...   \n",
              "957                                Towards immortality   Nov 16th 2006   \n",
              "958                           The king of cauliflowers   Feb 12th 2015   \n",
              "959  A blood test may help the diagnosis and treatm...    May 6th 2021   \n",
              "960                            Meet industrial biotech   Nov 20th 2003   \n",
              "961       Growing anti-malaria drugs in tobacco plants   Oct 20th 2016   \n",
              "\n",
              "                                                  link  \n",
              "0    https://www.economist.com/leaders/2019/04/04/t...  \n",
              "1    https://www.economist.com/science-and-technolo...  \n",
              "2    https://www.economist.com/science-and-technolo...  \n",
              "3    https://www.economist.com/leaders/2017/05/04/t...  \n",
              "4    https://www.economist.com/technology-quarterly...  \n",
              "..                                                 ...  \n",
              "957  https://www.economist.com/news/2006/11/16/towa...  \n",
              "958  https://www.economist.com/middle-east-and-afri...  \n",
              "959  https://www.economist.com/science-and-technolo...  \n",
              "960  https://www.economist.com/news/2003/11/20/meet...  \n",
              "961  https://www.economist.com/science-and-technolo...  \n",
              "\n",
              "[962 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0d12e650-07db-4a5e-b2f4-f8f09b50e463\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0.1</th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>text</th>\n",
              "      <th>title</th>\n",
              "      <th>date</th>\n",
              "      <th>link</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "      <td>The promise and perils of synthetic biology</td>\n",
              "      <td>Apr 4th 2019</td>\n",
              "      <td>https://www.economist.com/leaders/2019/04/04/t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>IN A former leatherworks just off Euston Road ...</td>\n",
              "      <td>Will artificial intelligence help to crack bio...</td>\n",
              "      <td>Jan 7th 2017</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>“How many cells are there in a human being?” I...</td>\n",
              "      <td>The idea of “holobionts” represents a paradigm...</td>\n",
              "      <td>Jun 14th 2023</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>LIVING creatures are jolly useful. Farmers rea...</td>\n",
              "      <td>The remarkable promise of cell-free biology</td>\n",
              "      <td>May 4th 2017</td>\n",
              "      <td>https://www.economist.com/leaders/2017/05/04/t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>4</td>\n",
              "      <td>A broken brain, hidden inside a skull, is hard...</td>\n",
              "      <td>Better brain biology will deliver better medic...</td>\n",
              "      <td>Sep 21st 2022</td>\n",
              "      <td>https://www.economist.com/technology-quarterly...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>957</th>\n",
              "      <td>957</td>\n",
              "      <td>957</td>\n",
              "      <td>Science can be a little scary. Its potential t...</td>\n",
              "      <td>Towards immortality</td>\n",
              "      <td>Nov 16th 2006</td>\n",
              "      <td>https://www.economist.com/news/2006/11/16/towa...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>958</th>\n",
              "      <td>958</td>\n",
              "      <td>958</td>\n",
              "      <td>IT MAY challenge the spirit to think of caulif...</td>\n",
              "      <td>The king of cauliflowers</td>\n",
              "      <td>Feb 12th 2015</td>\n",
              "      <td>https://www.economist.com/middle-east-and-afri...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>959</th>\n",
              "      <td>959</td>\n",
              "      <td>959</td>\n",
              "      <td>MAJOR DEPRESSION is a serious illness, but als...</td>\n",
              "      <td>A blood test may help the diagnosis and treatm...</td>\n",
              "      <td>May 6th 2021</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>960</th>\n",
              "      <td>960</td>\n",
              "      <td>960</td>\n",
              "      <td>Holt StudiosA beautiful crop of plasticAt the ...</td>\n",
              "      <td>Meet industrial biotech</td>\n",
              "      <td>Nov 20th 2003</td>\n",
              "      <td>https://www.economist.com/news/2003/11/20/meet...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>961</th>\n",
              "      <td>961</td>\n",
              "      <td>961</td>\n",
              "      <td>ONE of the most valuable weapons in the war on...</td>\n",
              "      <td>Growing anti-malaria drugs in tobacco plants</td>\n",
              "      <td>Oct 20th 2016</td>\n",
              "      <td>https://www.economist.com/science-and-technolo...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>962 rows × 6 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0d12e650-07db-4a5e-b2f4-f8f09b50e463')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0d12e650-07db-4a5e-b2f4-f8f09b50e463 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0d12e650-07db-4a5e-b2f4-f8f09b50e463');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-21826093-eb3a-4499-a806-2efa9a3f9b7d\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-21826093-eb3a-4499-a806-2efa9a3f9b7d')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-21826093-eb3a-4499-a806-2efa9a3f9b7d button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_e37403c8-bb3c-4e6e-b637-eefe4c009d53\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_e37403c8-bb3c-4e6e-b637-eefe4c009d53 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 962,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0.1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 277,\n        \"min\": 0,\n        \"max\": 961,\n        \"num_unique_values\": 962,\n        \"samples\": [\n          344,\n          381,\n          351\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 277,\n        \"min\": 0,\n        \"max\": 961,\n        \"num_unique_values\": 962,\n        \"samples\": [\n          344,\n          381,\n          351\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 961,\n        \"samples\": [\n          \"FOR a man who spent his career illuminating the vast, dim migrations of people in prehistory, Luca Cavalli-Sforza\\u2019s life was remarkably circular. He first became interested in his major field, genetics, in the house of the geneticist Adriano Buzzati at Belluno, in the hills north of Venice. There he helped to collect thousands of flies in search of mutant Y chromosomes; and though he subsequently travelled the world to study the makeup of its tribes and populations, it was in Belluno that he died.He had vaguely meant to be a doctor, but found his training useless in wartime hospitals. Instead, he took up the genetic study of E. coli bacteria, shutting himself away in the laboratory. But human beings still fascinated him: not least because he increasingly realised that the human genome contained not only the instructions for building future members of the species, but also its entire past. Migration, conquest and isolation had also left their traces in living populations. He set out to find them.The technology available was rudimentary. He pored over names in parish registers and learned to do statistics, convinced that with enough patience he could measure anything. For one early paper he pulled surnames from Italian phone directories as proxies for father-son transmission. By plotting the distribution of names in 91 provinces, he showed that the Apennines\\u2014the backbone of Italy\\u2014had formed a natural barrier to migration for millennia, resulting in genetically distinct populations on either side. And because he could also measure variation in blood proteins, such as that between A, B and O blood types, he also collected blood samples from townsfolk all over the Parma valley to show that marriage between related individuals had led to many of the genetic differences between those towns.For two decades from 1970, when he took up a professorship at Stanford, he devoted himself to this subject. By the 1990s he was able to study variation in DNA itself. When he and his colleagues instructed a computer to sort around 1,000 people from across the globe into five clusters by similarity of DNA, the clusters matched the labels by which humans had long grouped themselves intuitively: West Eurasians, East Asians, Native Americans, New Guineans and Africans. He soon concluded, though, that race was not a scientifically valid way to classify them. Europeans, for example, were about two-thirds Asian and one-third African, but after millennia of mixing there was no such thing as pure Asian or pure African either. Skin colour, or the shape of a nose, were just superficial adaptations to climate and place.He represented his genetic data as \\u201ctrees\\u201d branching over time: simplistic diagrams, but beautiful in their simplicity. They often agreed, as he had hoped, with the findings of linguists and archaeologists: suggesting, for example, that humanity arose in Africa, where it stayed for a long time before moving outwards, somewhere between 100,000 and 50,000 years ago. The earliest African migrants probably reached Asia first, moving on to Oceania, Europe and America, in that order.He particularly liked to tell his story of farmers. By checking variations between individuals based on their blood groups, he discerned a gradient in that variation that stretched south-east-to-north-west across Europe. This he saw as the genetic signature of farming after its invention around 9,000 years ago in the Fertile Crescent. As early farmers radiated out of the Near East, he argued, they bred with indigenous hunter-gatherers, until by the time they reached colder climates their genes were a good mix of both.This theory stood until the 2000s, when new technology reversed it. He had often carped about the public\\u2019s interest in fossils, so much less informative\\u2014to him\\u2014than genetic studies of the living. Once DNA could be extracted from ancient bones, however, it showed that although agriculture had indeed reached Europe from the Middle East, a wave of immigration from the north-east, starting about 5,000 years ago, had diluted those first farmers with tall herdsmen from the Pontic Steppe. People had moved and mixed in prehistory more than he thought; and not all ancient events, as he supposed, had left their mark in modern populations.Even more controversial was his Human Genome Diversity Project, which he set up in the 1990s. He wanted to study isolated populations in order to understand where all the others came from; to measure the background of drift, or genetic change, that takes place without marauding or migration. Some people thought this racist, and a more worldly man might have realised that. A few critics even brought up his membership of Benito Mussolini\\u2019s fascist youth organisation, compulsory before the war. All this dismayed him, as he had done so much to strike down \\u201cscientific\\u201d studies based on race. The project died in its original form, though it was resurrected as the Genographic Project, which goes on.From Belluno he witnessed the overturning of many other conclusions. Every evolutionary story he had touched became more complex by the day. Yet he could comfort himself that without his original vision for the study of human history through its genes, much of that great debate would not have happened at all.\",\n          \"T HE FIRST outbreak of a novel disease is the opening scene of a whodunnit. In 1976, when more than two dozen members of the American Legion died after a convention in Philadelphia, public-health officials spent months scouring the hotel they had met in before finally tracking down the culprit in the water tank on the roof: a new bacterium which, having caused the first known cases of Legionnaires\\u2019 disease, was named Legionella. In the 1980s it took years of hard work and acrimonious argument among epidemiologists and virologists to blame the terrible and varied symptoms of AIDS on HIV, a virus of a type never previously seen in humans.For covid-19, the mystery was solved almost as soon as it had begun. The novel pneumonia that doctors in Wuhan noticed in December 2019 immediately brought to mind Severe Acute Respiratory Syndrome (SARS), a disease caused by a coronavirus which broke out in 2002. As a result of SARS and the subsequent outbreak in 2012 of Middle East Respiratory Syndrome (MERS), also caused by a coronavirus, there were already established protocols for growing cells from the lining of the nose, throat and lungs in order to look for coronavirus infection. They were soon put to use.To identify the possible coronavirus responsible meant producing a sequence of its genome. The first step in this process was to extract RNA\\u2014the molecule on which coronavirus genomes are written\\u2014from the cell cultures. The genetic sequences in these RNA molecules then had to be transcribed into complementary bits of DNA, because that is what automated sequencing machines work with (see part A of diagram).Computer programs assembled the sequences those machines produced into a recognisable, if gappy, coronavirus genome. Researchers used this to make DNA \\u201cprimers\\u201d with which to fish out the not-yet-sequenced bits of the genome. Finished sequences were published less than two weeks after the process had begun. On January 12th 2020 the world knew its enemy\\u2014soon thereafter named SARS-CoV-2\\u2014down to the last letter of its genome.In terms of the science done, this was all routine; the appropriate use of standard laboratory techniques. In terms of its impact, it was enormous. Knowing the viral sequence was fundamental to vaccination efforts, made it possible to track the virus\\u2019s evolution and, most immediately, made it possible to test people with a cough and see if they were infected. The first AIDS tests were not available until four years after medical science became aware of the condition they tested for. For SARS it took six months. Procedures for testing swabs from the nose and throat for RNA from SARS-CoV-2 were published 11 days after the genome sequence, on January 23rd.There was, however, a drawback to the tests. They required suitably equipped laboratories. Most countries did not have nearly enough of the relevant lab capacity; in others, much of it was being used for different things. \\u201cPrior to March 23rd my lab had never performed a viral diagnostic,\\u201d says Stacey Gabriel, who runs genetic-sequencing operations at the Broad Institute in Cambridge, Massachusetts. But with public institutions swamped she and her colleagues created one of the largest testing shops on America\\u2019s east coast from scratch, reconfiguring the specialised robots that populate one of the world\\u2019s most advanced cancer-genetics labs to do the grunt work.Dr Gabriel learned two lessons in the process. The first is that uniformity matters a lot. The Broad started off testing samples from Massachusetts nursing homes which came in containers of varying size and with various amounts of liquid, some accompanied by handwritten forms, some by barcodes. Dealing with such messiness is no task for a robot, and so to begin with just a few thousand samples a day passed through machines capable of handling far more. The second is that you need software to track the whole process. Commercial software, dry swabs and barcoding soon had the lab firing on all cylinders. By March 2021 it could handle 200,000 tests a day and was serving customers as far afield as New York.The data such labs produce are not just for patients and doctors. In most countries covid-19 is a notifiable disease; the authorities have a legal right to know who has been found to be infected. When a sample tests positive the lab has to pass the identity of the person it came from on to public-health officials. At that point a new sort of detective work begins: where did that person\\u2014the index case, in public-health speak\\u2014pick up the virus? To whom might they have given it?Several East Asian countries demonstrated that, if started in the earliest days of an epidemic and pursued with vigour and persistence, such contact tracing can be a powerful tool. Some, such as Singapore and Taiwan, benefited in this from their experience with SARS in the mid-2000s; tracing systems set in place back then were put to use with an urgency born of experience. A level of invasiveness from which most Western authorities shied away was often employed. In South Korea, for instance, contact tracers were able to download a list of all financial transactions made by those who tested positive; they could then obtain CCTV footage from shops the index case had visited in search of other customers to check up on. \\u201c[Such snooping] has come up with discussions I\\u2019ve had with policymakers,\\u201d says Christophe Fraser, a digital epidemiologist at the University of Oxford. \\u201cWe got very hung up on the idea of contact tracing disrupting people\\u2019s lives.\\u201dWestern governments acted in a slower, less thoroughgoing way. They failed to track the initial spread, in part because of insufficient testing capacity (and, in America, dud tests from the Centres for Disease Control and Prevention). They ended up with much less impressive systems. Tom Frieden, a former director of the CDC, says he thinks that, at its current level of effort, America could plausibly trace about 15,000 cases a day\\u2014a level that has been handsomely exceeded every day since April 2nd 2020. Britain has earmarked \\u00a337bn for testing and tracing over the 2020 and 2021 financial years, and though it may well spend less, the shoddiness that has dogged some elements of the campaign, such as a database cock-up which lost thousands of results in September 2020, will be remembered.A basic problem is that contact tracing is a lot of work. Theoretical assessments based on analysing social networks and experience in Asia both suggest that some 30 contacts need to be identified for each index case. Digital tools can lessen the load. Resolve to Save Lives, a campaign run by Vital Strategies, an NGO, developed Locator, which taps into credit data to help tracers track down people who may have caught the virus from a specific index case. But the amount of work required remained enormous.A much discussed alternative to such programmes was to use the world\\u2019s most ubiquitous tracking devices: smartphones. Google and Apple worked together to develop a system which enabled phones to keep a list of occasions when they were near another phone for a significant period of time, and the identity of that second phone. When someone tests positive for SARS-CoV-2 they are asked to send a message from their phone which is used to notify all the other phones they had been near within a particular window of time. But everything is peer-to-peer; neither the big tech companies nor the public-health authorities get a list of contacts.Unfortunately this built-in privacy makes it hard to assess the technology\\u2019s efficacy. British and Swiss studies suggest such apps do reduce spread, but not enough to make them more than an also-ran technology. None of the successful contact-tracing systems in East Asia relies on such things to any significant extent.Even without good tracing, self isolation of those who tested positive helped slow the spread of the disease. But over time the shortcomings of the initial testing technology, reverse-transcriptase polymerase chain reaction (RT-PCR), became ever more apparent. It is conceptually elegant (see diagram) and easy for labs to use. But despite its familiarity, reliability and sensitivity, it has real disadvantages.One is that it needs labs, and is carried out most efficiently in big ones. This means samples may have to travel a long way. It also means that they can get held in queues. The Broad runs its RT-PCR tests in just three and a half hours; but the average sample takes 15 hours to process. Results from RT-PCR tests normally come in days not minutes.Another problem is that though the presence of viral RNA clearly shows that a person has been infected, it says very little about where they stand in the course of the disease; RNA is detectible from very soon after infection to long after the disease has run its course. In public-health terms what is needed is a test that spots people who are actually infectious\\u2014people with cells in their noses and throats actively churning out virus particles.To look for the SARS-CoV-2 particles themselves means looking for their distinctive protein components, not for the RNA that tells cells how to make them. The most detectible such component is the spike protein which studs the particles\\u2019 outer membranes. And one of the basic rules of modern biotechnology is that when you want to find a protein, use an antibody.Antibodies are large molecules that come in millions of varieties, each of which sticks to one target\\u2014known as that antibody\\u2019s \\u201cantigen\\u201d\\u2014and one target only. A handy technology called lateral-flow testing makes use of that specificity. A sample is placed at one end of a porous membrane and, as it seeps along to the other, encounters a line of antibodies designed to recognise it. When the sample is urine and the antigen is a hormone found in expectant women, you have a pregnancy test. When the sample is mucus from a swab and the antigen is the spike protein, it is a covid-19 test\\u2014a cheap, convenient one which can provide results with in half an hour. Such tests may not pick up all the people in whom RT-PCR might detect a trace of the virus. But if their nose and throat cells are not producing enough antigen for the test to detect, they are probably not producing enough to be infectious, either.At the beginning of the epidemic \\u201cthe supply chain for the lateral-flow tests wasn\\u2019t there,\\u201d says Dr Gabriel. Chris Hand, the chairman of Abingdon Health, a British contract manufacturer of lateral-flow tests, says the main bottleneck was the speciality membranes that are part of every test kit. \\u201cThey come on large reels of 100 metres plus, which go through automated equipment to add biochemicals by spraying them at low volumes,\\u201d he says. But once the biochemicals\\u2014the bespoke antibodies and some more generic bits and bobs\\u2014are ready, the production processes in place and the packaging sorted, the tests could be churned out by the million.New technologies now reaching the market will further change the dynamics of test and trace. QuantuMDx is one of a number of companies developing automated PCR-in-a-box systems that provide results within a couple of hours. Jonathan O\\u2019Halloran, the firm\\u2019s boss, says the British company has been relying on its own testing system for the past 22 weeks, testing its 90 staff members every morning (they are free to decline). About once a week lunchtime brings the news that someone has tested positive; they are immediately sent home to isolate. When things are done this fast, the fact that RNA is detectable before people are infectious is a plus; isolating on the basis of an early PCR test means no one ever turns up to work infectious. The company claims not to have lost a single day to infection.A combination of local, automated PCR and lateral-flow tests could be the basis of an ideal testing system\\u2014one which has a chance of keeping ahead of, and containing, a low level of disease rather than lagging behind one that is shooting up. Antigen tests would be used to scan the population for new infections. Those found would be referred to contact tracers; contacts who might have been infected could then be PCR-tested to find out which of them actually were.Great for public health. No benefit, in itself, to the index cases who would wait, isolated, to see what fate the virus and their immune systems had in store for them\\u2014painfully aware that their next encounter with the wonders of modern medical technology could be in a hospital bed. \\u25a0\",\n          \"FOR decades scientists have been trying to understand schizophrenia, a distressing disorder that afflicts one in a hundred people. Its destructive symptoms include hallucinations, delusions, muddled thoughts and changes in behaviour. The best drugs offer little more than the ability to target symptoms. There is presently no hope of a cure because its ultimate cause has long been a mystery.Schizophrenia is known to run in families, so it has always been thought that genes might shed light on the origins of the disease. But the genetics have proved harder to unpick than anyone imagined. It is only now, 16 years after the human genome was first sequenced, that scientists have homed in on the relevant genes.For more than five years researchers led by a group at the Broad Institute in Cambridge, Massachusetts, collected more than 100,000 DNA samples from around the world. They were looking for regions of the human genome that might be harbouring variants that increased the risk of schizophrenia. What they found implicated more than a hundred genes and provided a strong pointer towards a portion of the genome associated with infectious diseases. Many of these genes did not operate independently of each other, and further detective work revealed that the most important source of variation lay within a particular gene known as C4.Steven McCarroll, of Harvard Medical School, and his colleagues involved in the study, report in Nature this week that particular forms of the C4 gene led to a greater risk of developing schizophrenia. What makes C4 hard to identify is that it comes in many different forms and has an intriguing function. The gene produces a protein known as a complement factor, which helps to tag pathogens for removal by the body\\u2019s immune system. In the brain, however, it appears to be tied to a developmental process known as \\u201csynaptic pruning\\u201d, in which neurons are continuously eliminated throughout childhood and into early adulthood.In other words schizophrenics may be producing too much of a protein that creates a signal that synaptic connections should be removed. During late adolescence the brain undergoes widespread synaptic pruning, particularly in the cerebral cortex. This is also the age at which schizophrenia symptoms tend to begin.David Goldstein, director of the Institute for Genomic Medicine at Columbia University, says he is persuaded by the research, but reckons it is important for the medical community to examine the study carefully because the region of the genome under discussion is hard to work on and is \\u201cstuffed full of genetic variation\\u201d. That makes it difficult to point to a single gene and say \\u201cit\\u2019s this\\u201d, he adds.As to whether those with a family history of schizophrenia ought to have their genome sequenced to assess their susceptibility, Dr Goldstein is keen to point out that more work needs to be done on the risks that different forms of the genes bring. Moreover, other factors can be implicated in developing schizophrenia, ranging from taking psychoactive drugs to being exposed to viruses while in the womb.Synaptic pruning is so crucial in brain development researchers wonder what other mental disorders might be tied to its aberrant activation or regulation. This study is bound to initiate a flurry of similar work. Alzheimer\\u2019s disease, for one, has been tied to, among other things, excessive destruction of synapses. And it is already believed that autistic individuals have a surplus of synapses in the brain, due to a slowdown in normal brain pruning. Indeed, the onset of autism often follows a period when there is a burst of synapse formation. One big step forward in schizophrenia research could have an enormous impact on the way other neurological disorders are viewed and treated.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"title\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 962,\n        \"samples\": [\n          \"Genetic engineering may help control disease-carrying mosquitoes\",\n          \"Covid-19 has many faces\",\n          \"Studying broken chromosomes can illuminate neuroscience\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"date\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 783,\n        \"samples\": [\n          \" Aug 29th 2019\",\n          \" Aug 24th 2006\",\n          \" Jan 11th 2022\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"link\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 962,\n        \"samples\": [\n          \"https://www.economist.com/science-and-technology/2021/05/26/genetic-engineering-may-help-control-disease-carrying-mosquitoes\",\n          \"https://www.economist.com/science-and-technology/2020/05/09/covid-19-has-many-faces\",\n          \"https://www.economist.com/science-and-technology/2023/03/08/studying-broken-chromosomes-can-illuminate-neuroscience\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_sents = sent_split(df, 'text')"
      ],
      "metadata": {
        "id": "TBFw4ZROEOYl"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df_sents"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "kRZKXAxtESSt",
        "outputId": "135b8bbc-a793-418a-e47c-8d47a6e34ae8"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text\n",
              "0      FOR THE past four billion years or so the only...\n",
              "1      Sometimes the gene would be damaged or scrambl...\n",
              "2      From that raw material arose the glories of na...\n",
              "3      But beneath it all, gene begat gene.That is no...\n",
              "4      Now genes can be written from scratch and edit...\n",
              "...                                                  ...\n",
              "35972  He found that, after 15 days, parasite loads i...\n",
              "35973  Dr Kumar plans to try—though he will use genet...\n",
              "35974  But many, probably most, doctors are suspiciou...\n",
              "35975  Dr Kumar’s findings are, nevertheless, interes...\n",
              "35976  And regardless of whether his experiments in h...\n",
              "\n",
              "[35977 rows x 1 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-4e034105-e2c6-4b46-864b-ba438202d589\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Sometimes the gene would be damaged or scrambl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>From that raw material arose the glories of na...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>But beneath it all, gene begat gene.That is no...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Now genes can be written from scratch and edit...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35972</th>\n",
              "      <td>He found that, after 15 days, parasite loads i...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35973</th>\n",
              "      <td>Dr Kumar plans to try—though he will use genet...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35974</th>\n",
              "      <td>But many, probably most, doctors are suspiciou...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35975</th>\n",
              "      <td>Dr Kumar’s findings are, nevertheless, interes...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35976</th>\n",
              "      <td>And regardless of whether his experiments in h...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35977 rows × 1 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4e034105-e2c6-4b46-864b-ba438202d589')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-4e034105-e2c6-4b46-864b-ba438202d589 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-4e034105-e2c6-4b46-864b-ba438202d589');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-1c1bb14b-ac1d-4954-8d8e-91131d65150b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1c1bb14b-ac1d-4954-8d8e-91131d65150b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-1c1bb14b-ac1d-4954-8d8e-91131d65150b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_4bd33992-468b-4a21-b061-c437c2bf7b69\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_sents')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_4bd33992-468b-4a21-b061-c437c2bf7b69 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_sents');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_sents",
              "summary": "{\n  \"name\": \"df_sents\",\n  \"rows\": 35977,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35630,\n        \"samples\": [\n          \"Such an organisation would have the job of streamlining the impenetrable institutional tangle.\",\n          \"By 2050, according to the model, primary production is likely to have fallen in three of them, to be flat in one and to rise only in the Russian zone (the Kara Sea and part of the Barents Sea).\",\n          \"Now comes evidence that the same is true for another top predator: sharks.Jonathan Ruppert of the University of Toronto, in Canada, and his colleagues reached this conclusion by studying data on shark activity around two groups of coral reefs between Indonesia and Australia.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "id": "DWO5RClcFBLv"
      },
      "outputs": [],
      "source": [
        "def cleantext(df):\n",
        "    # удаляем имена из одной буквы и точки\n",
        "    df['cleaned_text'] = df['Text'].replace(r'[A-Z]\\.', \"\", regex=True)\n",
        "    # приводим к нижнему регистру\n",
        "    df['cleaned_text'] = df['cleaned_text'].str.lower()\n",
        "\n",
        "    # Внутритекстовые ссылки в квадратных скобках\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\[[0-9 ,– -]+\\]', \"\", regex=True) # от ссылок в квадратных скобках внутри текста\n",
        "\n",
        "    # Удаление дефиса\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r\"-\\n\", \"\", regex = True)\n",
        "\n",
        "    # удаляем всё, кроме заглавных и строчных букв английского алфавита\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'[^a-zA-Z0-9]', \" \", regex=True)\n",
        "\n",
        "    # Удаление doi | DOI\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'(/doi/10[.][0-9]{4,}.*)', '', regex=True)\n",
        "\n",
        "    # удаляем возможные ссылки URL(x):\n",
        "    df['cleaned_text']  = df['cleaned_text'].replace(r'https.*', \"\", regex = True)\n",
        "\n",
        "    # Удаление цифр\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\d+', '', regex=True)\n",
        "\n",
        "    # Удаление двух и более подряд идущих пробелов\n",
        "    df['cleaned_text'] = df['cleaned_text'].replace(r'\\s{2,}', ' ', regex=True)\n",
        "    return df\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df = cleantext(df_sents)"
      ],
      "metadata": {
        "id": "ZHQdzoscEdmZ"
      },
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "cA1W-wc7oJ2V",
        "outputId": "42cbfa64-49f3-487b-d23c-e270a738e869"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      FOR THE past four billion years or so the only...   \n",
              "1      Sometimes the gene would be damaged or scrambl...   \n",
              "2      From that raw material arose the glories of na...   \n",
              "3      But beneath it all, gene begat gene.That is no...   \n",
              "4      Now genes can be written from scratch and edit...   \n",
              "...                                                  ...   \n",
              "35972  He found that, after 15 days, parasite loads i...   \n",
              "35973  Dr Kumar plans to try—though he will use genet...   \n",
              "35974  But many, probably most, doctors are suspiciou...   \n",
              "35975  Dr Kumar’s findings are, nevertheless, interes...   \n",
              "35976  And regardless of whether his experiments in h...   \n",
              "\n",
              "                                            cleaned_text  \n",
              "0      for the past four billion years or so the only...  \n",
              "1      sometimes the gene would be damaged or scrambl...  \n",
              "2      from that raw material arose the glories of na...  \n",
              "3      but beneath it all gene begat gene that is no ...  \n",
              "4      now genes can be written from scratch and edit...  \n",
              "...                                                  ...  \n",
              "35972  he found that after days parasite loads in ani...  \n",
              "35973  dr kumar plans to try though he will use genet...  \n",
              "35974  but many probably most doctors are suspicious ...  \n",
              "35975  dr kumar s findings are nevertheless interesting   \n",
              "35976  and regardless of whether his experiments in h...  \n",
              "\n",
              "[35977 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-c078d013-453a-4d1d-8d9a-1609593fd048\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "      <td>for the past four billion years or so the only...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Sometimes the gene would be damaged or scrambl...</td>\n",
              "      <td>sometimes the gene would be damaged or scrambl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>From that raw material arose the glories of na...</td>\n",
              "      <td>from that raw material arose the glories of na...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>But beneath it all, gene begat gene.That is no...</td>\n",
              "      <td>but beneath it all gene begat gene that is no ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Now genes can be written from scratch and edit...</td>\n",
              "      <td>now genes can be written from scratch and edit...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35972</th>\n",
              "      <td>He found that, after 15 days, parasite loads i...</td>\n",
              "      <td>he found that after days parasite loads in ani...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35973</th>\n",
              "      <td>Dr Kumar plans to try—though he will use genet...</td>\n",
              "      <td>dr kumar plans to try though he will use genet...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35974</th>\n",
              "      <td>But many, probably most, doctors are suspiciou...</td>\n",
              "      <td>but many probably most doctors are suspicious ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35975</th>\n",
              "      <td>Dr Kumar’s findings are, nevertheless, interes...</td>\n",
              "      <td>dr kumar s findings are nevertheless interesting</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35976</th>\n",
              "      <td>And regardless of whether his experiments in h...</td>\n",
              "      <td>and regardless of whether his experiments in h...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35977 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c078d013-453a-4d1d-8d9a-1609593fd048')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-c078d013-453a-4d1d-8d9a-1609593fd048 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-c078d013-453a-4d1d-8d9a-1609593fd048');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-2c3b2594-6ec0-4607-95f3-c1e4bf0e4a2d\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-2c3b2594-6ec0-4607-95f3-c1e4bf0e4a2d')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-2c3b2594-6ec0-4607-95f3-c1e4bf0e4a2d button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_1b3d5daf-776a-4965-948b-99eaa0c0ba73\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_sents')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_1b3d5daf-776a-4965-948b-99eaa0c0ba73 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_sents');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_sents",
              "summary": "{\n  \"name\": \"df_sents\",\n  \"rows\": 35977,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35630,\n        \"samples\": [\n          \"Such an organisation would have the job of streamlining the impenetrable institutional tangle.\",\n          \"By 2050, according to the model, primary production is likely to have fallen in three of them, to be flat in one and to rise only in the Russian zone (the Kara Sea and part of the Barents Sea).\",\n          \"Now comes evidence that the same is true for another top predator: sharks.Jonathan Ruppert of the University of Toronto, in Canada, and his colleagues reached this conclusion by studying data on shark activity around two groups of coral reefs between Indonesia and Australia.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35600,\n        \"samples\": [\n          \"but the captains and the kings departed long ago \",\n          \"they are also usually more abundant than their hosts \",\n          \"mar starts up the cell s efflux system and also stops the cell from allowing any more threatening molecules in by halting the production of porin a membrane protein that acts as a channel into the cell \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def clean_long_sent(df):\n",
        "    df[\"len\"] = df[\"cleaned_text\"].apply(lambda x: len(x))\n",
        "    return df[df.len <= 512]\n",
        "\n",
        "cleaned_df = clean_long_sent(cleaned_df)"
      ],
      "metadata": {
        "id": "6lLelwyLaciF"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "ukKuR1EwFo1O",
        "outputId": "e35fd745-835b-4698-b197-22e089cbde69"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      FOR THE past four billion years or so the only...   \n",
              "1      Sometimes the gene would be damaged or scrambl...   \n",
              "2      From that raw material arose the glories of na...   \n",
              "3      But beneath it all, gene begat gene.That is no...   \n",
              "4      Now genes can be written from scratch and edit...   \n",
              "...                                                  ...   \n",
              "35972  He found that, after 15 days, parasite loads i...   \n",
              "35973  Dr Kumar plans to try—though he will use genet...   \n",
              "35974  But many, probably most, doctors are suspiciou...   \n",
              "35975  Dr Kumar’s findings are, nevertheless, interes...   \n",
              "35976  And regardless of whether his experiments in h...   \n",
              "\n",
              "                                            cleaned_text  len  \n",
              "0      for the past four billion years or so the only...  154  \n",
              "1      sometimes the gene would be damaged or scrambl...   96  \n",
              "2      from that raw material arose the glories of na...   62  \n",
              "3      but beneath it all gene begat gene that is no ...   58  \n",
              "4      now genes can be written from scratch and edit...   90  \n",
              "...                                                  ...  ...  \n",
              "35972  he found that after days parasite loads in ani...  269  \n",
              "35973  dr kumar plans to try though he will use genet...  137  \n",
              "35974  but many probably most doctors are suspicious ...  165  \n",
              "35975  dr kumar s findings are nevertheless interesting    49  \n",
              "35976  and regardless of whether his experiments in h...  188  \n",
              "\n",
              "[35852 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-41a44c06-2d69-4ad2-891b-90fdf5bce693\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "      <td>for the past four billion years or so the only...</td>\n",
              "      <td>154</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Sometimes the gene would be damaged or scrambl...</td>\n",
              "      <td>sometimes the gene would be damaged or scrambl...</td>\n",
              "      <td>96</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>From that raw material arose the glories of na...</td>\n",
              "      <td>from that raw material arose the glories of na...</td>\n",
              "      <td>62</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>But beneath it all, gene begat gene.That is no...</td>\n",
              "      <td>but beneath it all gene begat gene that is no ...</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Now genes can be written from scratch and edit...</td>\n",
              "      <td>now genes can be written from scratch and edit...</td>\n",
              "      <td>90</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35972</th>\n",
              "      <td>He found that, after 15 days, parasite loads i...</td>\n",
              "      <td>he found that after days parasite loads in ani...</td>\n",
              "      <td>269</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35973</th>\n",
              "      <td>Dr Kumar plans to try—though he will use genet...</td>\n",
              "      <td>dr kumar plans to try though he will use genet...</td>\n",
              "      <td>137</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35974</th>\n",
              "      <td>But many, probably most, doctors are suspiciou...</td>\n",
              "      <td>but many probably most doctors are suspicious ...</td>\n",
              "      <td>165</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35975</th>\n",
              "      <td>Dr Kumar’s findings are, nevertheless, interes...</td>\n",
              "      <td>dr kumar s findings are nevertheless interesting</td>\n",
              "      <td>49</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35976</th>\n",
              "      <td>And regardless of whether his experiments in h...</td>\n",
              "      <td>and regardless of whether his experiments in h...</td>\n",
              "      <td>188</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35852 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-41a44c06-2d69-4ad2-891b-90fdf5bce693')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-41a44c06-2d69-4ad2-891b-90fdf5bce693 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-41a44c06-2d69-4ad2-891b-90fdf5bce693');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-cac5060d-9cb6-4d38-b712-0a1c4af34e64\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-cac5060d-9cb6-4d38-b712-0a1c4af34e64')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-cac5060d-9cb6-4d38-b712-0a1c4af34e64 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_34792cff-4cce-4156-8e90-d755ea80aada\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('cleaned_df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_34792cff-4cce-4156-8e90-d755ea80aada button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('cleaned_df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "cleaned_df",
              "summary": "{\n  \"name\": \"cleaned_df\",\n  \"rows\": 35852,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35505,\n        \"samples\": [\n          \"(Cytokine storms, as it happens, are sometimes a cause of death in cases of covid-19.\",\n          \"Now her wait appears to be over.\",\n          \"Intensive fishing can rapidly deplete the number of smaller fish.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35475,\n        \"samples\": [\n          \"it and one of the other specimens are described in a paper that has just been published in nature ecology and evolution by wang bo of the nanjing institute of geology and palaeontology in china and his colleagues dr wang thinks chimerarachne yingi is a spider albeit an unusual one in that it has a tail \",\n          \"one characteristic but apocryphal tale tells of an american military system designed to translate russian into english which is said to have rendered the famous russian saying the spirit is willing but the flesh is weak into the vodka is good but the meat is rotten this sort of joke prompts a hollow laugh from those in the machine translation mt business \",\n          \"in a paper published in nature communications in april tiago costa a bacteriologist at imperial college london reported that conjugative pili hair like filaments that bacteria use to exchange genetic information can likewise be repurposed to bind them together the bacteria can also build barricades from other chemicals they find in their environment \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 86,\n        \"min\": 0,\n        \"max\": 512,\n        \"num_unique_values\": 503,\n        \"samples\": [\n          196,\n          49,\n          333\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 28
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "import string\n",
        "\n",
        "def stopwords_remover(words):\n",
        "\n",
        "  without_stopwords = \"\"\n",
        "  for word in words.split(\" \"):\n",
        "    if word in stops or word in string.punctuation:\n",
        "      continue\n",
        "    else:\n",
        "      without_stopwords += word + \" \"\n",
        "  return without_stopwords\n",
        "\"\"\""
      ],
      "metadata": {
        "id": "a2z4tdSuGFYQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df.loc[:, 'cleaned_text'] = cleaned_df['cleaned_text'].astype(str)"
      ],
      "metadata": {
        "id": "sEVhTyaKGnlE"
      },
      "execution_count": 29,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#cleaned_df.loc[:, 'cleaned_text'] = cleaned_df['cleaned_text'].apply(stopwords_remover)\n"
      ],
      "metadata": {
        "id": "ocSUfuyBGFYf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "XrHD33x1HQiF",
        "outputId": "c9a64707-09ce-4e03-91b0-744f1dcf0b18"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                    Text  \\\n",
              "0      FOR THE past four billion years or so the only...   \n",
              "1      Sometimes the gene would be damaged or scrambl...   \n",
              "2      From that raw material arose the glories of na...   \n",
              "3      But beneath it all, gene begat gene.That is no...   \n",
              "4      Now genes can be written from scratch and edit...   \n",
              "...                                                  ...   \n",
              "35972  He found that, after 15 days, parasite loads i...   \n",
              "35973  Dr Kumar plans to try—though he will use genet...   \n",
              "35974  But many, probably most, doctors are suspiciou...   \n",
              "35975  Dr Kumar’s findings are, nevertheless, interes...   \n",
              "35976  And regardless of whether his experiments in h...   \n",
              "\n",
              "                                            cleaned_text  len  \n",
              "0      for the past four billion years or so the only...  154  \n",
              "1      sometimes the gene would be damaged or scrambl...   96  \n",
              "2      from that raw material arose the glories of na...   62  \n",
              "3      but beneath it all gene begat gene that is no ...   58  \n",
              "4      now genes can be written from scratch and edit...   90  \n",
              "...                                                  ...  ...  \n",
              "35972  he found that after days parasite loads in ani...  269  \n",
              "35973  dr kumar plans to try though he will use genet...  137  \n",
              "35974  but many probably most doctors are suspicious ...  165  \n",
              "35975  dr kumar s findings are nevertheless interesting    49  \n",
              "35976  and regardless of whether his experiments in h...  188  \n",
              "\n",
              "[35852 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-53cebbcd-70b3-460e-8276-ef00a5205eaa\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Text</th>\n",
              "      <th>cleaned_text</th>\n",
              "      <th>len</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FOR THE past four billion years or so the only...</td>\n",
              "      <td>for the past four billion years or so the only...</td>\n",
              "      <td>154</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Sometimes the gene would be damaged or scrambl...</td>\n",
              "      <td>sometimes the gene would be damaged or scrambl...</td>\n",
              "      <td>96</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>From that raw material arose the glories of na...</td>\n",
              "      <td>from that raw material arose the glories of na...</td>\n",
              "      <td>62</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>But beneath it all, gene begat gene.That is no...</td>\n",
              "      <td>but beneath it all gene begat gene that is no ...</td>\n",
              "      <td>58</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Now genes can be written from scratch and edit...</td>\n",
              "      <td>now genes can be written from scratch and edit...</td>\n",
              "      <td>90</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35972</th>\n",
              "      <td>He found that, after 15 days, parasite loads i...</td>\n",
              "      <td>he found that after days parasite loads in ani...</td>\n",
              "      <td>269</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35973</th>\n",
              "      <td>Dr Kumar plans to try—though he will use genet...</td>\n",
              "      <td>dr kumar plans to try though he will use genet...</td>\n",
              "      <td>137</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35974</th>\n",
              "      <td>But many, probably most, doctors are suspiciou...</td>\n",
              "      <td>but many probably most doctors are suspicious ...</td>\n",
              "      <td>165</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35975</th>\n",
              "      <td>Dr Kumar’s findings are, nevertheless, interes...</td>\n",
              "      <td>dr kumar s findings are nevertheless interesting</td>\n",
              "      <td>49</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35976</th>\n",
              "      <td>And regardless of whether his experiments in h...</td>\n",
              "      <td>and regardless of whether his experiments in h...</td>\n",
              "      <td>188</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>35852 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-53cebbcd-70b3-460e-8276-ef00a5205eaa')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-53cebbcd-70b3-460e-8276-ef00a5205eaa button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-53cebbcd-70b3-460e-8276-ef00a5205eaa');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-08343273-7ac5-4121-b990-84a30831e84f\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-08343273-7ac5-4121-b990-84a30831e84f')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-08343273-7ac5-4121-b990-84a30831e84f button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_40a80f37-7c96-494b-9424-d49c8b049b35\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('cleaned_df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_40a80f37-7c96-494b-9424-d49c8b049b35 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('cleaned_df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "cleaned_df",
              "summary": "{\n  \"name\": \"cleaned_df\",\n  \"rows\": 35852,\n  \"fields\": [\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35505,\n        \"samples\": [\n          \"(Cytokine storms, as it happens, are sometimes a cause of death in cases of covid-19.\",\n          \"Now her wait appears to be over.\",\n          \"Intensive fishing can rapidly deplete the number of smaller fish.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cleaned_text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 35475,\n        \"samples\": [\n          \"it and one of the other specimens are described in a paper that has just been published in nature ecology and evolution by wang bo of the nanjing institute of geology and palaeontology in china and his colleagues dr wang thinks chimerarachne yingi is a spider albeit an unusual one in that it has a tail \",\n          \"one characteristic but apocryphal tale tells of an american military system designed to translate russian into english which is said to have rendered the famous russian saying the spirit is willing but the flesh is weak into the vodka is good but the meat is rotten this sort of joke prompts a hollow laugh from those in the machine translation mt business \",\n          \"in a paper published in nature communications in april tiago costa a bacteriologist at imperial college london reported that conjugative pili hair like filaments that bacteria use to exchange genetic information can likewise be repurposed to bind them together the bacteria can also build barricades from other chemicals they find in their environment \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 86,\n        \"min\": 0,\n        \"max\": 512,\n        \"num_unique_values\": 503,\n        \"samples\": [\n          196,\n          49,\n          333\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "cleaned_df.to_csv('prep_bio_Economist_SW.csv')"
      ],
      "metadata": {
        "id": "TUhIOvhUHX7Q"
      },
      "execution_count": 31,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}